{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Correlations\n",
    "This notebook displays the correlation between viral barcode expression in the transcriptome, supernatant, and second infection."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook setup\n",
    "Import python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display\n",
    "\n",
    "from dms_variants.constants import CBPALETTE\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import plotnine as p9\n",
    "\n",
    "import scipy\n",
    "\n",
    "import statsmodels.stats.multitest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get `snakemake` variables [as described here](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#jupyter-notebook-integration):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hardcoded for now\n",
    "viral_tag_by_cell_csv = 'results/viral_tags_bcs_in_cells/scProgenyProduction_trial2_cell_barcodes_with_viral_tags.csv.gz'\n",
    "viral_bc_by_cell_filtered_csv = 'results/viral_fastq10x/scProgenyProduction_trial2_viral_bc_by_cell_filtered.csv.gz'\n",
    "viral_bc_in_progeny_freq_csv = 'results/viral_progeny/scProgenyProduction_trial2_viral_bc_in_progeny_freq.csv.gz'\n",
    "expt = 'scProgenyProduction_trial2'\n",
    "plot = 'results/viral_fastq10x/scProgenyProduction_trial2_viral_bc_correlations.pdf'\n",
    "barcoded_viral_genes = ['fluHA', 'fluNA']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Organize data\n",
    "Read the cell barcodes and tags into a pandas dataframe. Only keep relevant columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCACAAGGCCTC</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCACACACACGC</td>\n",
       "      <td>True</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCATCGTGCATA</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCATCTACTGCC</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7436</th>\n",
       "      <td>TTTGGTTGTTAAGCAA</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7437</th>\n",
       "      <td>TTTGGTTTCGTCGCTT</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7438</th>\n",
       "      <td>TTTGTTGCATGTGGTT</td>\n",
       "      <td>True</td>\n",
       "      <td>wt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7439</th>\n",
       "      <td>TTTGTTGTCGTCGGGT</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7440</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>True</td>\n",
       "      <td>wt</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7441 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode  infected infecting_viral_tag\n",
       "0     AAACCCAAGTAGGTTA      True                 syn\n",
       "1     AAACCCACAAGGCCTC      True                 syn\n",
       "2     AAACCCACACACACGC      True                both\n",
       "3     AAACCCATCGTGCATA      True                 syn\n",
       "4     AAACCCATCTACTGCC     False                none\n",
       "...                ...       ...                 ...\n",
       "7436  TTTGGTTGTTAAGCAA     False                none\n",
       "7437  TTTGGTTTCGTCGCTT     False                none\n",
       "7438  TTTGTTGCATGTGGTT      True                  wt\n",
       "7439  TTTGTTGTCGTCGGGT     False                none\n",
       "7440  TTTGTTGTCTAGGAAA      True                  wt\n",
       "\n",
       "[7441 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_cells = pd.read_csv(viral_tag_by_cell_csv)\n",
    "all_cells = all_cells[['cell_barcode',\n",
    "                       'infected',\n",
    "                       'infecting_viral_tag']]\n",
    "display(all_cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the viral barcode frequencies from the transcriptome. Filter out viral barcodes that didn't pass significant threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>frac_viral_bc_UMIs</th>\n",
       "      <th>reject_uninfected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGAATCGACACATGTC</td>\n",
       "      <td>0.002609</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>CACGGATGGTGTACGA</td>\n",
       "      <td>0.002236</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>CTTAACTGTATATTCG</td>\n",
       "      <td>0.004844</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>TCTTAAGTATATCAGA</td>\n",
       "      <td>0.002422</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>TGTAAATAGAGTTCGC</td>\n",
       "      <td>0.003726</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58905</th>\n",
       "      <td>TTTGGTTCATCTGCGG</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTGTATAAAAATACAG</td>\n",
       "      <td>0.007323</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58906</th>\n",
       "      <td>TTTGGTTGTCGGTGTC</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>AATAAGCGGCTCTTTG</td>\n",
       "      <td>0.003318</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58913</th>\n",
       "      <td>TTTGTTGCATGTGGTT</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>CACCCCGTTAGTGGGG</td>\n",
       "      <td>0.007294</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58923</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GAACCCGATGGGGAAT</td>\n",
       "      <td>0.009800</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58926</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TGTTGGGCTTTAAGCA</td>\n",
       "      <td>0.001885</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16305 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           cell_barcode   gene     viral_barcode  frac_viral_bc_UMIs  \\\n",
       "1      AAACCCAAGTAGGTTA  fluHA  AGAATCGACACATGTC            0.002609   \n",
       "4      AAACCCAAGTAGGTTA  fluHA  CACGGATGGTGTACGA            0.002236   \n",
       "6      AAACCCAAGTAGGTTA  fluHA  CTTAACTGTATATTCG            0.004844   \n",
       "8      AAACCCAAGTAGGTTA  fluHA  TCTTAAGTATATCAGA            0.002422   \n",
       "12     AAACCCAAGTAGGTTA  fluHA  TGTAAATAGAGTTCGC            0.003726   \n",
       "...                 ...    ...               ...                 ...   \n",
       "58905  TTTGGTTCATCTGCGG  fluNA  TTGTATAAAAATACAG            0.007323   \n",
       "58906  TTTGGTTGTCGGTGTC  fluNA  AATAAGCGGCTCTTTG            0.003318   \n",
       "58913  TTTGTTGCATGTGGTT  fluNA  CACCCCGTTAGTGGGG            0.007294   \n",
       "58923  TTTGTTGTCTAGGAAA  fluNA  GAACCCGATGGGGAAT            0.009800   \n",
       "58926  TTTGTTGTCTAGGAAA  fluNA  TGTTGGGCTTTAAGCA            0.001885   \n",
       "\n",
       "       reject_uninfected  \n",
       "1                   True  \n",
       "4                   True  \n",
       "6                   True  \n",
       "8                   True  \n",
       "12                  True  \n",
       "...                  ...  \n",
       "58905               True  \n",
       "58906               True  \n",
       "58913               True  \n",
       "58923               True  \n",
       "58926               True  \n",
       "\n",
       "[16305 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "transcriptome_viral_bc_freqs = pd.read_csv(viral_bc_by_cell_filtered_csv)\n",
    "assert set(transcriptome_viral_bc_freqs['gene']) == set(barcoded_viral_genes), \\\n",
    "       \"Barcoded genes in barcode counts do not match expectation.\"\n",
    "transcriptome_viral_bc_freqs = (transcriptome_viral_bc_freqs\n",
    "                                .query('reject_uninfected == True'))\n",
    "display(transcriptome_viral_bc_freqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the viral barcode frequencies from the progeny. **For now just work with supernatant data. Remove second_infection frequencies.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>tag</th>\n",
       "      <th>gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>mean_freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>90716</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAAATCTTAATGA</td>\n",
       "      <td>2.414539e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90717</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAACGAATAAATT</td>\n",
       "      <td>2.483091e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90718</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAACGAATGGATT</td>\n",
       "      <td>3.352175e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90719</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAACGAGAAGAAT</td>\n",
       "      <td>1.277534e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90720</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAAGTTGAGATTT</td>\n",
       "      <td>1.205557e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130646</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTGTTAGCGTCCTG</td>\n",
       "      <td>3.067874e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130647</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTTTAGAAAACGTA</td>\n",
       "      <td>8.987181e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130648</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTTTAGAAAACGTC</td>\n",
       "      <td>1.680603e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130649</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTTTCACTGCCATT</td>\n",
       "      <td>8.987181e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130650</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTTTCGGCGAGTGC</td>\n",
       "      <td>1.204282e-04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39935 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             source  tag   gene     viral_barcode     mean_freq\n",
       "90716   supernatant  syn  fluHA  AAAAAAATCTTAATGA  2.414539e-04\n",
       "90717   supernatant  syn  fluHA  AAAAAACGAATAAATT  2.483091e-06\n",
       "90718   supernatant  syn  fluHA  AAAAAACGAATGGATT  3.352175e-04\n",
       "90719   supernatant  syn  fluHA  AAAAAACGAGAAGAAT  1.277534e-06\n",
       "90720   supernatant  syn  fluHA  AAAAAAGTTGAGATTT  1.205557e-06\n",
       "...             ...  ...    ...               ...           ...\n",
       "130646  supernatant   wt  fluNA  TTTTGTTAGCGTCCTG  3.067874e-04\n",
       "130647  supernatant   wt  fluNA  TTTTTTAGAAAACGTA  8.987181e-07\n",
       "130648  supernatant   wt  fluNA  TTTTTTAGAAAACGTC  1.680603e-04\n",
       "130649  supernatant   wt  fluNA  TTTTTTCACTGCCATT  8.987181e-07\n",
       "130650  supernatant   wt  fluNA  TTTTTTCGGCGAGTGC  1.204282e-04\n",
       "\n",
       "[39935 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "progeny_viral_bc_freqs = pd.read_csv(viral_bc_in_progeny_freq_csv)\n",
    "assert set(progeny_viral_bc_freqs['gene']) == set(barcoded_viral_genes), \\\n",
    "       \"Barcoded genes in barcode counts do not match expectation.\"\n",
    "progeny_viral_bc_freqs = (progeny_viral_bc_freqs\n",
    "                          .rename(columns={'barcode': 'viral_barcode'}))\n",
    "progeny_viral_bc_freqs = (progeny_viral_bc_freqs\n",
    "                          .query('source == \"supernatant\"'))\n",
    "display(progeny_viral_bc_freqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge dataframes to one. Structure of the dataframe should be as follows:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_bc_freqs = pd.merge()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
