{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter viral barcodes in transcripts\n",
    "This notebook filters viral barcodes in 10X transcriptome data to remove UMIs that are likely derived from leaked transcripts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook setup\n",
    "Import python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display\n",
    "\n",
    "from dms_variants.constants import CBPALETTE\n",
    "\n",
    "import kneed\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import plotnine as p9"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get `snakemake` variables [as described here](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#jupyter-notebook-integration):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_tag_by_cell_csv = 'results/viral_tags_bcs_in_cells/scProgenyProduction_trial2_cell_barcodes_with_viral_tags.csv.gz'\n",
    "viral_bc_by_cell_corrected_csv = 'results/viral_fastq10x/scProgenyProduction_trial2_viral_bc_by_cell_corrected.csv.gz'\n",
    "expt = 'scProgenyProduction_trial2'\n",
    "viral_bc_by_cell_filtered_csv = 'results/viral_fastq10x/scProgenyProduction_trial2_viral_bc_by_cell_filtered.csv.gz'\n",
    "plot = 'results/viral_fastq10x/scProgenyProduction_trial2_viral_bc_by_cell_filtered.pdf'\n",
    "barcoded_viral_genes = ['fluHA', 'fluNA']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Organize data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the viral barcode UMI counts data into a pandas dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>bc_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>ACGTTATTGATTGAGA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGAATCGACACATGTC</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGCCATAGTCTAAAGG</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGGATGATTTTTTTGT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>CACGGATGGTGTACGA</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55238</th>\n",
       "      <td>TTTGTTGCATGTGGTT</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTCCTCAGCCTTTAAG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55239</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GAACCCGATGGGGAAT</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55240</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TAAGGTAAAATAATAG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55241</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TAAGGTATAATTCTAG</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55242</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TGTTGGGCTTTAAGCA</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55243 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           cell_barcode   gene     viral_barcode  bc_UMIs\n",
       "0      AAACCCAAGTAGGTTA  fluHA  ACGTTATTGATTGAGA        1\n",
       "1      AAACCCAAGTAGGTTA  fluHA  AGAATCGACACATGTC       14\n",
       "2      AAACCCAAGTAGGTTA  fluHA  AGCCATAGTCTAAAGG        8\n",
       "3      AAACCCAAGTAGGTTA  fluHA  AGGATGATTTTTTTGT        5\n",
       "4      AAACCCAAGTAGGTTA  fluHA  CACGGATGGTGTACGA       12\n",
       "...                 ...    ...               ...      ...\n",
       "55238  TTTGTTGCATGTGGTT  fluNA  TTCCTCAGCCTTTAAG        3\n",
       "55239  TTTGTTGTCTAGGAAA  fluNA  GAACCCGATGGGGAAT       26\n",
       "55240  TTTGTTGTCTAGGAAA  fluNA  TAAGGTAAAATAATAG        1\n",
       "55241  TTTGTTGTCTAGGAAA  fluNA  TAAGGTATAATTCTAG        4\n",
       "55242  TTTGTTGTCTAGGAAA  fluNA  TGTTGGGCTTTAAGCA        5\n",
       "\n",
       "[55243 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_bc_counts = pd.read_csv(viral_bc_by_cell_corrected_csv)\n",
    "viral_bc_counts = (viral_bc_counts\n",
    "                   .rename(columns={'count': 'bc_UMIs'}))\n",
    "assert set(viral_bc_counts['gene']) == set(barcoded_viral_genes), \\\n",
    "       \"Barcoded genes in barcode counts do not match expectation.\"\n",
    "display(viral_bc_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the total number of UMIs per cell into a pandas dataframe. Only keep relevant columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>cellular_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "      <td>0.589901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCACAAGGCCTC</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>10304</td>\n",
       "      <td>1485</td>\n",
       "      <td>8819</td>\n",
       "      <td>0.144119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCACACACACGC</td>\n",
       "      <td>True</td>\n",
       "      <td>both</td>\n",
       "      <td>7395</td>\n",
       "      <td>3320</td>\n",
       "      <td>4075</td>\n",
       "      <td>0.448952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCATCGTGCATA</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>2764</td>\n",
       "      <td>1597</td>\n",
       "      <td>1167</td>\n",
       "      <td>0.577786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCATCTACTGCC</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>6313</td>\n",
       "      <td>3</td>\n",
       "      <td>6310</td>\n",
       "      <td>0.000475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7436</th>\n",
       "      <td>TTTGGTTGTTAAGCAA</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>8528</td>\n",
       "      <td>29</td>\n",
       "      <td>8499</td>\n",
       "      <td>0.003401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7437</th>\n",
       "      <td>TTTGGTTTCGTCGCTT</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>6492</td>\n",
       "      <td>6</td>\n",
       "      <td>6486</td>\n",
       "      <td>0.000924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7438</th>\n",
       "      <td>TTTGTTGCATGTGGTT</td>\n",
       "      <td>True</td>\n",
       "      <td>wt</td>\n",
       "      <td>3016</td>\n",
       "      <td>1680</td>\n",
       "      <td>1336</td>\n",
       "      <td>0.557029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7439</th>\n",
       "      <td>TTTGTTGTCGTCGGGT</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>11070</td>\n",
       "      <td>2</td>\n",
       "      <td>11068</td>\n",
       "      <td>0.000181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7440</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>True</td>\n",
       "      <td>wt</td>\n",
       "      <td>2653</td>\n",
       "      <td>1014</td>\n",
       "      <td>1639</td>\n",
       "      <td>0.382209</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7441 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode  infected infecting_viral_tag  total_UMIs  viral_UMIs  \\\n",
       "0     AAACCCAAGTAGGTTA      True                 syn        5367        3166   \n",
       "1     AAACCCACAAGGCCTC      True                 syn       10304        1485   \n",
       "2     AAACCCACACACACGC      True                both        7395        3320   \n",
       "3     AAACCCATCGTGCATA      True                 syn        2764        1597   \n",
       "4     AAACCCATCTACTGCC     False                none        6313           3   \n",
       "...                ...       ...                 ...         ...         ...   \n",
       "7436  TTTGGTTGTTAAGCAA     False                none        8528          29   \n",
       "7437  TTTGGTTTCGTCGCTT     False                none        6492           6   \n",
       "7438  TTTGTTGCATGTGGTT      True                  wt        3016        1680   \n",
       "7439  TTTGTTGTCGTCGGGT     False                none       11070           2   \n",
       "7440  TTTGTTGTCTAGGAAA      True                  wt        2653        1014   \n",
       "\n",
       "      cellular_UMIs  frac_viral_UMIs  \n",
       "0              2201         0.589901  \n",
       "1              8819         0.144119  \n",
       "2              4075         0.448952  \n",
       "3              1167         0.577786  \n",
       "4              6310         0.000475  \n",
       "...             ...              ...  \n",
       "7436           8499         0.003401  \n",
       "7437           6486         0.000924  \n",
       "7438           1336         0.557029  \n",
       "7439          11068         0.000181  \n",
       "7440           1639         0.382209  \n",
       "\n",
       "[7441 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_cells = pd.read_csv(viral_tag_by_cell_csv)\n",
    "all_cells = all_cells[['cell_barcode',\n",
    "                       'infected',\n",
    "                       'infecting_viral_tag',\n",
    "                       'total_UMIs',\n",
    "                       'viral_UMIs',\n",
    "                       'cellular_UMIs',\n",
    "                       'frac_viral_UMIs']]\n",
    "display(all_cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sanity check that `total_UMIs` is equal to `viral_UMIs + cellular_UMIs`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert bool((all_cells['total_UMIs'] ==\n",
    "             all_cells['viral_UMIs'] +\n",
    "             all_cells['cellular_UMIs'])\n",
    "            .all()), \"UMI counts do not add up\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge dataframes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>cellular_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "      <th>gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>bc_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "      <td>0.589901</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>ACGTTATTGATTGAGA</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "      <td>0.589901</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGAATCGACACATGTC</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "      <td>0.589901</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGCCATAGTCTAAAGG</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "      <td>0.589901</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGGATGATTTTTTTGT</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "      <td>0.589901</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>CACGGATGGTGTACGA</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58922</th>\n",
       "      <td>TTTGTTGTCGTCGGGT</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>11070</td>\n",
       "      <td>2</td>\n",
       "      <td>11068</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58923</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>True</td>\n",
       "      <td>wt</td>\n",
       "      <td>2653</td>\n",
       "      <td>1014</td>\n",
       "      <td>1639</td>\n",
       "      <td>0.382209</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GAACCCGATGGGGAAT</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58924</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>True</td>\n",
       "      <td>wt</td>\n",
       "      <td>2653</td>\n",
       "      <td>1014</td>\n",
       "      <td>1639</td>\n",
       "      <td>0.382209</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TAAGGTAAAATAATAG</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58925</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>True</td>\n",
       "      <td>wt</td>\n",
       "      <td>2653</td>\n",
       "      <td>1014</td>\n",
       "      <td>1639</td>\n",
       "      <td>0.382209</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TAAGGTATAATTCTAG</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58926</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>True</td>\n",
       "      <td>wt</td>\n",
       "      <td>2653</td>\n",
       "      <td>1014</td>\n",
       "      <td>1639</td>\n",
       "      <td>0.382209</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TGTTGGGCTTTAAGCA</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58927 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           cell_barcode  infected infecting_viral_tag  total_UMIs  viral_UMIs  \\\n",
       "0      AAACCCAAGTAGGTTA      True                 syn        5367        3166   \n",
       "1      AAACCCAAGTAGGTTA      True                 syn        5367        3166   \n",
       "2      AAACCCAAGTAGGTTA      True                 syn        5367        3166   \n",
       "3      AAACCCAAGTAGGTTA      True                 syn        5367        3166   \n",
       "4      AAACCCAAGTAGGTTA      True                 syn        5367        3166   \n",
       "...                 ...       ...                 ...         ...         ...   \n",
       "58922  TTTGTTGTCGTCGGGT     False                none       11070           2   \n",
       "58923  TTTGTTGTCTAGGAAA      True                  wt        2653        1014   \n",
       "58924  TTTGTTGTCTAGGAAA      True                  wt        2653        1014   \n",
       "58925  TTTGTTGTCTAGGAAA      True                  wt        2653        1014   \n",
       "58926  TTTGTTGTCTAGGAAA      True                  wt        2653        1014   \n",
       "\n",
       "       cellular_UMIs  frac_viral_UMIs   gene     viral_barcode  bc_UMIs  \n",
       "0               2201         0.589901  fluHA  ACGTTATTGATTGAGA      1.0  \n",
       "1               2201         0.589901  fluHA  AGAATCGACACATGTC     14.0  \n",
       "2               2201         0.589901  fluHA  AGCCATAGTCTAAAGG      8.0  \n",
       "3               2201         0.589901  fluHA  AGGATGATTTTTTTGT      5.0  \n",
       "4               2201         0.589901  fluHA  CACGGATGGTGTACGA     12.0  \n",
       "...              ...              ...    ...               ...      ...  \n",
       "58922          11068         0.000181  fluNA               NaN      NaN  \n",
       "58923           1639         0.382209  fluNA  GAACCCGATGGGGAAT     26.0  \n",
       "58924           1639         0.382209  fluNA  TAAGGTAAAATAATAG      1.0  \n",
       "58925           1639         0.382209  fluNA  TAAGGTATAATTCTAG      4.0  \n",
       "58926           1639         0.382209  fluNA  TGTTGGGCTTTAAGCA      5.0  \n",
       "\n",
       "[58927 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_bc_frac = pd.merge(\n",
    "                    left=pd.concat([all_cells.assign(gene=gene)\n",
    "                                    for gene in barcoded_viral_genes]),\n",
    "                    right=viral_bc_counts,\n",
    "                    how='outer',\n",
    "                    on=['cell_barcode', 'gene'],\n",
    "                    validate='one_to_many')\n",
    "assert (viral_bc_frac['cell_barcode'].unique() == \\\n",
    "    all_cells['cell_barcode'].unique()).all(), \\\n",
    "    \"Cell barcodes in merged dataframe don't \" \\\n",
    "    \"match barcodes in source data.\"\n",
    "assert viral_bc_frac['viral_barcode'].nunique() == \\\n",
    "    viral_bc_counts['viral_barcode'].nunique(), \\\n",
    "    \"Number of viral barcodes in merged dataframe don't \" \\\n",
    "    \"match number of barcodes in source data.\"\n",
    "\n",
    "# Make `infecting_viral_tag` column ordered category\n",
    "viral_bc_frac['infecting_viral_tag'] = (viral_bc_frac['infecting_viral_tag']\n",
    "                                        .astype('category')\n",
    "                                        .cat\n",
    "                                        .reorder_categories(['none',\n",
    "                                                             'wt',\n",
    "                                                             'syn',\n",
    "                                                             'both']))\n",
    "\n",
    "display(viral_bc_frac)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sum total number of barcoded UMIs per cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_bc_frac = pd.merge(left=viral_bc_frac,\n",
    "                         right=(viral_bc_frac\n",
    "                                .groupby('cell_barcode')\n",
    "                                ['bc_UMIs']\n",
    "                                .sum()\n",
    "                                .reset_index()\n",
    "                                .rename(columns={'bc_UMIs':\n",
    "                                                 'bc_UMIs_in_cell'})),\n",
    "                         on=['cell_barcode'])\n",
    "\n",
    "display(viral_bc_frac)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate fraction of all UMIs and fraction of viral UMIs that have barcode."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_bc_frac['frac_total_UMIs_wBC'] = (viral_bc_frac['bc_UMIs_in_cell'] /\n",
    "                                        viral_bc_frac['total_UMIs'])\n",
    "viral_bc_frac['frac_viral_UMIs_wBC'] = (viral_bc_frac['bc_UMIs_in_cell'] /\n",
    "                                        viral_bc_frac['viral_UMIs'])\n",
    "display(viral_bc_frac)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate **each barcode's** fraction of all UMIs per cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_bc_frac['frac_bc_UMIs'] = (viral_bc_frac['bc_UMIs'] /\n",
    "                                 viral_bc_frac['total_UMIs'])\n",
    "\n",
    "display(viral_bc_frac)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plots\n",
    "Set base plot style:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p9.theme_set(p9.theme_classic())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Per cell metrics\n",
    "Summary figures that address number and fraction of barcoded viral UMIs **in aggregate for each cell**.\n",
    "\n",
    "Plot number of viral barcode UMIs per cell and fraction of all UMIs with viral barcode per cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = (p9.ggplot((viral_bc_frac\n",
    "                  .groupby('cell_barcode')\n",
    "                  .max()\n",
    "                  .reset_index()), p9.aes(x='bc_UMIs_in_cell')) +\n",
    "       p9.geom_histogram(bins=20) +\n",
    "       p9.ggtitle('barcode UMI counts per cell\\n'\n",
    "                  f'{expt}') +\n",
    "       p9.theme(figure_size=(2, 2),\n",
    "                plot_title=p9.element_text(size=9),\n",
    "                axis_title=p9.element_text(size=9),\n",
    "                legend_title=p9.element_text(size=9),\n",
    "                legend_title_align='center'))\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = (p9.ggplot((viral_bc_frac\n",
    "                  .groupby('cell_barcode')\n",
    "                  .max()\n",
    "                  .reset_index()), p9.aes(x='frac_total_UMIs_wBC')) +\n",
    "       p9.geom_histogram(bins=20) +\n",
    "       p9.ggtitle('fraction of barcoded UMIs per cell\\n'\n",
    "                  f'{expt}') +\n",
    "       p9.theme(figure_size=(2, 2),\n",
    "                plot_title=p9.element_text(size=9),\n",
    "                axis_title=p9.element_text(size=9),\n",
    "                legend_title=p9.element_text(size=9),\n",
    "                legend_title_align='center'))\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot fraction of barcoded UMIs per cell as a function of viral UMI count and fraction:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = (p9.ggplot((viral_bc_frac\n",
    "                  .groupby('cell_barcode')\n",
    "                  .max()\n",
    "                  .reset_index()), p9.aes(x='viral_UMIs',\n",
    "                                          y='frac_total_UMIs_wBC')) +\n",
    "       p9.geom_point(alpha=0.3) +\n",
    "       p9.ggtitle('fraction of barcoded UMIs per cell\\n'\n",
    "                  'vs total viral UMI count\\n'\n",
    "                  f'{expt}') +\n",
    "       p9.theme(figure_size=(2, 2),\n",
    "                plot_title=p9.element_text(size=9),\n",
    "                axis_title=p9.element_text(size=9),\n",
    "                legend_title=p9.element_text(size=9),\n",
    "                legend_title_align='center'))\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = (p9.ggplot((viral_bc_frac\n",
    "                  .groupby('cell_barcode')\n",
    "                  .max()\n",
    "                  .reset_index()), p9.aes(x='frac_viral_UMIs',\n",
    "                                          y='frac_total_UMIs_wBC')) +\n",
    "       p9.geom_point(alpha=0.3) +\n",
    "       p9.ggtitle('fraction of barcoded UMIs per cell\\n'\n",
    "                  'vs fraction viral UMIs\\n'\n",
    "                  f'{expt}') +\n",
    "       p9.theme(figure_size=(2, 2),\n",
    "                plot_title=p9.element_text(size=9),\n",
    "                axis_title=p9.element_text(size=9),\n",
    "                legend_title=p9.element_text(size=9),\n",
    "                legend_title_align='center'))\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot relationship between viral UMI fraction and fraction of viral UMIs with barcode:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = (p9.ggplot((viral_bc_frac\n",
    "                  .groupby('cell_barcode')\n",
    "                  .max()\n",
    "                  .reset_index()), p9.aes(x='frac_viral_UMIs',\n",
    "                                          y='frac_viral_UMIs_wBC')) +\n",
    "       p9.geom_point(alpha=0.3) +\n",
    "       p9.ggtitle('fraction of barcoded UMIs per cell\\n'\n",
    "                  'vs fraction barcoded UMIs in viral UMIs\\n'\n",
    "                  f'{expt}') +\n",
    "       p9.theme(figure_size=(2, 2),\n",
    "                plot_title=p9.element_text(size=9),\n",
    "                axis_title=p9.element_text(size=9),\n",
    "                legend_title=p9.element_text(size=9),\n",
    "                legend_title_align='center'))\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Per barcode metrics\n",
    "Figures that show the outcomes **for each barcode individually.**\n",
    "\n",
    "Plot distribution of each barcode's fraction of a cell's total UMIs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = (p9.ggplot(viral_bc_frac, p9.aes(x='frac_bc_UMIs',\n",
    "                                       fill='infecting_viral_tag')) +\n",
    "       p9.geom_histogram(bins=200, position='stack') +\n",
    "       p9.facet_grid('infecting_viral_tag~') +\n",
    "       p9.ggtitle('fraction of all UMIs in cell\\n'\n",
    "                  'from each viral barcode\\n'\n",
    "                  f'{expt}') +\n",
    "       p9.scale_y_log10() +\n",
    "       p9.theme(figure_size=(5,\n",
    "                             1*viral_bc_frac['infecting_viral_tag'].nunique()),\n",
    "                plot_title=p9.element_text(size=10),\n",
    "                axis_title=p9.element_text(size=10),\n",
    "                legend_title=p9.element_text(size=9),\n",
    "                legend_title_align='center') +\n",
    "       p9.scale_fill_manual(CBPALETTE[0:]))\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = (p9.ggplot(viral_bc_frac, p9.aes(x='frac_bc_UMIs',\n",
    "                                       y='frac_viral_UMIs',\n",
    "                                       color='infecting_viral_tag')) +\n",
    "       p9.geom_point(alpha=0.3) +\n",
    "       p9.facet_grid('infecting_viral_tag~') +\n",
    "       p9.ggtitle('fraction of all UMIs in cell\\n'\n",
    "                  'from each viral barcode\\n'\n",
    "                  'vs fraction of viral UMIs in that cell\\n'\n",
    "                  f'{expt}') +\n",
    "       p9.scale_y_log10() +\n",
    "       p9.scale_x_log10() +\n",
    "       p9.theme(figure_size=(4,\n",
    "                             2*viral_bc_frac['infected'].nunique()),\n",
    "                plot_title=p9.element_text(size=10),\n",
    "                axis_title=p9.element_text(size=10),\n",
    "                legend_title=p9.element_text(size=9),\n",
    "                legend_title_align='center') +\n",
    "       p9.scale_color_manual(CBPALETTE[0:]))\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Knee plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rank order barcodes by fraction of total UMIs in their cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_bc_frac = (viral_bc_frac\n",
    "                 .sort_values('frac_bc_UMIs', ascending=True)\n",
    "                 .reset_index(drop=True))\n",
    "\n",
    "viral_bc_frac = (viral_bc_frac\n",
    "                 .assign(bc_rank=lambda x: (x['frac_bc_UMIs']\n",
    "                                            .rank(method='first',\n",
    "                                                  ascending=True))))\n",
    "\n",
    "display(viral_bc_frac)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate knee using [kneed](https://pypi.org/project/kneed/) package:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kl = kneed.KneeLocator(x=viral_bc_frac['bc_rank'].tolist(),\n",
    "                       y=viral_bc_frac['frac_bc_UMIs'].tolist(),\n",
    "                       curve='convex',\n",
    "                       direction='increasing',\n",
    "                       S=10\n",
    "                       )\n",
    "\n",
    "viral_bc_frac['below_knee'] = viral_bc_frac['frac_bc_UMIs'] <= kl.knee_y\n",
    "\n",
    "display(viral_bc_frac)\n",
    "\n",
    "print(f'knee rank: {kl.knee}')\n",
    "print(f'fraction of all UMIs at knee: {kl.knee_y}')\n",
    "print(f'barcodes retained: {len(viral_bc_frac.query(\"below_knee == False\"))}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot knee plot of each barcode's fraction of all UMIs in cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = (p9.ggplot(viral_bc_frac, p9.aes(x='bc_rank',\n",
    "                                       y='frac_bc_UMIs',\n",
    "                                       color='infecting_viral_tag')) +\n",
    "       p9.geom_point() +\n",
    "       p9.geom_vline(xintercept=kl.knee,\n",
    "                     linetype='dashed',\n",
    "                     color='#3A3B3C',\n",
    "                     size=0.5) +\n",
    "       p9.facet_grid('~infected') +\n",
    "       p9.ggtitle('knee plot of viral barcodes\\n'\n",
    "                  'fraction of all UMIs in cell\\n'\n",
    "                  f'{expt}') +\n",
    "       p9.xlab('barcode fraction rank') +\n",
    "       p9.ylab('fraction of all UMIs in cell\\n'\n",
    "               'assigned to barcode') +\n",
    "       p9.theme(figure_size=(6, 2),\n",
    "                plot_title=p9.element_text(size=10),\n",
    "                axis_title=p9.element_text(size=10),\n",
    "                legend_title=p9.element_text(size=9),\n",
    "                legend_title_align='center') +\n",
    "       p9.scale_color_manual(CBPALETTE[0:]))\n",
    "\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Break the plot out by infecting viral tag and viral gene:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = (p9.ggplot(viral_bc_frac, p9.aes(x='bc_rank',\n",
    "                                       y='frac_bc_UMIs',\n",
    "                                       color='below_knee')) +\n",
    "       p9.geom_point() +\n",
    "       p9.geom_vline(xintercept=kl.knee,\n",
    "                     linetype='dashed',\n",
    "                     color='#3A3B3C',\n",
    "                     size=0.5) +\n",
    "       p9.facet_grid('infecting_viral_tag~gene') +\n",
    "       p9.ggtitle('knee plot of viral barcodes\\n'\n",
    "                  'fraction of all UMIs in cell\\n'\n",
    "                  f'{expt}') +\n",
    "       p9.xlab('barcode fraction rank') +\n",
    "       p9.ylab('fraction of all UMIs in cell\\n'\n",
    "               'assigned to barcode') +\n",
    "       p9.theme(figure_size=(6, 5),\n",
    "                plot_title=p9.element_text(size=10),\n",
    "                axis_title=p9.element_text(size=10),\n",
    "                legend_title=p9.element_text(size=9),\n",
    "                legend_title_align='center') +\n",
    "       p9.scale_color_manual(CBPALETTE[0:]))\n",
    "\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Temporary cell to create files expected by snakemake.\n",
    "#  Not final version\n",
    "\n",
    "print(f\"Saving plots to {plot}\")\n",
    "p9.ggsave(plot=fig, filename=plot, verbose=False)\n",
    "\n",
    "print(f\"Saving filtered barcodes to {viral_bc_by_cell_filtered_csv}\")\n",
    "viral_bc_frac = (viral_bc_frac\n",
    "                 .query('viral_barcode != \"none\"')\n",
    "                 [['cell_barcode',\n",
    "                   'gene',\n",
    "                   'frac_bc_UMIs']]\n",
    "                 .drop_duplicates()\n",
    "                 .rename(columns={'mean_freq_corrected_bc':\n",
    "                                  'freq'}))\n",
    "\n",
    "\n",
    "viral_bc_frac.to_csv(viral_bc_by_cell_filtered_csv,\n",
    "                     columns=['cell_barcode',\n",
    "                              'gene',\n",
    "                              'frac_bc_UMIs'],\n",
    "                     index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
