{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter viral barcodes in transcripts\n",
    "This notebook filters viral barcodes in 10X transcriptome data to remove UMIs that are likely derived from leaked transcripts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook setup\n",
    "Import python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display\n",
    "\n",
    "from dms_variants.constants import CBPALETTE\n",
    "\n",
    "import kneed\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import plotnine as p9"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get `snakemake` variables [as described here](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#jupyter-notebook-integration):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_tag_by_cell_csv = 'results/viral_tags_bcs_in_cells/scProgenyProduction_trial2_cell_barcodes_with_viral_tags.csv.gz'\n",
    "viral_bc_by_cell_corrected_csv = 'results/viral_fastq10x/scProgenyProduction_trial2_viral_bc_by_cell_corrected.csv.gz'\n",
    "expt = 'scProgenyProduction_trial2'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Organize data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the viral barcode UMI counts data into a pandas dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>bc_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>ACGTTATTGATTGAGA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGAATCGACACATGTC</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGCCATAGTCTAAAGG</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGGATGATTTTTTTGT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>CACGGATGGTGTACGA</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55238</th>\n",
       "      <td>TTTGTTGCATGTGGTT</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTCCTCAGCCTTTAAG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55239</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GAACCCGATGGGGAAT</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55240</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TAAGGTAAAATAATAG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55241</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TAAGGTATAATTCTAG</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55242</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TGTTGGGCTTTAAGCA</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55243 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           cell_barcode   gene     viral_barcode  bc_UMIs\n",
       "0      AAACCCAAGTAGGTTA  fluHA  ACGTTATTGATTGAGA        1\n",
       "1      AAACCCAAGTAGGTTA  fluHA  AGAATCGACACATGTC       14\n",
       "2      AAACCCAAGTAGGTTA  fluHA  AGCCATAGTCTAAAGG        8\n",
       "3      AAACCCAAGTAGGTTA  fluHA  AGGATGATTTTTTTGT        5\n",
       "4      AAACCCAAGTAGGTTA  fluHA  CACGGATGGTGTACGA       12\n",
       "...                 ...    ...               ...      ...\n",
       "55238  TTTGTTGCATGTGGTT  fluNA  TTCCTCAGCCTTTAAG        3\n",
       "55239  TTTGTTGTCTAGGAAA  fluNA  GAACCCGATGGGGAAT       26\n",
       "55240  TTTGTTGTCTAGGAAA  fluNA  TAAGGTAAAATAATAG        1\n",
       "55241  TTTGTTGTCTAGGAAA  fluNA  TAAGGTATAATTCTAG        4\n",
       "55242  TTTGTTGTCTAGGAAA  fluNA  TGTTGGGCTTTAAGCA        5\n",
       "\n",
       "[55243 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_bc_counts = pd.read_csv(viral_bc_by_cell_corrected_csv)\n",
    "viral_bc_counts = (viral_bc_counts\n",
    "                   .rename(columns={'count': 'bc_UMIs'}))\n",
    "display(viral_bc_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the total number of UMIs per cell into a pandas dataframe. Only keep relevant columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>cellular_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCACAAGGCCTC</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>10304</td>\n",
       "      <td>1485</td>\n",
       "      <td>8819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCACACACACGC</td>\n",
       "      <td>True</td>\n",
       "      <td>both</td>\n",
       "      <td>7395</td>\n",
       "      <td>3320</td>\n",
       "      <td>4075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCATCGTGCATA</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>2764</td>\n",
       "      <td>1597</td>\n",
       "      <td>1167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCATCTACTGCC</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>6313</td>\n",
       "      <td>3</td>\n",
       "      <td>6310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7436</th>\n",
       "      <td>TTTGGTTGTTAAGCAA</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>8528</td>\n",
       "      <td>29</td>\n",
       "      <td>8499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7437</th>\n",
       "      <td>TTTGGTTTCGTCGCTT</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>6492</td>\n",
       "      <td>6</td>\n",
       "      <td>6486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7438</th>\n",
       "      <td>TTTGTTGCATGTGGTT</td>\n",
       "      <td>True</td>\n",
       "      <td>wt</td>\n",
       "      <td>3016</td>\n",
       "      <td>1680</td>\n",
       "      <td>1336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7439</th>\n",
       "      <td>TTTGTTGTCGTCGGGT</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>11070</td>\n",
       "      <td>2</td>\n",
       "      <td>11068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7440</th>\n",
       "      <td>TTTGTTGTCTAGGAAA</td>\n",
       "      <td>True</td>\n",
       "      <td>wt</td>\n",
       "      <td>2653</td>\n",
       "      <td>1014</td>\n",
       "      <td>1639</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7441 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode  infected infecting_viral_tag  total_UMIs  viral_UMIs  \\\n",
       "0     AAACCCAAGTAGGTTA      True                 syn        5367        3166   \n",
       "1     AAACCCACAAGGCCTC      True                 syn       10304        1485   \n",
       "2     AAACCCACACACACGC      True                both        7395        3320   \n",
       "3     AAACCCATCGTGCATA      True                 syn        2764        1597   \n",
       "4     AAACCCATCTACTGCC     False                none        6313           3   \n",
       "...                ...       ...                 ...         ...         ...   \n",
       "7436  TTTGGTTGTTAAGCAA     False                none        8528          29   \n",
       "7437  TTTGGTTTCGTCGCTT     False                none        6492           6   \n",
       "7438  TTTGTTGCATGTGGTT      True                  wt        3016        1680   \n",
       "7439  TTTGTTGTCGTCGGGT     False                none       11070           2   \n",
       "7440  TTTGTTGTCTAGGAAA      True                  wt        2653        1014   \n",
       "\n",
       "      cellular_UMIs  \n",
       "0              2201  \n",
       "1              8819  \n",
       "2              4075  \n",
       "3              1167  \n",
       "4              6310  \n",
       "...             ...  \n",
       "7436           8499  \n",
       "7437           6486  \n",
       "7438           1336  \n",
       "7439          11068  \n",
       "7440           1639  \n",
       "\n",
       "[7441 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_cells = pd.read_csv(viral_tag_by_cell_csv)\n",
    "all_cells = all_cells[['cell_barcode',\n",
    "                             'infected',\n",
    "                             'infecting_viral_tag',\n",
    "                             'total_UMIs',\n",
    "                             'viral_UMIs',\n",
    "                             'cellular_UMIs']]\n",
    "display(all_cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sanity check that `total_UMIs` is equal to `viral_UMIs + cellular_UMIs`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert bool((all_cells['total_UMIs'] ==\n",
    "             all_cells['viral_UMIs'] +\n",
    "             all_cells['cellular_UMIs'])\n",
    "            .all()), \"UMI counts do not add up\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge dataframes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>bc_UMIs</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>cellular_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>ACGTTATTGATTGAGA</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGAATCGACACATGTC</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGCCATAGTCTAAAGG</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGGATGATTTTTTTGT</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>CACGGATGGTGTACGA</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55238</th>\n",
       "      <td>TTGGATGGTGCACGCT</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GCATTTCCTCTCGCTT</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>5179</td>\n",
       "      <td>12</td>\n",
       "      <td>5167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55239</th>\n",
       "      <td>TTGTGGAGTCTGTCCT</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>AACACCTAATGGGGGT</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>6077</td>\n",
       "      <td>7</td>\n",
       "      <td>6070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55240</th>\n",
       "      <td>TTGTTTGCAAGGACAC</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTGTGTTTTACGAAAT</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>7249</td>\n",
       "      <td>5</td>\n",
       "      <td>7244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55241</th>\n",
       "      <td>TTTATGCAGAGTCGAC</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GGTATTATGTGAAGGA</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>7223</td>\n",
       "      <td>4</td>\n",
       "      <td>7219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55242</th>\n",
       "      <td>TTTGGTTGTTAAGCAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TCCTAAGTTAGTACCT</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>8528</td>\n",
       "      <td>29</td>\n",
       "      <td>8499</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55243 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           cell_barcode   gene     viral_barcode  bc_UMIs  infected  \\\n",
       "0      AAACCCAAGTAGGTTA  fluHA  ACGTTATTGATTGAGA        1      True   \n",
       "1      AAACCCAAGTAGGTTA  fluHA  AGAATCGACACATGTC       14      True   \n",
       "2      AAACCCAAGTAGGTTA  fluHA  AGCCATAGTCTAAAGG        8      True   \n",
       "3      AAACCCAAGTAGGTTA  fluHA  AGGATGATTTTTTTGT        5      True   \n",
       "4      AAACCCAAGTAGGTTA  fluHA  CACGGATGGTGTACGA       12      True   \n",
       "...                 ...    ...               ...      ...       ...   \n",
       "55238  TTGGATGGTGCACGCT  fluNA  GCATTTCCTCTCGCTT        1     False   \n",
       "55239  TTGTGGAGTCTGTCCT  fluNA  AACACCTAATGGGGGT        1     False   \n",
       "55240  TTGTTTGCAAGGACAC  fluNA  TTGTGTTTTACGAAAT        1     False   \n",
       "55241  TTTATGCAGAGTCGAC  fluNA  GGTATTATGTGAAGGA        1     False   \n",
       "55242  TTTGGTTGTTAAGCAA  fluNA  TCCTAAGTTAGTACCT        1     False   \n",
       "\n",
       "      infecting_viral_tag  total_UMIs  viral_UMIs  cellular_UMIs  \n",
       "0                     syn        5367        3166           2201  \n",
       "1                     syn        5367        3166           2201  \n",
       "2                     syn        5367        3166           2201  \n",
       "3                     syn        5367        3166           2201  \n",
       "4                     syn        5367        3166           2201  \n",
       "...                   ...         ...         ...            ...  \n",
       "55238                none        5179          12           5167  \n",
       "55239                none        6077           7           6070  \n",
       "55240                none        7249           5           7244  \n",
       "55241                none        7223           4           7219  \n",
       "55242                none        8528          29           8499  \n",
       "\n",
       "[55243 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_bc_frac = pd.merge(viral_bc_counts,\n",
    "                         all_cells,\n",
    "                         on=['cell_barcode'])\n",
    "display(viral_bc_frac)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate each barcode's fraction of all UMIs per cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>bc_UMIs</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>cellular_UMIs</th>\n",
       "      <th>bc_frac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>ACGTTATTGATTGAGA</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "      <td>0.000186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGAATCGACACATGTC</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "      <td>0.002609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGCCATAGTCTAAAGG</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "      <td>0.001491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGGATGATTTTTTTGT</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "      <td>0.000932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCAAGTAGGTTA</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>CACGGATGGTGTACGA</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>syn</td>\n",
       "      <td>5367</td>\n",
       "      <td>3166</td>\n",
       "      <td>2201</td>\n",
       "      <td>0.002236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55238</th>\n",
       "      <td>TTGGATGGTGCACGCT</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GCATTTCCTCTCGCTT</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>5179</td>\n",
       "      <td>12</td>\n",
       "      <td>5167</td>\n",
       "      <td>0.000193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55239</th>\n",
       "      <td>TTGTGGAGTCTGTCCT</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>AACACCTAATGGGGGT</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>6077</td>\n",
       "      <td>7</td>\n",
       "      <td>6070</td>\n",
       "      <td>0.000165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55240</th>\n",
       "      <td>TTGTTTGCAAGGACAC</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTGTGTTTTACGAAAT</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>7249</td>\n",
       "      <td>5</td>\n",
       "      <td>7244</td>\n",
       "      <td>0.000138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55241</th>\n",
       "      <td>TTTATGCAGAGTCGAC</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GGTATTATGTGAAGGA</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>7223</td>\n",
       "      <td>4</td>\n",
       "      <td>7219</td>\n",
       "      <td>0.000138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55242</th>\n",
       "      <td>TTTGGTTGTTAAGCAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TCCTAAGTTAGTACCT</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>none</td>\n",
       "      <td>8528</td>\n",
       "      <td>29</td>\n",
       "      <td>8499</td>\n",
       "      <td>0.000117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55243 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           cell_barcode   gene     viral_barcode  bc_UMIs  infected  \\\n",
       "0      AAACCCAAGTAGGTTA  fluHA  ACGTTATTGATTGAGA        1      True   \n",
       "1      AAACCCAAGTAGGTTA  fluHA  AGAATCGACACATGTC       14      True   \n",
       "2      AAACCCAAGTAGGTTA  fluHA  AGCCATAGTCTAAAGG        8      True   \n",
       "3      AAACCCAAGTAGGTTA  fluHA  AGGATGATTTTTTTGT        5      True   \n",
       "4      AAACCCAAGTAGGTTA  fluHA  CACGGATGGTGTACGA       12      True   \n",
       "...                 ...    ...               ...      ...       ...   \n",
       "55238  TTGGATGGTGCACGCT  fluNA  GCATTTCCTCTCGCTT        1     False   \n",
       "55239  TTGTGGAGTCTGTCCT  fluNA  AACACCTAATGGGGGT        1     False   \n",
       "55240  TTGTTTGCAAGGACAC  fluNA  TTGTGTTTTACGAAAT        1     False   \n",
       "55241  TTTATGCAGAGTCGAC  fluNA  GGTATTATGTGAAGGA        1     False   \n",
       "55242  TTTGGTTGTTAAGCAA  fluNA  TCCTAAGTTAGTACCT        1     False   \n",
       "\n",
       "      infecting_viral_tag  total_UMIs  viral_UMIs  cellular_UMIs   bc_frac  \n",
       "0                     syn        5367        3166           2201  0.000186  \n",
       "1                     syn        5367        3166           2201  0.002609  \n",
       "2                     syn        5367        3166           2201  0.001491  \n",
       "3                     syn        5367        3166           2201  0.000932  \n",
       "4                     syn        5367        3166           2201  0.002236  \n",
       "...                   ...         ...         ...            ...       ...  \n",
       "55238                none        5179          12           5167  0.000193  \n",
       "55239                none        6077           7           6070  0.000165  \n",
       "55240                none        7249           5           7244  0.000138  \n",
       "55241                none        7223           4           7219  0.000138  \n",
       "55242                none        8528          29           8499  0.000117  \n",
       "\n",
       "[55243 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_bc_frac['bc_frac'] = (viral_bc_frac['bc_UMIs'] /\n",
    "                            viral_bc_frac['total_UMIs'])\n",
    "display(viral_bc_frac)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
