{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# final_analysis.py.ipynb\n",
    "This notebook plots the figures for the manuscript. It starts with a single source of input data (`integrate_data.csv`). We restrict the analysis to our sample of interest (cells infected by a WT virus). Then, we calculate the necessary outcome metrics and plot the outcomes of infection."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook setup\n",
    "Import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display\n",
    "from pymodules.gini import *\n",
    "from scipy import stats\n",
    "from svgutils.compose import *\n",
    "import altair as alt\n",
    "import itertools\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import plotnine as p9\n",
    "import scikit_posthocs as sp\n",
    "import seaborn as sns\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set the theme for plots:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "p9.theme_set(p9.theme_classic())\n",
    "CBPALETTE = [\"#999999\", \"#E69F00\", \"#56B4E9\", \"#009E73\", \"#F0E442\", \"#0072B2\", \"#D55E00\", \"#CC79A7\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set output folder for figures:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory `results/figures` already exists.\n"
     ]
    }
   ],
   "source": [
    "figure_dir = 'results/figures'\n",
    "if not os.path.exists(figure_dir):\n",
    "    # Create a new directory because it does not exist \n",
    "    os.makedirs(figure_dir)\n",
    "    print(f\"Made `{figure_dir}` directory \")\n",
    "else:\n",
    "    print(f\"Directory `{figure_dir}` already exists.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input data\n",
    "### Read input data\n",
    "The general structure of the `integrated_data_csv` file is that each cell barcode is listed on a row, and features of that cell are listed in columns. Cells that have more than one valid viral barcode identified may have multiple rows--one for each valid viral barcode."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "      <th>n_viral_genes</th>\n",
       "      <th>transcriptome_UMIs_fluHA</th>\n",
       "      <th>transcriptome_UMIs_fluM</th>\n",
       "      <th>transcriptome_UMIs_fluNA</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio_UMIs_fluPB1</th>\n",
       "      <th>pacbio_UMIs_fluPB2</th>\n",
       "      <th>mutation_support_fluHA</th>\n",
       "      <th>mutation_support_fluM</th>\n",
       "      <th>mutation_support_fluNA</th>\n",
       "      <th>mutation_support_fluNP</th>\n",
       "      <th>mutation_support_fluNS</th>\n",
       "      <th>mutation_support_fluPA</th>\n",
       "      <th>mutation_support_fluPB1</th>\n",
       "      <th>mutation_support_fluPB2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACGAAAGCCACCGT</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>41743</td>\n",
       "      <td>451</td>\n",
       "      <td>0.010804</td>\n",
       "      <td>6</td>\n",
       "      <td>101</td>\n",
       "      <td>54</td>\n",
       "      <td>26</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACGCTAGTTCACTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>syn</td>\n",
       "      <td>11572</td>\n",
       "      <td>4702</td>\n",
       "      <td>0.406326</td>\n",
       "      <td>7</td>\n",
       "      <td>650</td>\n",
       "      <td>2075</td>\n",
       "      <td>101</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGCTAGTTCACTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>syn</td>\n",
       "      <td>11572</td>\n",
       "      <td>4702</td>\n",
       "      <td>0.406326</td>\n",
       "      <td>7</td>\n",
       "      <td>650</td>\n",
       "      <td>2075</td>\n",
       "      <td>101</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGCTAGTTCACTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>syn</td>\n",
       "      <td>11572</td>\n",
       "      <td>4702</td>\n",
       "      <td>0.406326</td>\n",
       "      <td>7</td>\n",
       "      <td>650</td>\n",
       "      <td>2075</td>\n",
       "      <td>101</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGCTAGTTCACTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>syn</td>\n",
       "      <td>11572</td>\n",
       "      <td>4702</td>\n",
       "      <td>0.406326</td>\n",
       "      <td>7</td>\n",
       "      <td>650</td>\n",
       "      <td>2075</td>\n",
       "      <td>101</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6387</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>63150</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6388</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>65941</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6389</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>31/31</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6390</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>31/31</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6391</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>135009</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6392 rows × 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode    infected infecting_viral_tag  total_UMIs  \\\n",
       "0     AAACGAAAGCCACCGT    infected                  wt       41743   \n",
       "1     AAACGCTAGTTCACTG    infected                 syn       11572   \n",
       "2     AAACGCTAGTTCACTG    infected                 syn       11572   \n",
       "3     AAACGCTAGTTCACTG    infected                 syn       11572   \n",
       "4     AAACGCTAGTTCACTG    infected                 syn       11572   \n",
       "...                ...         ...                 ...         ...   \n",
       "6387  TTTGATCTCCCGTTCA  uninfected                none       63150   \n",
       "6388  TTTGGAGAGTTGCCTA  uninfected                none       65941   \n",
       "6389  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "6390  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "6391  TTTGTTGAGTGCAGCA  uninfected                none      135009   \n",
       "\n",
       "      viral_UMIs  frac_viral_UMIs  n_viral_genes  transcriptome_UMIs_fluHA  \\\n",
       "0            451         0.010804              6                       101   \n",
       "1           4702         0.406326              7                       650   \n",
       "2           4702         0.406326              7                       650   \n",
       "3           4702         0.406326              7                       650   \n",
       "4           4702         0.406326              7                       650   \n",
       "...          ...              ...            ...                       ...   \n",
       "6387           3         0.000048              0                         0   \n",
       "6388          12         0.000182              1                         0   \n",
       "6389        3526         0.023486              8                        93   \n",
       "6390        3526         0.023486              8                        93   \n",
       "6391           8         0.000059              1                         2   \n",
       "\n",
       "      transcriptome_UMIs_fluM  transcriptome_UMIs_fluNA  ...  \\\n",
       "0                          54                        26  ...   \n",
       "1                        2075                       101  ...   \n",
       "2                        2075                       101  ...   \n",
       "3                        2075                       101  ...   \n",
       "4                        2075                       101  ...   \n",
       "...                       ...                       ...  ...   \n",
       "6387                        1                         0  ...   \n",
       "6388                        6                         1  ...   \n",
       "6389                     1668                        45  ...   \n",
       "6390                     1668                        45  ...   \n",
       "6391                        2                         1  ...   \n",
       "\n",
       "      pacbio_UMIs_fluPB1  pacbio_UMIs_fluPB2  mutation_support_fluHA  \\\n",
       "0                    NaN                 NaN                     NaN   \n",
       "1                    NaN                 NaN                     NaN   \n",
       "2                    NaN                 NaN                     NaN   \n",
       "3                    NaN                 NaN                     NaN   \n",
       "4                    NaN                 NaN                     NaN   \n",
       "...                  ...                 ...                     ...   \n",
       "6387                 0.0                 0.0          Not Applicable   \n",
       "6388                 0.0                 0.0          Not Applicable   \n",
       "6389                23.0                 7.0          Not Applicable   \n",
       "6390                23.0                 7.0          Not Applicable   \n",
       "6391                 0.0                 0.0          Not Applicable   \n",
       "\n",
       "      mutation_support_fluM  mutation_support_fluNA  mutation_support_fluNP  \\\n",
       "0                       NaN                     NaN                     NaN   \n",
       "1                       NaN                     NaN                     NaN   \n",
       "2                       NaN                     NaN                     NaN   \n",
       "3                       NaN                     NaN                     NaN   \n",
       "4                       NaN                     NaN                     NaN   \n",
       "...                     ...                     ...                     ...   \n",
       "6387         Not Applicable          Not Applicable          Not Applicable   \n",
       "6388         Not Applicable          Not Applicable          Not Applicable   \n",
       "6389         Not Applicable                   31/31          Not Applicable   \n",
       "6390         Not Applicable                   31/31          Not Applicable   \n",
       "6391         Not Applicable          Not Applicable          Not Applicable   \n",
       "\n",
       "      mutation_support_fluNS  mutation_support_fluPA  mutation_support_fluPB1  \\\n",
       "0                        NaN                     NaN                      NaN   \n",
       "1                        NaN                     NaN                      NaN   \n",
       "2                        NaN                     NaN                      NaN   \n",
       "3                        NaN                     NaN                      NaN   \n",
       "4                        NaN                     NaN                      NaN   \n",
       "...                      ...                     ...                      ...   \n",
       "6387          Not Applicable          Not Applicable           Not Applicable   \n",
       "6388          Not Applicable          Not Applicable           Not Applicable   \n",
       "6389          Not Applicable          Not Applicable           Not Applicable   \n",
       "6390          Not Applicable          Not Applicable           Not Applicable   \n",
       "6391          Not Applicable          Not Applicable           Not Applicable   \n",
       "\n",
       "      mutation_support_fluPB2  \n",
       "0                         NaN  \n",
       "1                         NaN  \n",
       "2                         NaN  \n",
       "3                         NaN  \n",
       "4                         NaN  \n",
       "...                       ...  \n",
       "6387           Not Applicable  \n",
       "6388           Not Applicable  \n",
       "6389           Not Applicable  \n",
       "6390           Not Applicable  \n",
       "6391           Not Applicable  \n",
       "\n",
       "[6392 rows x 66 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_samples_csv = 'results/viral_fastq10x/all_samples.csv'\n",
    "integrated_data = pd.read_csv(all_samples_csv)\n",
    "integrated_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine data for each cell into single row\n",
    "Process the dataframe so that each cell becomes a single row. Progeny contributions are summed for each cell, source, and barcoded gene (e.g. supernatant_freq_fluHA) and listed in a column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "      <th>n_viral_genes</th>\n",
       "      <th>transcriptome_UMIs_fluHA</th>\n",
       "      <th>transcriptome_UMIs_fluM</th>\n",
       "      <th>...</th>\n",
       "      <th>viral_barcodes_fluNA</th>\n",
       "      <th>freq_supernatant_fluHA</th>\n",
       "      <th>freq_supernatant_fluNA</th>\n",
       "      <th>freq_second_infection_fluHA</th>\n",
       "      <th>freq_second_infection_fluNA</th>\n",
       "      <th>all_viral_genes</th>\n",
       "      <th>missing_viral_genes</th>\n",
       "      <th>virus_fully_sequenced</th>\n",
       "      <th>all_barcoded_genes</th>\n",
       "      <th>sequenced_and_barcoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAACGAAAGCCACCGT</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>41743</td>\n",
       "      <td>451</td>\n",
       "      <td>0.010804</td>\n",
       "      <td>6</td>\n",
       "      <td>101</td>\n",
       "      <td>54</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>0.00001</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>NP; PB2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not sequenced or barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAACGCTAGTTCACTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>syn</td>\n",
       "      <td>11572</td>\n",
       "      <td>4702</td>\n",
       "      <td>0.406326</td>\n",
       "      <td>7</td>\n",
       "      <td>650</td>\n",
       "      <td>2075</td>\n",
       "      <td>...</td>\n",
       "      <td>AAATTCTTGGGGCGAC; TGATAGCGTAGACGTT; ATGAATTATA...</td>\n",
       "      <td>0.001942</td>\n",
       "      <td>0.002176</td>\n",
       "      <td>0.002218</td>\n",
       "      <td>0.001432</td>\n",
       "      <td>False</td>\n",
       "      <td>PA</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAAGAACGTAATTGGA</td>\n",
       "      <td>infected</td>\n",
       "      <td>syn</td>\n",
       "      <td>24525</td>\n",
       "      <td>3948</td>\n",
       "      <td>0.160979</td>\n",
       "      <td>7</td>\n",
       "      <td>456</td>\n",
       "      <td>1340</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>0.000077</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.000416</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>NA</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not sequenced or barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAAGGATCATTGCAAC</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>17304</td>\n",
       "      <td>888</td>\n",
       "      <td>0.051318</td>\n",
       "      <td>7</td>\n",
       "      <td>160</td>\n",
       "      <td>144</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>0.000013</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>NP</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not sequenced or barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAAGGGCGTGTATTCG</td>\n",
       "      <td>infected</td>\n",
       "      <td>both</td>\n",
       "      <td>12578</td>\n",
       "      <td>6213</td>\n",
       "      <td>0.493958</td>\n",
       "      <td>8</td>\n",
       "      <td>390</td>\n",
       "      <td>3719</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not sequenced or barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTGGGATGTAACGATA</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>161690</td>\n",
       "      <td>9470</td>\n",
       "      <td>0.058569</td>\n",
       "      <td>8</td>\n",
       "      <td>407</td>\n",
       "      <td>4600</td>\n",
       "      <td>...</td>\n",
       "      <td>TAAAAAATGTCTCATG</td>\n",
       "      <td>0.001014</td>\n",
       "      <td>0.001284</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>sequenced and barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>949</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTGGGATTCCGATTAG</td>\n",
       "      <td>infected</td>\n",
       "      <td>syn</td>\n",
       "      <td>137623</td>\n",
       "      <td>6652</td>\n",
       "      <td>0.048335</td>\n",
       "      <td>8</td>\n",
       "      <td>262</td>\n",
       "      <td>3656</td>\n",
       "      <td>...</td>\n",
       "      <td>GTTTCGTGAGCATGTA</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>sequenced and barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTTACCAGTCGCTTAA</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>114567</td>\n",
       "      <td>1434</td>\n",
       "      <td>0.012517</td>\n",
       "      <td>7</td>\n",
       "      <td>125</td>\n",
       "      <td>609</td>\n",
       "      <td>...</td>\n",
       "      <td>ATTTGGATGTAGGGGC; TTGGAGGAGACCCGTG</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>0.001222</td>\n",
       "      <td>0.00002</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>False</td>\n",
       "      <td>PA</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>sequenced and barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>951</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTTCACAAGCCAAGCA</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>68424</td>\n",
       "      <td>2308</td>\n",
       "      <td>0.033731</td>\n",
       "      <td>6</td>\n",
       "      <td>475</td>\n",
       "      <td>796</td>\n",
       "      <td>...</td>\n",
       "      <td>GGTATCAGTTATTGTT</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>False</td>\n",
       "      <td>NS; PB1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>sequenced and barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>...</td>\n",
       "      <td>ACATCTTATTTACACG</td>\n",
       "      <td>0.000695</td>\n",
       "      <td>0.00179</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>sequenced and barcoded</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>953 rows × 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample      cell_barcode  infected infecting_viral_tag  total_UMIs  \\\n",
       "0    trial1  AAACGAAAGCCACCGT  infected                  wt       41743   \n",
       "1    trial1  AAACGCTAGTTCACTG  infected                 syn       11572   \n",
       "2    trial1  AAAGAACGTAATTGGA  infected                 syn       24525   \n",
       "3    trial1  AAAGGATCATTGCAAC  infected                  wt       17304   \n",
       "4    trial1  AAAGGGCGTGTATTCG  infected                both       12578   \n",
       "..      ...               ...       ...                 ...         ...   \n",
       "948  trial3  TTGGGATGTAACGATA  infected                  wt      161690   \n",
       "949  trial3  TTGGGATTCCGATTAG  infected                 syn      137623   \n",
       "950  trial3  TTTACCAGTCGCTTAA  infected                  wt      114567   \n",
       "951  trial3  TTTCACAAGCCAAGCA  infected                  wt       68424   \n",
       "952  trial3  TTTGGAGGTATCGTTG  infected                  wt      150130   \n",
       "\n",
       "     viral_UMIs  frac_viral_UMIs  n_viral_genes  transcriptome_UMIs_fluHA  \\\n",
       "0           451         0.010804              6                       101   \n",
       "1          4702         0.406326              7                       650   \n",
       "2          3948         0.160979              7                       456   \n",
       "3           888         0.051318              7                       160   \n",
       "4          6213         0.493958              8                       390   \n",
       "..          ...              ...            ...                       ...   \n",
       "948        9470         0.058569              8                       407   \n",
       "949        6652         0.048335              8                       262   \n",
       "950        1434         0.012517              7                       125   \n",
       "951        2308         0.033731              6                       475   \n",
       "952        3526         0.023486              8                        93   \n",
       "\n",
       "     transcriptome_UMIs_fluM  ...  \\\n",
       "0                         54  ...   \n",
       "1                       2075  ...   \n",
       "2                       1340  ...   \n",
       "3                        144  ...   \n",
       "4                       3719  ...   \n",
       "..                       ...  ...   \n",
       "948                     4600  ...   \n",
       "949                     3656  ...   \n",
       "950                      609  ...   \n",
       "951                      796  ...   \n",
       "952                     1668  ...   \n",
       "\n",
       "                                  viral_barcodes_fluNA  \\\n",
       "0                                                        \n",
       "1    AAATTCTTGGGGCGAC; TGATAGCGTAGACGTT; ATGAATTATA...   \n",
       "2                                                        \n",
       "3                                                        \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "948                                   TAAAAAATGTCTCATG   \n",
       "949                                   GTTTCGTGAGCATGTA   \n",
       "950                 ATTTGGATGTAGGGGC; TTGGAGGAGACCCGTG   \n",
       "951                                   GGTATCAGTTATTGTT   \n",
       "952                                   ACATCTTATTTACACG   \n",
       "\n",
       "     freq_supernatant_fluHA  freq_supernatant_fluNA  \\\n",
       "0                   0.00001                    <NA>   \n",
       "1                  0.001942                0.002176   \n",
       "2                  0.000077                    <NA>   \n",
       "3                  0.000013                    <NA>   \n",
       "4                       NaN                     NaN   \n",
       "..                      ...                     ...   \n",
       "948                0.001014                0.001284   \n",
       "949                 0.00001                 0.00001   \n",
       "950                0.000106                0.001222   \n",
       "951                 0.00001                 0.00001   \n",
       "952                0.000695                 0.00179   \n",
       "\n",
       "     freq_second_infection_fluHA  freq_second_infection_fluNA  \\\n",
       "0                        0.00001                         <NA>   \n",
       "1                       0.002218                     0.001432   \n",
       "2                       0.000416                         <NA>   \n",
       "3                       0.000064                         <NA>   \n",
       "4                            NaN                          NaN   \n",
       "..                           ...                          ...   \n",
       "948                     0.000154                     0.000015   \n",
       "949                      0.00001                      0.00001   \n",
       "950                      0.00002                     0.000029   \n",
       "951                      0.00001                      0.00001   \n",
       "952                      0.00001                     0.000028   \n",
       "\n",
       "     all_viral_genes  missing_viral_genes  virus_fully_sequenced  \\\n",
       "0              False              NP; PB2                  False   \n",
       "1              False                   PA                  False   \n",
       "2              False                   NA                  False   \n",
       "3              False                   NP                  False   \n",
       "4               True                                       False   \n",
       "..               ...                  ...                    ...   \n",
       "948             True                                        True   \n",
       "949             True                                        True   \n",
       "950            False                   PA                   True   \n",
       "951            False              NS; PB1                   True   \n",
       "952             True                                        True   \n",
       "\n",
       "     all_barcoded_genes     sequenced_and_barcoded  \n",
       "0                 False  not sequenced or barcoded  \n",
       "1                  True                   barcoded  \n",
       "2                 False  not sequenced or barcoded  \n",
       "3                 False  not sequenced or barcoded  \n",
       "4                 False  not sequenced or barcoded  \n",
       "..                  ...                        ...  \n",
       "948                True     sequenced and barcoded  \n",
       "949                True     sequenced and barcoded  \n",
       "950                True     sequenced and barcoded  \n",
       "951                True     sequenced and barcoded  \n",
       "952                True     sequenced and barcoded  \n",
       "\n",
       "[953 rows x 69 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "min_viral_barcode_freq = 1e-5  # value assigned to unobserved viral barcodes\n",
    "\n",
    "# aggregate observed viral barcodes for each gene and cell\n",
    "processed_data = (\n",
    "    integrated_data\n",
    "    .groupby([\"sample\", \"cell_barcode\", \"barcoded_gene\"], as_index=False)\n",
    "    .aggregate(\n",
    "        n_viral_barcodes=pd.NamedAgg(\"viral_barcode\", lambda v: v.astype(bool).astype(int).sum()),\n",
    "        viral_barcodes=pd.NamedAgg(\"viral_barcode\", \"; \".join),\n",
    "        freq_supernatant=pd.NamedAgg(\"freq_supernatant\", \"sum\"),\n",
    "        freq_second_infection=pd.NamedAgg(\"freq_second_infection\", \"sum\"),\n",
    "    )\n",
    ")\n",
    "assert processed_data.notnull().all().all()\n",
    "\n",
    "# get barcoded viral genes\n",
    "barcoded_genes = processed_data[\"barcoded_gene\"].unique().tolist()\n",
    "\n",
    "# now pivot so just one row per cell\n",
    "pivoted_dfs = []\n",
    "for col_to_pivot, fill_value in [\n",
    "    (\"n_viral_barcodes\", 0),\n",
    "    (\"viral_barcodes\", \"\"),\n",
    "    (\"freq_supernatant\", pd.NA),\n",
    "    (\"freq_second_infection\", pd.NA),\n",
    "]:\n",
    "    pivoted_dfs.append(\n",
    "        processed_data\n",
    "        .pivot_table(\n",
    "            index=[\"sample\", \"cell_barcode\"],\n",
    "            columns=\"barcoded_gene\",\n",
    "            values=col_to_pivot,\n",
    "            fill_value=fill_value,\n",
    "            aggfunc='sum',\n",
    "        )\n",
    "        .rename_axis(None, axis=1)\n",
    "        .rename(columns={gene: f\"{col_to_pivot}_{gene}\" for gene in barcoded_genes})\n",
    "    )\n",
    "assert all((pivoted_dfs[0].index == df.index).all() for df in pivoted_dfs)\n",
    "processed_data = pd.concat(pivoted_dfs, axis=1).reset_index()\n",
    "\n",
    "# now add back in cell-specific information\n",
    "# Require infected cells since the above measurements only make sense for\n",
    "# infected cells.\n",
    "cell_info = (\n",
    "    integrated_data\n",
    "    [['sample', 'cell_barcode', 'infected', 'infecting_viral_tag',\n",
    "      'total_UMIs', 'viral_UMIs', 'frac_viral_UMIs', 'n_viral_genes'] +\n",
    "     [c for c in integrated_data.columns\n",
    "      if re.match(\"transcriptome_UMIs_flu|frac_UMIs_flu|present_flu|pacbio_UMIs|mutations_|mutation_support_\", c)]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    ")\n",
    "\n",
    "assert set(processed_data[[\"sample\", \"cell_barcode\"]]).issubset(cell_info[[\"sample\", \"cell_barcode\"]])\n",
    "processed_data = (\n",
    "    cell_info.query('infected == \"infected\"')\n",
    "    .merge(processed_data, on=[\"sample\", \"cell_barcode\"], how=\"left\", validate=\"one_to_one\")\n",
    ")\n",
    "\n",
    "# add columns explaining missing genes\n",
    "present_columns = {c: c.replace(\"present_flu\", \"\") for c in processed_data.columns\n",
    "                   if re.match(\"present_flu\", c)}\n",
    "\n",
    "processed_data = (\n",
    "    processed_data\n",
    "    .assign(\n",
    "        all_viral_genes=lambda x: x['n_viral_genes'] == len(present_columns),\n",
    "        missing_viral_genes=lambda x: x.apply(lambda row:\n",
    "            \"; \".join(val for (key, val) in present_columns.items() if not row[key]),\n",
    "            axis=1)\n",
    "    )\n",
    ")\n",
    "\n",
    "# check for complete pacbio data for all expressed viral genes\n",
    "def virus_fully_sequenced(row):\n",
    "    row_dict = row.to_dict()\n",
    "    for key, val in row_dict.items():\n",
    "        if re.fullmatch(\"present_flu\\w+\", key) and val:\n",
    "            gene = key.split(\"_\")[1]\n",
    "            if row_dict[f\"pacbio_UMIs_{gene}\"] >= 2:\n",
    "                return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "processed_data = (\n",
    "    processed_data\n",
    "    .assign(virus_fully_sequenced=lambda x: x.apply(virus_fully_sequenced, axis=1))\n",
    ")\n",
    "\n",
    "# check to see if cell expresses both barcoded viral genes\n",
    "processed_data[\"all_barcoded_genes\"] = np.logical_and.reduce(\n",
    "    [processed_data[f\"present_{gene}\"] for gene in barcoded_genes] +\n",
    "    [processed_data[f\"freq_supernatant_{gene}\"].notnull() for gene in barcoded_genes] +\n",
    "    [processed_data[f\"freq_second_infection_{gene}\"].notnull() for gene in barcoded_genes]\n",
    ")\n",
    "\n",
    "# annotate cells with both complete pacbio data and both barcoded viral genes\n",
    "def sequenced_and_barcoded(row):\n",
    "    if row['virus_fully_sequenced'] and row['all_barcoded_genes']:\n",
    "        return \"sequenced and barcoded\"\n",
    "    elif row['virus_fully_sequenced']:\n",
    "        return \"sequenced\"\n",
    "    elif row['all_barcoded_genes']:\n",
    "        return \"barcoded\"\n",
    "    else:\n",
    "        return \"not sequenced or barcoded\"\n",
    "processed_data[\"sequenced_and_barcoded\"] = (\n",
    "    processed_data.apply(sequenced_and_barcoded, axis=1)\n",
    ")\n",
    "\n",
    "display(processed_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Restrict analysis to WT infected cells\n",
    "We only want to plot our sample of interest, which is made up of cells infected by a WT virus. We will exclude uninfected cells, infected multiplets (transcriptomes derived from both WT and dblSyn cells), and infected dblSyn cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "      <th>n_viral_genes</th>\n",
       "      <th>transcriptome_UMIs_fluHA</th>\n",
       "      <th>transcriptome_UMIs_fluM</th>\n",
       "      <th>...</th>\n",
       "      <th>viral_barcodes_fluNA</th>\n",
       "      <th>freq_supernatant_fluHA</th>\n",
       "      <th>freq_supernatant_fluNA</th>\n",
       "      <th>freq_second_infection_fluHA</th>\n",
       "      <th>freq_second_infection_fluNA</th>\n",
       "      <th>all_viral_genes</th>\n",
       "      <th>missing_viral_genes</th>\n",
       "      <th>virus_fully_sequenced</th>\n",
       "      <th>all_barcoded_genes</th>\n",
       "      <th>sequenced_and_barcoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAACGAAAGCCACCGT</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>41743</td>\n",
       "      <td>451</td>\n",
       "      <td>0.010804</td>\n",
       "      <td>6</td>\n",
       "      <td>101</td>\n",
       "      <td>54</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>0.00001</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>NP; PB2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not sequenced or barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAAGGATCATTGCAAC</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>17304</td>\n",
       "      <td>888</td>\n",
       "      <td>0.051318</td>\n",
       "      <td>7</td>\n",
       "      <td>160</td>\n",
       "      <td>144</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>0.000013</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>NP</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not sequenced or barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAAGGGCGTTGCCTAA</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>36880</td>\n",
       "      <td>6704</td>\n",
       "      <td>0.181779</td>\n",
       "      <td>7</td>\n",
       "      <td>1079</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>0.00001</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>M</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not sequenced or barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAAGTGAAGGAAGTCC</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>19936</td>\n",
       "      <td>428</td>\n",
       "      <td>0.021469</td>\n",
       "      <td>6</td>\n",
       "      <td>186</td>\n",
       "      <td>27</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>0.00002</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.00002</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>False</td>\n",
       "      <td>NP; PB2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not sequenced or barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AACAACCGTGATTCTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>54425</td>\n",
       "      <td>2669</td>\n",
       "      <td>0.049040</td>\n",
       "      <td>8</td>\n",
       "      <td>240</td>\n",
       "      <td>1364</td>\n",
       "      <td>...</td>\n",
       "      <td>TCCTAAGTTAGTACCT</td>\n",
       "      <td>0.00002</td>\n",
       "      <td>0.002127</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>0.032429</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>947</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTGCGTCCACCCTCTA</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>160497</td>\n",
       "      <td>1816</td>\n",
       "      <td>0.011315</td>\n",
       "      <td>7</td>\n",
       "      <td>241</td>\n",
       "      <td>678</td>\n",
       "      <td>...</td>\n",
       "      <td>GAGCTCCTGTATTGCC</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>False</td>\n",
       "      <td>PA</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>sequenced and barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTGGGATGTAACGATA</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>161690</td>\n",
       "      <td>9470</td>\n",
       "      <td>0.058569</td>\n",
       "      <td>8</td>\n",
       "      <td>407</td>\n",
       "      <td>4600</td>\n",
       "      <td>...</td>\n",
       "      <td>TAAAAAATGTCTCATG</td>\n",
       "      <td>0.001014</td>\n",
       "      <td>0.001284</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>sequenced and barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTTACCAGTCGCTTAA</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>114567</td>\n",
       "      <td>1434</td>\n",
       "      <td>0.012517</td>\n",
       "      <td>7</td>\n",
       "      <td>125</td>\n",
       "      <td>609</td>\n",
       "      <td>...</td>\n",
       "      <td>ATTTGGATGTAGGGGC; TTGGAGGAGACCCGTG</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>0.001222</td>\n",
       "      <td>0.00002</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>False</td>\n",
       "      <td>PA</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>sequenced and barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>951</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTTCACAAGCCAAGCA</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>68424</td>\n",
       "      <td>2308</td>\n",
       "      <td>0.033731</td>\n",
       "      <td>6</td>\n",
       "      <td>475</td>\n",
       "      <td>796</td>\n",
       "      <td>...</td>\n",
       "      <td>GGTATCAGTTATTGTT</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>False</td>\n",
       "      <td>NS; PB1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>sequenced and barcoded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>...</td>\n",
       "      <td>ACATCTTATTTACACG</td>\n",
       "      <td>0.000695</td>\n",
       "      <td>0.00179</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>sequenced and barcoded</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>632 rows × 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample      cell_barcode  infected infecting_viral_tag  total_UMIs  \\\n",
       "0    trial1  AAACGAAAGCCACCGT  infected                  wt       41743   \n",
       "3    trial1  AAAGGATCATTGCAAC  infected                  wt       17304   \n",
       "5    trial1  AAAGGGCGTTGCCTAA  infected                  wt       36880   \n",
       "8    trial1  AAAGTGAAGGAAGTCC  infected                  wt       19936   \n",
       "10   trial1  AACAACCGTGATTCTG  infected                  wt       54425   \n",
       "..      ...               ...       ...                 ...         ...   \n",
       "947  trial3  TTGCGTCCACCCTCTA  infected                  wt      160497   \n",
       "948  trial3  TTGGGATGTAACGATA  infected                  wt      161690   \n",
       "950  trial3  TTTACCAGTCGCTTAA  infected                  wt      114567   \n",
       "951  trial3  TTTCACAAGCCAAGCA  infected                  wt       68424   \n",
       "952  trial3  TTTGGAGGTATCGTTG  infected                  wt      150130   \n",
       "\n",
       "     viral_UMIs  frac_viral_UMIs  n_viral_genes  transcriptome_UMIs_fluHA  \\\n",
       "0           451         0.010804              6                       101   \n",
       "3           888         0.051318              7                       160   \n",
       "5          6704         0.181779              7                      1079   \n",
       "8           428         0.021469              6                       186   \n",
       "10         2669         0.049040              8                       240   \n",
       "..          ...              ...            ...                       ...   \n",
       "947        1816         0.011315              7                       241   \n",
       "948        9470         0.058569              8                       407   \n",
       "950        1434         0.012517              7                       125   \n",
       "951        2308         0.033731              6                       475   \n",
       "952        3526         0.023486              8                        93   \n",
       "\n",
       "     transcriptome_UMIs_fluM  ...                viral_barcodes_fluNA  \\\n",
       "0                         54  ...                                       \n",
       "3                        144  ...                                       \n",
       "5                          6  ...                                       \n",
       "8                         27  ...                                       \n",
       "10                      1364  ...                    TCCTAAGTTAGTACCT   \n",
       "..                       ...  ...                                 ...   \n",
       "947                      678  ...                    GAGCTCCTGTATTGCC   \n",
       "948                     4600  ...                    TAAAAAATGTCTCATG   \n",
       "950                      609  ...  ATTTGGATGTAGGGGC; TTGGAGGAGACCCGTG   \n",
       "951                      796  ...                    GGTATCAGTTATTGTT   \n",
       "952                     1668  ...                    ACATCTTATTTACACG   \n",
       "\n",
       "     freq_supernatant_fluHA  freq_supernatant_fluNA  \\\n",
       "0                   0.00001                    <NA>   \n",
       "3                  0.000013                    <NA>   \n",
       "5                   0.00001                    <NA>   \n",
       "8                   0.00002                    <NA>   \n",
       "10                  0.00002                0.002127   \n",
       "..                      ...                     ...   \n",
       "947                 0.00001                 0.00001   \n",
       "948                0.001014                0.001284   \n",
       "950                0.000106                0.001222   \n",
       "951                 0.00001                 0.00001   \n",
       "952                0.000695                 0.00179   \n",
       "\n",
       "     freq_second_infection_fluHA  freq_second_infection_fluNA  \\\n",
       "0                        0.00001                         <NA>   \n",
       "3                       0.000064                         <NA>   \n",
       "5                        0.00001                         <NA>   \n",
       "8                        0.00002                         <NA>   \n",
       "10                      0.000058                     0.032429   \n",
       "..                           ...                          ...   \n",
       "947                      0.00001                      0.00001   \n",
       "948                     0.000154                     0.000015   \n",
       "950                      0.00002                     0.000029   \n",
       "951                      0.00001                      0.00001   \n",
       "952                      0.00001                     0.000028   \n",
       "\n",
       "     all_viral_genes  missing_viral_genes  virus_fully_sequenced  \\\n",
       "0              False              NP; PB2                  False   \n",
       "3              False                   NP                  False   \n",
       "5              False                    M                  False   \n",
       "8              False              NP; PB2                  False   \n",
       "10              True                                       False   \n",
       "..               ...                  ...                    ...   \n",
       "947            False                   PA                   True   \n",
       "948             True                                        True   \n",
       "950            False                   PA                   True   \n",
       "951            False              NS; PB1                   True   \n",
       "952             True                                        True   \n",
       "\n",
       "     all_barcoded_genes     sequenced_and_barcoded  \n",
       "0                 False  not sequenced or barcoded  \n",
       "3                 False  not sequenced or barcoded  \n",
       "5                 False  not sequenced or barcoded  \n",
       "8                 False  not sequenced or barcoded  \n",
       "10                 True                   barcoded  \n",
       "..                  ...                        ...  \n",
       "947                True     sequenced and barcoded  \n",
       "948                True     sequenced and barcoded  \n",
       "950                True     sequenced and barcoded  \n",
       "951                True     sequenced and barcoded  \n",
       "952                True     sequenced and barcoded  \n",
       "\n",
       "[632 rows x 69 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt_infected_cells = (\n",
    "    processed_data\n",
    "    .query('infected == \"infected\" and infecting_viral_tag == \"wt\"'))\n",
    "wt_infected_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 378 WT infected cells in the trial1 dataset.\n",
      "There are 254 WT infected cells in the trial3 dataset.\n"
     ]
    }
   ],
   "source": [
    "for sample in wt_infected_cells['sample'].unique():\n",
    "    n_wt_infected_cells = wt_infected_cells.query(f'sample == \"{sample}\"')[\"cell_barcode\"].nunique()\n",
    "    print(f'There are {n_wt_infected_cells} WT infected cells in the {sample} dataset.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## General transcriptional outcomes of infection\n",
    "### Plot viral transcription\n",
    "I will plot the histogram of viral transcription. I need to know where to set the limits and bins, so I will look up the maximum viral transcription fraction for these data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The maximum fraction of viral transcription in an infected cells is 0.8403226839106217\n"
     ]
    }
   ],
   "source": [
    "print(f'The maximum fraction of viral transcription '\n",
    "      f'in an infected cells is {wt_infected_cells[\"frac_viral_UMIs\"].max()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dbacsik/.conda/envs/barcoded_flu_pdmH1N1_final_analysis/lib/python3.8/site-packages/plotnine/ggplot.py:718: PlotnineWarning: Saving 5 x 3 in image.\n",
      "/home/dbacsik/.conda/envs/barcoded_flu_pdmH1N1_final_analysis/lib/python3.8/site-packages/plotnine/ggplot.py:719: PlotnineWarning: Filename: results/figures/viral_transcription.svg\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAecAAAE+CAYAAAC3PNizAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAA9hAAAPYQGoP6dpAABLQUlEQVR4nO3deViUVf8/8PfNvqMwLqAJCCooIoqKSwaY4v5NRUtxxaVMDTV7lGQRcn+iMs1yFwuX3DAjdwFtceHJKFB8MhVNEWNACwUU5Pz+8Mc8jjMiMwww6Pt1XVyXc86Zc3/uw8hnzr0dSQghQERERHrDoLYDICIiImVMzkRERHqGyZmIiEjPMDkTERHpGSZnIiIiPcPkTEREpGeYnImIiPQMkzMREZGeYXImIiLSM0zOz4khQ4bA3Nwcd+7ceWqbUaNGwdjYGLdu3UJcXBwkSUJWVpZO44iOjoYkSc9st3//fkRHR+t02/qqsmOiTnZ2NqKjo5GWlqbTfqvK2dkZ48eP11l/Vf08/vLLL/Dz84OtrS0kScLy5ct1Flu5xYsXY+/evTrvNysrC5IkIS4uTud9V+TJ32FtxUHqMTk/JyZOnIji4mJs3bpVbf3ff/+NhIQEDBw4EI0aNcKAAQNw8uRJODg41HCkj+zfvx8xMTG1su2aNmnSJJw8eVKr92ZnZyMmJkZtcq5Kv1WVkJCAyMhInfVX1c/jhAkTcPPmTWzfvh0nT57EiBEjdBZbuepKzkTqGNV2AKQb/fr1g6OjIzZu3IipU6eq1G/btg1FRUWYOHEiAKBBgwZo0KDBM/stLCyEhYWFzuPVhBACxcXFMDc3r9U4NFU+dk2bNkXTpk113n919VsZ7du312l/lf08Pk1GRgYmT56Mfv366TAqotrDmfNzwtDQEOPGjcPPP/+M9PR0lfpNmzbBwcFB8cdL3WFEf39/eHp64sSJE+jWrRssLCwwYcIEAMDXX3+NwMBAODg4wNzcHB4eHggLC8O9e/c0jnX8+PFYtWoVAECSJMVPeSySJGH69OlYvXo1PDw8YGpqis2bNwMAYmJi4OvrCzs7O9jY2KBDhw7YsGEDnly/xdnZGQMHDsTBgwfRoUMHmJubw93dHRs3blRqV1hYiPfeew8uLi4wMzODnZ0dOnbsiG3btim1O336NAYNGgR7e3uYmZnB1dUVM2fOVNSXH2I+e/Yshg0bhvr168PV1VWpTl18CQkJ8PLygpmZGZo3b44VK1Yo2qSkpKBTp04AgJCQEMU4lZ8OUNdvWVkZ/v3vf8Pd3R2mpqZo2LAhxo4di+vXryu1K/9dp6amokePHrCwsEDz5s2xdOlSlJWVVfj7K4//8UOiKSkpkCQJ27ZtQ3h4OBwdHWFjY4NevXrhv//97zP7q+jzWFGM5e8rLS3FF198oRijcjk5OXjrrbfQtGlTmJiYwMXFBTExMSgtLVXa/v379/HBBx/Aw8MDZmZmsLe3R0BAAH766ScAjz6T9+7dw+bNmxXb8Pf313g72dnZeP3112FtbQ1bW1u88cYbyMnJeeb4lLtx4wbefPNNvPTSSzAxMYGjoyOGDRuGW7duKdr8888/is+0iYkJmjRpgpkzZ2r1fzU3N1exPVNTUzRo0ADdu3fH0aNHNe6LNMOZ83NkwoQJWLp0KTZu3IhPPvlEUX7+/HmcOXMGYWFhMDQ0rLCPmzdvYvTo0ZgzZw4WL14MA4NH398uXryI/v37Y+bMmbC0tMSFCxewbNkynDlzBklJSRrFGRkZiXv37mHXrl1Kh2UfP6S5d+9efP/994iKikLjxo3RsGFDAI/Oi7311lto1qwZAODUqVN45513cOPGDURFRSlt59dff8Xs2bMRFhaGRo0aYf369Zg4cSLc3NzwyiuvAADeffddfPXVV1i4cCHat2+Pe/fuISMjA3l5eYp+Dh06hEGDBsHDwwMff/wxmjVrhqysLBw+fFhl34YOHYoRI0ZgypQpz/xjmJaWhpkzZyI6OhqNGzfGli1bMGPGDDx48ADvvfceOnTogE2bNiEkJAQREREYMGAAAFQ4W3777bexdu1aTJ8+HQMHDkRWVhYiIyORkpKCs2fPQiaTKdrm5ORg1KhRmD17NubPn4+EhAS8//77cHR0xNixYyuM/WnmzZuH7t27Y/369fjnn38wd+5cDBo0CJmZmc/87KnzrBjLD4d37doVw4YNw+zZs5Xe27lzZxgYGCAqKgqurq44efIkFi5ciKysLGzatAkAUFpain79+uH777/HzJkz0bNnT5SWluLUqVO4du0aunXrhpMnT6Jnz54ICAhQHM63sbHRaDtFRUXo1asXsrOzsWTJErRs2RLfffcd3njjjUqNxY0bN9CpUyeUlJRg3rx58PLyQl5eHg4dOoTbt2+jUaNGKCwshJ+fH65fv65oc+7cOURFRSE9PR1Hjx7V6DqFMWPG4OzZs1i0aBFatmyJO3fu4OzZs0r/P6iaCHqu+Pn5CZlMJh48eKAomz17tgAgfv/9d0XZpk2bBABx5coVpfcCEMeOHatwG2VlZaKkpEQcP35cABC//vqrom7+/PmiMh+radOmPbUdAGFrayvy8/Mr7OPhw4eipKREfPDBB8Le3l6UlZUp6pycnISZmZm4evWqoqyoqEjY2dmJt956S1Hm6ekpBg8eXOF2XF1dhaurqygqKnpqm/L9joqKemrd45ycnIQkSSItLU2pvHfv3sLGxkbcu3dPCCFEamqqACA2bdr0zH4zMzMFADF16lSldqdPnxYAxLx58xRl5b/r06dPK7Vt3bq16NOnz1P38/H4x40bp3idnJwsAIj+/fsrtduxY4cAIE6ePFlhfxV9HisTIwAxbdo0pbK33npLWFlZKX0GhBAiNjZWABDnzp0TQgjx5ZdfCgBi3bp1FcZoaWmptM+abueLL74QAMQ333yj1G7y5MlP/R0/bsKECcLY2FicP3/+qW2WLFkiDAwMRGpqqlL5rl27BACxf/9+RdmTv8MrV66oxGFlZSVmzpxZYVxUPXhY+zkzceJEyOVy7Nu3D8CjWUF8fDx69OiBFi1aPPP99evXR8+ePVXKL1++jODgYDRu3BiGhoYwNjaGn58fACAzM1O3OwGgZ8+eqF+/vkp5UlISevXqBVtbW0UcUVFRyMvLw19//aXU1tvbWzHDBgAzMzO0bNkSV69eVZR17twZBw4cQFhYGFJSUlBUVKTUx++//45Lly5h4sSJMDMze2bcQUFBld7HNm3aoF27dkplwcHB+Oeff3D27NlK91MuOTkZAFSuou7cuTM8PDxw7NgxpfLGjRujc+fOSmVeXl5K46Op//u//1PpD4DWfVYlxsTERAQEBMDR0RGlpaWKn/JTO8ePHwcAHDhwAGZmZopTOJqq7HaSk5NhbW2tMkbBwcGV2s6BAwcQEBAADw+PCmPx9PSEt7e3Uix9+vSBJElISUnRaN86d+6MuLg4LFy4EKdOnUJJSYlG7yftMTk/Z4YNGwZbW1vFobT9+/fj1q1bigvBnkXd1bJ3795Fjx49cPr0aSxcuBApKSlITU3Fnj17AEAloemCujjOnDmDwMBAAMC6devw448/IjU1FeHh4WrjsLe3V+nD1NRUqd2KFSswd+5c7N27FwEBAbCzs8PgwYNx8eJFAI/OuQEVH0p+VtxP07hx46eWaXPYsPw96mJwdHRU6bMy46OpJ/s0NTUFoP1npCox3rp1C99++y2MjY2Vftq0aQMAkMvlAB79jh0dHRWncDRV2e3k5eWhUaNGKu9X9zlQJzc395mfw1u3buG3335TicXa2hpCCEUslfX1119j3LhxWL9+Pbp27Qo7OzuMHTtWo/PkpB2ec37OmJubY+TIkVi3bh1u3ryJjRs3wtraGsOHD6/U+9Wdj0pKSkJ2djZSUlIUs2UAFd5TXVXq4ti+fTuMjY2RmJioNIutyu0tlpaWiImJQUxMDG7duqWYRQ8aNAgXLlxQXEH85AVVmsT9NOr+wJWXqUtKz1L+nps3b6r8Ec/OzlY63/wikMlk8PLywqJFi9TWOzo6Anh0pfgPP/yAsrIyrRJ0Zbdjb2+PM2fOqNRXNtE1aNDgmZ9DmUwGc3NzlQsfH6/XhEwmw/Lly7F8+XJcu3YN+/btQ1hYGP766y8cPHhQo75IM5w5P4cmTpyIhw8f4sMPP8T+/fsxYsSIKt0OVZ5wymdB5dasWaN1n9rMqCRJgpGRkdKFRUVFRfjqq6+0juNxjRo1wvjx4zFy5Ej897//RWFhIVq2bAlXV1ds3LgR9+/f18l2yp07dw6//vqrUtnWrVthbW2NDh06ANBsnMpPR8THxyuVp6amIjMzE6+++qouwq4zBg4ciIyMDLi6uqJjx44qP+VJs1+/figuLn7mwzeeNmOv7HYCAgJQUFCgOOVU7mnPJnhSv379kJycXOHV7wMHDsSlS5dgb2+vNhZnZ+dKbUudZs2aYfr06ejdu7dWp11IM5w5P4c6duwILy8vLF++HEKISh/Sfppu3bqhfv36mDJlCubPnw9jY2Ns2bJFJbFoom3btgCAZcuWoV+/fjA0NISXlxdMTEye+p4BAwbg448/RnBwMN58803k5eUhNjZW5UuDJnx9fTFw4EB4eXmhfv36yMzMxFdffYWuXbsqvtCsWrUKgwYNQpcuXTBr1iw0a9YM165dw6FDh7Blyxatt+3o6Ij/+7//Q3R0NBwcHBAfH48jR45g2bJlim27urrC3NwcW7ZsgYeHB6ysrODo6Kj4g/+4Vq1a4c0338TKlSthYGCAfv36Ka7WfumllzBr1iytY62LPvjgAxw5cgTdunVDaGgoWrVqheLiYmRlZWH//v1YvXo1mjZtipEjR2LTpk2YMmUK/vvf/yIgIABlZWU4ffo0PDw8FA80adu2LVJSUvDtt9/CwcEB1tbWaNWqVaW3M3bsWHzyyScYO3YsFi1ahBYtWmD//v04dOhQpffnwIEDeOWVVzBv3jy0bdsWd+7cwcGDB/Huu+/C3d0dM2fOxO7du/HKK69g1qxZ8PLyQllZGa5du4bDhw9j9uzZ8PX1rdT2/v77bwQEBCA4OBju7u6wtrZGamoqDh48iKFDh2r9e6FKqu0r0qh6fPrppwKAaN26tdr6p10d26ZNG7Xtf/rpJ9G1a1dhYWEhGjRoICZNmiTOnj2rcnVnZa/Wvn//vpg0aZJo0KCBkCRJKRaoufK23MaNG0WrVq2EqampaN68uViyZInYsGGDyr44OTmJAQMGqLzfz89P+Pn5KV6HhYWJjh07ivr16yv6nDVrlpDL5UrvO3nypOjXr5+wtbUVpqamwtXVVcyaNUtlv3Nzc1W2+bSrtQcMGCB27dol2rRpI0xMTISzs7P4+OOPVd6/bds24e7uLoyNjQUAMX/+/Kf2+/DhQ7Fs2TLRsmVLYWxsLGQymRg9erT4888/VcZB3e963LhxwsnJSaX8SU+7Wnvnzp1K7dRdAayOJp9HdTE+7TOTm5srQkNDhYuLizA2NhZ2dnbCx8dHhIeHi7t37yraFRUViaioKNGiRQthYmIi7O3tRc+ePcVPP/2kaJOWlia6d+8uLCwsBAClz1Flt3P9+nURFBQkrKyshLW1tQgKChI//fRTpcZICCH+/PNPMWHCBNG4cWNhbGwsHB0dxeuvvy5u3bqlaHP37l0REREhWrVqJUxMTIStra1o27atmDVrlsjJyVG0e9bV2sXFxWLKlCnCy8tL2NjYCHNzc9GqVSsxf/58xd0EVH0kIZ54egMRVTtnZ2d4enoiMTGxtkMhIj3Ec85ERER6hsmZiIhIz/CwNhERkZ7hzJmIiEjPMDkTERHpGb1IzgUFBZgzZw4CAwPRoEEDpWXxHjd+/HilJQbLf9zd3dX2u3LlSsXSeeVLuPHZsEREpO/04iEkeXl5WLt2Ldq1a4fBgwdj/fr1T21rbm6uskShubm5SrtFixYhMjISYWFhCAwMRGpqKiIiInDjxg2sXbtW5/tARESkK3qRnJ2cnHD79m1IkgS5XF5hcjYwMECXLl0q7C8vLw8LFy7E5MmTsXjxYgCPFm4vKSlBREQEZs6cidatW+t0H4iIiHRFLw5rlx+e1pWDBw+iuLgYISEhSuUhISEQQlRpoQQiIqLqphfJWRNFRUWKNYWbNm2K6dOnIz8/X6lNRkYGgP89v7mcg4MDZDKZop6IiEgf6cVh7cpq164d2rVrB09PTwCPFjH/5JNPcOzYMaSmpsLKygrAo8PapqamsLS0VOnDzs7umWvlFhYW4sKFC3B3d6/Sak5ERFS3Xbt2TeN1sCsik8nQrFmzZ7arU8n5yVV1evfujfbt22PYsGFYt26dUn1Fh8mfdQj9woUL8PHxwc8//6xYuo+IiF4s165dg4eHBwoLC3XWp4WFBTIzM5+ZoOtUclZnyJAhsLS0xKlTpxRl9vb2KC4uRmFhocrMNz8/Hz4+Pir9yOVyxbejK1euVG/QRESk9+RyOQoLCxEfHw8PD48q95eZmYnRo0dDLpc//8kZAIQQMDD43+nz8nPN6enpSmuX5uTkQC6XKw6LP2737t1Yt24dAOj0WxIREdVtHh4eNX4Utc4n5127dqGwsFDp9qq+ffvCzMwMcXFxSsk5Li4OkiRh8ODBKv0EBQXBz88PAHDu3DkMGzas2mMnIiJSR2+S84EDB3Dv3j0UFBQAAM6fP49du3YBAPr374/c3FwEBwdjxIgRcHNzgyRJOH78OJYvX442bdpg0qRJir7s7OwQERGByMhI2NnZKR5CEh0djUmTJqm9x1kmk0EmkwHgzJmIiGqX3iTnt99+G1evXlW83rlzJ3bu3Ang0TlgW1tbNGrUCB9//DFu3bqFhw8fwsnJCaGhoZg3b57Kldnh4eGwtrbGqlWrEBsbi8aNGyMsLAzh4eE1ul9ERESa0pvknJWV9cw2e/bs0ajP0NBQhIaGahkRERFR7ahzDyEhIiJ63jE5ExER6RkmZyIiIj3D5ExERKRnmJyJiIj0DJMzERGRnmFyJiIi0jNMzkRERHqGyZmIiEjPMDkTERHpGSZnIiIiPcPkTEREpGeYnImIiPQMkzMREZGeYXImIiLSM0zOREREeobJmYiISM8wORMREekZJmciIiI9w+RMRESkZ5iciYiI9AyTMxERkZ7Ri+RcUFCAOXPmIDAwEA0aNIAkSYiOjlbb9uzZs+jVqxesrKxQr149DB06FJcvX1bbduXKlXB3d4epqSlcXFwQExODkpKSatwTIiKiqtOL5JyXl4e1a9fi/v37GDx48FPbXbhwAf7+/njw4AF27NiBjRs34vfff0ePHj2Qm5ur1HbRokWYMWMGhg4dikOHDmHq1KlYvHgxpk2bVs17Q0REVDVGtR0AADg5OeH27duQJAlyuRzr169X2y4qKgqmpqZITEyEjY0NAMDHxwctWrRAbGwsli1bBuBRsl+4cCEmT56MxYsXAwD8/f1RUlKCiIgIzJw5E61bt66ZnSMiItKQXsycJUmCJEkVtiktLUViYiKCgoIUiRl4lNgDAgKQkJCgKDt48CCKi4sREhKi1EdISAiEENi7d69O4yciItIlvUjOlXHp0iUUFRXBy8tLpc7Lywt//PEHiouLAQAZGRkAgLZt2yq1c3BwgEwmU9QTERHpozqTnPPy8gAAdnZ2KnV2dnYQQuD27duKtqamprC0tFTbtrwvIiIifVRnknO5ig5/P15X2XZERET6Ri8uCKsMe3t7AFA7683Pz4ckSahXr56ibXFxMQoLC2FhYaHS1sfHR6UPuVwOuVwOALhy5YqOoyciIqq8OpOcXV1dYW5ujvT0dJW69PR0uLm5wczMDMD/zjWnp6fD19dX0S4nJwdyuRyenp4qfezevRvr1q0DABQWFlbHLhAREVVKnTmsbWRkhEGDBmHPnj0oKChQlF+7dg3JyckYOnSooqxv374wMzNDXFycUh9xcXGQJEntvdRBQUGIj49HfHw8FixYUF27QURE9Ex6M3M+cOAA7t27p0i858+fx65duwAA/fv3h4WFBWJiYtCpUycMHDgQYWFhKC4uRlRUFGQyGWbPnq3oy87ODhEREYiMjISdnR0CAwORmpqK6OhoTJo0Se09zjKZDDKZDABnzkREVLv0Jjm//fbbuHr1quL1zp07sXPnTgCPzgE7OzvD3d0dKSkpmDt3LoYNGwYjIyP07NkTsbGxaNCggVJ/4eHhsLa2xqpVqxAbG4vGjRsjLCwM4eHhNbpfREREmtKb5JyVlVWpdj4+Pjh69Gil2oaGhiI0NLQKUREREWkuNDQUK1as0Pr9epOciYiI6prDhw9j5cqVkMvlEEIoym/cuIEzZ84AAE6dOqVxv0zOREREWoqMjMTUqVPh7OysKJMkCRMmTFCs96ANJmciIiIt3b17F+PGjVMpNzQ0hJ+fn9b91plbqYiIiPTN+PHj1ZZPmDChSv1y5kxERKSlFStW4NNPP1VbN3fuXK37ZXImIiLSUkpKiuLfQghkZ2dj5cqVGDBgQJX6ZXImIiLSkqurq9JrNzc3eHp6YvDgwU895F0ZPOdMRESkQ5aWlsjNzUVpaanWfXDmTEREpEOmpqbIzMysUh9MzkRERFrq2rWr0sNHHlf+8JGePXsiKSlJo36ZnImIiLS0dOnSZ7b54IMPNO6XyZmIiEhLlXnQyMsvv6xxv0zOdUxwcLDO+9y6davO+yQiehGcPn0aCxYsgKWlJZYsWYLmzZsjIyMDZWVl8PLy0rpfXq1NRESkpTlz5iAgIAAeHh6IiYkBANy/fx+RkZFV6pczZyIiIi3l5ORg9uzZKCwsRPfu3QEAHTp0wMWLF6vUL2fOREREWvL09MSlS5dgYWGB4uJiAEB2djbMzMyq1C9nzkRERFrq1q0bJk6ciDFjxqC0tBTr16/Hli1b0Ldv3yr1y+RMRESkpePHj8PGxgbffPMNWrdujR9//BEjRozgqlRERES1Zd++fdXSL885ExER6RnOnImIiLT00ksvPfXxnQAgSRKEELh+/bpG/TI5ExERaenx9Zx1qU4l55SUFAQEBKitO3nyJLp06aJ4ffbsWcyZMwenTp2CkZERevbsidjYWDRv3rymwiUioufck+s560qdSs7lFi9erJKkPT09Ff++cOEC/P394e3tjR07dqC4uBhRUVHo0aMH0tLS0KBBg5oOmYiInkPlE0NJklTq0tLSUFhYiG7dumncb51Mzi1atFCaJT8pKioKpqamSExMhI2NDQDAx8cHLVq0QGxsLJYtW1ZToRIR0XNs+PDhuHTpEkxNTVXqLl++jG3btr04ybkipaWlSExMxNixYxWJGQCcnJwQEBCAhIQEJmciItIJSZJw6tQpmJiYqNSVlJQgIyNDq37rZHKeNm0aRowYAQsLC3Tt2hWRkZGKJbkuXbqEoqIitauBeHl54ciRIyguLq7yo9WIiIgAYObMmWoPawOAubm5Vn3WqeRsa2uLGTNmwN/fH/b29vjjjz/w4Ycfwt/fH9999x369OmDvLw8AICdnZ3K++3s7CCEwO3bt+Hg4FDt8VbH8o5ERKQ/hBA4deqU2sPaVVGnknP79u3Rvn17xesePXpgyJAhaNu2LebMmYM+ffoo6p72LeZZdURERJX10ksvVUtOqVPJWZ169eph4MCBWL16NYqKimBvbw8Aihn04/Lz8yFJEurVq6dSJ5fLIZfLAQBXrlyp1piJiOj5cPLkyWrpt84nZwCKp7NIkgRXV1eYm5sjPT1dpV16ejrc3NzUnm/evXs31q1bBwAoLCys3oCJiIgqUOefrX379m0kJibC29sbZmZmMDIywqBBg7Bnzx4UFBQo2l27dg3JyckYOnSo2n6CgoIQHx+P+Ph4LFiwoKbCJyIiUlGnZs7BwcFo1qwZOnbsCJlMhosXL+Kjjz7CrVu3EBcXp2gXExODTp06YeDAgQgLC1M8hEQmk2H27Nlq+5bJZJDJZAA4cyYiotpVp2bOXl5eOHToECZNmoRevXohPDwcrVu3xk8//YRevXop2rm7uyMlJQXGxsYYNmwYxo8fDzc3N5w4cYJPByMiIr2nk5lzcXExsrKy0KJFCxgaGuqiS7XCwsIQFhZWqbY+Pj44evRotcVCRERUXTSeOa9cuVLpnOzPP/+Ml156CW3atEHLli3x559/6jRAIiKiF43GyXn9+vVKtyLNnTsXdnZ2+OSTTyCEwMKFC3UZHxER0QtH48Pa165dg7u7OwCgoKAAJ06cwPbt2zF06FDUr18fUVFROg+SiIjoRaLxzPn+/fswNjYG8Ojm67KyMsXFWM7OzsjJydFthERERC8YjZNzs2bN8P333wMAvvnmG3h7eytWf8rNzVVaCYqIiIg0p/Fh7dGjRyMmJgZ79+7Fr7/+itjYWEXdf/7zH7Rs2VKnARIREb1oNE7O4eHhMDIywk8//YQhQ4YgNDRUUZeRkYGgoCCdBkhUXapj1bCtW7fqvE8ievFonJwlSXrqvcb79u2rckBEREQvujr1hDAiIqIXQaVmzhMmTKh0h5IkYcOGDVoHRETKdH34nYfeifRfpZJzUlJSpReTro5Fp6l6Vce5V11jQiGiF0mlknNWVlY1h0FERETleM6ZiIhIzzA5ExER6ZlKHdZ2cXHR6JzzpUuXqhQUERHRi6xSydnPz48XelGtqgsXrdUVfPgKkf6rVHKOi4ur5jCIng/8EkFEuqDxE8KIiOoiHjGgukSrC8Jyc3Px/vvvo2vXrmjRogXOnTsHAFizZg1++eUXnQZIRET0otE4OV+5cgXt2rXDihUrIEkSLl++jPv37wMAfvvtN6xYsULnQRIREb1INE7Oc+bMQb169XDx4kWcOHECQghF3csvv4wff/xRpwESERG9aDQ+53zs2DF88cUXcHR0xMOHD5XqHBwckJ2drbPgiIiIXkQaz5yLi4thZ2entu7evXswMOBzTYiIiKpC45lzq1atcPToUfTu3Vul7sSJE/D09NRJYFV19+5dREREYMeOHcjPz4e7uzvCwsIwYsSI2g6N6LnDK6GJdEvj5Dx58mS8++67cHR0xKhRowAADx48wK5du/D555/js88+03mQ2hg6dChSU1OxdOlStGzZElu3bsXIkSNRVlbGe1GJ6oAX9f8pv+gQoEVynjp1KtLS0jBr1izMnj0bwKMLwYQQmDx5MsaNG6fzIDW1f/9+HDlyRJGQASAgIABXr17Fv/71L7zxxhswNDSs5SiJiIjU0+ohJGvXrsWECRPw3Xff4datW5DJZBg4cCC6deum6/i0kpCQACsrKwwfPlypPCQkBMHBwTh9+rTexEpEVNdwdl/9tH5CWJcuXdClSxddxqIzGRkZ8PDwgJGR8u55eXkp6pmciaiq6sqh97oQJxO+Mo0vrT516hR27Nihtm7Hjh04ffp0lYOqqry8PLVXlJeX5eXl1XRIRERElaZxcp43bx7S09PV1p0/fx4RERFVDkoXKlpFS12dXC7HhQsXcOHCBVy5cqU6QyMiIqqQxoe1f/vtN8WFYE/y9fXF559/XuWgqsre3l7t7Dg/Px8A1M6qd+/ejXXr1gEACgsLdRJHXT6kQkREtUfj5Hzv3j2Vc7nlDAwMUFBQUOWgqqpt27bYtm0bSktLlWItn/Gruxc7KCgIfn5+AIBz585h2LBhNRMsERHREzQ+rO3i4oLk5GS1dcnJyXBycqpyUFU1ZMgQ3L17F7t371Yq37x5MxwdHeHr66vyHplMBnd3d7i7u8PFxaWmQiUiIlKh8cx5xIgRWLRoEVq1aoWQkBBFeVxcHJYvX473339fpwFqo1+/fujduzfefvtt/PPPP3Bzc8O2bdtw8OBBxMfH8x5nIiLSaxon57CwMKSkpGDixImYPn06HB0dkZ2djeLiYvj7++tFcgaAPXv2IDw8HFFRUYrHd27btq1Sj+8sKioCAGRmZlZ3mEREpKfKc0BaWppO+6sMSTy+5mMlPXz4EFu3bsXBgweRm5uLBg0aoF+/fhg5cuRzMSvdsmULRo8eXdthEBHRc8bCwgKZmZlo1qxZhe20Ss66VlBQgAULFiAtLQ2//PIL5HI55s+fj+joaKV248ePx+bNm1Xe36pVK1y4cEGlfOXKlVi1ahWuXLkCR0dHjB8/HvPmzYOxsXGF8cjlchw6dAjOzs4wNzev0r4REVHddfPmTdSrV09nuUAmkz0zMQNVeEKYLuXl5WHt2rVo164dBg8ejPXr1z+1rbm5OZKSklTKnrRo0SJERkYiLCwMgYGBSE1NRUREBG7cuIG1a9dWGI9MJlMs6kFERFTT9CI5Ozk54fbt25AkCXK5vMLkbGBg8MzHhubl5WHhwoWYPHkyFi9eDADw9/dHSUkJIiIiMHPmTLRu3Vqn+0BERKQrGt9KVR0kSarwiV6aOnjwIIqLi5WuJgceLXwhhMDevXt1ti0iIiJd04vkrImioiI0btwYhoaGaNq0KaZPn6548le5jIwMAI8eRvI4BwcHyGQyRT0REZE+qtRh7RUrVmDEiBFo2LBhdcdToXbt2qFdu3aKJ3wdP34cn3zyCY4dO4bU1FRYWVkBeHRY29TUFJaWlip92NnZPXPhi8LCQly4cAHu7u6wsLDQ/Y4QEVGdcO3aNcjlcp31p9MLwmbNmoUuXbqgYcOGMDQ0xMmTJ9G5c+cqB6mpWbNmKb3u3bs32rdvj2HDhmHdunVK9ZoufPG4CxcuwMfHBz///DM6dOhQtaCJiKhOunbtGjw8PHS23gJQ+VupKpWcra2tcefOHQCAHtx5pWTIkCGwtLTEqVOnFGX29vYoLi5GYWGhysw3Pz8fPj4+Kv3I5XLFtyOuSkVERHK5HIWFhYiPj4eHh0eV+8vMzMTo0aMhl8t1k5y9vb0RGhqKXr16AQA++ugjNGrUSG1bSZLw6aefahhy1QghYGDwv9Pn5eea09PTlZ6jnZOTA7lcrnbhi+pYlYqIiOo+Dw+PGj+KWqnkvHz5coSEhGDt2rWQJAk7d+58atuaTs67du1CYWGh0u1Vffv2hZmZGeLi4pSSc1xcHCRJwuDBg1X64apURESkLyo9c/7ll18APLrP+NSpUzo/53zgwAHcu3dPseTk+fPnsWvXLgBA//79kZubi+DgYIwYMQJubm6QJAnHjx/H8uXL0aZNG0yaNEnRl52dHSIiIhAZGQk7OzvFQ0iio6MxadIktfc4y2QyyGQyAJw5ExFR7dL4ISSbNm2Cq6urzgN5++23cfXqVcXrnTt3KmboV65cga2tLRo1aoSPP/4Yt27dwsOHD+Hk5ITQ0FDMmzdP5crs8PBwWFtbY9WqVYiNjUXjxo0RFhaG8PBwncf+NMHBwTrvc+vWrTrvk4iI9IvGyXncuHEAHj0P++TJk8jLy4NMJkOXLl1gbW2tdSBZWVnPbLNnzx6N+gwNDUVoaKiWEREREdUOrR7fGRsbi5iYGBQWFkIIAUmSYGFhgZiYGLz77ru6jpGIiOiFonFy/vLLLzFnzhz069cP48ePV6znvHnzZvzrX/9CgwYNMGbMmOqIlYiI6IWgcXL+5JNPEBwcjPj4eKXy4cOHY/To0fjkk0+YnImIiKpA42drX7hwAaNHj1ZbN3r0aGRmZlY5KCIioheZxsnZ3NxcZaGJcvn5+TpbkJqIiOhFpXFy7tGjB6Kjo5Gdna1UnpOTgw8++ACvvPKKzoIjIiJ6EWl8znnx4sXo1q0b3Nzc8Oqrr8LBwQE3b95EUlISjI2NNb7diYiIiJRpPHNu06YNUlNT8dprryE1NRWbNm1CamoqBg8ejDNnzqh9+hYRERFVnlb3Obds2RLbtm3TdSxEREQELWbOREREVL2YnImIiPQMkzMREZGeYXImIiLSM0zOREREeobJmYiISM9olZwnTJiA69ev6zoWIiIigpbJefPmzYrna5eVlaFnz564ePGiTgMjIiJ6UVXqISQuLi7w9fWFr68vOnXqBCEEJEkCAAghkJKSgoKCgmoNlIiI6EVRqeQcGhqK06dP49NPP8W1a9cgSRKmTZuGwMBAdOvWDQAUyZqIiIiqplKHtWfNmoXt27cjKysL2dnZEEJAJpNh+/bt6N27NyRJwrvvvovVq1fzXDQREVEVVSo5P34+uXHjxgCA6OhoZGRk4ObNmxBCwMbGBosXL4aTk1P1REpERPSCqNRh7VatWqF+/fro3LkzOnfuDEmS8PDhQwCAvb09AGD+/Pno0KED0tLSqi1YIiKiF0GlZs5paWn497//DScnJyQmJkIIgW7duuGVV15BZGQkJElCWVkZAMDb21vjIAoKCjBnzhwEBgaiQYMGkCQJ0dHRatuePXsWvXr1gpWVFerVq4ehQ4fi8uXLatuuXLkS7u7uMDU1hYuLC2JiYlBSUqJxfERERDWpUsnZy8sLEydOxOrVq/Hzzz8DABYsWICOHTvi0KFDEELA398fr732GjZs2KBxEHl5eVi7di3u37+PwYMHP7XdhQsX4O/vjwcPHmDHjh3YuHEjfv/9d/To0QO5ublKbRctWoQZM2Zg6NChOHToEKZOnYrFixdj2rRpGsdHRERUk7RazxkAAgMD4eXlhdLSUpiYmCAiIgK///47wsPDMXHiRI36cnJywu3btyFJEuRyOdavX6+2XVRUFExNTZGYmAgbGxsAgI+PD1q0aIHY2FgsW7YMwKNkv3DhQkyePBmLFy8GAPj7+6OkpAQRERGYOXMmWrdure2uExERVSutHkLi5OQEExMTAP+7hSowMBAbN27EzZs3Ne5PkqRn3opVWlqKxMREBAUFKRJzeSwBAQFISEhQlB08eBDFxcUICQlR6iMkJARCCOzdu1fjGImIiGqKVsn5ypUrcHd3B/Aosfr5+cHa2lrxujpcunQJRUVF8PLyUqnz8vLCH3/8geLiYgBARkYGAKBt27ZK7RwcHCCTyRT1RERE+kjrw9rlDAwMkJycrItYKpSXlwcAsLOzU6mzs7ODEAK3b9+Gg4MD8vLyYGpqCktLS7Vty/siIiLSR3VuVaqKZuaP11W2HRERkb6p8sy5ppTfT61u1pufnw9JklCvXj1F2+LiYhQWFsLCwkKlrY+Pj0ofcrkccrkcwKPD9kRERLWlziRnV1dXmJubIz09XaUuPT0dbm5uMDMzA/C/c83p6enw9fVVtMvJyYFcLoenp6dKH7t378a6desAAIWFhdWxC0RERJVSZw5rGxkZYdCgQdizZ4/SCljXrl1DcnIyhg4dqijr27cvzMzMEBcXp9RHXFwcJElSey91UFAQ4uPjER8fjwULFlTXbhARET2T3sycDxw4gHv37ikS7/nz57Fr1y4AQP/+/WFhYYGYmBh06tQJAwcORFhYGIqLixEVFQWZTIbZs2cr+rKzs0NERAQiIyNhZ2eHwMBApKamIjo6GpMmTVJ7j7NMJoNMJgPAmTMREdUuvUnOb7/9Nq5evap4vXPnTuzcuRPAo3PAzs7OcHd3R0pKCubOnYthw4bByMgIPXv2RGxsLBo0aKDUX3h4OKytrbFq1SrExsaicePGCAsLQ3h4eI3uFxERkab0JjlnZWVVqp2Pjw+OHj1aqbahoaEIDQ2tQlRERESaCw0NxYoVK7R+v94kZyIiorrm8OHDWLlyJeRyOYQQivIbN27gzJkzAIBTp05p3C+TMxERkZYiIyMxdepUODs7K8okScKECRMU6z1og8mZiIhIS3fv3sW4ceNUyg0NDeHn56d1v3XmVioiIiJ9M378eLXlEyZMqFK/nDkTERFpacWKFfj000/V1s2dO1frfpmciYiItJSSkqL4txAC2dnZWLlyJQYMGFClfpmciYiItOTq6qr02s3NDZ6enhg8ePBTD3lXBs85ExER6ZClpSVyc3NRWlqqdR+cORMREemQqakpMjMzq9QHkzMREZGWunbtqvTwkceVP3ykZ8+eSEpK0qhfJmciIiItLV269JltPvjgA437ZXImIiLSUmUeNPLyyy9r3C8vCCMiItLS6dOnMXDgQLzxxhu4fPkyACAjIwO//fZblfplciYiItLSnDlzEBAQAA8PD8TExAAA7t+/j8jIyCr1y8PaREREWsrJycHs2bNRWFiI7t27AwA6dOiAixcvVqlfzpyJiIi05OnpiUuXLsHCwgLFxcUAgOzsbJiZmVWpX86ciYiItNStWzdMnDgRY8aMQWlpKdavX48tW7agb9++VeqXyZmIiEhLx48fh42NDb755hu0bt0aP/74I0aMGMFVqYiIiGrLvn37qqVfnnMmIiLSM5w5ExERaemll1566uM7AUCSJAghcP36dY36ZXImIiLS0uPrOetSnUrOKSkpCAgIUFt38uRJdOnSRfH67NmzmDNnDk6dOgUjIyP07NkTsbGxaN68eU2FS0REz7kn13PWlTqVnMstXrxYJUl7enoq/n3hwgX4+/vD29sbO3bsQHFxMaKiotCjRw+kpaWhQYMGNR2yzgQHB+u8z61bt+q8TyKiF0F4eDi8vLzwxhtvqNRdv34dKSkpGD16tMb91snk3KJFC6VZ8pOioqJgamqKxMRE2NjYAAB8fHzQokULxMbGYtmyZTUVKhERPccOHDiAGTNmKJUdO3YMr776KqytrfHFF19olZyfu6u1S0tLkZiYiKCgIEViBgAnJycEBAQgISGhFqMjIqLnyd27d9GwYUOlsunTpwMAbG1tkZ+fr1W/dTI5T5s2DUZGRrCxsUGfPn3www8/KOouXbqEoqIieHl5qbzPy8sLf/zxh+IRa0RERFVRr149XL16VfE6OzsbBQUFKCgoQGFhodaP8axTh7VtbW0xY8YM+Pv7w97eHn/88Qc+/PBD+Pv747vvvkOfPn2Ql5cHALCzs1N5v52dHYQQuH37NhwcHGo6fCIies706tULM2bMwL/+9S8YGxvj+PHjsLCwwLx58yCEeOpFzM9Sp5Jz+/bt0b59e8XrHj16YMiQIWjbti3mzJmDPn36KOokSXpqPxXVERERVdbcuXMxZ84cTJ48GU5OTvj888/h6+uLNWvWwMXFBe+//75W/dap5KxOvXr1MHDgQKxevRpFRUWwt7cHAMUM+nH5+fmQJAn16tVTqZPL5ZDL5QCAK1euVGvMRET0fLC1tcWaNWuUylxcXPDKK69Uqd86n5wBKJ7OIkkSXF1dYW5ujvT0dJV26enpcHNzU3sOYPfu3Vi3bh0AoLCwsHoDJiIiqkCdvCDscbdv30ZiYiK8vb1hZmYGIyMjDBo0CHv27EFBQYGi3bVr15CcnIyhQ4eq7ScoKAjx8fGIj4/HggULaip8IiIiFXVq5hwcHIxmzZqhY8eOkMlkuHjxIj766CPcunULcXFxinYxMTHo1KkTBg4ciLCwMMVDSGQyGWbPnq22b5lMBplMBoAzZyIiql11aubs5eWFQ4cOYdKkSejVqxfCw8PRunVr/PTTT+jVq5einbu7O1JSUmBsbIxhw4Zh/PjxcHNzw4kTJ+r008GIiOjFUKdmzmFhYQgLC6tUWx8fHxw9erSaIyIiItK9OjVzJiIiehEwORMREekZJmciIiI9w+RMRESkZ5iciYiI9AyTMxERkZ5hciYiItIzTM5ERER6hsmZiIhIzzA5ExER6RkmZyIiIj3D5ExERKRnmJyJiIj0DJMzERGRnmFyJiIi0jNMzkRERHrGqLYDoNoXHBys8z63bt2q8z6JiF4UnDkTERHpGSZnIiIiPcPkTEREpGeYnImIiPQMkzMREZGe4dXaVC10fQU4r/4mohfJcztzvnv3LmbOnAlHR0eYmZnB29sb27dvr+2wiIiInum5nTkPHToUqampWLp0KVq2bImtW7di5MiRKCsrq5b7eomIiHTluUzO+/fvx5EjRxQJGQACAgJw9epV/Otf/8Ibb7wBQ0PDWo6SiIhIvecyOSckJMDKygrDhw9XKg8JCUFwcDBOnz6Nbt261VJ0pC/4ZDQi0lfPZXLOyMiAh4cHjIyUd8/Ly0tRz+Rct/BUhO7wSwmR/nsuk3NeXh6aN2+uUm5nZ6eoJyKqKn7RoeryXCZnAJAkSaM6uVwOuVwOALhy5Uq1xUWkKR41eLEw4RPwnCZne3t7tbPj/Px8AP+bQT9u9+7dWLduHQCgsLBQJ3HwPwTpAj9H+ou/G6ouz2Vybtu2LbZt24bS0lKl887p6ekAAE9PT5X3BAUFwc/PDwBw7tw5DBs2rGaCJSIiesJz+RCSIUOG4O7du9i9e7dS+ebNm+Ho6AhfX1+V98hkMri7u8Pd3R0uLi41FSoREZGK53Lm3K9fP/Tu3Rtvv/02/vnnH7i5uWHbtm04ePAg4uPjeY8zERHptecyOQPAnj17EB4ejqioKOTn58Pd3R3btm3DiBEjnvneoqIiAEBmZmZ1h0lERHqqPAekpaXptL/KkIQQQidbfY5s2bIFo0ePru0wiIjoOWNhYYHMzEw0a9aswnZMzmrI5XIcOnQIzs7OMDc316qPK1euIDIyEgsWLOA57CriWOoOx1J3OJa6o89jefPmTdSrV0/rXPAkmUz2zMQMPMeHtatCJpNh1KhRVerDwsICFhYWaNOmDdzd3XUU2YuJY6k7HEvd4VjqDsdS1XN5tTYREVFdxuRcTWQyGSZPngyZTFbbodR5HEvd4VjqDsdSdziWqnjOmYiISM9w5kxERKRnmJx17O7du5g5cyYcHR1hZmYGb29vbN++vbbD0kvajtWePXswcuRIuLm5wdzcHM7Ozhg1ahQuXrxYA1HrJ1197iIiIiBJktpH3L4oqjqW33zzDfz8/GBjYwNLS0u0adMGa9eurcaI9VdVxjI5ORm9e/dGw4YNYWVlBS8vL6xYsQIPHz6s5qj1hCCd6t27t6hXr55YvXq1SEpKEpMmTRIAxJYtW2o7NL2j7Vh17txZ/N///Z/YuHGjSElJEV999ZXw8PAQVlZWIiMjo4ai1y+6+Nz98ssvwtTUVDRq1Ei0adOmGqPVb1UZyyVLlggDAwMxdepUceDAAXH06FHx2WefiZUrV9ZA5PpH27E8cuSIMDAwEP7+/mLv3r3iyJEj4p133hEARGhoaA1FX7uYnHXou+++EwDE1q1blcp79+4tHB0dRWlpaS1Fpn+qMla3bt1SKbtx44YwNjYWEydO1Hms+k4Xn7uSkhLh7e0tQkNDhZ+f3wubnKsylv/5z3+EgYGBWLZsWXWHWSdUZSxHjRolTE1Nxd27d5XKAwMDhY2NTbXEq294WFuHEhISYGVlheHDhyuVh4SEIDs7G6dPn66lyPRPVcaqYcOGKmWOjo5o2rQp/vzzT53Hqu908blbunQp8vPzsWjRouoKs06oylh+9tlnMDU1xTvvvFPdYdYJVRlLY2NjmJiYqDz4o169ejAzM6uWePUNk7MOZWRkwMPDQ2mZSgDw8vJS1NMjuh6ry5cv4+rVq2jTpo3OYqwrqjqW58+fx8KFC/HFF1/Aysqq2uKsC6oylidOnICHhwd2796NVq1awdDQEE2bNkVYWBgePHhQrXHro6qM5ZQpU/DgwQOEhoYiOzsbd+7cwVdffYWEhATMmTOnWuPWF0zOOpSXlwc7OzuV8vKyvLy8mg5Jb+lyrEpLSzFx4kRYWVlh1qxZOouxrqjKWJaVlWHChAkYOnQo+vfvX20x1hVVGcsbN27g4sWLCA0NRWhoKI4ePYrx48cjNjYWISEh1RazvqrKWPr6+iIpKQkJCQlo0qQJ6tevj5CQECxatAizZ8+utpj1CR/fqWOSJGlV9yLSxVgJITBx4kR8//332L17N1566SVdhVenaDuWH3/8MS5evIh9+/ZVR1h1krZjWVZWhoKCAqXV7wICAnDv3j0sX74cMTExcHNz03m8+kzbsfz5558xZMgQ+Pr6Ys2aNbC0tERSUhIiIiJQXFyMyMjI6ghXrzA565C9vb3ab4P5+fkAoPZb5ItKF2MlhMCkSZMQHx+PzZs347XXXtN5nHWBtmN57do1REVFYenSpTAxMcGdO3cAPDoSUVZWhjt37sDU1FRnD/yvC6ryubS3t0dOTg769OmjVN6vXz8sX74cZ8+efaGSc1XGctq0aWjUqBESEhJgaGgI4NEXHQMDA0RHR2PUqFFo3rx59QSuJ3hYW4fatm2LzMxMlJaWKpWnp6cDwAt97+iTqjpW5Yl506ZNWL9+/Qu9xKe2Y3n58mUUFRVhxowZqF+/vuLnxx9/RGZmJurXr4/333+/2uPXJ1X5XJafS32S+P8PYTQweLH+3FZlLNPS0uDj46NIzOU6deqEsrIyjdZFrrNq92Lx58v+/fsFALF9+3al8r59+/JWqidUZazKysrExIkThSRJYu3atdUdqt7Tdixv374tkpOTVX7atWsnnJ2dRXJysrh48WJN7ILeqMrncs2aNWrv4Q0NDRUGBgYiKyurWmLWV1UZSxcXF+Hp6anSZt68eQKASEtLq5aY9QmTs4717t1b1K9fX6xdu1YkJSWJyZMnCwAiPj6+tkPTO5UZqwkTJghDQ0OlP2zTp08XAMSECRPEyZMnlX7Onj1bG7tS67QdS3Ve5PuchdB+LB88eCA6dOggbG1txaeffiqOHDki5s6dKwwNDcX06dNrY1dqnbZjuWLFCgFA9OvXT+zdu1ccPnxYzJ07VxgZGYlevXrVxq7UOCZnHSsoKBChoaGicePGwsTERHh5eYlt27bVdlh6qTJjNW7cOAFAXLlyRVHm5OQkAKj9cXJyqtmd0BPajqU6L3pyrspY5uXlibfeeks0atRIGBsbi5YtW4oPP/xQPHz4sAb3QH9UZSx3794tXn75ZSGTyYSlpaVo06aNWLBggcqDSZ5XXJWKiIhIz7xYVygQERHVAUzOREREeobJmYiISM8wORMREekZJmciIiI9w+RMRESkZ5iciYiI9AyTMxERkZ5hciatff3112jTpg3Mzc0hSRLS0tJqbNvnz59HdHQ0srKyVOrGjx8PZ2fnGotFG/n5+RgxYgQaNmwISZIwePDgattWdHQ0JElS/BgbG6NZs2aYPHkycnJyVNo7OztDkiRMmTJFpS4lJQWSJGHXrl1qt7VixQpIkqTxIi8PHjzAlClT4ODgAENDQ3h7e2v0fn3l7OyM8ePH13YYVAdxyUjSSm5uLsaMGYO+ffvi888/h6mpKVq2bFlj2z9//jxiYmLg7++vkogjIyMxY8aMGotFGwsWLEBCQgI2btwIV1fXGllO9ODBg7C1tcXdu3dx+PBhfPTRR/jpp5+QlpYGY2NjlfYbNmzArFmz0KpVq0pvY+PGjQCAc+fO4fTp0/D19a3U+7744gusWbMGK1euhI+PD6ysrCq9TX2WkJAAGxub2g6D6iAmZ9LK77//jpKSEowePRp+fn4Vti0sLISFhUUNRQa4urrW2La0lZGRAVdXV4waNarGtunj4wOZTAYA6NWrF+RyOTZt2oQffvgBAQEBSm27du2K8+fPY968edi9e3el+v/Pf/6DX3/9FQMGDMB3332HDRs2VDo5Z2RkwNzcHNOnT6+wnRACxcXFdWaN6fbt2z+zTUlJCSRJgpER/xzT//CwNmls/PjxePnllwEAb7zxBiRJgr+/v6LOysoK6enpCAwMhLW1NV599VUAwJEjR/Daa6+hadOmMDMzg5ubG9566y3I5XKVbVy4cAEjR45Eo0aNYGpqimbNmmHs2LG4f/8+4uLiMHz4cACPFmAvP1wbFxeniOHJ2XRxcTHef/99uLi4wMTEBE2aNMG0adNw584dpXbOzs4YOHAgDh48iA4dOsDc3Bzu7u6KGeGz5OfnY+rUqWjSpAlMTEzQvHlzhIeH4/79+wCArKwsSJKEo0ePIjMzUxF7SkrKU/ssjykxMRHt27eHubk5PDw8kJiYCACIi4uDh4cHLC0t0blzZ/znP/+pVKwdO3YEANy6dUulzs7ODmFhYdizZw9OnTpVqf42bNgAAFi6dCm6deuG7du3o7Cw8JnvkyQJ69evR1FRkcrvUpIkTJ8+HatXr4aHhwdMTU2xefNmAMAPP/yAV199FdbW1rCwsEC3bt3w3XffKfUdFxcHSZKQlJSEyZMnw97eHjY2Nhg7dizu3buHnJwcvP7666hXrx4cHBzw3nvvoaSkpMJ4Bw8eDCcnJ5SVlanU+fr6okOHDorXTx7WLj8t8NVXX2H27Nlo0qQJTE1N8ccffyhOPzypfB8eP4WTlJQEf39/2Nvbw9zcHM2aNUNQUFClxpvqiFpeeIPqoD/++EOsWrVKABCLFy8WJ0+eFOfOnRNCPFphxtjYWDg7O4slS5aIY8eOiUOHDgkhhPjiiy/EkiVLxL59+8Tx48fF5s2bRbt27USrVq3EgwcPFP2npaUJKysr4ezsLFavXi2OHTsm4uPjxeuvvy7++ecf8ddff4nFixcLAGLVqlWK5SL/+usvRQyPr05VVlYm+vTpI4yMjERkZKQ4fPiwiI2NFZaWlqJ9+/aiuLhY0dbJyUk0bdpUtG7dWnz55Zfi0KFDYvjw4QKAOH78eIXjUlRUJLy8vISlpaWIjY0Vhw8fFpGRkcLIyEj0799fCCFEcXGxOHnypGjfvr1o3ry5Iva///77qf2Wx+Tp6Sm2bdsm9u/fL3x9fYWxsbGIiooS3bt3F3v27BEJCQmiZcuWolGjRqKwsFDx/vnz5wsAIjc3V6nf9957TwAQP//8s8r2BgwYIAoLC0WTJk1Ejx49FHXJyckCgNi5c6fSewoLC4Wtra3o1KmTEEKI9evXCwAiLi6uwjETQoiTJ0+K/v37C3Nzc5XfJQDRpEkT4eXlJbZu3SqSkpJERkaGSElJEcbGxsLHx0d8/fXXYu/evSIwMFBIkqS0fvCmTZsEAOHi4iJmz54tDh8+LJYtWyYMDQ3FyJEjRYcOHcTChQsVyzsCEB999FGF8X7zzTcCgDhy5IhSeWZmpgAgVqxYoTSW48aNUxm/Jk2aiGHDhol9+/aJxMREkZeXp/g9Pal8H8pXbbpy5YowMzMTvXv3Fnv37hUpKSliy5YtYsyYMeL27dvPHG+qG5icSStP+yNdvvzbxo0bK3x/WVmZKCkpEVevXhUAxDfffKOo69mzp6hXr57iD7Q6O3fuFABEcnKySt2TyfngwYMCgPj3v/+t1O7rr78WAMTatWsVZU5OTsLMzExcvXpVUVZUVCTs7OzEW2+9VeE+rV69WgAQO3bsUCpftmyZACAOHz6sKNNkWUYnJydhbm4url+/rihLS0sTAISDg4O4d++eonzv3r0CgNi3b5+irPyPfk5OjigpKRG3b98WO3bsEJaWlmLkyJFqtzdgwAAhhBDr1q0TAMS3334rhHj67/3LL78UAMTq1auFEI+WCrSyslJK7BUZN26csLS0VCkHIGxtbUV+fr5SeZcuXUTDhg1FQUGBoqy0tFR4enqKpk2birKyMiHE/xLbO++8o/T+wYMHCwDi448/Vir39vYWHTp0qDDWkpIS0ahRIxEcHKxUPmfOHGFiYiLkcrmi7GnJ+ZVXXlHpt7LJedeuXQKASEtLqzBOqtt4WJuqRVBQkErZX3/9hSlTpuCll16CkZERjI2N4eTkBADIzMwE8Oj89PHjx/H666+jQYMGOoklKSkJAFSumh0+fDgsLS1x7NgxpXJvb280a9ZM8drMzAwtW7bE1atXn7kdS0tLDBs2TKm8fLtPbkcT3t7eaNKkieK1h4cHAMDf31/pfH55ubpYGzduDGNjY9SvXx+vv/46fHx8FIeInyYkJAStW7dGWFiY2sO45TZs2ABzc3OMGDECAGBlZYXhw4fj+++/x8WLFyu/o2r07NkT9evXV7y+d+8eTp8+jWHDhildOGZoaIgxY8bg+vXr+O9//6vUx8CBA5Vel4/TgAEDVMqf9Xs2MjLC6NGjsWfPHvz9998AgIcPH+Krr77Ca6+9Bnt7+2fuk7r/H5Xl7e0NExMTvPnmm9i8eTMuX76sdV+kv5icSecsLCxUrlAtKytDYGAg9uzZgzlz5uDYsWM4c+aM4nxmUVERAOD27dt4+PAhmjZtqrN48vLyYGRkpJLsJUlC48aNkZeXp1Su7o+rqampIsaKttO4cWOV84YNGzaEkZGRynY08eTV3CYmJhWWFxcXq/Rx9OhRpKam4tChQwgKCsKJEyfwzjvvVLhdQ0NDLF68GOfOnXtqIv/jjz9w4sQJDBgwAEII3LlzB3fu3FF8Sans+fqncXBwUHp9+/ZtCCFUygHA0dERAFTGWpPxUzd2T5owYQKKi4uxfft2AMChQ4dw8+ZNhISEPPO9gOo+acLV1RVHjx5Fw4YNMW3aNLi6usLV1RWffvqp1n2S/mFyJp1Td1FLRkYGfv31V3z44Yd455134O/vj06dOqkkQjs7OxgaGuL69es6i8fe3h6lpaXIzc1VKhdCICcnR3EFsy62c+vWLQghlMr/+usvlJaW6mw72mrXrh06duyIwMBA7Ny5E71798batWuRmppa4ftee+01dO/eHfPnz1ebuDZu3AghBHbt2oX69esrfspnpZs3b8bDhw+1jvvJz1P9+vVhYGCAmzdvqrTNzs4GgGof69atW6Nz587YtGkTAGDTpk1wdHREYGBgpd6v7v+ImZkZACguHiyn7oLJHj164Ntvv8Xff/+NU6dOoWvXrpg5c6biywLVfUzOVCPK/xiZmpoqla9Zs0bptbm5Ofz8/LBz5061f5TKlffzrNksAMXV4vHx8Urlu3fvxr179xT1VfXqq6/i7t272Lt3r1L5l19+qRSHPpAkCatWrYKhoSEiIiKe2X7ZsmX4888/sWLFCqXyhw8fYvPmzXB1dUVycrLKz+zZs3Hz5k0cOHBAZ7FbWlrC19cXe/bsUfr9l5WVIT4+Hk2bNq2Re+5DQkJw+vRp/PDDD/j2228xbtw4GBoaat1f+R0Gv/32m1L5t99++9T3GBoawtfXF6tWrQIAnD17Vuvtk37hjXVUI9zd3eHq6oqwsDAIIWBnZ4dvv/0WR44cUWn78ccf4+WXX4avry/CwsLg5uaGW7duYd++fVizZg2sra0VT6Bau3YtrK2tYWZmBhcXF7WHpHv37o0+ffpg7ty5+Oeff9C9e3f89ttvmD9/Ptq3b48xY8boZB/Hjh2LVatWYdy4ccjKykLbtm3xww8/YPHixejfvz969eqlk+3oSosWLfDmm2/i888/xw8//KC4PU6d7t2747XXXsM333yjVH7gwAFkZ2dj2bJlitvpHufp6YnPPvsMGzZsUDnvWxVLlixB7969ERAQgPfeew8mJib4/PPPkZGRgW3btqmdmerayJEj8e6772LkyJG4f/9+lZ8E1r9/f9jZ2WHixIn44IMPYGRkhLi4OPz5559K7VavXo2kpCQMGDAAzZo1Q3FxseLUgb59xkh7nDlTjTA2Nsa3336Lli1b4q233sLIkSPx119/4ejRoypt27VrhzNnzsDHxwfvv/8++vbti7lz58LU1FRxrtDFxQXLly/Hr7/+qjhE/rQZhiRJ2Lt3L959911s2rQJ/fv3R2xsLMaMGYOkpCSV2by2zMzMkJycjFGjRuHDDz9Ev379EBcXh/feew979uzRyTZ0bf78+bCyskJUVNQz2y5ZskRlZrhhwwaYmJg89VyrTCbDkCFDkJiYqPZ+am35+fkpLsAbP348RowYgb///hv79u3DG2+8obPtVMTW1hZDhgzB9evX0b179yrP1m1sbHDw4EFYW1tj9OjRmDJlCjw9PREeHq7UztvbG6WlpZg/fz769euHMWPGIDc3F/v27av0YXXSf5J48gQZERER1SrOnImIiPQMkzMREZGeYXImIiLSM0zOREREeobJmYiISM8wORMREekZJmciIiI9w+RMRESkZ5iciYiI9AyTMxERkZ5hciYiItIzTM5ERER65v8BZpn6UpHZ0uEAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 500x300 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<ggplot: (8780487676639)>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_transcription_file = figure_dir+'/'+'viral_transcription.svg'\n",
    "\n",
    "viral_transcription_plot = (\n",
    "    p9.ggplot(\n",
    "        wt_infected_cells,\n",
    "        p9.aes(x='frac_viral_UMIs')) +\n",
    "    p9.geom_histogram(breaks=[0,0.05,0.1,0.15,0.2,0.25,0.3,0.35,\n",
    "                              0.4,0.45,0.5,0.55,0.6,0.65,0.7,0.75,0.8,0.85]) +\n",
    "    p9.facet_grid('sample~') +\n",
    "    p9.ggtitle(f'Viral transcription in infected cells') +\n",
    "    p9.labs(x='fraction of mRNA from virus',\n",
    "            y='# of cells') +\n",
    "    p9.theme(figure_size=(5, 3),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text=p9.element_text(size=12, color=\"black\")))\n",
    "viral_transcription_plot.save(filename = viral_transcription_file)\n",
    "display(viral_transcription_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot viral genes per infected cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dbacsik/.conda/envs/barcoded_flu_pdmH1N1_final_analysis/lib/python3.8/site-packages/plotnine/ggplot.py:718: PlotnineWarning: Saving 3 x 3 in image.\n",
      "/home/dbacsik/.conda/envs/barcoded_flu_pdmH1N1_final_analysis/lib/python3.8/site-packages/plotnine/ggplot.py:719: PlotnineWarning: Filename: results/figures/genes_per_cell.svg\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAUwAAAE+CAYAAADmhCmVAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAA9hAAAPYQGoP6dpAABN20lEQVR4nO3dd1RUV/s24PsgZQaQOhRBAQV1RAWMEbEFVFRsEZEYxIaoyWv5WWIiRIqAYjTBFt/YYmwRNRbUSIzYAJM3qChq1EhiA2NBaRqkqMD+/vDjhOMMMIODQ3mutVjL2bNnn32G4fG0uQ/HGGMghBBSIw11T4AQQhoKKpiEEKIgKpiEEKIgKpiEEKIgKpiEEKIgKpiEEKIgKpiEEKIgKpiEEKIgKpiEEKKgJlEwt27dCo7jIBKJkJmZKfO8h4cHOnXqpIaZAUlJSeA4Dvv27VPL8pWVkZGBoUOHwsTEBBzHYc6cOW88JsdxiIiIeONxKsvIyADHcdi6datKxyXqVfH3kpSUxLdFRESA47i3snzNt7KUeuL58+cIDQ3F999/r+6pNFhz587F2bNnsXnzZlhaWqJFixZvPGZKSgpatmypgtkRUreaxBZmBS8vL+zcuROXL19W91TeuuLiYqgiNuDq1atwdXWFt7c33NzcYGtr+8Zjurm51VgwVTV/UjeKi4vVPYW3okkVzPnz58PU1BRBQUHV9qtud+713ceK3YHff/8dH3zwAQwNDWFiYoJPPvkEpaWl+PPPP+Hl5YXmzZvDzs4OX375pdxllpSU4JNPPoGlpSXEYjHc3d1x8eJFmX7nz5/H+++/DxMTE4hEInTp0gV79uwR9Kk4BHHs2DEEBgbCzMwMurq6eP78eZXrfPfuXYwbNw7m5ubQ0dFBhw4dsHz5cpSXlwP4d1fo5s2b+Pnnn8FxHDiOQ0ZGhtzxunTpgj59+si0l5WVwdraGj4+PlW+p9XN/+bNm5g0aRLatm0LXV1dWFtbY/jw4bhy5UqV61aTa9euYeDAgdDV1YWZmRlmzJiBn376SWbXDwBOnDiB/v37w8DAALq6uujVqxdOnjwp6FPxmbh27RrGjBkDQ0NDWFhYIDAwEE+fPhX0ZYxh7dq1cHFxgVgshrGxMXx9fXH79m1Bv4sXL2LYsGH878fKygpDhw7FvXv3ql23isNNv/zyC9zc3CAWi2FtbY2wsDCUlZUJ+r548QKLFy+GVCqFjo4OzMzMMGnSJGRnZwv62dnZYdiwYYiLi0OXLl0gEokQGRlZ7TyOHj2K/v37w9DQELq6uujQoQO++OILQR9FPtvq1qQKZvPmzREaGoqEhAScOnVKpWOPHj0azs7O2L9/P6ZOnYqVK1di7ty58Pb2xtChQ3HgwAH069cPQUFBiIuLk3n9ggULcPv2bWzatAmbNm3CgwcP4OHhIfjDSUxMRK9evfDkyROsX78ehw4dgouLCz788EO5xT0wMBBaWlr4/vvvsW/fPmhpacmde3Z2Nnr27Iljx45h0aJF+PHHH+Hp6YlPP/0UM2fOBAC88847SElJgaWlJXr16oWUlBSkpKRUuUs+adIk/Prrr7hx44ag/dixY3jw4AEmTZpU43sqb/4PHjyAqakpli5diqNHj+Kbb76BpqYmunfvjj///LPGMV/38OFDuLu7488//8S6deuwfft2FBQU8Otd2Y4dOzBw4EAYGBhg27Zt2LNnD0xMTDBo0CCZogkAo0aNQrt27bB//34EBwdj586dmDt3rqDPxx9/jDlz5sDT0xMHDx7E2rVrce3aNfTs2ROPHj0CABQWFmLAgAF49OgRvvnmGxw/fhyrVq2CjY0NCgoKalzHrKws+Pn5YezYsTh06BB8fX2xePFizJ49m+9TXl6OESNGYOnSpfD398dPP/2EpUuX4vjx4/Dw8JDZgkxLS8Nnn32GWbNm4ejRoxg1alSVy//uu+8wZMgQlJeXY/369Th8+DBmzZolKPbKfrbVhjUBW7ZsYQBYamoqe/78OWvTpg179913WXl5OWOMMXd3d9axY0e+/507dxgAtmXLFpmxALCFCxfyjxcuXMgAsOXLlwv6ubi4MAAsLi6Ob3v58iUzMzNjPj4+fFtiYiIDwN555x1+PowxlpGRwbS0tNiUKVP4NqlUyrp06cJevnwpWNawYcNYixYtWFlZmWB9J0yYoND7ExwczACws2fPCtqnTZvGOI5jf/75J99ma2vLhg4dWuOYOTk5TFtbmy1YsEDQPnr0aGZhYSFYh9ffU2XmX1payl68eMHatm3L5s6dy7dX9zus7LPPPmMcx7Fr164J2gcNGsQAsMTERMYYY4WFhczExIQNHz5c0K+srIw5OzszV1dXvq3iM/Hll18K+k6fPp2JRCL+95ySkiL3s/P3338zsVjM5s+fzxhj7Pz58wwAO3jwYI3vx+vc3d0ZAHbo0CFB+9SpU5mGhgbLzMxkjDG2a9cuBoDt379f0C81NZUBYGvXruXbbG1tWbNmzQSfi6oUFBQwAwMD1rt3b8Hn+3WKfrYr/l4qfi+M/ft+vw1NagsTALS1tbF48WKcP39epZv7w4YNEzzu0KEDOI7D4MGD+TZNTU04ODjIPVPv7+8vONNna2uLnj17IjExEQBw8+ZNpKenY+zYsQCA0tJS/mfIkCF4+PChzBZWdf/rV3bq1Ck4OjrC1dVV0B4QEADGWK22xk1NTTF8+HBs27aN363Pz8/HoUOHMGHCBGhq1ny+Ud78S0tLsWTJEjg6OkJbWxuamprQ1tbGjRs3cP36daXnmZycjE6dOsHR0VHQPmbMGMHj3377DXl5eZg4caLgvS8vL4eXlxdSU1NRWFgoeM37778veOzk5ISSkhI8fvwYABAfHw+O4zBu3DjBmJaWlnB2duYPBzg4OMDY2BhBQUFYv349/vjjD6XWsXnz5jJz8ff3R3l5OU6fPs3PxcjICMOHDxfMxcXFBZaWljKHJpycnNCuXbsal/3bb7/hn3/+wfTp06s8k12bz7a6NLmCCQB+fn545513EBISgpcvX6pkTBMTE8FjbW1t6OrqQiQSybSXlJTIvN7S0lJuW25uLgDwu2effvoptLS0BD/Tp08HAOTk5Aher+gZ7NzcXLl9rays+OdrIzAwEPfv38fx48cBALt27cLz588REBCg0OvlzemTTz5BWFgYvL29cfjwYZw9exapqalwdnau1YmH3NxcWFhYyLS/3lbx/vv6+sq8/8uWLQNjDHl5eYLXmJqaCh7r6OgA+PcEyaNHj8AYg4WFhcyYZ86c4X+fhoaGSE5OhouLCxYsWICOHTvCysoKCxcuVOjzK2/9Kj5vlT9fT548gba2tsxcsrKyav3Zqjj+Wd1Jvdp8ttWlSV1WVIHjOCxbtgwDBgzAxo0bZZ6vKHKvnySpbeFQRFZWlty2ij86iUQCAPj8888FJ0wqa9++veCxotemmZqa4uHDhzLtDx48ECxbWYMGDYKVlRW2bNmCQYMGYcuWLejevbvM1lxV5M1/x44dmDBhApYsWSJoz8nJgZGRkdJzNDU15f9gK3v991HxHqxZswZubm5yx5JXmKojkUjAcRx++eUXvphWVrmtc+fO2L17Nxhj+P3337F161ZERUVBLBYjODi42uVUt36VP1+mpqY4evSo3DGaN28ueKzoZ8vMzAwAqj05VZvPtro0yYIJAJ6enhgwYACioqLQqlUrwXMWFhYQiUT4/fffBe2HDh2qs/ns2rULn3zyCf9BzMzMxG+//YYJEyYAePWBadu2LS5fvixTLN5U//798cUXXyAtLQ3vvPMO3759+3ZwHIe+ffvWatxmzZph/PjxWLVqFX755RecP38eGzZseKO5chwnU1x++ukn3L9/Hw4ODkqP5+7ujpiYGPzxxx+CQr57925Bv169esHIyAh//PGH3BNCtTFs2DAsXboU9+/fx+jRoxV6DcdxcHZ2xsqVK7F161akpaXV+JqCggL8+OOPgt3ynTt3QkNDA++99x4/l927d6OsrAzdu3ev3QrJ0bNnTxgaGmL9+vXw8/OTW2jr8rOtak22YALAsmXL0LVrVzx+/BgdO3bk2yuOK23evBn29vZwdnbGuXPnsHPnzjqby+PHjzFy5EhMnToVT58+xcKFCyESifD555/zfTZs2IDBgwdj0KBBCAgIgLW1NfLy8nD9+nWkpaVh7969tVr23LlzsX37dgwdOhRRUVGwtbXFTz/9hLVr12LatGkKHauqSmBgIJYtWwZ/f3+IxWJ8+OGHtR4LePWHvXXrVkilUjg5OeHChQv46quvan3h+5w5c7B582YMHjwYUVFRsLCwwM6dO5Geng4A0NB4ddRKX18fa9aswcSJE5GXlwdfX1+Ym5sjOzsbly9fRnZ2NtatW6fUsnv16oWPPvoIkyZNwvnz5/Hee+9BT08PDx8+xK+//orOnTtj2rRpiI+Px9q1a+Ht7Y02bdqAMYa4uDg8efIEAwYMqHE5pqammDZtGu7evYt27drhyJEj+PbbbzFt2jTY2NgAeHWYKjY2FkOGDMHs2bPh6uoKLS0t3Lt3D4mJiRgxYgRGjhyp5Lv76n1bvnw5pkyZAk9PT0ydOhUWFha4efMmLl++jP/+978A6u6zrXJv5dSSmlU+S/46f39/BkBwlpwxxp4+fcqmTJnCLCwsmJ6eHhs+fDjLyMio8ix5dna24PUTJ05kenp6Mst7/Yx8xVm/77//ns2aNYuZmZkxHR0d1qdPH3b+/HmZ11++fJmNHj2amZubMy0tLWZpacn69evH1q9fr9D6ViUzM5P5+/szU1NTpqWlxdq3b8+++uor/uxkBUXPklfWs2dPBoCNHTtW7vOvv6fVzT8/P59NnjyZmZubM11dXda7d2/2yy+/MHd3d+bu7s73U/QsOWOMXb16lXl6ejKRSMRMTEzY5MmT2bZt2xgAdvnyZUHf5ORkNnToUGZiYsK0tLSYtbU1Gzp0KNu7dy/fp6rPRMV63blzR9C+efNm1r17d6anp8fEYjGzt7dnEyZM4H//6enpbMyYMcze3p6JxWJmaGjIXF1d2datW2tct4rPW1JSEnv33XeZjo4Oa9GiBVuwYIHMGemXL1+ymJgY5uzszEQiEdPX12dSqZR9/PHH7MaNG3y/2nwGjhw5wtzd3Zmenh7T1dVljo6ObNmyZYI+iny21X2WnGOMvj5ByOs++ugj7Nq1C7m5udDW1lb3dGrNw8MDOTk5uHr1qrqn0ig06V1yQgAgKioKVlZWaNOmDZ49e4b4+Hhs2rQJoaGhDbpYEtWjgkmaPC0tLXz11Ve4d+8eSktL0bZtW6xYsULwTRhCAIB2yQkhREFN8sJ1QgipDSqYhBCiICqYhBCiICqYhBCiICqYhBCiICqYhBCiILUXzEuXLmHo0KGwsbGBWCyGiYkJevTogR07dsj0TUtLg6enJ/T19WFkZAQfHx+ZKP8Ka9as4aP2W7dujcjISJVFuRFCmia1X7j+5MkTtGrVCmPGjIG1tTUKCwsRGxuL8ePHIyMjA6GhoQCA9PR0eHh4wMXFBXv27EFJSQnCw8PRp08fXLp0iY+RAoDo6GiEhYUhODgYAwcORGpqKkJDQ3H//n25cW6vKyoqQnp6OqRSKXR1dets3QkhtXP37l2VZmRKJBI+iKRab+Ub67XQvXt31qpVK/7xBx98wCQSCXv69CnfVnEbh4oof8Ze3RpBJBKxjz76SDBedHS03FsRyHPhwgUGgF24cEEFa0IIUaXMzEymq6vLAKjsR1dXl79dR3XUvoVZFYlEwkf5l5aWIj4+HhMmTICBgQHfx9bWFn379sWBAwewbNkyAK/uTldSUiJzk61JkyYhJCQEBw8eVDjAlhBS/+Tk5KCoqAg7duxAhw4d3ni869evY9y4ccjJyalxK7PeFMzy8nKUl5cjPz8fe/fuRUJCAp+Vd+vWLRQXF8PJyUnmdU5OTjh+/DhKSkogEon4VJbOnTsL+rVo0QISiYRSWwhpJDp06CAIvH4b6k3BnD59Op/Gra2tja+//hoff/wxgH9vDfH6fXMq2hhjyM/PR4sWLZCbmwsdHR3o6enJ7VuXt5kghDRu9aZgLliwAFOmTMHjx49x+PBhzJw5E4WFhfj000/5PtXdR6Tyc4r2qywnJ4c/iHznzh1lp08IaQLqTcG0sbHhjx8MGTIEwKubIk2cOJG/UZO8rcO8vDxwHMffAMvU1BQlJSUoKiqSOcOdl5eHrl27yl3+/v378e233wJ4dZacEEJep/brMKvi6uqK0tJS3L59G/b29hCLxbhy5YpMvytXrsDBwYG/02PFscvX+1bcKrRTp05ylzdq1Cjs2LEDO3bswKJFi1S8NoSQxqDeFszExERoaGigTZs20NTUxPDhwxEXF4eCggK+z927d5GYmCi4NaeXlxdEIhG2bt0qGG/r1q3gOA7e3t5ylyeRSCCVSiGVStG6deu6WCVCSAOn9l3yjz76CAYGBnB1dYWFhQVycnKwd+9e/PDDD/jss8/4C9IjIyPRrVs3DBs2DMHBwfyF6xKJBPPmzePHMzExQWhoKMLCwmBiYsJfuB4REYEpU6bQJUWEkFpTe8Hs0aMHtmzZgm3btuHJkyfQ19eHs7Mzvv/+e4wbN47vJ5VKkZSUhKCgIPj6+kJTUxP9+vVDTEyM4Fs+ABASEoLmzZvjm2++QUxMDCwtLREcHIyQkJC3vXqEkEZE7QVz0qRJMheZV6Vr1644ceKEQn1nzZqFWbNmvcnUCCFEoN4ewySEkPqGCiYhhCiICiYhhCiICiYhhChI7QXz1KlTCAwMhFQqhZ6eHqytrTFixAhcuHBB0C8gIAAcx8n8SKVSueNSgDAhRNXUfpZ83bp1yM3NxezZs+Ho6Ijs7GwsX74cbm5uSEhIQL9+/fi+YrEYp06dErxeLBbLjPmmAcKEECKP2gvmN998A3Nzc0Gbl5cXHBwcsGTJEkHB1NDQgJubW7Xj5ebmYvHixZg6dSqWLFkCAPDw8MDLly8RGhqKOXPm0MXrhCjI39//rS5v586db3V5ylL7LvnrxRIA9PX14ejoiL///lvp8aoLEGaM4eDBg7WdKiGkiVN7wZTn6dOnSEtLQ8eOHQXtxcXFsLS0RLNmzdCyZUvMnDkTeXl5gj4UIEwIqStq3yWXZ8aMGSgsLBR8ldHZ2RnOzs582lBycjJWrlyJkydPIjU1Ffr6+gBAAcKEkDpT7wpmWFgYYmNjsWbNGkF25dy5cwX9BgwYgC5dusDX1xfffvut4HkKECaE1IV6VTAjIyOxePFiREdHY+bMmTX2HzlyJPT09HDmzBm+jQKECSF1pd4UzMjISERERCAiIgILFixQ+HWMMWho/HsotnKAcPfu3fl2RQKE3d3dAQDXrl2Dr69vbVaDENKI1YuCuWjRIkRERCA0NBQLFy5U+HX79u1DUVGR4FKjygHClQumIgHCEokEAG1hEkLkU3vBXL58OcLDw+Hl5YWhQ4cKdq8BwM3NDZmZmfD394efnx8cHBzAcRySk5OxatUqdOzYEVOmTOH7U4AwIaSuqL1gHj58GMCr6yePHj0q8zxjDAYGBrCwsMCKFSvw6NEjlJWVwdbWFrNmzcKCBQtkzohTgDAhpC6ovWAmJSXV2MfY2BhxcXFKjUsBwoQQVauXF64TQkh9RAWTEEIURAWTEEIURAWTEEIUpPaCqWiAMACkpaXB09MT+vr6MDIygo+PD27fvi13XAoQJoSomtoL5rp165CRkYHZs2fjyJEjWL16NR4/fgw3NzdBWHB6ejo8PDzw4sUL7NmzB5s3b8Zff/2FPn36IDs7WzBmdHQ0Zs+eDR8fHyQkJGD69OlYsmQJZsyY8bZXjxDSiKj9siJFA4TDw8Oho6OD+Ph4GBgYAHh1n/K2bdsiJiYGy5YtA0ABwoSQuqP2LUxFAoRLS0sRHx+PUaNG8cUSAGxtbdG3b18cOHCAb6MAYUJIXVF7wZTn9QDhW7duobi4GE5OTjJ9nZyccPPmTZSUlACgAGFCSN2plwXz9QDhitBfExMTmb4mJiZgjCE/P5/vSwHChJC6oPZjmK+rKkAYUDwYmAKECSF1oV4VzKoChE1NTQFA7tZhXl4eOI6DkZER35cChAkhdaHeFMzqAoTt7e0hFotx5coVmddduXIFDg4OEIlEAChAmBBSd+rFMcyaAoQ1NTUxfPhwxMXFoaCggG+/e/cuEhMT4ePjw7dVDhCuTJEAYalUCqlUitatW6tkvQgh9cubJpipfQtTkQBh4NUWaLdu3TBs2DAEBwejpKQE4eHhkEgkmDdvHt+fAoQJIceOHcOaNWuQk5MDxhjffv/+fZw7dw4AZGqNItReMBUJEAYAqVSKpKQkBAUFwdfXF5qamujXrx9iYmJgZmYmeA0FCBPStIWFhWH69Omws7Pj2ziOQ2BgIP8ll9pQe8FUJEC4QteuXXHixAmF+lKAMCFN17NnzzBx4kSZ9mbNmvHnKmqjXhzDJIQQVQoICJDbHhgY+Ebjqn0LkxBCVO3rr7/G6tWr5T4XFBRU63GpYBJCGp3Kh/oYY3jw4AHWrFmDoUOHvtG4VDAJIY2Ovb294LGDgwM6deoEb2/vKnfXFVEvjmEWFBRg/vz5GDhwIMzMzMBxHCIiImT6BQQEgOM4mR+pVCp3XAoRJoRU0NPTQ3Z2NkpLS2s9Rr3YwszNzcXGjRvh7OwMb29vbNq0qcq+YrFYECxc0fa66OhohIWFITg4mL8WMzQ0FPfv38fGjRtVvg6EkPpNR0cH169ff6Mx6kXBtLW1RX5+PjiOQ05OTrUFU0NDg7+YvSoUIkxI09ajRw/BBeuVVVyw3q9fP5mNr5rUi4JZXbpQbVQXIhwSEoKDBw9SwSSkEVu6dGmNfaKiopQet14UTGUUFxfD0tIS2dnZaNGiBby9vREVFSXIyqQQYUKaNkUuTu/du7fS4zaoguns7AxnZ2c+cSg5ORkrV67EyZMnkZqaCn19fQAUIkxIU3f27FksWrQIenp6+OKLL9CmTRtcvXoV5eXlcu/coKgGVTDnzp0reDxgwAB06dIFvr6++PbbbwXPKxsiTAHChDQe8+fPx/vvv4+CggJERkZi27ZteP78OaKionDo0KFaj9ugCqY8I0eOhJ6eniB5pDYhwhQgTEjjkZWVhXnz5qGoqAi9evUCALzzzju4cePGG43b4Asm8OpKfg2Nfy8prU2IMAUIE9J4dOrUCbdu3YK9vT1/g8QHDx7wQeO11eAL5r59+1BUVCS41KhyiHDlglldiLBEIoFEIgFAW5iENHQ9e/bE5MmTMX78eJSWlmLTpk2IjY2Fl5fXG41bbwrmzz//jMLCQj5R/Y8//sC+ffsAAEOGDEF2djb8/f3h5+cHBwcHcByH5ORkrFq1Ch07dsSUKVP4sShEmJCmLTk5GQYGBjh06BAcHR3xv//9D35+fo0nrWjatGnIzMzkH+/duxd79+4F8OokjKGhISwsLLBixQo8evQIZWVlsLW1xaxZs7BgwQKZM+IUIkxI0/Xjjz/Wybj1pmBmZGTU2CcuLk6pMSlEmBCiSvWmYBJCiKq0atWqyq9GAq8uLWSM4d69e0qNSwWTENLoKHPrG2VQwSSENDqv52GqSr3IwySEEFVKSUmpcpf80qVL+O2332o1br0omIoGCANAWloaPD09oa+vDyMjI/j4+OD27dty+1KAMCFN0wcffIAXL17Ife727dtYuXJlrcatF7vkigYIp6enw8PDAy4uLtizZw9KSkoQHh6OPn364NKlS4L7k1OAMCFNF8dxOHPmDLS1tWWee/nyZa0Ty+pFwVQ0QDg8PBw6OjqIj4+HgYEBgFf3Km/bti1iYmL4G7RTgDAhZM6cOVWG8Mi7S4Mi6kXBVCRAuLS0FPHx8ZgwYQJfLIFXxbZv3744cOAAXzApQJiQpo0xhjNnzkBHR0el49aLY5iKuHXrFoqLi+Vm2Tk5OeHmzZv8l+wpQJiQpq1Vq1Yqv5MD0IAKZkXob+Vk9QomJiZgjCE/P5/vSwHChDRdKSkpco9fvql6sUuuDEWDgSlAmBCiag2mYJqamgKA3K3DvLw8cBwHIyMjvi8FCBNCVK3BFEx7e3uIxWJcuXJF5rkrV67AwcGBDwelAGFCSF1oMMcwNTU1MXz4cMTFxfGZmQBw9+5dJCYmwsfHh2+rHCBcWU0BwlKpFFKpFK1bt66r1SCENGAq2cIsKSlBRkYG2rZti2bNmtVqjJoChHV1dREZGYlu3bph2LBhCA4O5i9cl0gkmDdvHj8WBQgTQuqC0gVzzZo1ePLkCcLCwgAAFy5cgJeXF/Ly8mBnZ4ekpCS0atVK6YnUFCBsZ2cHqVSKpKQkBAUFwdfXF5qamujXrx9iYmIE3/IBKECYEKJ6Su+Sb9q0iT+5AgBBQUEwMTHBypUrwRjD4sWLazWRjIwMMMbk/tjZ2fH9unbtihMnTqCwsBBPnz7FgQMHqkwmmTVrFv788088f/4cmZmZiIiIgJaWVq3mRwghSm9h3r17F1KpFMCr0IzTp09j9+7d8PHxgbGxMcLDw1U+SUIIqQ+U3sJ8/vw5v5WWkpKC8vJyeHp6AgDs7OyQlZWl2hkSQkg9oXTBtLGxwS+//AIAOHToEFxcXPjvdmdnZwu+500IIY2J0rvk48aNQ2RkJA4ePIjLly8jJiaGf+78+fNo166dSidICCH1hdJbmCEhIVi0aBGsra0REREhuCvj1atXMWrUKJVOsLKkpCRwHCf358yZM4K+ygQNE0KIIpTewuQ4DsHBwXKfq6t7Ab9uyZIl6Nu3r6Ct8rd3lAkaJoQQRTWYr0ZW1rZtW7i5uVX5vKJBw4QQogyFCmZgYKDCA3Ich++++67WE3pTygQNE9JQ+Pv7v7Vl7dy5860tq6FRqGCeOnVK4TDOugjtfN2MGTPg5+cHXV1d9OjRA2FhYejduzeAmoOGjx8/jpKSEj6ogxBCFKVQwczIyKjjaSjG0NAQs2fPhoeHB0xNTXHz5k189dVX8PDwwE8//YRBgwYpHDTcokWLtz19QkgD16COYXbp0gVdunThH/fp0wcjR45E586dMX/+fAwaNIh/jgKECSGq1qAKpjxGRkYYNmwY1q9fj+LiYqWChiujAGFCSE0UKpitW7dW6hjmrVu33mhSymKM8ctWJmi4MgoQJoTURKGC6e7u/lZO5tRGfn4+4uPj4eLiwhfCiqDhL7/8Es2bNwfwb9Dw3Llz5Y4jkUggkUgA0BYmIUQ+hQrm68nl6uLv7w8bGxu8++67kEgkuHHjBpYvX45Hjx4J5qho0DAhhCijwdyiAnh1WVBCQgKmTJkCT09PhISEwNHREb/99hufmASADxrW0tKCr68vAgIC4ODggNOnT9O3fAghtVarkz7Z2dlYsWIFkpKSkJOTg4MHD6Jjx47YsGEDXF1dBWeyVSk4OLjKr2W+riJomBBCVEXpLcw7d+7A2dkZX3/9NTiOw+3bt/H8+XMAwO+//46vv/5a5ZMkhJD6QOmCOX/+fBgZGeHGjRs4ffo0f4YaAHr37o3//e9/Kp0gIYTUF0rvkp88eRLr1q2DlZUVysrKBM+1aNECDx48UNnkCKlP3ub3uQH6Tnd9pPQWZklJidyvHQJAYWEhNDQa1HkkQghRmNLVrX379lWeTDl9+rQgl1Kdnj17hjlz5sDKygoikQguLi7YvXu3uqdFCGnAlN4lnzp1Kj755BNYWVlh7NixAIAXL15g3759WLt2Lf773/+qfJK14ePjg9TUVCxduhTt2rXDzp07MWbMGJSXl7/1XStCSOOgdMGcPn06Ll26hLlz5/IXgffu3RuMMUydOhUTJ05U+SSVdeTIERw/fpwvkgDQt29fZGZm4rPPPsOHH36IZs2aqXmWhJCGplbXYW7cuBGBgYH46aef8OjRI0gkEgwbNgw9e/ZU9fxq5cCBA9DX18cHH3wgaJ80aRL8/f1x9uzZejNXohw68ULUqdZpRW5ubtXeJkKdrl69ig4dOkBTU7h6FaHCV69epYL5higBnDRFSp/0OXPmDPbs2SP3uT179uDs2bNvPKk3lZubW2WAcMXzhBCiLKW3MBcsWIBevXph9OjRMs/98ccf+Pbbb3H8+HGVTO5NNIUAYdo9JeTtUrpg/v7771Um/nTv3h1r165940m9KVNT0yoDhAH5t69oiAHC6ixg6lp2U1xndS6b/pMUUrpgFhYWyhwbrKChoYGCgoI3ntSb6ty5M3bt2oXS0lLBXCtCheVdK0oBwoSQmih9DLN169ZITEyU+1xiYiJsbW3feFJvauTIkXj27Bn2798vaN+2bRusrKzQvXt3mddIJBJIpVJIpVK0bt36bU2VENKAKL2F6efnh+joaLRv3x6TJk3i27du3YpVq1bh888/V+kEa2Pw4MEYMGAApk2bhn/++QcODg7YtWsXjh49ih07dtA1mISQWlG6YAYHByMpKQmTJ0/GzJkzYWVlhQcPHqCkpAQeHh71omACQFxcHEJCQhAeHo68vDxIpVLs2rULfn5+Nb62uLgYAHD9+vW6niYhREkVf5eXLl1S6XiK4FjlfDYFlZWVYefOnTh69Ciys7NhZmaGwYMHY8yYMY1i6y02Nhbjxo1T9zQIIW+Jrq4url+/Dhsbm2r71apgNnY5OTlISEiAnZ0dxGKxuqdDCHnNw4cPYWRkpLK/T4lEUmOxBKhgEkKIwii8khBCFEQFkxBCFKT2gnnp0iUMHToUNjY2EIvFMDExQY8ePbBjxw6ZvmlpafD09IS+vj6MjIzg4+OD27dvyx13zZo1kEql0NHRQevWrREZGYmXL1/W9eoQQhoxhS4r+vrrr+Hn5wdzc3OVT+DJkydo1aoVxowZA2traxQWFiI2Nhbjx49HRkYGQkNDAQDp6enw8PCAi4sL9uzZg5KSEoSHh6NPnz64dOmS4H7j0dHRCAsLQ3BwMAYOHIjU1FSEhobi/v372LhxY41zKioqQnp6OqRSKXR1dVW+zoSQN3P37l0++0EVFD3pA6YADQ0NdvbsWZl/16Xu3buzVq1a8Y8/+OADJpFI2NOnT/m2jIwMpqWlxebPn8+35eTkMJFIxD766CPBeNHR0YzjOHbt2rUal33hwgUGgF24cEEFa0IIUaXMzEymq6vLAKjsR1dXl2VmZta4bIW2MJs3b44nT55UFFhFi/YbkUgkePz4MQCgtLQU8fHxmDBhAgwMDPg+tra26Nu3Lw4cOIBly5YBAI4ePYqSkhLBt5CAV+HBISEhOHjwIBwdHd/KOhBCVC8nJwdFRUXYsWMHOnTo8MbjXb9+HePGjUNOTk6NW5kKFUwXFxfMmjULnp6eAIDly5fDwsJCbl+O47B69WolpwyUl5ejvLwc+fn52Lt3LxISEvj7A926dQvFxcV8AHBlTk5OOH78OEpKSiASiXD16lUArwI4KmvRogUkEgn/PCGkYevQoQPeeeedt7pMhQrmqlWrMGnSJGzcuBEcx2Hv3r1V9q1twZw+fTo2bNgAANDW1sbXX3+Njz/+GMC/gb9VhQIzxpCfn48WLVogNzcXOjo60NPTk9uXwoMJIbWl8BbmxYsXAbyKcDtz5gxcXV1VOpEFCxZgypQpePz4MQ4fPoyZM2eisLAQn376Kd9H0VBgZcODgYYZIEwIebuUDt/YsmUL7O3tVT4RGxsb/vjBkCFDAACff/45Jk6cCFNTUwDyby2Rl5cHjuNgZGQE4FV4cElJCYqKimTOcOfl5aFr165yl98QA4QJIW+X0gWz4ja6BQUFSElJQW5uLiQSCdzc3NC8eXOVTczV1RXr16/H7du30bVrV4jFYj4AuLIrV67AwcEBIpEIwL/HLq9cuSLIvczKykJOTo7c8GCAAoQJITWr1YXrMTExsLKywuDBgzF27Fh4eXnBysoKK1asUNnEEhMToaGhgTZt2kBTUxPDhw9HXFycINH97t27SExMhI+PD9/m5eUFkUiErVu3CsbbunUrOI6Dt7e33OVRgDAhpCZKb2Fu374d8+fPx+DBgxEQEMDnYW7btg2fffYZzMzMMH78eIXH++ijj2BgYABXV1dYWFggJycHe/fuxQ8//MCPBwCRkZHo1q0bhg0bhuDgYP7CdYlEIrjHkImJCUJDQxEWFgYTExP+wvWIiAhMmTKFLikihNSesheNuri4sLFjx8p9buzYsaxLly5Kjbd582bWp08fJpFImKamJjMyMmLu7u7s+++/l+l7/vx51r9/f6arq8sMDAyYt7c3u3nzptxxV69ezdq1a8e0tbWZjY0NW7hwIXvx4oVCc6IL1wmpv1T996nMeEpvYaanp+OLL76Q+9y4ceMwcuRIpcabNGmSzEXmVenatStOnDihUN9Zs2Zh1qxZSs2FEEKqo/QxTLFYzN+u9nV5eXkUuEsIabSULph9+vRBREQEHjx4IGjPyspCVFQU3nvvPZVNjhBC6hOld8mXLFmCnj17wsHBAf3790eLFi3w8OFDnDp1ClpaWoiLi6uLeRJCiNopvYXZsWNHpKamYsSIEUhNTcWWLVuQmpoKb29vnDt3js5CE0IarVpdh9muXTvs2rULWVlZePnyJbKyshAbG4t27dopPdapU6cQGBgIqVQKPT09WFtbY8SIEbhw4YKgX0BAADiOk/mRSqVyx6UAYUKIqim9S65q69atQ25uLmbPng1HR0dkZ2dj+fLlcHNzQ0JCAvr168f3FYvFOHXqlOD18k4yvWmAMCGEyKP2gvnNN9/IJLl7eXnBwcEBS5YsERRMDQ0NuLm5VTtebm4uFi9ejKlTp2LJkiUAAA8PD7x8+RKhoaGYM2cOHTYghNSK2u/pI++2F/r6+nB0dMTff/+t9HjVBQgzxnDw4MHaTpUQ0sSpvWDK8/TpU6SlpaFjx46C9uLiYlhaWqJZs2Zo2bIlZs6cKXNNKAUIE0Lqitp3yeWZMWMGCgsLERISwrc5OzvD2dmZTxtKTk7GypUrcfLkSaSmpkJfXx8AKECYEFJn6l3BDAsLQ2xsLNasWSPIrpw7d66g34ABA9ClSxf4+vri22+/FTxPAcKEkLpQq4IZGBiIqKgotGzZUqWTiYyMxOLFixEdHY2ZM2fW2H/kyJHQ09PDmTNn+DYKECaE1JVaHcPctm0bf+ywvLwc/fr1w40bN95oIpGRkYiIiEBERAQWLFig8OsYY9DQ+Hc1KgcIV6ZIgPCOHTuwY8cOLFq0qBZrQAhp7BTawmzdujW6d++O7t27o1u3bmCM8bu2jDEkJSUJgn2VtWjRIkRERCA0NBQLFy5U+HX79u1DUVGR4FKjygHClRPXFQkQlkgkAGgLkxAin0IFc9asWTh79ixWr16Nu3fvguM4zJgxAwMHDkTPnj0BVH/csDrLly9HeHg4vLy8MHToUMHuNQC4ubkhMzMT/v7+8PPzg4ODAziOQ3JyMlatWoWOHTtiypQpfH8KECaE1BllwzYfPnzIOI5jI0eOZB07dmQaGhpMQ0ODeXh4sHXr1rG///5bqfHc3d0ZgCp/GGMsLy+PjRw5ktnZ2TGxWMy0tbVZ27Zt2fz589mTJ0/kjksBwoQ0TvU+QPjGjRto27YtAMDS0hIAEBERAScnJzx+/BiWlpYwMDDAkiVLMGPGDJSVlSlcsJOSkmrsY2xsrHQKEgUIE0JUTaGC2b59exgbG8PV1RWurq7gOI4vihW3wF24cCHeeecdXLp0qc4mSwgh6qRQwbx06RJSU1ORmpqK+Ph4MMbQs2dPdOvWDb179wbHcSgvLwcAuLi41OV8CSFEbRQqmE5OTnBycsLkyZMBvArBWLRoER48eICEhAQwxuDh4YH+/fvj/fff5/sRQkhjUuvvkg8cOBArVqzA2bNnAQChoaEwNTUVfJ2REEIak1oVTFtbW2hrawP493KigQMHYvPmzXj48KFSYykaIAwAaWlp8PT0hL6+PoyMjODj44Pbt2/LHZcChAkhqlargnnnzh0+6ZzjOLi7u6N58+b8Y2WsW7cOGRkZmD17No4cOYLVq1fj8ePHcHNzE4QFp6enw8PDAy9evMCePXuwefNm/PXXX+jTpw+ys7MFY0ZHR2P27Nnw8fFBQkICpk+fzp/BJ4SQWlPJhUxv4NGjRzJtBQUFzMLCgvXv359v++CDD5hEImFPnz7l2zIyMpiWlhabP38+35aTk8NEIhH76KOPBGNGR0czjuPYtWvXapwTXYdJSP2lzusw1Z6HqUiAcGlpKeLj4zFq1CgYGBjw/WxtbdG3b18cOHCAb6MAYUJIXVF7wZTn9QDhW7duobi4GE5OTjJ9nZyccPPmTZSUlACgAGFCSN2plwXz9QDhitBfExMTmb4mJiZgjCE/P5/vSwHChJC60GAChAHFg4EpQJgQUhfqVcGsKkC44uuX8rYO8/LywHEcjIyM+L4UIEwIqQv1pmBWFyBsb28PsVgsEwoMvAoKdnBwgEgkAiAMEK6ch6lIgLC7uzsA4Nq1a/D19VXJehFCGo96cQyzpgBhTU1NDB8+HHFxcYKg4rt37yIxMRE+Pj58W+UA4coUCRCWSqWQSqVo3bq1StaLEFK/vGmCmdq3MBUJEAZebYF269YNw4YNQ3BwMEpKShAeHg6JRIJ58+bx/SlAmBBy7NgxrFmzBjk5OWCM8e3379/HuXPnAECm1ihC7QXz8OHDAF5dP3n06FGZ5ytWViqVIikpCUFBQfD19YWmpib69euHmJgYmJmZCV4TEhKC5s2b45tvvkFMTAwsLS0RHBxM33MnpIkICwvD9OnTYWdnx7dxHIfAwEAsW7as1uOqvWAqEiBcoWvXrjhx4oRCfSlAmJCm69mzZ5g4caJMe7NmzfhzFbVRL45hEkKIKgUEBMhtDwwMfKNx1b6FSQghqvb1119j9erVcp8LCgqq9bhUMAkhjU7lQ32MMTx48ABr1qzB0KFD32hcKpiEkEbH3t5e8NjBwQGdOnWCt7d3lbvriqgXxzALCgowf/58DBw4EGZmZuA4DhERETL9AgICwHGczE9FNufrKESYEFJBT08P2dnZKC0trfUY9WILMzc3Fxs3boSzszO8vb2xadOmKvuKxWJBsHBF2+uio6MRFhaG4OBg/lrM0NBQ3L9/Hxs3blT5OhBC6jcdHR1cv379jcaoFwXT1tYW+fn54DgOOTk51RZMDQ0N/mL2quTm5mLx4sWYOnUqlixZAgDw8PDAy5cvERoaijlz5tAF7IQ0Yj169BBcsF5ZxQXr/fr1k9n4qkm9KJjK3taiJtWFCIeEhODgwYNUMAlpxJYuXVpjn6ioKKXHrRcFUxnFxcWwtLREdnY2WrRoAW9vb0RFRQmyMilEmJCmTZGL03v37q30uA2qYDo7O8PZ2ZlPHEpOTsbKlStx8uRJpKamQl9fHwCFCBOiKv7+/m91eTt37lTJOGfPnsWiRYugp6eHL774Am3atMHVq1dRXl4u984NimpQBXPu3LmCxwMGDECXLl3g6+uLb7/9VvC8siHCFCBMSOMxf/58vP/++ygoKEBkZCS2bduG58+fIyoqCocOHar1uA2qYMozcuRI6OnpCZJHahMiTAHChDQeWVlZmDdvHoqKitCrVy8AwDvvvIMbN2680bgNvmACr67k19D495LS2oQIU4AwIY1Hp06dcOvWLdjb2/M3SHzw4AEfNF5bDb5g7tu3D0VFRYJLjSqHCFcumNWFCEskEkgkEgC0hUlIQ9ezZ09MnjwZ48ePR2lpKTZt2oTY2Fh4eXm90bj1pmD+/PPPKCws5BPV//jjD+zbtw8AMGTIEGRnZ8Pf3x9+fn5wcHAAx3FITk7GqlWr0LFjR0yZMoUfi0KECWnakpOTYWBggEOHDsHR0RH/+9//4Ofn13jSiqZNm4bMzEz+8d69e7F3714Ar07CGBoawsLCAitWrMCjR49QVlYGW1tbzJo1CwsWLJA5I04hwoQ0XT/++GOdjFtvCmZGRkaNfeLi4pQak0KECSGqVG8KJiGEqEqrVq2q/Gok8OrSQsYY7t27p9S4VDAJIY2OMre+UQYVTEJIo/N6Hqaq1Is8TEIIUaWQkBD88MMPcp+7d+8eduzYUatx60XBVDRAGADS0tLg6ekJfX19GBkZwcfHB7dv35bblwKECWmafv75Z/Tt21fQdvLkSQBA8+bNsW7dulqNWy8KZkWA8PPnz+VeVF4hPT0dHh4eePHiBfbs2YPNmzfjr7/+Qp8+fZCdnS3oGx0djdmzZ8PHxwcJCQmYPn06lixZghkzZtTx2hBC1O3Zs2cwNzcXtM2cORMAYGhoiLy8vFqNWy+OYSoaIBweHg4dHR3Ex8fDwMAAwKt7lbdt2xYxMTH8DdopQJiQps3IyAiZmZmwtbUF8OprkQUFBSgoKECzZs1q/RXJerGFWXFvnuqUlpYiPj4eo0aN4osl8KrY9u3bFwcOHODbqgsQZozh4MGDKp0/IaR+8fT0xOzZs/G///0P586dQ2xsLHR1dbFgwQLMnz9fZnddUfWiYCri1q1bKC4ulptl5+TkhJs3b/JfsqcAYUKatqCgIFhYWGDq1KlYuHAhfH19sWnTJuTl5cHIyAiLFi2q1bj1YpdcERWhv5WT1SuYmJiAMYb8/Hy0aNGCAoQJaeIMDQ2xYcMGQVvr1q3x3nvvvdG4DaZgVlA0GJgChAkhqtZgCqapqSkAyN06zMvLA8dxMDIy4vtSgDAhRNUaTMG0t7eHWCzGlStXZJ67cuUKHBwc+DNfFCBMCKkLDeakj6amJoYPH464uDg+MxMA7t69i8TERPj4+PBtlQOEK6spQFgqlUIqlaJ169Z1tRqEkAas3mxh1hQgrKuri8jISHTr1g3Dhg1DcHAwSkpKEB4eDolEgnnz5vFjUYAwIaQu1JuCWVOAsJ2dHaRSKZKSkhAUFARfX19oamqiX79+iImJgZmZmWA8ChAmhKhavSmYigQIA6++2XPixAmF+lKAMCFElRrMMUxCCFE3KpiEEKIgKpiEEKIgKpiEEKKgenPSRxFJSUlVpoykpKTAzc2Nf5yWlob58+fjzJkzgrPpbdq0eVvTJURl/P3939qydu7c+daW1dA0qIJZYcmSJTKFs/K3dyqChl1cXLBnzx7+es0+ffrg0qVLMpcgEUKIIhpkwWzbtq1ga/J1igYNE0KIMhrdMUxlgoYJIUQZDbJgzpgxA5qamjAwMMCgQYPw66+/8s8pEzRMCCHKaFAF09DQELNnz8aGDRuQmJiI1atX4++//4aHhwcSEhIAKB40TAghympQxzC7dOmCLl268I/79OmDkSNHonPnzpg/fz4GDRrEP0cBwoQQVWtQBVMeIyMjDBs2DOvXr0dxcbFSQcOVUYAwIaQmDb5gAgBjDMCrLUdlgoYrowBhQkhNGtQxTHny8/MRHx8PFxcXiEQipYKGK6MAYUJITRrUFqa/vz9sbGzw7rvvQiKR4MaNG1i+fDkePXokSFdXNGiYEEKU0aC2MJ2cnJCQkIApU6bA09MTISEhcHR0xG+//QZPT0++X0XQsJaWFnx9fREQEAAHBwecPn2avuVDCKm1BrWFGRwcjODgYIX6KhM0TAghimhQW5iEEKJOVDAJIURBVDAJIURBVDAJIURBjbZgPnv2DHPmzIGVlRVEIhFcXFywe/dudU+LENKANaiz5Mrw8fFBamoqli5dinbt2mHnzp0YM2YMysvL32p6NSGk8WiUBfPIkSM4fvw4XyQBoG/fvsjMzMRnn32GDz/8EM2aNVPzLAkhDU2jLJgHDhyAvr4+PvjgA0H7pEmT4O/vj7Nnz6Jnz55qmh1pqN72ngndW6f+aZQF8+rVq+jQoQM0NYWrVxEqfPXqVSqYDRQVLaJOjfKkT25ubpUBwhXPE0KIshrlFibw9gKE1bnF01SWTVt5pL5olAXT1NS0ygBhQP7tK2obIKzOP+amuOymuM7qXDb9ZyXUKAtm586dsWvXLpSWlgqOY1aECle+h3kFChAmhNSkUR7DHDlyJJ49e4b9+/cL2rdt2wYrKyt0795d5jUUIEwIqUmj3MIcPHgwBgwYgGnTpuGff/6Bg4MDdu3ahaNHj2LHjh10DSYhpFYaZcEEgLi4OISEhCA8PBx5eXmQSqXYtWsX/Pz8anxtcXExAOD69et1PU1CiJIq/i4vXbqk0vEUwbGKO4gRXmxsLMaNG6fuaRBC3hJdXV1cv34dNjY21fajgilHTk4OEhISYGdnB7FYrPLx79y5g7CwMCxatKjJHC9tiusMNM31fhvr/PDhQxgZGans71MikdRYLIFGvEv+JiQSCcaOHVtn4+vq6kJXVxcdO3aEVCqts+XUJ01xnYGmud6NeZ0b5VlyQgipC1Qw1UAikWDq1KmQSCTqnspb0xTXGWia692Y15mOYRJCiIJoC5MQQhREBfMtOXXqFAIDAyGVSqGnpwdra2uMGDECFy5cUPfU3qpNmzaB4zjo6+ureyp17tdff8WQIUNgbGwMsViMtm3bYtGiReqeVp25ePEivL29YWVlBV1dXUilUkRFRSmVzVDf0Vnyt2TdunXIzc3F7Nmz4ejoiOzsbCxfvhxubm5ISEhAv3791D3FOnf//n18+umnsLKywtOnT9U9nTq1c+dOjB8/HqNHj8b27duhr6+PW7du4cGDB+qeWp34448/0LNnT7Rv3x6rVq2CRCLB6dOnERUVhQsXLuDQoUPqnqJK0DHMt+Tx48cwNzcXtD179gwODg7o1KkTTpw4oaaZvT3Dhw8Hx3EwMTHBvn378OzZM3VPqU7cv38f7du3x4QJE7B27Vp1T+etCA0NRXR0NG7evAl7e3u+/eOPP8bGjRuRl5cHY2NjNc5QNWiX/C15vVgCgL6+PhwdHfH333+rYUZv144dO5CcnNwkCsimTZtQWFiIoKAgdU/lrdHS0gIAGBoaCtqNjIygoaEBbW1tdUxL5ahgqtHTp0+RlpaGjh07qnsqderx48eYM2cOli5dipYtW6p7OnXu9OnTMDExQXp6OlxcXKCpqQlzc3P85z//wT///KPu6dWJiRMnwsjICNOmTcPt27dRUFCA+Ph4bNiwATNmzICenp66p6gajKjN2LFjmaamJjt//ry6p1KnRo0axXr27MnKy8sZY4xNnDiR6enpqXlWdad9+/ZMJBKx5s2bsyVLlrDExET25ZdfMrFYzHr16sW/D43N9evXmVQqZQD4n1mzZjWq9aWTPmoSFhaG2NhYrFmzBl27dlX3dOrM/v37cfjwYVy8eLHa24Y0JuXl5SgpKcHChQsRHBwMAPDw8IC2tjbmzJmDkydPwtPTU82zVK2MjAwMHz4cFhYW2LdvH8zMzHD27FksXrwYz549w3fffafuKaqGuit2UxQREcEAsOjoaHVPpU4VFBQwCwsLNm/ePJafn8//jBkzhunp6bH8/Hz27NkzdU9T5dzc3BgAlpaWJmj/888/GQC2bNkyNc2s7nz44YfM3Nxc5ve5efNmBoAlJSWpaWaqRccw37LIyEhEREQgIiICCxYsUPd06lROTg4ePXqE5cuXw9jYmP/ZtWsXCgsLYWxsXKchJ+pScTvn17H/f0GKhkbj+7O7dOkSHB0dZY5VduvWDcCrW1s3BrRL/hYtWrQIERERCA0NxcKFC9U9nTpnaWmJxMREmfalS5ciOTkZP//8c6P8vvGoUaOwceNG/Pzzz+jSpQvffuTIEQCAm5ubuqZWZ6ysrHD16lU8e/ZM8KWElJQUAGg0J/voOsy3ZPny5fj000/h5eUlt1g2xj+iqgQEBDTq6zAB4P3338exY8cQGhoKNzc3nD9/HpGRkfD09MThw4fVPT2V+/HHH+Ht7Y3u3btj7ty5kEgkOHPmDL744gvY2Njg4sWLjePSIjUfEmgy3N3dBWcPX/9pShr7WXLGGCsqKmJBQUGsVatWTFNTk9nY2LDPP/+clZSUqHtqdebUqVNs4MCBzNLSkonFYtauXTs2b948lpOTo+6pqQxtYRJCiIIa39FnQgipI1QwCSFEQVQwCSFEQVQwCSFEQVQwCSFEQVQwCSFEQVQwCSFEQVQwCSFEQVQwVeDy5cvgOA5//vknAGDlypWws7NT+XIuXrwId3d3GBoaguM4rFq16o3Gy8jIAMdx2Lp1q0rmVyEpKQkcxyEpKUml4zYWdnZ2CAgIUPp1RUVFiIiIqPP3tS6X09A/GxS+oQKpqakwMjJCu3btAABnzpyBq6urypcTGBiIwsJC7N69G8bGxm9clFu0aIGUlBTBPVhI/VVUVITIyEgAr/I1G/pyGiIqmCqQmpoKV1dXPiD3zJkz+L//+z+VL+fq1auYOnUqBg8erJLxdHR0FAr9KCoqgq6urkqWSUhDRrvkKlBRMAEgKysLd+/eVWoL8+rVqxgxYgSMjY0hEong4uKCbdu28c9v3boVHMehtLQU69atA8dxVaaXv3z5Eubm5hg/frzMc0+ePIFYLMYnn3wCQP4ueUREBDiOQ1paGnx9fWFsbMxvgZ4/fx5+fn6ws7ODWCyGnZ0dxowZg8zMTIXX9XW//vorevToAZFIBGtra4SFhfH3Ls/IyBD0/eGHH9CjRw/o6elBX18fgwYNwsWLFwV9AgICoK+vj5s3b2LIkCHQ19dHq1atMG/ePDx//lzQ98WLF1i8eDGkUil0dHRgZmaGSZMmITs7W9Dv1KlT8PDwgKmpKcRiMWxsbDBq1Kga77f98uVLzJ8/H5aWltDV1UXv3r1x7tw5uX2zsrLw8ccfo2XLltDW1kbr1q0RGRmJ0tJSAK9+V2ZmZgBeZapWfAYq79rfuHED/v7+MDc3h46ODjp06IBvvvlGZllPnjzBvHnz0KZNG+jo6MDc3BxDhgxBenq6SpeTnp4OLy8v6OrqQiKR4D//+Q8KCgqqfc/qPXWnfzRUtra21aYPVfxMnDix2nHS09NZ8+bNmb29Pdu+fTv76aef2JgxYwTJ3I8fP2YpKSkMAPP19WUpKSksJSWlyjHnzp3LxGIxe/r0qaB97dq1DAD7/fffGWOM3blzhwFgW7Zs4fssXLiQAWC2trYsKCiIHT9+nB08eJAxxtjevXtZeHg4O3DgAEtOTma7d+9m7u7uzMzMjGVnZ/NjJCYmMgAsMTGx2nW/fPkyE4lEzMnJie3evZv9+OOPbMiQIczOzo4BYHfu3OH7RkdHM47jWGBgIIuPj2dxcXGsR48eTE9Pj127do3vN3HiRKatrc06dOjAYmJi2IkTJ1h4eDjjOI5FRkby/crKypiXlxfT09NjkZGR7Pjx42zTpk3M2tqaOTo6sqKiIv49EolEbMCAAezgwYMsKSmJxcbGsvHjx7P8/Pxq12/ixImM4zj22WefsWPHjrEVK1Ywa2trZmBgIPhcPHz4kLVq1YrZ2tqyDRs2sBMnTrBFixYxHR0dFhAQwBhjrKSkhB09epQBYJMnT+Y/Azdv3mSMMXbt2jVmaGjIOnfuzLZv386OHTvG5s2bxzQ0NFhERAS/rH/++Yd17NiR6enpsaioKJaQkMD279/PZs+ezU6dOqWy5WRlZTFzc3NmbW3NtmzZwo4cOcLGjh3LbGxsFPps1FdUMGvp2rVr7OLFi+zLL79k2traLDU1lV28eJF5e3uzXr16sYsXL7KLFy+yzMzMasfx8/NjOjo67O7du4L2wYMHM11dXfbkyRO+DQCbMWNGjXP7/fffGQC2ceNGQburqyvr2rUr/7i6ghkeHl7jckpLS9mzZ8+Ynp4eW716Nd+uaMH84IMPmJ6enqDYlpWVMUdHR0HBvHv3LtPU1GT/93//J3h9QUEBs7S0ZKNHj+bbJk6cyACwPXv2CPoOGTKEtW/fnn+8a9cuBoDt379f0C81NZUBYGvXrmWMMbZv3z4GgF26dKnG96Oy69evMwBs7ty5gvbY2FiZ/0g//vhjpq+vL/NZiYmJYQD4/xCys7MZALZw4UKZ5Q0aNIi1bNlS5j/JmTNnMpFIxPLy8hhjjEVFRTEA7Pjx41XOXRXLCQoKYhzHybxvAwYMaNAFk3bJa8nR0REuLi548OABunXrhnfffRcuLi7466+/MGDAALi4uMDFxQU2NjbVjnPq1Cn0798frVq1ErQHBASgqKiIT6xWRufOndG1a1ds2bKFb7t+/TrOnTuHwMBAhcYYNWqUTNuzZ88QFBQEBwcHaGpqQlNTE/r6+igsLMT169eVnmdycjL69esnSF3X0NDA6NGjBf0SEhJQWlqKCRMmoLS0lP8RiURwd3eXOePKcRyGDx8uaHNychIcOoiPj4eRkRGGDx8uGNPFxQWWlpb8mC4uLtDW1sZHH32Ebdu24fbt2wqtW0XS/Ou34Bg9ejQ0NYWnDuLj49G3b19YWVkJ5lJxrDo5ObnaZZWUlODkyZMYOXIkdHV1BWMMGTIEJSUlOHPmDADg559/Rrt27Wp1EzZllpOYmIiOHTvC2dlZMIa/v7/Sy61PqGDWQllZGf9BSU5ORu/evVFaWorHjx/j+vXr6NWrF0pLS1FWVlbjWLm5uWjRooVMu5WVFf98bQQGBiIlJQXp6ekAgC1btkBHRwdjxoxR6PXy5uTv74///ve/mDJlChISEnDu3DmkpqbCzMwMxcXFSs8xNzcXFhYWMu2vtz169AjAq/vDaGlpCX5++OEH5OTkCPrr6upCJBIJ2nR0dFBSUiIY88mTJ9DW1pYZMysrix/T3t4eJ06cgLm5OWbMmAF7e3vY29tj9erVNa4b8Oo2HZVpamrC1NRUZv0OHz4sM4+K+9W/vn7yllVaWoo1a9bIjDFkyBDBGNnZ2bW+XYQyy8nNzZVZd0D2/Who6Cx5LfTv31/wv/7FixexbNky/vGAAQMAQO7Wz+tMTU3x8OFDmfYHDx4AQK3veTNmzBh88skn2Lp1K6Kjo/H999/D29sbxsbGCr3+9ZNKT58+RXx8vODWsQDw/Plz5OXl1WqOpqamfDGsLCsrS/C44j3Yt28fbG1ta7Ws10kkEpiamuLo0aNyn2/evDn/7z59+qBPnz4oKyvD+fPnsWbNGsyZMwcWFhbw8/OT+/qKopiVlQVra2u+vbS0VOY/QYlEAicnJ0RHR8sdq+I/z6oYGxujWbNmGD9+PGbMmCG3T+vWrQEAZmZmuHfvXrXjqWI5pqamMr9HQPZ329BQwayFDRs2oKCgAMePH8eiRYuQmJiIZs2aISoqCk+fPsXy5csBCP/oqtK/f38cOHAADx48EPxhbN++Hbq6urW+14+xsTG8vb2xfft29OjRA1lZWQrvjsvDcRwYY9DR0RG0b9q0SaEtaXnc3d1x5MgR5OTk8EWxvLwce/fuFfQbNGgQNDU1cevWLbmHCmpj2LBh2L17N8rKytC9e3eFXtOsWTN0794dUqkUsbGxSEtLq7JgVly/GBsbK7jv/J49e/gz35XncuTIEdjb21f7H1rFe//61ryuri769u2LixcvwsnJqdp75wwePBjh4eE4deoU+vXrV2fL6du3L7788ktcvnxZsFu+c+fOKl/TEFDBrIX27dsDANauXYv+/fvzf3AXL15EcHAw3n33XYXHWrhwIX8MKzw8HCYmJoiNjcVPP/2EL7/8EoaGhrWeZ2BgIH744QfMnDkTLVu2rNVxqwoGBgZ477338NVXX0EikcDOzg7Jycn47rvvYGRkVKsxQ0JCcPjwYfTv3x8hISEQi8VYv349CgsLAfx7O1o7OztERUUhJCQEt2/fhpeXF4yNjfHo0SOcO3cOenp6/IXWivLz80NsbCyGDBmC2bNnw9XVFVpaWrh37x4SExMxYsQIjBw5EuvXr8epU6cwdOhQ2NjYoKSkBJs3bwaAat/PDh06YNy4cVi1ahW0tLTg6emJq1evIiYmBgYGBoK+UVFROH78OHr27IlZs2ahffv2KCkpQUZGBo4cOYL169ejZcuWaN68OWxtbXHo0CH0798fJiYm/O9i9erV6N27N/r06YNp06bBzs4OBQUFuHnzJg4fPoxTp04BAObMmYMffvgBI0aMQHBwMFxdXVFcXIzk5GQMGzYMffv2VdlyNm/ejKFDh2Lx4sWwsLBAbGwsf4iowVL3WaeGqqysjJmZmbENGzYwxhhLS0tjAGo8Ky7PlStX2PDhw5mhoSHT1tZmzs7OgjPXFaDgWfLKc2zVqhUDwEJCQmSer+4seeUz1xXu3bvHRo0axYyNjVnz5s2Zl5cXu3r1KrO1tRWc9VX0LDljjP3yyy+se/fuTEdHh1laWrLPPvuMLVu2jAEQXCHAGGMHDx5kffv2ZQYGBkxHR4fZ2toyX19fduLECb5PVTdYq1ivyl6+fMliYmKYs7MzE4lETF9fn0mlUvbxxx+zGzduMMYYS0lJYSNHjmS2trZMR0eHmZqaMnd3d/bjjz/WuG7Pnz9n8+bNY+bm5kwkEjE3NzeWkpIi834x9urM9KxZs1jr1q2ZlpYWMzExYV27dmUhISHs2bNnfL8TJ06wLl26MB0dHZmz7Xfu3GGBgYHM2tqaaWlpMTMzM9azZ0+2ePFiwbLy8/PZ7NmzmY2NDdPS0mLm5uZs6NChLD09XaXL+eOPP9iAAQOYSCRiJiYmbPLkyezQoUMN+iw53QSN1DsDBw5ERkYG/vrrL3VPhRAB2iUnavXJJ5+gS5cuaNWqFfLy8hAbG4vjx4/ju+++U/fUCJFBBZOoVVlZGcLDw5GVlQWO4+Do6Ijvv/8e48aNU/fUCJFBu+SEEKIgunCdEEIURAWTEEIURAWTEEIURAWTEEIURAWTEEIURAWTEEIURAWTEEIURAWTEEIURAWTEEIU9P8Aq5LlQXYMB7cAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 300x300 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<ggplot: (8780486590196)>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_per_cell_file = figure_dir+'/'+'genes_per_cell.svg'\n",
    "\n",
    "genes_per_cell_plot = (\n",
    "    p9.ggplot(\n",
    "        wt_infected_cells[['sample', 'cell_barcode', 'n_viral_genes']].drop_duplicates(),\n",
    "        p9.aes(x='n_viral_genes')) +\n",
    "    p9.geom_bar(stat='count', position='dodge') +\n",
    "    p9.facet_grid('sample~') +\n",
    "    p9.ggtitle('Number of viral genes per cell') +\n",
    "    p9.labs(x='# of viral genes detected',\n",
    "            y='# of cells') +\n",
    "    p9.scale_y_continuous(limits=[0,300], breaks=[0,50,100,150,200,250,300]) +\n",
    "    p9.theme(figure_size=(3, 3),\n",
    "                plot_title=p9.element_text(size=12),\n",
    "                axis_title=p9.element_text(size=12),\n",
    "                axis_text=p9.element_text(size=12, color=\"black\")))\n",
    "genes_per_cell_plot.save(filename = genes_per_cell_file)\n",
    "display(genes_per_cell_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot viral gene presence or absence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To generate this plot, I need to tabulate how many cells express or do not express each viral gene. I will make a new dataframe with cell barcode, frac_viral_UMIs, and gene as columns. For each cell, I will list whether the gene is present and the fraction of gene UMIs in the cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_present</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "      <th>frac_gene_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAACGAAAGCCACCGT</td>\n",
       "      <td>PB2</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.010804</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAAGGATCATTGCAAC</td>\n",
       "      <td>PB2</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.051318</td>\n",
       "      <td>0.000347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAAGGGCGTTGCCTAA</td>\n",
       "      <td>PB2</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.181779</td>\n",
       "      <td>0.002359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AAAGTGAAGGAAGTCC</td>\n",
       "      <td>PB2</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.021469</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>trial1</td>\n",
       "      <td>AACAACCGTGATTCTG</td>\n",
       "      <td>PB2</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.049040</td>\n",
       "      <td>0.000220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5051</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTGCGTCCACCCTCTA</td>\n",
       "      <td>NS</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.011315</td>\n",
       "      <td>0.002885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5052</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTGGGATGTAACGATA</td>\n",
       "      <td>NS</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.058569</td>\n",
       "      <td>0.022673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5053</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTTACCAGTCGCTTAA</td>\n",
       "      <td>NS</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.012517</td>\n",
       "      <td>0.004836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5054</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTTCACAAGCCAAGCA</td>\n",
       "      <td>NS</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.033731</td>\n",
       "      <td>0.000015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5055</th>\n",
       "      <td>trial3</td>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>NS</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>0.008193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5056 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample      cell_barcode gene  gene_present  frac_viral_UMIs  \\\n",
       "0     trial1  AAACGAAAGCCACCGT  PB2   Gene absent         0.010804   \n",
       "1     trial1  AAAGGATCATTGCAAC  PB2  Gene present         0.051318   \n",
       "2     trial1  AAAGGGCGTTGCCTAA  PB2  Gene present         0.181779   \n",
       "3     trial1  AAAGTGAAGGAAGTCC  PB2   Gene absent         0.021469   \n",
       "4     trial1  AACAACCGTGATTCTG  PB2  Gene present         0.049040   \n",
       "...      ...               ...  ...           ...              ...   \n",
       "5051  trial3  TTGCGTCCACCCTCTA   NS  Gene present         0.011315   \n",
       "5052  trial3  TTGGGATGTAACGATA   NS  Gene present         0.058569   \n",
       "5053  trial3  TTTACCAGTCGCTTAA   NS  Gene present         0.012517   \n",
       "5054  trial3  TTTCACAAGCCAAGCA   NS   Gene absent         0.033731   \n",
       "5055  trial3  TTTGGAGGTATCGTTG   NS  Gene present         0.023486   \n",
       "\n",
       "      frac_gene_UMIs  \n",
       "0           0.000000  \n",
       "1           0.000347  \n",
       "2           0.002359  \n",
       "3           0.000000  \n",
       "4           0.000220  \n",
       "...              ...  \n",
       "5051        0.002885  \n",
       "5052        0.022673  \n",
       "5053        0.004836  \n",
       "5054        0.000015  \n",
       "5055        0.008193  \n",
       "\n",
       "[5056 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_present</th>\n",
       "      <th>fraction_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>trial1</td>\n",
       "      <td>PB2</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.949735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>trial1</td>\n",
       "      <td>PB2</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.050265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>trial1</td>\n",
       "      <td>PB1</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.952381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>trial1</td>\n",
       "      <td>PB1</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.047619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>trial1</td>\n",
       "      <td>PA</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.920635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>trial1</td>\n",
       "      <td>PA</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.079365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>trial1</td>\n",
       "      <td>HA</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.902116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>trial1</td>\n",
       "      <td>HA</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.097884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>trial1</td>\n",
       "      <td>NP</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.955026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>trial1</td>\n",
       "      <td>NP</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.044974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>trial1</td>\n",
       "      <td>NA</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.962963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>trial1</td>\n",
       "      <td>NA</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.037037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>trial1</td>\n",
       "      <td>M</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.917989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>trial1</td>\n",
       "      <td>M</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.082011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>trial1</td>\n",
       "      <td>NS</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.912698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>trial1</td>\n",
       "      <td>NS</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.087302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>trial3</td>\n",
       "      <td>PB2</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.933071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>trial3</td>\n",
       "      <td>PB2</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.066929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>trial3</td>\n",
       "      <td>PB1</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.960630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>trial3</td>\n",
       "      <td>PB1</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.039370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>trial3</td>\n",
       "      <td>PA</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.913386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>trial3</td>\n",
       "      <td>PA</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.086614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>trial3</td>\n",
       "      <td>HA</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.854331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>trial3</td>\n",
       "      <td>HA</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.145669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>trial3</td>\n",
       "      <td>NP</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.960630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>trial3</td>\n",
       "      <td>NP</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.039370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>trial3</td>\n",
       "      <td>NA</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.956693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>trial3</td>\n",
       "      <td>NA</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.043307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>trial3</td>\n",
       "      <td>M</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.933071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>trial3</td>\n",
       "      <td>M</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.066929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>trial3</td>\n",
       "      <td>NS</td>\n",
       "      <td>Gene present</td>\n",
       "      <td>0.881890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>trial3</td>\n",
       "      <td>NS</td>\n",
       "      <td>Gene absent</td>\n",
       "      <td>0.118110</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample gene  gene_present  fraction_cells\n",
       "0   trial1  PB2  Gene present        0.949735\n",
       "1   trial1  PB2   Gene absent        0.050265\n",
       "2   trial1  PB1  Gene present        0.952381\n",
       "3   trial1  PB1   Gene absent        0.047619\n",
       "4   trial1   PA  Gene present        0.920635\n",
       "5   trial1   PA   Gene absent        0.079365\n",
       "6   trial1   HA  Gene present        0.902116\n",
       "7   trial1   HA   Gene absent        0.097884\n",
       "8   trial1   NP  Gene present        0.955026\n",
       "9   trial1   NP   Gene absent        0.044974\n",
       "10  trial1   NA  Gene present        0.962963\n",
       "11  trial1   NA   Gene absent        0.037037\n",
       "12  trial1    M  Gene present        0.917989\n",
       "13  trial1    M   Gene absent        0.082011\n",
       "14  trial1   NS  Gene present        0.912698\n",
       "15  trial1   NS   Gene absent        0.087302\n",
       "16  trial3  PB2  Gene present        0.933071\n",
       "17  trial3  PB2   Gene absent        0.066929\n",
       "18  trial3  PB1  Gene present        0.960630\n",
       "19  trial3  PB1   Gene absent        0.039370\n",
       "20  trial3   PA  Gene present        0.913386\n",
       "21  trial3   PA   Gene absent        0.086614\n",
       "22  trial3   HA  Gene present        0.854331\n",
       "23  trial3   HA   Gene absent        0.145669\n",
       "24  trial3   NP  Gene present        0.960630\n",
       "25  trial3   NP   Gene absent        0.039370\n",
       "26  trial3   NA  Gene present        0.956693\n",
       "27  trial3   NA   Gene absent        0.043307\n",
       "28  trial3    M  Gene present        0.933071\n",
       "29  trial3    M   Gene absent        0.066929\n",
       "30  trial3   NS  Gene present        0.881890\n",
       "31  trial3   NS   Gene absent        0.118110"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene_present_by_cell = (\n",
    "    wt_infected_cells\n",
    "    [['sample', 'cell_barcode',\n",
    "      'present_fluPB2', 'present_fluPB1',\n",
    "      'present_fluPA', 'present_fluHA',\n",
    "      'present_fluNP', 'present_fluNA',\n",
    "      'present_fluM', 'present_fluNS']]\n",
    "    .drop_duplicates()\n",
    "    .melt(\n",
    "        id_vars=['sample', 'cell_barcode'],\n",
    "        var_name='gene',\n",
    "        value_name='gene_present')\n",
    ")\n",
    "gene_present_by_cell['gene'] = (\n",
    "    gene_present_by_cell['gene']\n",
    "    .str.split('_').str[-1])\n",
    "\n",
    "gene_frac_by_cell = (\n",
    "    wt_infected_cells\n",
    "    [['sample', 'cell_barcode', 'frac_viral_UMIs',\n",
    "      'frac_UMIs_fluPB2', 'frac_UMIs_fluPB1',\n",
    "      'frac_UMIs_fluPA', 'frac_UMIs_fluHA',\n",
    "      'frac_UMIs_fluNP', 'frac_UMIs_fluNA',\n",
    "      'frac_UMIs_fluM', 'frac_UMIs_fluNS']]\n",
    "    .drop_duplicates()\n",
    "    .melt(\n",
    "        id_vars=['sample', 'cell_barcode', 'frac_viral_UMIs'],\n",
    "        var_name='gene',\n",
    "        value_name='frac_gene_UMIs')\n",
    ")\n",
    "gene_frac_by_cell['gene'] = (\n",
    "    gene_frac_by_cell['gene']\n",
    "    .str.split('_').str[-1])\n",
    "\n",
    "viral_genes_by_cell = pd.merge(\n",
    "    left=gene_present_by_cell,\n",
    "    right=gene_frac_by_cell,\n",
    "    on=['sample', 'cell_barcode','gene'],\n",
    "    how='outer'\n",
    ")\n",
    "viral_genes_by_cell['gene'] = (\n",
    "    viral_genes_by_cell['gene']\n",
    "    .replace(['fluPB2', 'fluPB1', 'fluPA', 'fluHA',\n",
    "              'fluNP', 'fluNA', 'fluM', 'fluNS'],\n",
    "             ['PB2', 'PB1', 'PA', 'HA',\n",
    "              'NP', 'NA', 'M', 'NS']))\n",
    "viral_genes_by_cell['gene'] = (\n",
    "    viral_genes_by_cell['gene'].astype('category'))\n",
    "viral_genes_by_cell['gene'] = (\n",
    "    viral_genes_by_cell['gene']\n",
    "    .cat.reorder_categories(\n",
    "        ['PB2', 'PB1', 'PA', 'HA',\n",
    "         'NP', 'NA', 'M', 'NS']))\n",
    "viral_genes_by_cell['gene_present'] = (\n",
    "    viral_genes_by_cell['gene_present']\n",
    "    .replace([True,False],\n",
    "             [\"Gene present\", \"Gene absent\"]))\n",
    "\n",
    "for sample in wt_infected_cells['sample'].unique():\n",
    "    assert viral_genes_by_cell.query(f'sample == \"{sample}\"')['cell_barcode'].unique().all() == \\\n",
    "           wt_infected_cells.query(f'sample == \"{sample}\"')['cell_barcode'].unique().all(), \\\n",
    "           f\"Cell barcodes don't match for {sample}\"\n",
    "\n",
    "display(viral_genes_by_cell)\n",
    "\n",
    "gene_absence_rate = pd.DataFrame(\n",
    "    viral_genes_by_cell\n",
    "    .groupby(['sample', 'gene'])\n",
    "    ['gene_present']\n",
    "    .value_counts(normalize=True)\n",
    "    .reset_index(name='fraction_cells'))\n",
    "\n",
    "\n",
    "assert(gene_absence_rate['gene'].nunique() == 8)\n",
    "\n",
    "display(gene_absence_rate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the fraction of infected cells missing each viral gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dbacsik/.conda/envs/barcoded_flu_pdmH1N1_final_analysis/lib/python3.8/site-packages/plotnine/ggplot.py:718: PlotnineWarning: Saving 4 x 3 in image.\n",
      "/home/dbacsik/.conda/envs/barcoded_flu_pdmH1N1_final_analysis/lib/python3.8/site-packages/plotnine/ggplot.py:719: PlotnineWarning: Filename: results/figures/gene_absence_plot.svg\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZcAAAFfCAYAAACC3YHUAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAA9hAAAPYQGoP6dpAABcQUlEQVR4nO3dd1xTZ/s/8E8YIWzEKKCyXAxxMFSQUqVO3IqrVRSFWoutuz76tVZxWxfKUxyIoqUqWqyTqjgQ6xaBWhx1IY4ghqECIhDu3x/+yENMgCQmBOV6v155wbnPfe5znXOSXLnP5DDGGAghhBAV0tJ0AIQQQj49lFwIIYSoHCUXQgghKkfJhRBCiMpRciGEEKJylFwIIYSoHCUXQgghKkfJhRBCiMpRciGEEKJylFwqGTJkCPT19ZGfn19lndGjR0NXVxfPnz9HdHQ0OBwOMjIyVBrHwoULweFwVNom+fQkJiaCw+EgMTFRZW1269YN3bp1U3r68PBwtGzZElwuFxwOp9rPkjJu3ryJhQsXqvwzB2jmcydrG34qn39KLpUEBQWhuLgYu3btkjn+5cuX+OOPP9C/f39YWFigX79+uHjxIqysrGo5UkIANzc3XLx4EW5ubiprMyIiAhEREUpNm5qaiilTpsDX1xenT5/GxYsXYWxsrLLYgHfJJTQ0VC3JhaiWjqYDqEv8/PzQpEkTbNu2DSEhIVLjd+/ejTdv3iAoKAgA0KhRIzRq1KjGdouKimBgYKDyeOur0tJScDgc6OjU77eviYkJPD09Vdqms7Oz0tOmp6cDAL7++mt06tRJVSGRjxT1XCrR1tbGuHHjkJycjBs3bkiN3759O6ysrODn5wcAMneLdevWDS4uLkhKSkKXLl1gYGCACRMmAABiY2PRq1cvWFlZQV9fH05OTpgzZw4KCwuVjjkyMhKtW7eGnp4enJ2dsWvXLgQGBsLOzk6iXklJCZYsWQJHR0fo6emhUaNGGD9+PF68eCFRz87ODv3798exY8fg5uYGfX19ODo6Ytu2bVLzzsrKwjfffINmzZqBy+XC3t4eoaGhKCsrqzHuivn88ccfaNeuHXg8Hpo3b44NGzZI1KvYbfDrr79i5syZaNq0KfT09HDv3j0AwMmTJ9G9e3eYmJjAwMAA3t7eOHXqlEQbL168wMSJE2FtbS1edm9vb5w8eVKinjxtVeyySE9Px5dffglTU1NYWFhgwoQJePnypUTd8vJyhIeHo0OHDtDX14eZmRk8PT1x6NAhiXqxsbHw8vKCoaEhjIyM0Lt3b6SkpNS4DmXtUgkMDISRkRHu3buHvn37wsjICNbW1pg5cybevn1bY5vv7xbLyMgAh8PB6tWrsXbtWtjb28PIyAheXl64dOmSxHRjxowBAHTu3BkcDgeBgYHi8fKsWwC4ffs2vvzyS1hYWEBPTw82NjYYO3Ys3r59i+joaAwfPhwA4OvrCw6HAw6Hg+joaIXnc/ToUXTo0AF6enqwt7fH6tWra1w3lR07dgzdu3eHqakpDAwM4OTkhOXLl0vUuXbtGgYOHAhzc3PweDy4urpi7969Cs2nwunTp9GtWzc0bNgQ+vr6sLGxgb+/P4qKipRqr1YwIuHu3buMw+GwadOmSZSnp6czAGzOnDnisu3btzMA7OHDh+Kyrl27MnNzc2Ztbc3Cw8PZmTNn2NmzZxljjC1evJitW7eOHT16lCUmJrJNmzYxe3t75uvrKzGvBQsWMHk2zebNmxkA5u/vz44cOcJ+++031rp1a2Zra8tsbW3F9UQiEevTpw8zNDRkoaGhLCEhgW3dupU1bdqUOTs7s6KiInFdW1tb1qxZM+bs7Mx27tzJjh8/zoYPH84AiJeDMcYEAgGztrZmtra2bPPmzezkyZNs8eLFTE9PjwUGBtYYu62tLWvatCmzsbFh27ZtY/Hx8Wz06NEMAFu1apW43pkzZxgA1rRpUzZs2DB26NAhduTIEZaTk8N+/fVXxuFw2ODBg9n+/fvZ4cOHWf/+/Zm2tjY7efKkuI3evXuzRo0asS1btrDExER24MAB9tNPP7E9e/aI68jbVsW2cXBwYD/99BNLSEhga9euZXp6emz8+PESyxgQEMA4HA4LDg5mBw8eZH/++SdbunQpW79+vbjO0qVLGYfDYRMmTGBHjhxh+/fvZ15eXszQ0JClp6dXuw4r1s2ZM2fEZePGjWNcLpc5OTmx1atXs5MnT7KffvqJcTgcFhoaWuN26dq1K+vatat4+OHDhwwAs7OzY3369GEHDhxgBw4cYG3btmUNGjRg+fn5jLF3n48ff/yRAWDbt29nFy9eZPfu3VNo3aampjIjIyNmZ2fHNm3axE6dOsViYmLYiBEj2KtXr1h2djZbtmwZA8B++eUXdvHiRXbx4kWWnZ2t0HxOnjzJtLW12Weffcb279/P9u3bxzp27MhsbGzk+txt3bqVcTgc1q1bN7Zr1y528uRJFhERwUJCQsR1Tp8+zbhcLvPx8WGxsbHs2LFjLDAwULx+qtuG73/+Hz58yHg8HuvZsyc7cOAAS0xMZL/99hsLCAhgeXl5NcarKZRcZOjatSvj8/mspKREXDZz5kwGgP3777/isqqSCwB26tSpaudRXl7OSktL2dmzZxkAlpaWJh4nT3IRiUTM0tKSde7cWaL80aNHTFdXVyK57N69mwFgcXFxEnWvXr3KALCIiAhxma2tLePxeOzRo0fisjdv3jBzc3P2zTffiMu++eYbZmRkJFGPMcZWr17NANT4xWhra8s4HA5LTU2VKO/ZsyczMTFhhYWFjLH/ffg+//xziXqFhYXM3NycDRgwQGq9tG/fnnXq1ElcZmRkJPVjQdm2KrbNzz//LFE3JCSE8Xg8Vl5ezhhjLCkpiQFg8+bNq3K+mZmZTEdHh33//fcS5a9fv2aWlpZsxIgRVU7LWNXJBQDbu3evRN2+ffsyBweHattjrOrk0rZtW1ZWViYuv3LlCgPAdu/eLS6r+DxcvXpVXKbIuv3iiy+YmZmZOFnIsm/fPqllVnQ+nTt3Zk2aNGFv3rwRl7169YqZm5vX+Ll7/fo1MzExYZ999pl4W8vi6OjIXF1dWWlpqUR5//79mZWVFROJRIwx+ZLL77//zgBIfVbqOtotJkNQUBCEQqF490VZWRliYmLg4+ODVq1a1Th9gwYN8MUXX0iVP3jwAF999RUsLS2hra0NXV1ddO3aFQBw69YthWK8c+cOsrKyMGLECIlyGxsbeHt7S5QdOXIEZmZmGDBgAMrKysSvDh06wNLSUupsow4dOsDGxkY8zOPx0Lp1azx69EiiTV9fXzRp0kSizYpdhmfPnq1xGdq0aYP27dtLlH311Vd49eoVrl+/LlHu7+8vMXzhwgXk5uZi3LhxEvMvLy9Hnz59cPXqVfHuxk6dOiE6OhpLlizBpUuXUFpaqnRbFQYOHCgx3K5dOxQXFyM7OxsA8OeffwIAJk+eXOXyHz9+HGVlZRg7dqzEfHk8Hrp27ar0WWAcDgcDBgyQiq/y9lNUv379oK2tLdEegBrblHfdFhUV4ezZsxgxYoRcxzGVnU9hYSGuXr2KoUOHgsfjiac3NjaWWmdVzefVq1cICQmp8oyue/fu4fbt2xg9ejQASMTTt29fCAQC3LlzR+5l69ChA7hcLiZOnIgdO3bgwYMHck+rSfX7iGgVhg0bhu+//x7bt2+Hv78/4uPj8fz5c6xcuVKu6WWdPVZQUAAfHx/weDwsWbIErVu3hoGBAR4/foyhQ4fizZs3CsWYk5MDALCwsJAaZ2FhgYcPH4qHnz9/jvz8fHC5XJltCYVCieGGDRtK1dHT05OI8fnz5zh8+DB0dXXlalMWS0vLKssqlq/C++v0+fPnAN5tq6rk5ubC0NAQsbGxWLJkCbZu3Yr58+fDyMgIQ4YMwc8//wxLS0uF2qrw/jrS09MDAPE6evHiBbS1tWUu4/vL0LFjR5njtbSU++1nYGAg8cVZEV9xcbFS7QE1L29V5F23WlpaEIlEaNasmVLxyTsfDoeD8vLyat971ak4RlldnBWxzJo1C7NmzZJZR57PR4UWLVrg5MmT+PnnnzF58mQUFhaiefPmmDJlCqZOnSp3O7WNkosM+vr6+PLLLxEZGQmBQIBt27bB2NhYfDCxJrJ+0Zw+fRrPnj1DYmKiuLcCQOnrACo+7BVv5MqysrIkhvl8Pho2bIhjx47JbEuZ00X5fD7atWuHpUuXyhzfpEmTGtt4P87KZe9/mb2/Tvl8PoB311VUdcZUReLl8/kICwtDWFgYMjMzcejQIcyZMwfZ2dk4duyYQm3Jq1GjRhCJRMjKyqryVPWK+f7++++wtbVVqP2PhbzrViQSQVtbG0+ePFHrfCrONKzuvVedil5VdXFWxDJ37lwMHTpUZh0HB4ca51WZj48PfHx8IBKJcO3aNYSHh2PatGmwsLDAqFGjFGqrtlByqUJQUBA2bdqEVatWIT4+HoGBgR90OnHFl2PFL74KmzdvVqo9BwcHWFpaYu/evZgxY4a4PDMzExcuXJD4cu/fvz/27NkDkUiEzp07KzW/9/Xv3x/x8fFo0aIFGjRooFQb6enpSEtLk9g1tmvXLhgbG9d47Ya3tzfMzMxw8+ZNfPfdd3LP08bGBt999x1OnTqF8+fPf1Bb1fHz88Py5cuxceNGLFq0SGad3r17Q0dHB/fv35fa7fepUGTddu3aFfv27cPSpUvFX9Dvq6rHJO98uFwuOnXqhP3792PVqlXiHt7r169x+PDhGpenS5cuMDU1xaZNmzBq1CiZPyQdHBzQqlUrpKWlYdmyZTW2qQhtbW107twZjo6O+O2333D9+nVKLh8bDw8PtGvXDmFhYWCMia9tUVaXLl3QoEEDTJo0CQsWLICuri5+++03pKWlKdWelpYWQkND8c0332DYsGGYMGEC8vPzERoaCisrK4ldKqNGjcJvv/2Gvn37YurUqejUqRN0dXXx5MkTnDlzBoMGDcKQIUMUmv+iRYuQkJCALl26YMqUKXBwcEBxcTEyMjIQHx+PTZs21biLo0mTJhg4cCAWLlwIKysrxMTEICEhAStXrqwxkRsZGSE8PBzjxo1Dbm4uhg0bhsaNG+PFixdIS0vDixcvsHHjRrx8+RK+vr746quv4OjoCGNjY1y9ehXHjh0T/6qUty1F+Pj4ICAgAEuWLMHz58/Rv39/6OnpISUlBQYGBvj+++9hZ2eHRYsWYd68eXjw4AH69OmDBg0a4Pnz57hy5QoMDQ0RGhqq0HzrGkXW7dq1a/HZZ5+hc+fOmDNnDlq2bInnz5/j0KFD2Lx5M4yNjeHi4gIA2LJlC4yNjcHj8WBvb4+GDRvKPZ/FixejT58+6NmzJ2bOnAmRSISVK1fC0NAQubm5NS7PmjVrEBwcjB49euDrr7+GhYUF7t27h7S0NPz3v/8F8O5Ho5+fH3r37o3AwEA0bdoUubm5uHXrFq5fv459+/bJvQ43bdqE06dPo1+/frCxsUFxcbH40oAePXoovE1qjabPKKjL1q9fzwAwZ2dnmeOrOlusTZs2MutfuHCBeXl5MQMDA9aoUSMWHBzMrl+/LnV6orynIjPG2JYtW1jLli0Zl8tlrVu3Ztu2bWODBg1irq6uEvVKS0vZ6tWrWfv27RmPx2NGRkbM0dGRffPNN+zu3bviera2tqxfv35S83n/LCLGGHvx4gWbMmUKs7e3Z7q6uszc3Jy5u7uzefPmsYKCgmrjrpjP77//ztq0acO4XC6zs7Nja9eulahXcTbNvn37ZLZz9uxZ1q9fP2Zubs50dXVZ06ZNWb9+/cT1i4uL2aRJk1i7du2YiYkJ09fXZw4ODmzBggXiM9LkbYux/22bFy9eSEwr670gEonYunXrmIuLC+NyuczU1JR5eXmxw4cPS0x74MAB5uvry0xMTJienh6ztbVlw4YNkzh9VpaqzhYzNDSUqivve6qqs8Uqnx5eAQBbsGCBeFjW2WIV5Fm3jDF28+ZNNnz4cNawYUPG5XKZjY0NCwwMZMXFxeI6YWFhzN7enmlra0t9duSdz6FDh1i7du3E81ixYoVCn7v4+HjWtWtXZmhoyAwMDJizszNbuXKlRJ20tDQ2YsQI1rhxY6arq8ssLS3ZF198wTZt2iSuI8/ZYhcvXmRDhgxhtra2TE9PjzVs2JB17dqVHTp0SK5YNYXDGGO1n9KIuuTn56N169YYPHgwtmzZoulwqmRnZwcXFxccOXJE06EQQtSAdot9xLKysrB06VL4+vqiYcOGePToEdatW4fXr1/X6bNICCGfPkouHzE9PT1kZGQgJCQEubm5MDAwgKenJzZt2oQ2bdpoOjxCSD1Gu8UIIYSoHF2hTwghROUouRBCCFE5Si6EEEJUjpILIYQQlaPkQgghROUouRBCCFE5Si6EEEJUji6irKSoqAi3b9+Go6PjB90BmRBC6pLMzEyFniFTEz6fL/FAQVkouVRy+/ZtuLu7Izk5ucZbvhNCyMcgMzMTTk5OKCoqUlmbBgYGuHXrVrUJhpILIYR8woRCIYqKihATEwMnJ6cPbu/WrVsYM2YMhEIhJRdCCKnvnJycanWPDB3QJ4QQonLUc1FS7qkI5MSvQtlLAfSatIHF6DAYOvhUWb/w9lk83zUDb5+lQ8esCRr2nQ3zLyaJx+efi8azreOlpnOMfAMtLo/iei8uAMg5Hoa80xtRmpMJbWM+TDyGofHw5XLHpUhMpfkCPN89E8UZySh5fhfmPafAcnSYVL1XV+OQvX8+SrPvQ7dxCzT2XwoTD8We8qmOdVXh5aU9eLrxSxi7DYL11AMUl5reW+qIi5WVQnhkOfL/2oGy/KfgWjrAYsRKGLXrI3dMtUqzzyqrW5KTkxkAlpycXG29/Et7WPp4XZZ7JpIVP73JBDFT2c2vDVmJ8JHM+m+zH7CbwQZMEDOVFT+9yXLPRLL08brs5ZXfxXXykrazW9+YsNI8gcRLEfUprvzzMexmkB7Lv/Abe5v9kL3++zi7M8WKCWKmqSmmh0zw6xSWd24Hu/9jByaImSpVp/DuBZYeqM1eHFrGip/eYi8OLWPp43VY4b1LcsWkXFw1rytx3RcZ7M7UpuzhEh+WGTZI7pjqW1wf+t5SV1xZe2azO1OasFepR9nb5/dZzskIdjOIx4oyrlcbi7zfa/KStz1KLpXIu9LuL+zEnm2fJFF29z+OLCt2jsz6WXtms7v/cZQoe7b9G/Yg1FM8nJe0nd2aZKpc4PUwrmc7JrOHK76QqCPYNYM9WPKZWmKq7OGyrjKTy+P/jmAZq/pIlGWs6s0e/zJKrpiUiUuedcUYY+WiMvZgsTfLTdzKnmwZp/CXeH2K60PfW+qK684UK5aT8F+JOplhg9iTjaOrjUVTyYWOuSiIlZWgOCMZhi69JMqNXHrhzb0LMqd5c+8ijN6rb+jSG28yroGVlYrLyosLcHeGLf6d1gyZa/vjzaMUiquKuAxaf4bijGS8uX8FAFCS/QAFafEwbt9PLTHJo0hG3EYuveVuU53b8MWBRdAxboQGXYPkiqU+x/Uh7y11xsVK34KjK7lbjqOrj6K7f8kVV22jYy4KKnstBMpF0DG1kCjXNrVA2css2dO8zIL2e/V1TC0AURnKCoTQNbMC18oRTYKjwbNuC9GbV8g9sR4ZS7zRfHEa9CxbUVzvxWXqOQqi1y/wcOlnABggKkODL74Fv/8ctcQkj7KXWVJt6ijQprrWVdG/55GfFIXmi1PlX5h6HNeHvLfUGZdh297IPbYWBg6fg9u4BQpvnsLrlINAuUiuuGobJRdlcTiSw4xJl1WuDhn1K5UbtPSEQUtP8WiDVt54sMANeSfDYTlmA8X1XlyFtxLx4vBSWI2NgH6Lzih5fg9Zv02FjpkVGg2ar5aY5Gz0w9tU4boSvXmNp5vHwGp8JHSM+YrFUQ/jAlT03lJDXJaj10Ow/Wvcn+MIcDjgNm4BM5/xyD+3Xf6YahElFwXpGPMBLW2U5Uv+AhG9yoaOiYXsaUwtpX6xlL3KBrR1oG3UUOY0HC0t6Nt3xNusuxSXjLiy98+HWZcANOgWDADgWbdF+dtCCKIngj9gHjhaVe/xVSYmeVQVt7xtqmNdvX2ajlJhBh6HDfhfBVYOALg5XgctV9wB16IFxaWi95Y649IxaQTrqQdQXlIMUUEOdBo0QfbeOeDy7auNR1PomIuCODpc8OzcUZieIFFekJ4A/ZZdZE6j39ILBe/VL/znBPTtPMDR0ZU5DWMMxZmp0DWzorhkxMXeFgEcybcvR0v7///aYyqPSR4GLb2k2/znhNxtqmNdca0c0XzpDTRfnCp+GbsOhIGTL5ovToVuQ2uKS4XvLXXGVUGLy4OueVNAVIZX1+Jg5Daoxpg0gZKLEhr2mYG8s1uRl7QNb5/dQtZv01Gak4kG//+c9Od75+Lp5rHi+g2+mIRS4SNk7ZqBt89uIS9pG/KSotDQb5a4zos/QlFw4zhKsh+g+FEqBFFBKM5MRQNf2dcF1Pe4jFwHIO/0Rry8tAclLx6i4J8EZO+fD2PXge++CFQcEwAUP0pF8aNUlBcXoOz1CxQ/SsXbpzfF4817TUXBPycgPLoSb5/dhvDoShTePAnz3tM0tq60uDzwmrlIvLQMzKDNMwavmQs4OlyKS8XvLXXFVXT/Ml5d24+S7AcovHMOj9b0AVg5+H1nyxVTbftkd4slJSVh1apVSE5OhkAgwB9//IHBgwerpG3TziMhKsiB8OAilOULoNfUBTYz4sHl2wIAyl4KUJqbKa7PbWQPm5nxeL5rOvJO/QIdsyawHLMBJh39xXVERfkQbJ+IspdZ0NI3Bc/WFXb/lwT9Fp0oLhlxNRr4IzjgIDvuR5TlPYW2cSMYuw5AY/+laokJAB785Cr+vzgjGa8u7oIu3xat1mQAAAxadUGzkD3IjvsR2XHzwW3cAs1CYmHQorNG15Uq1Ke4PvS9pa64WGkxsuN+ROmLB9DSM4JRu75oOvFXaBuaKbv61IrDGKu5n/cR+vPPP3H+/Hm4ubnB399fruRy/fp1uisyIeSTourvNXnb+2R7Ln5+fvDz89N0GIQQUi/RMRdCCCEq98n2XOQlFArFT2h7+PChhqMhhJBPQ71PLnFxcYiMjASAap/UdnPch15cpxrOO/53iKwuxgRQXDWhuBRTF9/zQN2M6/1tqEn1Prn4+/uja9euAID09HQMGzZMwxERQsjHr94nFz6fDz7/3e0nVPmMaUIIqc8+2eRSUFCAe/fuiYcfPnyI1NRUmJubV/vcZ0IIIR/uk00u165dg6+vr3h4xowZAIBx48YhOjpaQ1ERIu0v782aDgEA4PzecF2Ni3wcPtnk0q1bN6jy+tC6+EGrizERQghA17kQQghRA0ouhBBCVO6T3S1GNId21xF1qSvvLYB2UdeEkgupN+gLgJDaQ7vFCCGEqBwlF0IIISpHyYUQQojKUXIhhBCicpRcCCGEqBwlF0IIISpHyYUQQojK1YnkEhERAXt7e/B4PLi7u+PcuXNV1g0MDASHw5F6tWnTRlwnOjpaZp3i4uLaWBxCCKn3NJ5cYmNjMW3aNMybNw8pKSnw8fGBn58fMjMzZdZfv349BAKB+PX48WOYm5tj+PDhEvVMTEwk6gkEAvB4vNpYJEIIqfc0nlzWrl2LoKAgBAcHw8nJCWFhYbC2tsbGjRtl1jc1NYWlpaX4de3aNeTl5WH8+PES9TgcjkQ9S0vL2lgcQggh0HByKSkpQXJyMnr16iVR3qtXL1y4cEGuNqKiotCjRw/Y2tpKlBcUFMDW1hbNmjVD//79kZKSorK4CSGEVE+jyUUoFEIkEsHCwkKi3MLCAllZWTVOLxAI8OeffyI4OFii3NHREdHR0Th06BB2794NHo8Hb29v3L17V6XxE0IIka1O3LiSw+FIDDPGpMpkiY6OhpmZGQYPHixR7unpCU9PT/Gwt7c33NzcEB4ejg0bNkjUFQqFEAqFAN49CpkQQsiH02hy4fP50NbWluqlZGdnS/Vm3scYw7Zt2xAQEAAul1ttXS0tLXTs2FFmzyUuLg6RkZEAgKKiIgWXgBBCPl1TpkyR+kEuL40mFy6XC3d3dyQkJGDIkCHi8oSEBAwaNKjaac+ePYt79+4hKCioxvkwxpCamoq2bdtKjfP390fXrl0BAOnp6Rg2bJiCS0EIIR+3EydOIDw8HEKhUOLx8E+fPsWVK1cAAJcuXVKoTY3vFpsxYwYCAgLg4eEBLy8vbNmyBZmZmZg0aRIAYO7cuXj69Cl27twpMV1UVBQ6d+4MFxcXqTZDQ0Ph6emJVq1a4dWrV9iwYQNSU1Pxyy+/SNXl8/ng8/kAqOdCCKmf5s+fj5CQENjZ2YnLOBwOJkyYgJUrVyrVpsaTy8iRI5GTk4NFixZBIBDAxcUF8fHx4rO/BAKB1DUvL1++RFxcHNavXy+zzfz8fEycOBFZWVkwNTWFq6srkpKS0KlTJ7UvDyGEfGwKCgowbtw4qXJtbW3xnh1FaTy5AEBISAhCQkJkjouOjpYqMzU1rbaXsW7dOqxbt05V4RFCyCctMDBQZvmECROUbrNOJBdCCCGas2HDhir3BP3nP/9Rqk1KLoQQUs8lJiaK/2eM4dmzZwgPD0e/fv2UbpOSCyGE1HMtWrSQGG7ZsiVcXFwwePDgKneZ1UTj9xYjhBBS9xgaGuLFixcoKytTanrquRBCCJGip6eHW7duKT09JRdCCKnnvLy8JC6erKzi4skvvvgCp0+flrtNSi6EEFLPrVixosY6ixYtUqhNSi6EEFLPyXOh5GeffaZQm3RAnxBC6rnLly+jf//+GDlyJB48eAAA+Oeff/D3338r3SYlF0IIqedmz54NX19fODk5ITQ0FADw9u1bzJ8/X+k2abcYIYTUc1lZWZg5cyaKiorg7e0NAHBzc/ugByzWiZ5LREQE7O3twePx4O7ujnPnzlVZNzExERwOR+p1+/ZtiXpxcXFwdnaGnp4enJ2d8ccff6h7MQgh5KPk4uKC+/fvw8DAAMXFxQCAZ8+egcfjKd2mxpNLbGwspk2bhnnz5iElJQU+Pj7w8/OTuhPy++7cuQOBQCB+tWrVSjzu4sWLGDlyJAICApCWloaAgACMGDECly9fVvfiEELIR6dLly4ICgpCVFQUysrKsHXrVowZMwZ9+vRRuk2NJ5e1a9ciKCgIwcHBcHJyQlhYGKytrbFx48Zqp2vcuDEsLS3FL21tbfG4sLAw9OzZE3PnzoWjoyPmzp2L7t27IywsTM1LQwghH5+zZ8/CxMQEBw8ehLOzM86fP49Ro0aJj78oQ6PHXEpKSpCcnIw5c+ZIlPfq1QsXLlyodlpXV1cUFxfD2dkZP/74I3x9fcXjLl68iOnTp0vU7927NyUXQgiR4dChQypvU6M9F6FQCJFIBAsLC4lyCwsLZGVlyZzGysoKW7ZsQVxcHPbv3w8HBwd0794dSUlJ4jpZWVkKtUkIIUS1PrjnIhKJcOPGDdja2qJBgwZKtcHhcCSGGWNSZRUcHBzg4OAgHvby8sLjx4+xevVqfP755wq3KRQKIRQKAQAPHz5UKn5CCPmYWVtbV3n7F+Dd9yljDE+ePJG7TYWTy7Rp09C2bVsEBQVBJBKha9euuHDhAgwMDHDkyBF069ZN7rb4fD60tbWlehTZ2dlSPY/qeHp6IiYmRjxsaWkpd5txcXGIjIwEgGqfbkkIIZ+qys9zURWFk8vvv/+OMWPGAAAOHz6Mhw8f4vbt29i5cyfmzZuH8+fPy90Wl8uFu7s7EhISMGTIEHF5QkICBg0aJHc7KSkpsLKyEg97eXkhISFB4rjLiRMn0KVLF6lp/f39xbc+SE9Px7Bhw+SeLyGEfAref56LKiicXIRCISwtLQEA8fHxGD58OFq3bo2goCBs2LBB4QBmzJiBgIAAeHh4wMvLC1u2bEFmZiYmTZoEAJg7dy6ePn2KnTt3Anh3JpidnR3atGmDkpISxMTEIC4uDnFxceI2p06dis8//xwrV67EoEGDcPDgQZw8eRJ//fWX1Pz5fD74fD4A6rkQQuqnixcvwtPTU+ahg9TUVBQVFcn8cV4dhZOLhYUFbt68CSsrKxw7dgwREREA3n0xVz4dWF4jR45ETk4OFi1aBIFAABcXF8THx8PW1hYAIBAIJK55KSkpwaxZs/D06VPo6+ujTZs2OHr0KPr27Suu06VLF+zZswc//vgj5s+fjxYtWiA2NhadO3dWOD5CCPnUDR8+HPfv34eenp7UuAcPHmD37t3qTy7jx4/HiBEjYGVlBQ6Hg549ewJ4d+MzR0dHRZsDAISEhCAkJETmuOjoaInh2bNnY/bs2TW2OWzYMNrFRQghcuBwOLh06RK4XK7UuNLSUvzzzz8Kt6lwclm4cCFcXFzw+PFjDB8+XJzptLW1pa5XIYQQ8nGYNm1alWfp6uvrK9yeUqciy+oRjBs3TpmmCCGEaBhjDJcuXZK5W0xZciUXRQ7UT5kyRelgCCGE1D5ra+sqey3Kkiu5rFu3Tq7GOBwOJRdCCPnIXLx4UeVtypVc6Mp1QgghitD4XZEJIYR8euTqucyYMUPuBteuXat0MIQQQj4NciWXlJQUuRpT9QEhQgghHye5ksuZM2fUHQchhJBPiNLHXO7du4fjx4/jzZs3AFDt7ZoJIYTULwonl5ycHHTv3h2tW7dG3759IRAIAADBwcGYOXOmygMkhBDy8VE4uUyfPh26urrIzMyEgYGBuHzkyJE4duyYUkFERETA3t4ePB4P7u7uOHfuXJV19+/fj549e6JRo0YwMTGBl5cXjh8/LlEnOjoaHA5H6lVcXKxUfIQQQhSjcHI5ceIEVq5ciWbNmkmUt2rVCo8ePVI4gNjYWEybNg3z5s1DSkoKfHx84OfnJ3En5MqSkpLQs2dPxMfHIzk5Gb6+vhgwYIDUSQcmJiYQCAQSLx6Pp3B8hBBCFKfwvcUKCwsleiwVhEKhUvelWbt2LYKCghAcHAzg3fNajh8/jo0bN2L58uVS9cPCwiSGly1bhoMHD+Lw4cNwdXUVl3M4HPFzZwghhNQuhXsun3/+ufjBXcC7L/Hy8nKsWrUKvr6+CrVVUlKC5ORk9OrVS6K8V69euHDhglxtlJeX4/Xr1zA3N5coLygogK2tLZo1a4b+/fvLfTo1IYSQD6dwz2XVqlXo1q0brl27hpKSEsyePRvp6enIzc1V6BHHwLvejkgkknq2vYWFBbKysuRqY82aNSgsLMSIESPEZY6OjoiOjkbbtm3x6tUrrF+/Ht7e3khLS0OrVq0UipEQQojiFE4uzs7O+Pvvv7Fx40Zoa2ujsLAQQ4cOxeTJkyWeY6+I9y++ZIzJdUHm7t27sXDhQhw8eBCNGzcWl3t6esLT01M87O3tDTc3N4SHh0vd4VkoFEIoFAKge6gRQoiqKPU8F0tLS4SGhn7wzPl8PrS1taV6KdnZ2VK9mffFxsYiKCgI+/btQ48ePaqtq6WlhY4dO+Lu3btS4+Li4hAZGQng3aOaCSGEfDiFj7ls374d+/btkyrft28fduzYoVBbXC4X7u7uSEhIkChPSEio9nnNu3fvRmBgIHbt2oV+/frVOB/GGFJTU2X2rPz9/RETE4OYmBgsXrxYofgJIYTIpnByWbFiBfh8vlR548aNsWzZMoUDmDFjBrZu3Ypt27bh1q1bmD59OjIzMzFp0iQAwNy5czF27Fhx/d27d2Ps2LFYs2YNPD09kZWVhaysLLx8+VJcJzQ0FMePH8eDBw+QmpqKoKAgpKamitusjM/nw9HREY6OjrC3t1c4fkIIIdIU3i326NEjmV/Ctra2VV6bUp2RI0ciJycHixYtgkAggIuLC+Lj42FrawsAEAgEEu1u3rwZZWVlmDx5MiZPniwuHzduHKKjowEA+fn5mDhxIrKysmBqagpXV1ckJSWhU6dOCsdHCCFEcQonl8aNG+Pvv/+GnZ2dRHlaWhoaNmyoVBAhISEICQmROa4iYVRITEyssb1169bJ/fRMQgghqqfwbrFRo0ZhypQpOHPmDEQiEUQiEU6fPo2pU6di1KhR6oiREELIR0bhnsuSJUvw6NEjdO/eHTo67yYvLy/H2LFjlTrmQggh5NOjcHLhcrmIjY3FkiVLkJqaCn19fbRt21Z8jIQQQghR6joX4N2NKulqd0IIIbIo/bAwQgghpCqUXAghhKgcJRdCCCEqJ1dyGTp0KF69egUA2LlzJ96+favWoAghhHzc5EouR44cQWFhIQBg/PjxErdaIYQQQt4n19lijo6OmDt3Lnx9fcEYw969e2FiYiKzbuX7gBFCCKmf5EoumzZtwowZM3D06FFwOBz8+OOPMp+3wuFwKLkQQgiRb7dYly5dcOnSJbx48QKMMfz777/Iy8uTeuXm5qo7XoVERETA3t4ePB4P7u7uOHfunKZDIoSQekHhs8UePnyIRo0aqSMWlYqNjcW0adMwb948pKSkwMfHB35+fkrduZkQQohiFL5C39bWFvn5+YiKisKtW7fA4XDg5OSEoKAgmJqaqiNGpaxduxZBQUEIDg4GAISFheH48ePYuHEjli9fruHoCCHk06Zwz+XatWto0aIF1q1bh9zcXAiFQqxbtw4tWrTA9evX1RGjwkpKSpCcnIxevXpJlPfq1QsXLlzQUFSEEFJ/KNxzmT59OgYOHIjIyEjxXZHLysoQHByMadOmISkpSeVBKkooFEIkEsHCwkKi3MLCAllZWRqKihBC6g+Fk8u1a9ckEgsA6OjoYPbs2fDw8FBpcB/q/TPaGGNSZUKhEEKhEMC740mEEEI+nMLJxcTEBJmZmXB0dJQof/z4MYyNjVUW2Ifg8/nQ1taW6qVkZ2dL9Wbi4uIQGRkJACgqKqqyzYkTJ6o+0A9UF2MCKC5FUVzyq4sxAXU3Lk1S+JjLyJEjERQUhNjYWDx+/BhPnjzBnj17EBwcjC+//FIdMSqMy+XC3d0dCQkJEuUJCQno0qWLRJm/vz9iYmIQExODxYsX12aYhBDyyVK457J69WrxxZJlZWUAAF1dXXz77bdYsWKFygNU1owZMxAQEAAPDw94eXlhy5YtyMzMxKRJkyTq8fl88Pl8ANX3XAghhMhPqSdRrl+/HsuXL8f9+/fBGEPLli1hYGCgjviUNnLkSOTk5GDRokUQCARwcXFBfHw8PTGTEEJqgdJPojQwMEDbtm1VGYvKhYSEICQkRO76b968AQDcunVLXSERQkitqvg+S01NVWl7NVE6uXyKMjIyAABjxozRbCCEEKJiQUFBKmvLwMBAfDihKhzGGFPZHD9yQqEQx48fh52dHfT19TUdDiGEqIRAIICZmZnKvtf4fD5sbGyqrUPJhRBCiMrRY44JIYSonFLHXP79918kJiYiOzsb5eXlEuN++uknlQRGCCHk46XwbrHIyEh8++234PP5sLS0lLidCofDqTM3rySEEKI5CicXW1tbhISE4D//+Y+6YtKYoqIi3L59G46OjnXuuh1CCFFWZmam+B6KqiDPAX2Fd4vl5eVh+PDhSgdVl92+fRvu7u5ITk6Gm5ubpsMhhJAPlpmZCScnJ5XegcTAwAC3bt2qNsEonFyGDx+OEydOSN1GhRBCSN0jFApRVFSEmJgYODk5fXB7t27dwpgxYyAUClWbXFq2bIn58+fj0qVLaNu2LXR1dSXGT5kyRfFoCSGEqJWTk1Ot7pFROLls2bIFRkZGOHv2LM6ePSsxjsPhUHIhhBCieHKhB2q9k3sqAjnxq1D2UgC9Jm1gMToMhg4+VdYvvH0Wz3fNwNtn6dAxa4KGfWfD/Iv/7VrMWN4NRbfPSk1n1L4vbGYc1Vhclb28tAdPN34JY7dBsJ56QO6Y1BEXKyuF8Mhy5P+1A2X5T8G1dIDFiJUwatdHLTGV5gvwfPdMFGcko+T5XZj3nALL0WESdfLPRePZ1vFS0zpGvoEWl6eWuICPdxvW1bg08VmU5/2VlxiJ/PM78fbJPwAAfTt3NB62DPotOskdU61iH6C8vJyVl5d/SBN1SnJyMgPAkpOTq62Xf2kPSx+vy3LPRLLipzeZIGYqu/m1ISsRPpJZ/232A3Yz2IAJYqay4qc3We6ZSJY+Xpe9vPK7uE7Z6xxWmicQv948/oelB2qzvKTtcsevjrjEdV9ksDtTm7KHS3xYZtgguWNSV1xZe2azO1OasFepR9nb5/dZzskIdjOIx4oyrqsppodM8OsUlnduB7v/YwcmiJkqVScvaTu79Y2JxHYszRPIFY/ycX2827CuxqWZz2LN76/HG79iOSd/YW8yUljx01vsaeR4dmuSKSvJeVJtLPJ+r8lL3vaUSi47duxgLi4uTE9Pj+np6bG2bduynTt3KhVoXSLvSru/sBN7tn2SRNnd/ziyrNg5Mutn7ZnN7v7HUaLs2fZv2INQzyrnITy2jt2aaMxExQVyRq++uMpFZezBYm+Wm7iVPdkyTuEvAHXEdWeKFctJ+K9EncywQezJxtFqiamyh8u6Vp1cJpnKNX9VxfUxb8O6HFdltfFZrKyq99f7ykVl7NZEY5Z3bke19TSVXBS+/cvatWvx7bffom/fvti7dy9iY2PRp08fTJo0CevWrVNtt6oOYmUlKM5IhqFLL4lyI5deeHPvgsxp3ty7CKP36hu69MabjGtgZaUyp8lPioJJ51HQ0jPUeFwvDiyCjnEjNOiq+F1V1RUXK30Ljq7kriaOrj6K7v6llpjkVV5cgLszbPHvtGbIXNsfbx6lyD1tfduGdTmuymrjs6iM8rdFYKJSaBuZq6xNVVL4mEt4eDg2btyIsWPHissGDRqENm3aYOHChZg+fbpKA6xryl4LgXIRdEwtJMq1TS1Q9jJL9jQvs6D9Xn0dUwtAVIayAiF0zawkxr25fwVvn/yDJhOiNB5X0b/nkZ8UheaLU+WOpTbiMmzbG7nH1sLA4XNwG7dA4c1TeJ1yECgXqSUmeXCtHNEkOBo867YQvXmF3BPrkbHEG80Xp0HPspVa4vqYt2Fdjauy2vosKiN73xzoNGgKQ+ceKmtTlRTuuQgEAqnn0ANAly5dIBAIVBLUR6HSbW8AAIxJl1WuDhn1ZZUDyEuKgl4zF+UO1KkwLtGb13i6eQysxkdCx7j6ZzfUZlwAYDl6PbiWrXB/jiNuBXGR9et3MPMZD2hpqy2mmhi09ISZ9xjwbNrD0MEHzSbvBdeyNfJOhivWUD3YhnU1rvfV5mdREcKjP+Plpd2w/n6/QieL1CalrnPZu3cv/u///k+iPDY2Fq1a1fzr7GOnY8wHtLRRli/5C0T0Khs6JhaypzG1lPrFUvYqG9DWgbZRQ4ny8rdFeHV5DxoNXaTxuN4+TUepMAOPwwb8rwJ7d6PSm+N10HLFHXAtWtR6XACgY9II1lMPoLykGKKCHOg0aILsvXPA5dtXG4+yMSmDo6UFffuOeJt1V6769Wkb1tW4KqvNz6IihPGrITyyDLazT4Jn0+6D21MXhZNLaGgoRo4ciaSkJHh7e4PD4eCvv/7CqVOnsHfvXnXEWKdwdLjg2bmjMD0BJh5DxOUF6Qkwdh0kcxr9ll54nXpYoqzwnxPQt/MAR0fyItRXV/aClb2FaRfFnoapjri4Vo5ovvSGxPgXcT9CVPwalqPXQ7ehtUbiqkyLy4OWeVOwslK8uhYHk04j1BKTMhhjKM5MBa+ZfI8Dr0/bsK7GVVltfhblJYxfBeGhJbCZdRz69h4f1Ja6KZxc/P39cfnyZaxbtw4HDhwAYwzOzs64cuUKXF1d1RFjndOwzww83RwAnr0HDFp6Ie/MFpTmZKLB/z9X/vneuSjLe4qm3+wEADT4YhJyT/4XWbtmoEG3r1F07yLykqLQ7NvdUm3nJUXB2G0wdN77FaWJuLS4PPCauUjMQ8vADACkymszLgAoun8ZZXlPwbPpgNK8p3hxYCHAysHvO1stMQFA8aNUAO8O2pe9foHiR6ng6HCh19QZAPDij1Dot/QE16IVyt+8Qm7CBhRnpsIq4BeNrau6ug3ralyV1eZnEaj5/SU8+jNe7J+PppN2gcu3E/eMtHhG0OIZKRyjuin1PBd3d3fExMSoOhaVSkpKwqpVq5CcnAyBQIA//vgDgwcPVknbpp1HQlSQA+HBRSjLF0CvqQtsZsSDy7cFAJS9FKA0N1Ncn9vIHjYz4/F813TknfoFOmZNYDlmA0w6+ku0+zbrX7z59y80+uFEnYrrQ6kjLlZajOy4H1H64gG09Ixg1K4vmk78FdqGZmqJCQAe/PS/H0/FGcl4dXEXdPm2aLUmAwAgKsqHYPtElL3Mgpa+KXi2rrD7vySF9tfXp21Yl+Oq7c8iUPP7K+90BFhZCZ78d5jEdPzBC9B4yEKl4lQnuW65/+rVK5iYmIj/r05FPU37888/cf78ebi5ucHf31+u5HL9+nW6KzIh5JOi6u81eduTq+fSoEEDCAQCNG7cGGZmZhIPCKvAGAOHw4FIVPOpoLXBz88Pfn5+mg6DEELqJbmSy+nTp2Fu/u5CnTNnzqg1IEIIIR8/uZJL165dxf/b29vD2tpaqvfCGMPjx49VG10tEAqF4ie00U05CSFENRQ+oG9vby/eRVZZbm4u7O3t68xuMXnFxcUhMjISAKp9UtvNcaq5+OlDOe/43yGyuhgTQHHVhOJSTF18zwN1M673t6EmKZxcKo6tvK+goAA8Xt28UrQ6/v7+4p5Zeno6hg0bVsMUhBBCaiJ3cpkxYwaAdw8Emz9/PgwMDMTjRCIRLl++jA4dOqg8QHXj8/ng89/dfkKVz5gmhJD6TO7kkpLy7u6ujDHcuHEDXC5XPI7L5aJ9+/aYNWuW6iNUUkFBAe7duycefvjwIVJTU2Fubl7tc58JIYR8OLmTS8VZYuPHj8f69evrzPUsVbl27Rp8fX3FwxU9r3HjxiE6OlpDUanWX96bNR0CAMBZ0wF85OrqdqyrcZGPg8LHXMLCwlBWViZVnpubCx0dnTqTdLp16wY5rg8lhBCiBgonl1GjRmHAgAEICQmRKN+7dy8OHTqE+Ph4lQVXl9CvOEI0r658DgH6LNZE4ee5XL58WWJ3U4Vu3brh8uXLKgmKEELIx03hnsvbt29l7hYrLS3FmzdvVBIU+bjVlV+X9MuSEM1RuOfSsWNHbNmyRap806ZNcHd3V0lQhBBCPm4K91yWLl2KHj16IC0tDd27dwcAnDp1ClevXsWJE8rdnpoQQsinReHk4u3tjYsXL2LVqlXYu3cv9PX10a5dO0RFRdWLxxwTQsj7aFewNKUeFtahQwf89ttvqo6FEELIJ0Kp5FLhzZs3KC0tlSirK9e5EPI++nVJSO1R+IB+UVERvvvuOzRu3BhGRkZo0KCBxIsQQghROLn88MMPOH36NCIiIqCnp4etW7ciNDQUTZo0wc6dO9URIyGEkI+MwrvFDh8+jJ07d6Jbt26YMGECfHx80LJlS9ja2uK3337D6NGj1REnIYSQj4jCPZeKh4IB746v5ObmAgA+++wzJCUlKRVEREQE7O3twePx4O7ujnPnzlVZNzAwEBwOR+rVpk0bcZ3o6GiZdYqLi5WKjxBCiGIUTi7NmzdHRkYGAMDZ2Rl79+4F8K5HY2ZmpnAAsbGxmDZtGubNm4eUlBT4+PjAz88PmZmZMuuvX78eAoFA/Hr8+DHMzc0xfPhwiXomJiYS9QQCwUf5MDNCCPkYKZxcxo8fj7S0NADA3Llzxcdepk+fjh9++EHhANauXYugoCAEBwfDyckJYWFhsLa2xsaNG2XWNzU1haWlpfh17do15OXlYfz48RL1OByORD1LS0uFYyOEEKIchY+5TJ8+Xfy/r68vbt++jWvXrqFFixZo3769Qm2VlJQgOTkZc+bMkSjv1asXLly4IFcbUVFR6NGjB2xtbSXKCwoKYGtrC5FIhA4dOmDx4sVwdXVVKD5CCCHKUajnUlpaCl9fX/z777/iMhsbGwwdOlThxAIAQqEQIpEIFhYWEuUWFhbIysqqcXqBQIA///wTwcHBEuWOjo6Ijo7GoUOHsHv3bvB4PHh7e+Pu3bsKx0gIIURxCvVcdHV18c8//4DD4ag0iPfbY4zJNY/o6GiYmZlh8ODBEuWenp7w9PQUD3t7e8PNzQ3h4eHYsGGDRF2hUAihUAjg3aOQCSGEfDiFj7mMHTsWUVFRKpk5n8+Htra2VC8lOztbqjfzPsYYtm3bhoCAAHC53GrramlpoWPHjjJ7LnFxcRgzZgzGjBmD+fPnK74QhBDyiZoyZYrS0yp8zKWkpARbt25FQkICPDw8YGhoKDF+7dq1crfF5XLh7u6OhIQEDBkyRFyekJCAQYMGVTvt2bNnce/ePQQFBdU4H8YYUlNT0bZtW6lx/v7+6Nq1KwAgPT0dw4YNkzt+Qgj5FJw4cQLh4eEQCoUSj4d/+vQprly5AgC4dOmSQm0qnFz++ecfuLm5AYDEsRdAeveWPGbMmIGAgAB4eHjAy8sLW7ZsQWZmJiZNmgTg3RlpT58+lbr6PyoqCp07d4aLi4tUm6GhofD09ESrVq3w6tUrbNiwAampqfjll1+k6vL5fPD5fADvbm1DCCH1zfz58xESEgI7OztxGYfDwYQJE7By5Uql2pQrufz9999wcXGBlpYWzpw5o9SMqjJy5Ejk5ORg0aJFEAgEcHFxQXx8vPjsL4FAIHXNy8uXLxEXF4f169fLbDM/Px8TJ05EVlYWTE1N4erqiqSkJHTq1EmlsRNCyKegoKAA48aNkyrX1tYW79lRlFzJxdXVFQKBAI0bN0bz5s1x9epVNGzYUKkZyhISEoKQkBCZ46Kjo6XKTE1Nq+1lrFu3DuvWrVNVeIQQ8kkLDAyUWT5hwgSl25TrgL6ZmZn4TKqMjAyUl5crPUNCCCF1y4YNG9CsWTOpV3h4uNJtytVzqTjobWVlBQ6HAw8PD2hra8us++DBA6WDIYQQUvsSExPF/zPG8OzZM4SHh6Nfv35KtylXctmyZQuGDh2Ke/fuYcqUKfj6669hbGys9EwJIYTUHS1atJAYbtmyJVxcXDB48OAqd5nVRO6zxfr06QMASE5OxtSpUym5EELIJ8zQ0BAvXrxAWVkZdHQUf2ixwlNs375d4ZkQQgj5uOjp6eHWrVtKT694OiKEEPJJ8fLykrh4srKKiye/+OILnD59Wu42KbkQQkg9t2LFihrrLFq0SKE2KbkQQkg9J8+Fkp999plCbSp840pCCCGflsuXL6N///4YOXKk+HKSf/75B3///bfSbVJyIYSQem727Nnw9fWFk5MTQkNDAQBv3779oDvF024xQgip57KysjBz5kwUFRXB29sbAODm5vZBD1isEz2XiIgI2Nvbg8fjwd3dHefOnauybmJiIjgcjtTr9u3bEvXi4uLg7OwMPT09ODs7448//lD3YhBCyEfJxcUF9+/fh4GBAYqLiwEAz549A4/HU7pNjSeX2NhYTJs2DfPmzUNKSgp8fHzg5+cndSfk9925cwcCgUD8atWqlXjcxYsXMXLkSAQEBCAtLQ0BAQEYMWIELl++rO7FIYSQj06XLl0QFBSEqKgolJWVYevWrRgzZoz44nllaDy5rF27FkFBQQgODoaTkxPCwsJgbW2NjRs3Vjtd48aNYWlpKX5VvtdZWFgYevbsiblz58LR0RFz585F9+7dERYWpualIYSQj8/Zs2dhYmKCgwcPwtnZGefPn8eoUaPEx1+UodFjLiUlJUhOTsacOXMkynv16oULFy5UO62rqyuKi4vh7OyMH3/8Eb6+vuJxFy9exPTp0yXq9+7dm5ILIYTIcOjQIZW3qdGei1AohEgkgoWFhUS5hYUFsrKyZE5jZWWFLVu2IC4uDvv374eDgwO6d++OpKQkcZ2srCyF2iSEEKJadeJssfcfj8wYq/KRyQ4ODnBwcBAPe3l54fHjx1i9ejU+//xzhdsUCoUQCoUAIH5mDSGE1CfW1tZV3v4FePd9yhjDkydP5G5To8mFz+dDW1tbqkeRnZ0t1fOojqenJ2JiYsTDlpaWcrcZFxeHyMhIAKj26ZaEEPKpqvw8F1XRaHLhcrlwd3dHQkIChgwZIi5PSEjAoEGD5G4nJSUFVlZW4mEvLy8kJCRIHHc5ceIEunTpIjVtxYPQACA9PR3Dhg1TZlEIIeSj9f7zXFRB47vFZsyYgYCAAHh4eMDLywtbtmxBZmYmJk2aBACYO3cunj59ip07dwJ4dyaYnZ0d2rRpg5KSEsTExCAuLg5xcXHiNqdOnYrPP/8cK1euxKBBg3Dw4EGcPHkSf/31l9T8+Xw++Hw+AOq5EELqp3nz5qFdu3YYOXKk1LgnT54gMTERY8aMUahNjZ+KPHLkSISFhWHRokXo0KEDkpKSEB8fD1tbWwCAQCCQuOalpKQEs2bNQrt27eDj44O//voLR48exdChQ8V1unTpgj179mD79u1o164doqOjERsbi86dO9f68hFCSF33559/SpxxCwCnTp0CABgbG9d4aYgsGu+5AEBISAhCQkJkjouOjpYYnj17NmbPnl1jm8OGDaNdXIQQIoeCggI0btxYouy7777DrVu3YGpqitzcXIXb1HjPhRBCiGaZmZnh0aNH4uFnz57h9evXeP36NYqKipS6DUyd6LkQQgjRnB49emDq1Kn44YcfoKuri7Nnz8LAwAD/93//B8aY1C4zeVDPhRBC6rn//Oc/sLCwwNdff40FCxZg2LBh2Lp1K3Jzc2FmZobFixcr3Cb1XAghpJ4zNTXF5s2bJcrs7e0lLkxXFPVcCCGEqBwlF0IIISpHyYUQQojKUXIhhBCicpRcCCGEqBwlF0IIISpHyYUQQojK1YnkEhERAXt7e/B4PLi7u+PcuXNV1t2/fz969uyJRo0awcTEBF5eXjh+/LhEnejoaHA4HKlXcXGxuheFEEII6kByiY2NxbRp0zBv3jykpKTAx8cHfn5+EndCriwpKQk9e/ZEfHw8kpOT4evriwEDBiAlJUWinomJCQQCgcRLmfvjEEIIUZzGr9Bfu3YtgoKCEBwcDODd81qOHz+OjRs3Yvny5VL1w8LCJIaXLVuGgwcP4vDhw3B1dRWXczgcWFpaqjV2Qgghsmm051JSUoLk5GT06tVLorxXr164cOGCXG2Ul5fj9evXMDc3lygvKCiAra0tmjVrhv79+0v1bAghhKiPRpOLUCiESCSSera9hYUFsrKy5GpjzZo1KCwsxIgRI8Rljo6OiI6OxqFDh7B7927weDx4e3vj7t27Ko2fEEKIbBrfLQa824VVGWNMqkyW3bt3Y+HChTh48KDEg248PT3h6ekpHvb29oabmxvCw8OxYcMGiTaEQiGEQiEA4OHDhx+yGIQQQv4/jSYXPp8PbW1tqV5Kdna2VG/mfbGxsQgKCsK+ffvQo0ePautqaWmhY8eOMnsucXFxiIyMBAAUFRUpuASEEEJk0ehuMS6XC3d3dyQkJEiUJyQkoEuXLlVOt3v3bgQGBmLXrl3o169fjfNhjCE1NRVWVlZS4/z9/RETE4OYmBilnllACCFEmsZ3i82YMQMBAQHw8PCAl5cXtmzZgszMTEyaNAkAMHfuXDx9+hQ7d+4E8C6xjB07FuvXr4enp6e416Ovrw9TU1MAQGhoKDw9PdGqVSu8evUKGzZsQGpqKn755Rep+fP5fPD5fADUcyGEEFXReHIZOXIkcnJysGjRIggEAri4uCA+Ph62trYAAIFAIHHNy+bNm1FWVobJkydj8uTJ4vJx48YhOjoaAJCfn4+JEyciKysLpqamcHV1RVJSEjp16lSry0YIIfWVxpMLAISEhCAkJETmuIqEUSExMbHG9tatW4d169apIDJCCCHK0PgV+oQQQj49lFwIIYSoHCUXQgghKkfJhRBCiMpRciGEEKJylFwIIYSoHCUXQgghKkfJhRBCiMpRciGEEKJylFwIIYSoHCUXQgghKvdJJ5eIiAjY29uDx+PB3d0d586d03RIhBBSL3yyySU2NhbTpk3DvHnzkJKSAh8fH/j5+UncYZkQQoh6fLLJZe3atQgKCkJwcDCcnJwQFhYGa2trbNy4UdOhEULIJ++TTC4lJSVITk5Gr169JMp79eqFCxcuaCgqQgipPz7J5CIUCiESiWBhYSFRbmFhIX5yJSGEEPWpEw8LUxcOhyMxzBiTKhMKhRAKhQCAhw8f1lpshBDyKfskkwufz4e2trZULyU7O1uqNxMXF4fIyEgAQFFRUZVtTpw4UfWBfqC6GBNAcSmK4pJfXYwJqLtxadInmVy4XC7c3d2RkJCAIUOGiMsTEhIwaNAgibr+/v7o2rUrACA9PR3Dhg2r1VgJIeRT9EkmFwCYMWMGAgIC4OHhAS8vL2zZsgWZmZmYNGmSRD0+nw8+nw+g+p4LIYQQ+X2yyWXkyJHIycnBokWLIBAI4OLigvj4eNja2mo6NEII+eR9sskFAEJCQhASEiJ3/Tdv3gAAbt26pa6QCCGkVlV8n6Wmpqq0vZp80slFURkZGQCAMWPGaDYQQghRsaCgIJW1ZWBgID6cUBUOY4ypbI4fOaFQiOPHj8POzg76+voqbfvhw4eYP38+Fi9eDHt7e5W2ray6GBNAcSmK4pJfXYwJUH9cAoEAZmZmKvte4/P5sLGxqbYO9Vwq4fP5GD16tFraNjAwgIGBAdq0aQNHR0e1zENRdTEmgOJSFMUlv7oYE1B34/oQn+QV+oQQQjSLkkst4fP5+Prrr2vcT1mb6mJMAMWlKIpLfnUxJqDuxvUh6JgLIYQQlaOeCyGEEJWj5EIIIUTlKLl8gMDAQHA4HHA4HOjq6qJ58+aYNWsWCgsLkZGRIR7H4XDA5XLRsmVLLFmyBJX3REZGRsLHxwcNGjRAgwYN0KNHD1y5ckWjMaWnp8Pf3x92dnbgcDgICwv7kNWkVKwVJk6cCG1tbezZs0ctMciKafDgwVLliYmJ4HA4yM/Plyh3cHAAl8vF06dP1R4Xh8PBihUrJMoPHDggvtN3RYwVr0aNGsHPzw9paWkai6my2lpX76uI8/1bPwHvLrTmcDgIDAystThqWl+bN29G+/btYWhoCDMzM7i6umLlypVqj0/VKLl8oD59+kAgEODBgwdYsmQJIiIiMGvWLPH4kydPQiAQ4O7duwgNDcXSpUuxbds28fjExER8+eWXOHPmDC5evAgbGxv06tXrgz6AHxpTUVERmjdvjhUrVsDS0lLpOD401qKiIsTGxuKHH35AVFSUWuNQxl9//YXi4mIMHz4c0dHRap8fj8fDypUrkZeXV229O3fuQCAQ4OjRo8jLy0OfPn3w8uVLjcZU2+vqfdbW1tizZ4/4LhwAUFxcjN27d9d4vYYq1bS+oqKiMGPGDEyZMgVpaWk4f/48Zs+ejYKCglqLUVUouXwgPT09WFpawtraGl999RVGjx6NAwcOiMc3bNgQlpaWsLW1xejRo9GlSxdcv35dPP63335DSEgIOnToAEdHR0RGRqK8vBynTp3SWEwdO3bEqlWrMGrUKOjp6Skdx4fGum/fPjg7O2Pu3Lk4f/68+A4KdUVUVBS++uorBAQEYNu2bVD3uTE9evSApaUlli9fXm29xo0bw9LSEp06dcKaNWuQlZWFS5cuaTSm2l5X73Nzc4ONjQ32798vLtu/fz+sra3h6upaa3HUtL4OHz6MESNGICgoCC1btkSbNm3w5ZdfYvHixbUWo6pQclExfX19lJaWyhx37do1XL9+HZ07d65y+qKiIpSWlsLc3LzOxFSbKscaFRWFMWPGwNTUFH379sX27ds1HN3/vH79Gvv27cOYMWPQs2dPFBYWIjExUa3z1NbWxrJlyxAeHo4nT57INU3FFdlVbf/aiEkT60qW8ePHS7yHtm3bhgkTJtRqDDWtL0tLS1y6dAmPHj2q1bjUgZKLCl25cgW7du1C9+7dxWVdunSBkZERuFwuOnbsiBEjRmDs2LFVtjFnzhw0bdoUPXr0qDMx1ZbKsd69exeXLl3CyJEjAby739v27dtRXl6u9jiOHDkCIyMjiZefn59EnT179qBVq1Zo06YNtLW1MWrUqFrZdTdkyBB06NABCxYsqLFuTk4OQkNDYWxsjE6dOmksJk2tq/cFBATgr7/+QkZGBh49eoTz589r5D6C1a2vBQsWwMzMDHZ2dnBwcEBgYCD27t1bK+97VaPk8oEqvoh4PB68vLzw+eefIzw8XDw+NjYWqampSEtLQ2xsLA4ePIg5c+bIbOvnn3/G7t27sX//fvB4vDoRk7pVFWtUVBR69+4tvqisb9++KCwsxMmTJ9Uek6+vL1JTUyVeW7dulahT0auqMGbMGOzfv1/qgL86rFy5Ejt27MDNmzdljm/WrBmMjIzA5/Nx69Yt7Nu3D40bN9ZYTJpcV5Xx+Xz069cPO3bswPbt29GvXz+NXbRY1fqysrLCxYsXcePGDUyZMgWlpaUYN24c+vTp8/ElGEaUNm7cONajRw929+5dlpGRwUpKSsTjHj58yACwlJQUiWmWL1/OdHR02Js3byTKV61axUxNTdnVq1frTEyMMWZra8vWrVv3QTEpGmtZWRmzsrJiHA6HaWtri18A2IgRI9QSS+WYBg0aJFV+5swZBoDl5eWx9PR0BoBpaWlJxRcREVErcfXt25cNGjSI/fHHH6ziY1wR4/Xr19m9e/fYy5cv1RKLIjFpYl1VF+eRI0eYnZ0ds7OzY0ePHmWMMTZo0CA2bty4Wo2DMdnrS5Zz584xAOz06dNqj1GV6MaVH8jQ0BAtW7aUu762tjbKyspQUlIi7p2sWrUKS5YswfHjx+Hh4VEnYqotsmKNj4/H69evkZKSAm1tbXH57du3MXr0aOTk5KBhw4a1GmdlUVFR+Pzzz/HLL79IlP/666+IiorCt99+q/YYVqxYgQ4dOqB169ZS4+zt7WFmZqb2GOSJqS6sq8r69OmDkpISAEDv3r1rdd7vq24bVubs7AwAEqfofwwouahZTk4OsrKyUFZWhhs3bmD9+vXw9fWFiYkJgHe7wubPn49du3bBzs4OWVlZACDe16+JmEpKSsTd9ZKSEjx9+hSpqakwMjJSKGkpKyoqCv369UP79u0lytu0aYNp06YhJiYGU6dOVXscspSWluLXX3/FokWL4OLiIjEuODgYP//8M9LS0qRiV7W2bdti9OjRErs7Ne39mOrKuqpMW1tb/LCryj9cNEHWNvz222/RpEkTfPHFF2jWrBkEAgGWLFmCRo0awcvLS4PRKo6OuahZjx49YGVlBTs7O0ycOBF9+/ZFbGyseHxERARKSkowbNgwWFlZiV+rV6/WWEzPnj2Dq6srXF1dIRAIsHr1ari6uiI4OFhtMVV4/vw5jh49Cn9/f6lxHA4HQ4cO1eg1L0lJScjJycGQIUOkxrVq1Qpt27attfgWL15c66f01qRyTIcOHaoz66oyExMT8Q8pTXt/G/bo0QOXLl3C8OHD0bp1a/j7+4PH4+HUqVMa7a0rg25cSQghROWo50IIIUTlKLkQQghROUouhBBCVI6SCyGEEJWj5EIIIUTlKLkQQghROUouhBBCVI6SCyGEEJWj5EIIIUTlKLkQQghROUouhBBCVI6SCyGEEJWj5EIIIUTlKLkQQghROUouhBBCVI6SCyGEEJWj5EIIIUTlKLkQQghROR1NB0CAm+M4amnXeQc9wbou2bJli1ranThxolraJeRDUM+FEEKIylFyIYQQonKUXAghhKgcJRfyUUpMTASHw0F+fr6mQyGEyEDJpZ7KysrC1KlT0bJlS/B4PFhYWOCzzz7Dpk2bUFRUpOnwPhocDgcHDhzQdBgy0TZWn8DAQAwePFjTYdRpdLZYPfTgwQN4e3vDzMwMy5YtQ9u2bVFWVoZ///0X27ZtQ5MmTTBw4EBNh0k+wKe6jUtLS6Grq6vpMIgcqOdSD4WEhEBHRwfXrl3DiBEj4OTkhLZt28Lf3x9Hjx7FgAEDxHVfvnyJiRMnonHjxjAxMcEXX3yBtLQ08fiFCxeiQ4cO+PXXX2FnZwdTU1OMGjUKr1+/FtdhjOHnn39G8+bNoa+vj/bt2+P333+vNsaYmBh4eHjA2NgYlpaW+Oqrr5CdnS1V7/z582jfvj14PB46d+6MGzduiMc9evQIAwYMQIMGDWBoaIg2bdogPj5ePP7mzZvo27cvjIyMYGFhgYCAAAiFQvH4bt26YcqUKZg9ezbMzc1haWmJhQsXisfb2dkBAIYMGQIOhyMergs+hm1sZ2eHxYsX46uvvoKRkRGaNGmC8PBwiTocDgebNm3CoEGDYGhoiCVLlgAADh8+DHd3d/B4PDRv3hyhoaEoKyuTiNnGxgZ6enpo0qQJpkyZIh5XUlKC2bNno2nTpjA0NETnzp2RmJgoHh8dHQ0zMzMcP34cTk5OMDIyQp8+fSAQCMRt79ixAwcPHgSHwwGHw5GYnrxDyaWeycnJwYkTJzB58mQYGhrKrMPhvLvuhjGGfv36ISsrC/Hx8UhOToabmxu6d++O3Nxccf379+/jwIEDOHLkCI4cOYKzZ89ixYoV4vE//vgjtm/fjo0bNyI9PR3Tp0/HmDFjcPbs2SrjLCkpweLFi5GWloYDBw7g4cOHCAwMlKr3ww8/YPXq1bh69SoaN26MgQMHorS0FAAwefJkvH37FklJSbhx4wZWrlwJIyMjAIBAIEDXrl3RoUMHXLt2DceOHcPz588xYsQIifZ37NgBQ0NDXL58GT///DMWLVqEhIQEAMDVq1cBANu3b4dAIBAPa9rHso0BYNWqVWjXrh2uX7+OuXPnYvr06eL1W2HBggUYNGgQbty4gQkTJuD48eMYM2YMpkyZgps3b2Lz5s2Ijo7G0qVLAQC///471q1bh82bN+Pu3bs4cOAA2rZtK25v/PjxOH/+PPbs2YO///4bw4cPR58+fXD37l1xnaKiIqxevRq//vorkpKSkJmZiVmzZgEAZs2ahREjRogTjkAgQJcuXeTZNPUK7RarZ+7duwfGGBwcHCTK+Xw+iouLAbz7Ul65ciXOnDmDGzduIDs7G3p6egCA1atX48CBA/j999/FF++Vl5cjOjoaxsbGAICAgACcOnUKS5cuRWFhIdauXYvTp0/Dy8sLANC8eXP89ddf2Lx5M7p27SozzgkTJoj/b968OTZs2IBOnTqhoKBAnCCAd188PXv2BPAuETRr1gx//PEHRowYgczMTPj7+4u/WJo3by6ebuPGjXBzc8OyZcvEZdu2bYO1tTX+/fdftG7dGgDQrl07LFiwAADQqlUr/Pe//8WpU6fQs2dPNGrUCABgZmYGS0tL+TeCmn0s2xgAvL29MWfOHABA69atcf78eaxbt068TQHgq6++kng/BAQEYM6cORg3bpx4XosXL8bs2bOxYMECZGZmwtLSEj169ICuri5sbGzQqVMnAO+S5O7du/HkyRM0adIEwLtkcezYMWzfvl38figtLcWmTZvQokULAMB3332HRYsWAQCMjIygr6+Pt2/f1qntXtdQcqmnKn65Vrhy5QrKy8sxevRovH37FgCQnJyMgoICNGzYUKLumzdvcP/+ffGwnZ2d+EsHAKysrMS7sG7evIni4mKJLwvgXc/E1dW1yvhSUlKwcOFCpKamIjc3F+Xl5QCAzMxMODs7i+tVfJkBgLm5ORwcHHDr1i0AwJQpU/Dtt9/ixIkT6NGjB/z9/dGuXTvxsp05c0YiUVW4f/++RHKprPKy1XV1fRsDktuvYjgsLEyizMPDQ2I4OTkZV69eFfdUAEAkEqG4uBhFRUUYPnw4wsLC0Lx5c/Tp0wd9+/bFgAEDoKOjg+vXr4MxJt6+Fd6+fSuxDgwMDMSJ5f3lJfKh5FLPtGzZEhwOB7dv35Yor/hVr6+vLy4rLy+HlZWVzP3JZmZm4v/fP8DK4XDEyaDi79GjR9G0aVOJehW/lN9XWFiIXr16oVevXoiJiUGjRo2QmZmJ3r17o6SkpMZlrPhSDQ4ORu/evXH06FGcOHECy5cvx5o1a/D999+jvLwcAwYMwMqVK6Wmt7KykmvZ6qqPYRtX5/2k+P6uvfLycoSGhmLo0KFS0/J4PFhbW+POnTtISEjAyZMnERISglWrVuHs2bMoLy+HtrY2kpOToa2tLTFt5R8aspaXMbqdkiIoudQzDRs2RM+ePfHf//4X33//fZX75AHAzc0NWVlZ0NHRUfpgtbOzM/T09JCZmVnt7pHKbt++DaFQiBUrVsDa2hoAcO3aNZl1L126BBsbGwBAXl4e/v33Xzg6OorHW1tbY9KkSZg0aRLmzp2LyMhIfP/993Bzc0NcXBzs7Oygo6P8x0BXVxcikUjp6dXhY9jGFS5duiQ1XHn7yeLm5oY7d+6gZcuWVdbR19fHwIEDMXDgQEyePBmOjo64ceMGXF1dIRKJkJ2dDR8fH4VirYzL5da57V7XUHKphyIiIuDt7Q0PDw8sXLgQ7dq1g5aWFq5evYrbt2/D3d0dANCjRw94eXlh8ODBWLlyJRwcHPDs2TPEx8dj8ODBUrsrZDE2NsasWbMwffp0lJeX47PPPsOrV69w4cIFGBkZifebV2ZjYwMul4vw8HBMmjQJ//zzDxYvXiyz/UWLFqFhw4awsLDAvHnzwOfzxdcfTJs2DX5+fmjdujXy8vJw+vRpODk5AXh3zCEyMhJffvklfvjhB/D5fNy7dw979uxBZGSk1K/aqtjZ2eHUqVPw9vaGnp4eGjRoINd06lbXt3GF8+fP4+eff8bgwYORkJCAffv24ejRo9XO76effkL//v1hbW2N4cOHQ0tLC3///Tdu3LiBJUuWIDo6GiKRCJ07d4aBgQF+/fVX6Ovrw9bWFg0bNsTo0aMxduxYrFmzBq6urhAKhTh9+jTatm2Lvn37yrV+7ezscPz4cdy5cwcNGzaEqakpnSL9HjpbrB5q0aIFUlJS0KNHD8ydOxft27eHh4cHwsPDMWvWLPEXOYfDQXx8PD7//HNMmDABrVu3xqhRo5CRkQELCwu557d48WL89NNPWL58OZycnNC7d28cPnwY9vb2Mus3atQI0dHR2LdvH5ydnbFixQqsXr1aZt0VK1Zg6tSpcHd3h0AgwKFDh8DlcgG82w8/efJkODk5oU+fPnBwcEBERAQAoEmTJjh//jxEIhF69+4NFxcXTJ06FaamptDSkv9jsWbNGiQkJMDa2rrG4wu1qa5v4wozZ85EcnIyXF1dsXjxYqxZswa9e/eudprevXvjyJEjSEhIQMeOHeHp6Ym1a9fC1tYWwLvdeZGRkfD29ka7du1w6tQpHD58WHxMZfv27Rg7dixmzpwJBwcHDBw4EJcvXxb3kuXx9ddfw8HBAR4eHmjUqBHOnz8v97T1BYfRjkSNo1vu1w90y31JdnZ2mDZtGqZNm6bpUIgaUM+FEEKIylFyIYQQonJ0QJ8QohEZGRmaDoGoEfVcCCGEqBwlF0IIISpHyYUQQojK0anIhBBCVI56LoQQQlSOkgshhBCVo+RCCCFE5Si5EEIIUTlKLoQQQlSOkgshhBCVo+RCCCFE5Si5EEIIUTlKLoQQQlSOkgshhBCVo+RCCCFE5Si5EEIIUTlKLoQQQlSOkgshhBCVo+RCCCFE5Si5EEIIUTlKLoQQQlSOkgshhBCVo+RCCCFE5Si5EEIIUTlKLoQQQlSOkgshhBCVo+RCCCFE5Si5EEIIUTlKLoQQQlSOkgshhBCV+3+SiieJ36+sXQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 400x300 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<ggplot: (8780481888690)>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene_absence_file = figure_dir+'/'+'gene_absence_plot.svg'\n",
    "\n",
    "gene_absence_plot = (\n",
    "    p9.ggplot(gene_absence_rate,\n",
    "              p9.aes(x='gene',\n",
    "                     y='fraction_cells',\n",
    "                     fill='gene_present')) +\n",
    "    p9.geom_bar(stat='identity') +\n",
    "    p9.geom_text(data=gene_absence_rate.query('gene_present == \"Gene absent\"'),\n",
    "           mapping=p9.aes(x='gene',\n",
    "                          y=1.1,\n",
    "                          label='fraction_cells'),\n",
    "                          format_string='{:.2f}',\n",
    "                          size=10,\n",
    "                          color=CBPALETTE[6]) +\n",
    "    p9.facet_grid('sample~') +\n",
    "    p9.ggtitle('Viral gene presence in infected cells') +\n",
    "    p9.labs(x = '', y='fraction of cells') +\n",
    "    p9.scale_y_continuous(limits=[0,1.15],\n",
    "                          breaks=[0,0.25,0.5,0.75,1])+\n",
    "    p9.theme(figure_size=(4, 3),\n",
    "       plot_title=p9.element_text(size=12),\n",
    "       axis_title=p9.element_text(size=10),\n",
    "       axis_text=p9.element_text(size=10, color=\"black\"),\n",
    "       legend_position='bottom',\n",
    "       legend_box_spacing=0.2,\n",
    "       legend_title=p9.element_blank(),\n",
    "       legend_text=p9.element_text(size=10)) +\n",
    "    p9.scale_fill_manual([CBPALETTE[6],CBPALETTE[0]]))\n",
    "\n",
    "gene_absence_plot.save(filename = gene_absence_file)\n",
    "display(gene_absence_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make composite Fig 2\n",
    "Make a composite figure showing viral transcription, number of viral genes per cell, and rate of viral gene absence.\n",
    "\n",
    "**N.b. For some reason this cell and the ones immediatebly above need to be run manually to generate a correct figure. When run automatically, panel A is oversized.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version='1.0' encoding='ASCII' standalone='yes'?>\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" version=\"1.1\" width=\"32.0cm\" viewBox=\"0 0 32.0cm 26.0cm\" height=\"26.0cm\">\n",
       "  <g>\n",
       "    <g transform=\"translate(0, 0) scale(2) \">\n",
       "      <g>\n",
       "        <path d=\"M0 0 L0 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M20 0 L20 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M40 0 L40 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M60 0 L60 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M80 0 L80 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M100 0 L100 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M120 0 L120 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M140 0 L140 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M160 0 L160 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M180 0 L180 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M200 0 L200 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M220 0 L220 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M240 0 L240 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M260 0 L260 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M280 0 L280 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M300 0 L300 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M320 0 L320 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M340 0 L340 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M360 0 L360 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M380 0 L380 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M400 0 L400 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M420 0 L420 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M440 0 L440 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M460 0 L460 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M480 0 L480 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M500 0 L500 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M520 0 L520 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M540 0 L540 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M560 0 L560 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M580 0 L580 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M600 0 L600 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M620 0 L620 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M640 0 L640 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M660 0 L660 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M680 0 L680 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M700 0 L700 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M720 0 L720 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M740 0 L740 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M760 0 L760 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M780 0 L780 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M800 0 L800 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M820 0 L820 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M840 0 L840 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M860 0 L860 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M880 0 L880 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M900 0 L900 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M920 0 L920 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M940 0 L940 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M960 0 L960 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M980 0 L980 1000\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 0 L1000 0\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 20 L1000 20\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 40 L1000 40\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 60 L1000 60\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 80 L1000 80\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 100 L1000 100\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 120 L1000 120\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 140 L1000 140\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 160 L1000 160\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 180 L1000 180\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 200 L1000 200\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 220 L1000 220\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 240 L1000 240\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 260 L1000 260\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 280 L1000 280\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 300 L1000 300\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 320 L1000 320\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 340 L1000 340\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 360 L1000 360\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 380 L1000 380\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 400 L1000 400\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 420 L1000 420\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 440 L1000 440\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 460 L1000 460\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 480 L1000 480\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 500 L1000 500\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 520 L1000 520\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 540 L1000 540\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 560 L1000 560\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 580 L1000 580\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 600 L1000 600\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 620 L1000 620\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 640 L1000 640\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 660 L1000 660\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 680 L1000 680\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 700 L1000 700\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 720 L1000 720\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 740 L1000 740\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 760 L1000 760\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 780 L1000 780\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 800 L1000 800\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 820 L1000 820\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 840 L1000 840\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 860 L1000 860\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 880 L1000 880\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 900 L1000 900\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 920 L1000 920\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 940 L1000 940\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 960 L1000 960\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <path d=\"M0 980 L1000 980\" stroke-width=\"0.5\" stroke=\"black\"/>\n",
       "        <text x=\"0\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">0</text>\n",
       "        <text x=\"20\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">20</text>\n",
       "        <text x=\"40\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">40</text>\n",
       "        <text x=\"60\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">60</text>\n",
       "        <text x=\"80\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">80</text>\n",
       "        <text x=\"100\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">100</text>\n",
       "        <text x=\"120\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">120</text>\n",
       "        <text x=\"140\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">140</text>\n",
       "        <text x=\"160\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">160</text>\n",
       "        <text x=\"180\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">180</text>\n",
       "        <text x=\"200\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">200</text>\n",
       "        <text x=\"220\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">220</text>\n",
       "        <text x=\"240\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">240</text>\n",
       "        <text x=\"260\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">260</text>\n",
       "        <text x=\"280\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">280</text>\n",
       "        <text x=\"300\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">300</text>\n",
       "        <text x=\"320\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">320</text>\n",
       "        <text x=\"340\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">340</text>\n",
       "        <text x=\"360\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">360</text>\n",
       "        <text x=\"380\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">380</text>\n",
       "        <text x=\"400\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">400</text>\n",
       "        <text x=\"420\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">420</text>\n",
       "        <text x=\"440\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">440</text>\n",
       "        <text x=\"460\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">460</text>\n",
       "        <text x=\"480\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">480</text>\n",
       "        <text x=\"500\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">500</text>\n",
       "        <text x=\"520\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">520</text>\n",
       "        <text x=\"540\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">540</text>\n",
       "        <text x=\"560\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">560</text>\n",
       "        <text x=\"580\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">580</text>\n",
       "        <text x=\"600\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">600</text>\n",
       "        <text x=\"620\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">620</text>\n",
       "        <text x=\"640\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">640</text>\n",
       "        <text x=\"660\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">660</text>\n",
       "        <text x=\"680\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">680</text>\n",
       "        <text x=\"700\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">700</text>\n",
       "        <text x=\"720\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">720</text>\n",
       "        <text x=\"740\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">740</text>\n",
       "        <text x=\"760\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">760</text>\n",
       "        <text x=\"780\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">780</text>\n",
       "        <text x=\"800\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">800</text>\n",
       "        <text x=\"820\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">820</text>\n",
       "        <text x=\"840\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">840</text>\n",
       "        <text x=\"860\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">860</text>\n",
       "        <text x=\"880\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">880</text>\n",
       "        <text x=\"900\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">900</text>\n",
       "        <text x=\"920\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">920</text>\n",
       "        <text x=\"940\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">940</text>\n",
       "        <text x=\"960\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">960</text>\n",
       "        <text x=\"980\" y=\"10.0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">980</text>\n",
       "        <text x=\"0\" y=\"0\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">0</text>\n",
       "        <text x=\"0\" y=\"20\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">20</text>\n",
       "        <text x=\"0\" y=\"40\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">40</text>\n",
       "        <text x=\"0\" y=\"60\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">60</text>\n",
       "        <text x=\"0\" y=\"80\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">80</text>\n",
       "        <text x=\"0\" y=\"100\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">100</text>\n",
       "        <text x=\"0\" y=\"120\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">120</text>\n",
       "        <text x=\"0\" y=\"140\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">140</text>\n",
       "        <text x=\"0\" y=\"160\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">160</text>\n",
       "        <text x=\"0\" y=\"180\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">180</text>\n",
       "        <text x=\"0\" y=\"200\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">200</text>\n",
       "        <text x=\"0\" y=\"220\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">220</text>\n",
       "        <text x=\"0\" y=\"240\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">240</text>\n",
       "        <text x=\"0\" y=\"260\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">260</text>\n",
       "        <text x=\"0\" y=\"280\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">280</text>\n",
       "        <text x=\"0\" y=\"300\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">300</text>\n",
       "        <text x=\"0\" y=\"320\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">320</text>\n",
       "        <text x=\"0\" y=\"340\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">340</text>\n",
       "        <text x=\"0\" y=\"360\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">360</text>\n",
       "        <text x=\"0\" y=\"380\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">380</text>\n",
       "        <text x=\"0\" y=\"400\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">400</text>\n",
       "        <text x=\"0\" y=\"420\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">420</text>\n",
       "        <text x=\"0\" y=\"440\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">440</text>\n",
       "        <text x=\"0\" y=\"460\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">460</text>\n",
       "        <text x=\"0\" y=\"480\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">480</text>\n",
       "        <text x=\"0\" y=\"500\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">500</text>\n",
       "        <text x=\"0\" y=\"520\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">520</text>\n",
       "        <text x=\"0\" y=\"540\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">540</text>\n",
       "        <text x=\"0\" y=\"560\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">560</text>\n",
       "        <text x=\"0\" y=\"580\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">580</text>\n",
       "        <text x=\"0\" y=\"600\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">600</text>\n",
       "        <text x=\"0\" y=\"620\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">620</text>\n",
       "        <text x=\"0\" y=\"640\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">640</text>\n",
       "        <text x=\"0\" y=\"660\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">660</text>\n",
       "        <text x=\"0\" y=\"680\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">680</text>\n",
       "        <text x=\"0\" y=\"700\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">700</text>\n",
       "        <text x=\"0\" y=\"720\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">720</text>\n",
       "        <text x=\"0\" y=\"740\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">740</text>\n",
       "        <text x=\"0\" y=\"760\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">760</text>\n",
       "        <text x=\"0\" y=\"780\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">780</text>\n",
       "        <text x=\"0\" y=\"800\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">800</text>\n",
       "        <text x=\"0\" y=\"820\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">820</text>\n",
       "        <text x=\"0\" y=\"840\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">840</text>\n",
       "        <text x=\"0\" y=\"860\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">860</text>\n",
       "        <text x=\"0\" y=\"880\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">880</text>\n",
       "        <text x=\"0\" y=\"900\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">900</text>\n",
       "        <text x=\"0\" y=\"920\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">920</text>\n",
       "        <text x=\"0\" y=\"940\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">940</text>\n",
       "        <text x=\"0\" y=\"960\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">960</text>\n",
       "        <text x=\"0\" y=\"980\" font-size=\"8\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">980</text>\n",
       "      </g>\n",
       "      <g><metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2022-11-03T14:15:02.135659</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.5.2, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 231.084201  L 351.56875 231.084201  L 351.56875 -0  L 0 -0  z \" style=\"fill: #ffffff; stroke: #ffffff; stroke-linejoin: miter\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 51.06875 105.626701  L 330.06875 105.626701  L 330.06875 24.986701  L 51.06875 24.986701  z \" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\"/>\n",
       "    <g id=\"xtick_2\"/>\n",
       "    <g id=\"xtick_3\"/>\n",
       "    <g id=\"xtick_4\"/>\n",
       "    <g id=\"xtick_5\"/>\n",
       "    <g id=\"xtick_6\"/>\n",
       "    <g id=\"xtick_7\"/>\n",
       "    <g id=\"xtick_8\"/>\n",
       "    <g id=\"xtick_9\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\">\n",
       "     <g id=\"line2d_1\">\n",
       "      <path d=\"M 51.06875 101.961246  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <defs>\n",
       "       <path id=\"m2ce59a0e6b\" d=\"M 0 0  L -2.75 0  \" style=\"stroke: #333333\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ce59a0e6b\" x=\"51.06875\" y=\"101.961246\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_1\">\n",
       "      <!-- 0 -->\n",
       "      <g transform=\"translate(38.48375 105.272496)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-30\" d=\"M 2034 4250  Q 1547 4250 1301 3770  Q 1056 3291 1056 2328  Q 1056 1369 1301 889  Q 1547 409 2034 409  Q 2525 409 2770 889  Q 3016 1369 3016 2328  Q 3016 3291 2770 3770  Q 2525 4250 2034 4250  z M 2034 4750  Q 2819 4750 3233 4129  Q 3647 3509 3647 2328  Q 3647 1150 3233 529  Q 2819 -91 2034 -91  Q 1250 -91 836 529  Q 422 1150 422 2328  Q 422 3509 836 4129  Q 1250 4750 2034 4750  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_2\">\n",
       "     <g id=\"line2d_2\">\n",
       "      <path d=\"M 51.06875 76.682249  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ce59a0e6b\" x=\"51.06875\" y=\"76.682249\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_2\">\n",
       "      <!-- 50 -->\n",
       "      <g transform=\"translate(30.84875 79.993499)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-35\" d=\"M 691 4666  L 3169 4666  L 3169 4134  L 1269 4134  L 1269 2991  Q 1406 3038 1543 3061  Q 1681 3084 1819 3084  Q 2600 3084 3056 2656  Q 3513 2228 3513 1497  Q 3513 744 3044 326  Q 2575 -91 1722 -91  Q 1428 -91 1123 -41  Q 819 9 494 109  L 494 744  Q 775 591 1075 516  Q 1375 441 1709 441  Q 2250 441 2565 725  Q 2881 1009 2881 1497  Q 2881 1984 2565 2268  Q 2250 2553 1709 2553  Q 1456 2553 1204 2497  Q 953 2441 691 2322  L 691 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_3\">\n",
       "     <g id=\"line2d_3\">\n",
       "      <path d=\"M 51.06875 51.403252  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ce59a0e6b\" x=\"51.06875\" y=\"51.403252\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_3\">\n",
       "      <!-- 100 -->\n",
       "      <g transform=\"translate(23.21375 54.714502)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-31\" d=\"M 794 531  L 1825 531  L 1825 4091  L 703 3866  L 703 4441  L 1819 4666  L 2450 4666  L 2450 531  L 3481 531  L 3481 0  L 794 0  L 794 531  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_4\">\n",
       "     <g id=\"line2d_4\">\n",
       "      <path d=\"M 51.06875 26.124256  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ce59a0e6b\" x=\"51.06875\" y=\"26.124256\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_4\">\n",
       "      <!-- 150 -->\n",
       "      <g transform=\"translate(23.21375 29.435506)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_5\"/>\n",
       "    <g id=\"ytick_6\"/>\n",
       "    <g id=\"ytick_7\"/>\n",
       "    <g id=\"text_5\">\n",
       "     <!-- # of cells -->\n",
       "     <g transform=\"translate(16.318125 135.142013)rotate(-90)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-23\" d=\"M 3272 2816  L 2363 2816  L 2100 1772  L 3016 1772  L 3272 2816  z M 2803 4594  L 2478 3297  L 3391 3297  L 3719 4594  L 4219 4594  L 3897 3297  L 4872 3297  L 4872 2816  L 3775 2816  L 3519 1772  L 4513 1772  L 4513 1294  L 3397 1294  L 3072 0  L 2572 0  L 2894 1294  L 1978 1294  L 1656 0  L 1153 0  L 1478 1294  L 494 1294  L 494 1772  L 1594 1772  L 1856 2816  L 850 2816  L 850 3297  L 1978 3297  L 2297 4594  L 2803 4594  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-20\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6f\" d=\"M 1959 3097  Q 1497 3097 1228 2736  Q 959 2375 959 1747  Q 959 1119 1226 758  Q 1494 397 1959 397  Q 2419 397 2687 759  Q 2956 1122 2956 1747  Q 2956 2369 2687 2733  Q 2419 3097 1959 3097  z M 1959 3584  Q 2709 3584 3137 3096  Q 3566 2609 3566 1747  Q 3566 888 3137 398  Q 2709 -91 1959 -91  Q 1206 -91 779 398  Q 353 888 353 1747  Q 353 2609 779 3096  Q 1206 3584 1959 3584  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-66\" d=\"M 2375 4863  L 2375 4384  L 1825 4384  Q 1516 4384 1395 4259  Q 1275 4134 1275 3809  L 1275 3500  L 2222 3500  L 2222 3053  L 1275 3053  L 1275 0  L 697 0  L 697 3053  L 147 3053  L 147 3500  L 697 3500  L 697 3744  Q 697 4328 969 4595  Q 1241 4863 1831 4863  L 2375 4863  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-63\" d=\"M 3122 3366  L 3122 2828  Q 2878 2963 2633 3030  Q 2388 3097 2138 3097  Q 1578 3097 1268 2742  Q 959 2388 959 1747  Q 959 1106 1268 751  Q 1578 397 2138 397  Q 2388 397 2633 464  Q 2878 531 3122 666  L 3122 134  Q 2881 22 2623 -34  Q 2366 -91 2075 -91  Q 1284 -91 818 406  Q 353 903 353 1747  Q 353 2603 823 3093  Q 1294 3584 2113 3584  Q 2378 3584 2631 3529  Q 2884 3475 3122 3366  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-65\" d=\"M 3597 1894  L 3597 1613  L 953 1613  Q 991 1019 1311 708  Q 1631 397 2203 397  Q 2534 397 2845 478  Q 3156 559 3463 722  L 3463 178  Q 3153 47 2828 -22  Q 2503 -91 2169 -91  Q 1331 -91 842 396  Q 353 884 353 1716  Q 353 2575 817 3079  Q 1281 3584 2069 3584  Q 2775 3584 3186 3129  Q 3597 2675 3597 1894  z M 3022 2063  Q 3016 2534 2758 2815  Q 2500 3097 2075 3097  Q 1594 3097 1305 2825  Q 1016 2553 972 2059  L 3022 2063  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6c\" d=\"M 603 4863  L 1178 4863  L 1178 0  L 603 0  L 603 4863  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-73\" d=\"M 2834 3397  L 2834 2853  Q 2591 2978 2328 3040  Q 2066 3103 1784 3103  Q 1356 3103 1142 2972  Q 928 2841 928 2578  Q 928 2378 1081 2264  Q 1234 2150 1697 2047  L 1894 2003  Q 2506 1872 2764 1633  Q 3022 1394 3022 966  Q 3022 478 2636 193  Q 2250 -91 1575 -91  Q 1294 -91 989 -36  Q 684 19 347 128  L 347 722  Q 666 556 975 473  Q 1284 391 1588 391  Q 1994 391 2212 530  Q 2431 669 2431 922  Q 2431 1156 2273 1281  Q 2116 1406 1581 1522  L 1381 1569  Q 847 1681 609 1914  Q 372 2147 372 2553  Q 372 3047 722 3315  Q 1072 3584 1716 3584  Q 2034 3584 2315 3537  Q 2597 3491 2834 3397  z \" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-23\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"83.789062\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"115.576172\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-66\" x=\"176.757812\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"211.962891\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-63\" x=\"243.75\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"298.730469\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6c\" x=\"360.253906\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6c\" x=\"388.037109\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-73\" x=\"415.820312\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PolyCollection_1\">\n",
       "    <path d=\"M 63.750568 101.961246  L 63.750568 74.659929  L 78.670354 74.659929  L 78.670354 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 78.670354 101.961246  L 78.670354 63.537171  L 93.59014 63.537171  L 93.59014 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 93.59014 101.961246  L 93.59014 87.299428  L 108.509926 87.299428  L 108.509926 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 108.509926 101.961246  L 108.509926 87.299428  L 123.429713 87.299428  L 123.429713 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 123.429713 101.961246  L 123.429713 91.344067  L 138.349499 91.344067  L 138.349499 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 138.349499 101.961246  L 138.349499 84.771528  L 153.269285 84.771528  L 153.269285 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 153.269285 101.961246  L 153.269285 91.849647  L 168.189071 91.849647  L 168.189071 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 168.189071 101.961246  L 168.189071 86.793848  L 183.108857 86.793848  L 183.108857 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 183.108857 101.961246  L 183.108857 91.344067  L 198.028643 91.344067  L 198.028643 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 198.028643 101.961246  L 198.028643 96.905447  L 212.948429 96.905447  L 212.948429 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 212.948429 101.961246  L 212.948429 97.916607  L 227.868215 97.916607  L 227.868215 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 227.868215 101.961246  L 227.868215 92.355227  L 242.788001 92.355227  L 242.788001 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 242.788001 101.961246  L 242.788001 99.938926  L 257.707787 99.938926  L 257.707787 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 257.707787 101.961246  L 257.707787 97.411027  L 272.627574 97.411027  L 272.627574 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 272.627574 101.961246  L 272.627574 99.938926  L 287.54736 99.938926  L 287.54736 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 287.54736 101.961246  L 287.54736 98.927766  L 302.467146 98.927766  L 302.467146 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 302.467146 101.961246  L 302.467146 99.938926  L 317.386932 99.938926  L 317.386932 101.961246  z \" clip-path=\"url(#p7bbc686492)\" style=\"fill: #595959\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 330.06875 105.626701  L 344.36875 105.626701  L 344.36875 24.986701  L 330.06875 24.986701  L 330.06875 105.626701  z \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 51.06875 105.626701  L 51.06875 24.986701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 51.06875 105.626701  L 330.06875 105.626701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_6\">\n",
       "    <!-- trial1 -->\n",
       "    <g style=\"fill: #1a1a1a\" transform=\"translate(334.7905 53.832326)rotate(-270)scale(0.088 -0.088)\">\n",
       "     <defs>\n",
       "      <path id=\"DejaVuSans-74\" d=\"M 1172 4494  L 1172 3500  L 2356 3500  L 2356 3053  L 1172 3053  L 1172 1153  Q 1172 725 1289 603  Q 1406 481 1766 481  L 2356 481  L 2356 0  L 1766 0  Q 1100 0 847 248  Q 594 497 594 1153  L 594 3053  L 172 3053  L 172 3500  L 594 3500  L 594 4494  L 1172 4494  z \" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"DejaVuSans-72\" d=\"M 2631 2963  Q 2534 3019 2420 3045  Q 2306 3072 2169 3072  Q 1681 3072 1420 2755  Q 1159 2438 1159 1844  L 1159 0  L 581 0  L 581 3500  L 1159 3500  L 1159 2956  Q 1341 3275 1631 3429  Q 1922 3584 2338 3584  Q 2397 3584 2469 3576  Q 2541 3569 2628 3553  L 2631 2963  z \" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"DejaVuSans-69\" d=\"M 603 3500  L 1178 3500  L 1178 0  L 603 0  L 603 3500  z M 603 4863  L 1178 4863  L 1178 4134  L 603 4134  L 603 4863  z \" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"DejaVuSans-61\" d=\"M 2194 1759  Q 1497 1759 1228 1600  Q 959 1441 959 1056  Q 959 750 1161 570  Q 1363 391 1709 391  Q 2188 391 2477 730  Q 2766 1069 2766 1631  L 2766 1759  L 2194 1759  z M 3341 1997  L 3341 0  L 2766 0  L 2766 531  Q 2569 213 2275 61  Q 1981 -91 1556 -91  Q 1019 -91 701 211  Q 384 513 384 1019  Q 384 1609 779 1909  Q 1175 2209 1959 2209  L 2766 2209  L 2766 2266  Q 2766 2663 2505 2880  Q 2244 3097 1772 3097  Q 1472 3097 1187 3025  Q 903 2953 641 2809  L 641 3341  Q 956 3463 1253 3523  Q 1550 3584 1831 3584  Q 2591 3584 2966 3190  Q 3341 2797 3341 1997  z \" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#DejaVuSans-74\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-72\" x=\"39.208984\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-69\" x=\"80.322266\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-61\" x=\"108.105469\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-6c\" x=\"169.384766\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-31\" x=\"197.167969\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       "  <g id=\"axes_2\">\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 51.06875 191.306701  L 330.06875 191.306701  L 330.06875 110.666701  L 51.06875 110.666701  z \" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_3\">\n",
       "    <g id=\"xtick_10\">\n",
       "     <g id=\"line2d_5\">\n",
       "      <path d=\"M 63.750568 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <defs>\n",
       "       <path id=\"m0a831aec53\" d=\"M 0 0  L 0 2.75  \" style=\"stroke: #333333\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0a831aec53\" x=\"63.750568\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_7\">\n",
       "      <!-- 0 -->\n",
       "      <g transform=\"translate(59.933068 205.374826)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_11\">\n",
       "     <g id=\"line2d_6\">\n",
       "      <path d=\"M 123.429713 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0a831aec53\" x=\"123.429713\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_8\">\n",
       "      <!-- 0.2 -->\n",
       "      <g transform=\"translate(113.887838 205.374826)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-2e\" d=\"M 684 794  L 1344 794  L 1344 0  L 684 0  L 684 794  z \" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-32\" d=\"M 1228 531  L 3431 531  L 3431 0  L 469 0  L 469 531  Q 828 903 1448 1529  Q 2069 2156 2228 2338  Q 2531 2678 2651 2914  Q 2772 3150 2772 3378  Q 2772 3750 2511 3984  Q 2250 4219 1831 4219  Q 1534 4219 1204 4116  Q 875 4013 500 3803  L 500 4441  Q 881 4594 1212 4672  Q 1544 4750 1819 4750  Q 2544 4750 2975 4387  Q 3406 4025 3406 3419  Q 3406 3131 3298 2873  Q 3191 2616 2906 2266  Q 2828 2175 2409 1742  Q 1991 1309 1228 531  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-32\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_12\">\n",
       "     <g id=\"line2d_7\">\n",
       "      <path d=\"M 183.108857 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0a831aec53\" x=\"183.108857\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_9\">\n",
       "      <!-- 0.4 -->\n",
       "      <g transform=\"translate(173.566982 205.374826)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-34\" d=\"M 2419 4116  L 825 1625  L 2419 1625  L 2419 4116  z M 2253 4666  L 3047 4666  L 3047 1625  L 3713 1625  L 3713 1100  L 3047 1100  L 3047 0  L 2419 0  L 2419 1100  L 313 1100  L 313 1709  L 2253 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-34\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_13\">\n",
       "     <g id=\"line2d_8\">\n",
       "      <path d=\"M 242.788001 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0a831aec53\" x=\"242.788001\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_10\">\n",
       "      <!-- 0.6 -->\n",
       "      <g transform=\"translate(233.246126 205.374826)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-36\" d=\"M 2113 2584  Q 1688 2584 1439 2293  Q 1191 2003 1191 1497  Q 1191 994 1439 701  Q 1688 409 2113 409  Q 2538 409 2786 701  Q 3034 994 3034 1497  Q 3034 2003 2786 2293  Q 2538 2584 2113 2584  z M 3366 4563  L 3366 3988  Q 3128 4100 2886 4159  Q 2644 4219 2406 4219  Q 1781 4219 1451 3797  Q 1122 3375 1075 2522  Q 1259 2794 1537 2939  Q 1816 3084 2150 3084  Q 2853 3084 3261 2657  Q 3669 2231 3669 1497  Q 3669 778 3244 343  Q 2819 -91 2113 -91  Q 1303 -91 875 529  Q 447 1150 447 2328  Q 447 3434 972 4092  Q 1497 4750 2381 4750  Q 2619 4750 2861 4703  Q 3103 4656 3366 4563  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-36\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_14\">\n",
       "     <g id=\"line2d_9\">\n",
       "      <path d=\"M 302.467146 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m0a831aec53\" x=\"302.467146\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_11\">\n",
       "      <!-- 0.8 -->\n",
       "      <g transform=\"translate(292.925271 205.374826)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-38\" d=\"M 2034 2216  Q 1584 2216 1326 1975  Q 1069 1734 1069 1313  Q 1069 891 1326 650  Q 1584 409 2034 409  Q 2484 409 2743 651  Q 3003 894 3003 1313  Q 3003 1734 2745 1975  Q 2488 2216 2034 2216  z M 1403 2484  Q 997 2584 770 2862  Q 544 3141 544 3541  Q 544 4100 942 4425  Q 1341 4750 2034 4750  Q 2731 4750 3128 4425  Q 3525 4100 3525 3541  Q 3525 3141 3298 2862  Q 3072 2584 2669 2484  Q 3125 2378 3379 2068  Q 3634 1759 3634 1313  Q 3634 634 3220 271  Q 2806 -91 2034 -91  Q 1263 -91 848 271  Q 434 634 434 1313  Q 434 1759 690 2068  Q 947 2378 1403 2484  z M 1172 3481  Q 1172 3119 1398 2916  Q 1625 2713 2034 2713  Q 2441 2713 2670 2916  Q 2900 3119 2900 3481  Q 2900 3844 2670 4047  Q 2441 4250 2034 4250  Q 1625 4250 1398 4047  Q 1172 3844 1172 3481  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-38\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_15\"/>\n",
       "    <g id=\"xtick_16\"/>\n",
       "    <g id=\"xtick_17\"/>\n",
       "    <g id=\"xtick_18\"/>\n",
       "    <g id=\"text_12\">\n",
       "     <!-- fraction of mRNA from virus -->\n",
       "     <g transform=\"translate(102.431562 221.388576)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-6e\" d=\"M 3513 2113  L 3513 0  L 2938 0  L 2938 2094  Q 2938 2591 2744 2837  Q 2550 3084 2163 3084  Q 1697 3084 1428 2787  Q 1159 2491 1159 1978  L 1159 0  L 581 0  L 581 3500  L 1159 3500  L 1159 2956  Q 1366 3272 1645 3428  Q 1925 3584 2291 3584  Q 2894 3584 3203 3211  Q 3513 2838 3513 2113  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6d\" d=\"M 3328 2828  Q 3544 3216 3844 3400  Q 4144 3584 4550 3584  Q 5097 3584 5394 3201  Q 5691 2819 5691 2113  L 5691 0  L 5113 0  L 5113 2094  Q 5113 2597 4934 2840  Q 4756 3084 4391 3084  Q 3944 3084 3684 2787  Q 3425 2491 3425 1978  L 3425 0  L 2847 0  L 2847 2094  Q 2847 2600 2669 2842  Q 2491 3084 2119 3084  Q 1678 3084 1418 2786  Q 1159 2488 1159 1978  L 1159 0  L 581 0  L 581 3500  L 1159 3500  L 1159 2956  Q 1356 3278 1631 3431  Q 1906 3584 2284 3584  Q 2666 3584 2933 3390  Q 3200 3197 3328 2828  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-52\" d=\"M 2841 2188  Q 3044 2119 3236 1894  Q 3428 1669 3622 1275  L 4263 0  L 3584 0  L 2988 1197  Q 2756 1666 2539 1819  Q 2322 1972 1947 1972  L 1259 1972  L 1259 0  L 628 0  L 628 4666  L 2053 4666  Q 2853 4666 3247 4331  Q 3641 3997 3641 3322  Q 3641 2881 3436 2590  Q 3231 2300 2841 2188  z M 1259 4147  L 1259 2491  L 2053 2491  Q 2509 2491 2742 2702  Q 2975 2913 2975 3322  Q 2975 3731 2742 3939  Q 2509 4147 2053 4147  L 1259 4147  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-4e\" d=\"M 628 4666  L 1478 4666  L 3547 763  L 3547 4666  L 4159 4666  L 4159 0  L 3309 0  L 1241 3903  L 1241 0  L 628 0  L 628 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-41\" d=\"M 2188 4044  L 1331 1722  L 3047 1722  L 2188 4044  z M 1831 4666  L 2547 4666  L 4325 0  L 3669 0  L 3244 1197  L 1141 1197  L 716 0  L 50 0  L 1831 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-76\" d=\"M 191 3500  L 800 3500  L 1894 563  L 2988 3500  L 3597 3500  L 2284 0  L 1503 0  L 191 3500  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-75\" d=\"M 544 1381  L 544 3500  L 1119 3500  L 1119 1403  Q 1119 906 1312 657  Q 1506 409 1894 409  Q 2359 409 2629 706  Q 2900 1003 2900 1516  L 2900 3500  L 3475 3500  L 3475 0  L 2900 0  L 2900 538  Q 2691 219 2414 64  Q 2138 -91 1772 -91  Q 1169 -91 856 284  Q 544 659 544 1381  z M 1991 3584  L 1991 3584  z \" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-66\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-72\" x=\"35.205078\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-61\" x=\"76.318359\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-63\" x=\"137.597656\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"192.578125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"231.787109\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"259.570312\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" x=\"320.751953\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"384.130859\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"415.917969\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-66\" x=\"477.099609\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"512.304688\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6d\" x=\"544.091797\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-52\" x=\"641.503906\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-4e\" x=\"710.986328\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-41\" x=\"785.791016\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"854.199219\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-66\" x=\"885.986328\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-72\" x=\"921.191406\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"960.054688\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6d\" x=\"1021.236328\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"1118.648438\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-76\" x=\"1150.435547\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"1209.615234\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-72\" x=\"1237.398438\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-75\" x=\"1278.511719\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-73\" x=\"1341.890625\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_4\">\n",
       "    <g id=\"ytick_8\">\n",
       "     <g id=\"line2d_10\">\n",
       "      <path d=\"M 51.06875 187.641246  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ce59a0e6b\" x=\"51.06875\" y=\"187.641246\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_13\">\n",
       "      <!-- 0 -->\n",
       "      <g transform=\"translate(38.48375 190.952496)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_9\">\n",
       "     <g id=\"line2d_11\">\n",
       "      <path d=\"M 51.06875 162.362249  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ce59a0e6b\" x=\"51.06875\" y=\"162.362249\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_14\">\n",
       "      <!-- 50 -->\n",
       "      <g transform=\"translate(30.84875 165.673499)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_10\">\n",
       "     <g id=\"line2d_12\">\n",
       "      <path d=\"M 51.06875 137.083252  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ce59a0e6b\" x=\"51.06875\" y=\"137.083252\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_15\">\n",
       "      <!-- 100 -->\n",
       "      <g transform=\"translate(23.21375 140.394502)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_11\">\n",
       "     <g id=\"line2d_13\">\n",
       "      <path d=\"M 51.06875 111.804256  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ce59a0e6b\" x=\"51.06875\" y=\"111.804256\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_16\">\n",
       "      <!-- 150 -->\n",
       "      <g transform=\"translate(23.21375 115.115506)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_12\"/>\n",
       "    <g id=\"ytick_13\"/>\n",
       "    <g id=\"ytick_14\"/>\n",
       "   </g>\n",
       "   <g id=\"PolyCollection_2\">\n",
       "    <path d=\"M 63.750568 187.641246  L 63.750568 114.332155  L 78.670354 114.332155  L 78.670354 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 78.670354 187.641246  L 78.670354 157.30645  L 93.59014 157.30645  L 93.59014 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 93.59014 187.641246  L 93.59014 177.529647  L 108.509926 177.529647  L 108.509926 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 108.509926 187.641246  L 108.509926 182.079867  L 123.429713 182.079867  L 123.429713 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 123.429713 187.641246  L 123.429713 185.618926  L 138.349499 185.618926  L 138.349499 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 138.349499 187.641246  L 138.349499 185.113346  L 153.269285 185.113346  L 153.269285 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 153.269285 187.641246  L 153.269285 187.641246  L 168.189071 187.641246  L 168.189071 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 168.189071 187.641246  L 168.189071 187.135666  L 183.108857 187.135666  L 183.108857 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 183.108857 187.641246  L 183.108857 185.618926  L 198.028643 185.618926  L 198.028643 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 198.028643 187.641246  L 198.028643 187.641246  L 212.948429 187.641246  L 212.948429 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 212.948429 187.641246  L 212.948429 187.641246  L 227.868215 187.641246  L 227.868215 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 227.868215 187.641246  L 227.868215 187.641246  L 242.788001 187.641246  L 242.788001 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 242.788001 187.641246  L 242.788001 186.124506  L 257.707787 186.124506  L 257.707787 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 257.707787 187.641246  L 257.707787 187.135666  L 272.627574 187.135666  L 272.627574 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 272.627574 187.641246  L 272.627574 187.641246  L 287.54736 187.641246  L 287.54736 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 287.54736 187.641246  L 287.54736 187.641246  L 302.467146 187.641246  L 302.467146 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 302.467146 187.641246  L 302.467146 187.641246  L 317.386932 187.641246  L 317.386932 187.641246  z \" clip-path=\"url(#p44559af3cf)\" style=\"fill: #595959\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_7\">\n",
       "    <path d=\"M 330.06875 191.306701  L 344.36875 191.306701  L 344.36875 110.666701  L 330.06875 110.666701  L 330.06875 191.306701  z \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_8\">\n",
       "    <path d=\"M 51.06875 191.306701  L 51.06875 110.666701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_9\">\n",
       "    <path d=\"M 51.06875 191.306701  L 330.06875 191.306701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_17\">\n",
       "    <!-- trial3 -->\n",
       "    <g style=\"fill: #1a1a1a\" transform=\"translate(334.7905 139.512326)rotate(-270)scale(0.088 -0.088)\">\n",
       "     <defs>\n",
       "      <path id=\"DejaVuSans-33\" d=\"M 2597 2516  Q 3050 2419 3304 2112  Q 3559 1806 3559 1356  Q 3559 666 3084 287  Q 2609 -91 1734 -91  Q 1441 -91 1130 -33  Q 819 25 488 141  L 488 750  Q 750 597 1062 519  Q 1375 441 1716 441  Q 2309 441 2620 675  Q 2931 909 2931 1356  Q 2931 1769 2642 2001  Q 2353 2234 1838 2234  L 1294 2234  L 1294 2753  L 1863 2753  Q 2328 2753 2575 2939  Q 2822 3125 2822 3475  Q 2822 3834 2567 4026  Q 2313 4219 1838 4219  Q 1578 4219 1281 4162  Q 984 4106 628 3988  L 628 4550  Q 988 4650 1302 4700  Q 1616 4750 1894 4750  Q 2613 4750 3031 4423  Q 3450 4097 3450 3541  Q 3450 3153 3228 2886  Q 3006 2619 2597 2516  z \" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#DejaVuSans-74\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-72\" x=\"39.208984\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-69\" x=\"80.322266\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-61\" x=\"108.105469\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-6c\" x=\"169.384766\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-33\" x=\"197.167969\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       "  <g id=\"text_18\">\n",
       "   <!-- Viral transcription in infected cells -->\n",
       "   <g transform=\"translate(83.49125 16.318125)scale(0.12 -0.12)\">\n",
       "    <defs>\n",
       "     <path id=\"DejaVuSans-56\" d=\"M 1831 0  L 50 4666  L 709 4666  L 2188 738  L 3669 4666  L 4325 4666  L 2547 0  L 1831 0  z \" transform=\"scale(0.015625)\"/>\n",
       "     <path id=\"DejaVuSans-70\" d=\"M 1159 525  L 1159 -1331  L 581 -1331  L 581 3500  L 1159 3500  L 1159 2969  Q 1341 3281 1617 3432  Q 1894 3584 2278 3584  Q 2916 3584 3314 3078  Q 3713 2572 3713 1747  Q 3713 922 3314 415  Q 2916 -91 2278 -91  Q 1894 -91 1617 61  Q 1341 213 1159 525  z M 3116 1747  Q 3116 2381 2855 2742  Q 2594 3103 2138 3103  Q 1681 3103 1420 2742  Q 1159 2381 1159 1747  Q 1159 1113 1420 752  Q 1681 391 2138 391  Q 2594 391 2855 752  Q 3116 1113 3116 1747  z \" transform=\"scale(0.015625)\"/>\n",
       "     <path id=\"DejaVuSans-64\" d=\"M 2906 2969  L 2906 4863  L 3481 4863  L 3481 0  L 2906 0  L 2906 525  Q 2725 213 2448 61  Q 2172 -91 1784 -91  Q 1150 -91 751 415  Q 353 922 353 1747  Q 353 2572 751 3078  Q 1150 3584 1784 3584  Q 2172 3584 2448 3432  Q 2725 3281 2906 2969  z M 947 1747  Q 947 1113 1208 752  Q 1469 391 1925 391  Q 2381 391 2643 752  Q 2906 1113 2906 1747  Q 2906 2381 2643 2742  Q 2381 3103 1925 3103  Q 1469 3103 1208 2742  Q 947 2381 947 1747  z \" transform=\"scale(0.015625)\"/>\n",
       "    </defs>\n",
       "    <use xlink:href=\"#DejaVuSans-56\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-69\" x=\"66.158203\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-72\" x=\"93.941406\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-61\" x=\"135.054688\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6c\" x=\"196.333984\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"224.117188\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-74\" x=\"255.904297\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-72\" x=\"295.113281\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-61\" x=\"336.226562\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6e\" x=\"397.505859\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-73\" x=\"460.884766\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-63\" x=\"512.984375\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-72\" x=\"567.964844\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-69\" x=\"609.078125\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-70\" x=\"636.861328\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-74\" x=\"700.337891\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-69\" x=\"739.546875\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6f\" x=\"767.330078\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6e\" x=\"828.511719\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"891.890625\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-69\" x=\"923.677734\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6e\" x=\"951.460938\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"1014.839844\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-69\" x=\"1046.626953\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6e\" x=\"1074.410156\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-66\" x=\"1137.789062\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"1172.994141\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-63\" x=\"1234.517578\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-74\" x=\"1289.498047\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"1328.707031\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-64\" x=\"1390.230469\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"1453.707031\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-63\" x=\"1485.494141\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"1540.474609\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6c\" x=\"1601.998047\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6c\" x=\"1629.78125\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-73\" x=\"1657.564453\"/>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"p7bbc686492\">\n",
       "   <rect x=\"51.06875\" y=\"24.986701\" width=\"279\" height=\"80.64\"/>\n",
       "  </clipPath>\n",
       "  <clipPath id=\"p44559af3cf\">\n",
       "   <rect x=\"51.06875\" y=\"110.666701\" width=\"279\" height=\"80.64\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</g>\n",
       "      <text x=\"5\" y=\"15\" font-size=\"14\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">A</text>\n",
       "    </g>\n",
       "    <g transform=\"translate(720, 0) scale(1) translate(0, 0) scale(2) \">\n",
       "      <g><metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2022-11-03T14:17:25.630890</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.5.2, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 231.084201  L 239.96875 231.084201  L 239.96875 -0  L 0 -0  z \" style=\"fill: #ffffff; stroke: #ffffff; stroke-linejoin: miter\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 51.06875 105.626701  L 218.46875 105.626701  L 218.46875 24.986701  L 51.06875 24.986701  z \" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\"/>\n",
       "    <g id=\"xtick_2\"/>\n",
       "    <g id=\"xtick_3\"/>\n",
       "    <g id=\"xtick_4\"/>\n",
       "    <g id=\"xtick_5\"/>\n",
       "    <g id=\"xtick_6\"/>\n",
       "    <g id=\"xtick_7\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\">\n",
       "     <g id=\"line2d_1\">\n",
       "      <path d=\"M 51.06875 101.961246  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <defs>\n",
       "       <path id=\"m2ff2f96fda\" d=\"M 0 0  L -2.75 0  \" style=\"stroke: #333333\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"101.961246\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_1\">\n",
       "      <!-- 0 -->\n",
       "      <g transform=\"translate(38.48375 105.272496)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-30\" d=\"M 2034 4250  Q 1547 4250 1301 3770  Q 1056 3291 1056 2328  Q 1056 1369 1301 889  Q 1547 409 2034 409  Q 2525 409 2770 889  Q 3016 1369 3016 2328  Q 3016 3291 2770 3770  Q 2525 4250 2034 4250  z M 2034 4750  Q 2819 4750 3233 4129  Q 3647 3509 3647 2328  Q 3647 1150 3233 529  Q 2819 -91 2034 -91  Q 1250 -91 836 529  Q 422 1150 422 2328  Q 422 3509 836 4129  Q 1250 4750 2034 4750  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_2\">\n",
       "     <g id=\"line2d_2\">\n",
       "      <path d=\"M 51.06875 89.743064  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"89.743064\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_2\">\n",
       "      <!-- 50 -->\n",
       "      <g transform=\"translate(30.84875 93.054314)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-35\" d=\"M 691 4666  L 3169 4666  L 3169 4134  L 1269 4134  L 1269 2991  Q 1406 3038 1543 3061  Q 1681 3084 1819 3084  Q 2600 3084 3056 2656  Q 3513 2228 3513 1497  Q 3513 744 3044 326  Q 2575 -91 1722 -91  Q 1428 -91 1123 -41  Q 819 9 494 109  L 494 744  Q 775 591 1075 516  Q 1375 441 1709 441  Q 2250 441 2565 725  Q 2881 1009 2881 1497  Q 2881 1984 2565 2268  Q 2250 2553 1709 2553  Q 1456 2553 1204 2497  Q 953 2441 691 2322  L 691 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_3\">\n",
       "     <g id=\"line2d_3\">\n",
       "      <path d=\"M 51.06875 77.524882  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"77.524882\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_3\">\n",
       "      <!-- 100 -->\n",
       "      <g transform=\"translate(23.21375 80.836132)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-31\" d=\"M 794 531  L 1825 531  L 1825 4091  L 703 3866  L 703 4441  L 1819 4666  L 2450 4666  L 2450 531  L 3481 531  L 3481 0  L 794 0  L 794 531  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_4\">\n",
       "     <g id=\"line2d_4\">\n",
       "      <path d=\"M 51.06875 65.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"65.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_4\">\n",
       "      <!-- 150 -->\n",
       "      <g transform=\"translate(23.21375 68.617951)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_5\">\n",
       "     <g id=\"line2d_5\">\n",
       "      <path d=\"M 51.06875 53.088519  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"53.088519\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_5\">\n",
       "      <!-- 200 -->\n",
       "      <g transform=\"translate(23.21375 56.399769)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-32\" d=\"M 1228 531  L 3431 531  L 3431 0  L 469 0  L 469 531  Q 828 903 1448 1529  Q 2069 2156 2228 2338  Q 2531 2678 2651 2914  Q 2772 3150 2772 3378  Q 2772 3750 2511 3984  Q 2250 4219 1831 4219  Q 1534 4219 1204 4116  Q 875 4013 500 3803  L 500 4441  Q 881 4594 1212 4672  Q 1544 4750 1819 4750  Q 2544 4750 2975 4387  Q 3406 4025 3406 3419  Q 3406 3131 3298 2873  Q 3191 2616 2906 2266  Q 2828 2175 2409 1742  Q 1991 1309 1228 531  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_6\">\n",
       "     <g id=\"line2d_6\">\n",
       "      <path d=\"M 51.06875 40.870337  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"40.870337\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_6\">\n",
       "      <!-- 250 -->\n",
       "      <g transform=\"translate(23.21375 44.181587)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_7\">\n",
       "     <g id=\"line2d_7\">\n",
       "      <path d=\"M 51.06875 28.652155  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"28.652155\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_7\">\n",
       "      <!-- 300 -->\n",
       "      <g transform=\"translate(23.21375 31.963405)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-33\" d=\"M 2597 2516  Q 3050 2419 3304 2112  Q 3559 1806 3559 1356  Q 3559 666 3084 287  Q 2609 -91 1734 -91  Q 1441 -91 1130 -33  Q 819 25 488 141  L 488 750  Q 750 597 1062 519  Q 1375 441 1716 441  Q 2309 441 2620 675  Q 2931 909 2931 1356  Q 2931 1769 2642 2001  Q 2353 2234 1838 2234  L 1294 2234  L 1294 2753  L 1863 2753  Q 2328 2753 2575 2939  Q 2822 3125 2822 3475  Q 2822 3834 2567 4026  Q 2313 4219 1838 4219  Q 1578 4219 1281 4162  Q 984 4106 628 3988  L 628 4550  Q 988 4650 1302 4700  Q 1616 4750 1894 4750  Q 2613 4750 3031 4423  Q 3450 4097 3450 3541  Q 3450 3153 3228 2886  Q 3006 2619 2597 2516  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_8\"/>\n",
       "    <g id=\"ytick_9\"/>\n",
       "    <g id=\"ytick_10\"/>\n",
       "    <g id=\"ytick_11\"/>\n",
       "    <g id=\"ytick_12\"/>\n",
       "    <g id=\"ytick_13\"/>\n",
       "    <g id=\"text_8\">\n",
       "     <!-- # of cells -->\n",
       "     <g transform=\"translate(16.318125 135.142013)rotate(-90)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-23\" d=\"M 3272 2816  L 2363 2816  L 2100 1772  L 3016 1772  L 3272 2816  z M 2803 4594  L 2478 3297  L 3391 3297  L 3719 4594  L 4219 4594  L 3897 3297  L 4872 3297  L 4872 2816  L 3775 2816  L 3519 1772  L 4513 1772  L 4513 1294  L 3397 1294  L 3072 0  L 2572 0  L 2894 1294  L 1978 1294  L 1656 0  L 1153 0  L 1478 1294  L 494 1294  L 494 1772  L 1594 1772  L 1856 2816  L 850 2816  L 850 3297  L 1978 3297  L 2297 4594  L 2803 4594  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-20\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6f\" d=\"M 1959 3097  Q 1497 3097 1228 2736  Q 959 2375 959 1747  Q 959 1119 1226 758  Q 1494 397 1959 397  Q 2419 397 2687 759  Q 2956 1122 2956 1747  Q 2956 2369 2687 2733  Q 2419 3097 1959 3097  z M 1959 3584  Q 2709 3584 3137 3096  Q 3566 2609 3566 1747  Q 3566 888 3137 398  Q 2709 -91 1959 -91  Q 1206 -91 779 398  Q 353 888 353 1747  Q 353 2609 779 3096  Q 1206 3584 1959 3584  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-66\" d=\"M 2375 4863  L 2375 4384  L 1825 4384  Q 1516 4384 1395 4259  Q 1275 4134 1275 3809  L 1275 3500  L 2222 3500  L 2222 3053  L 1275 3053  L 1275 0  L 697 0  L 697 3053  L 147 3053  L 147 3500  L 697 3500  L 697 3744  Q 697 4328 969 4595  Q 1241 4863 1831 4863  L 2375 4863  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-63\" d=\"M 3122 3366  L 3122 2828  Q 2878 2963 2633 3030  Q 2388 3097 2138 3097  Q 1578 3097 1268 2742  Q 959 2388 959 1747  Q 959 1106 1268 751  Q 1578 397 2138 397  Q 2388 397 2633 464  Q 2878 531 3122 666  L 3122 134  Q 2881 22 2623 -34  Q 2366 -91 2075 -91  Q 1284 -91 818 406  Q 353 903 353 1747  Q 353 2603 823 3093  Q 1294 3584 2113 3584  Q 2378 3584 2631 3529  Q 2884 3475 3122 3366  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-65\" d=\"M 3597 1894  L 3597 1613  L 953 1613  Q 991 1019 1311 708  Q 1631 397 2203 397  Q 2534 397 2845 478  Q 3156 559 3463 722  L 3463 178  Q 3153 47 2828 -22  Q 2503 -91 2169 -91  Q 1331 -91 842 396  Q 353 884 353 1716  Q 353 2575 817 3079  Q 1281 3584 2069 3584  Q 2775 3584 3186 3129  Q 3597 2675 3597 1894  z M 3022 2063  Q 3016 2534 2758 2815  Q 2500 3097 2075 3097  Q 1594 3097 1305 2825  Q 1016 2553 972 2059  L 3022 2063  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6c\" d=\"M 603 4863  L 1178 4863  L 1178 0  L 603 0  L 603 4863  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-73\" d=\"M 2834 3397  L 2834 2853  Q 2591 2978 2328 3040  Q 2066 3103 1784 3103  Q 1356 3103 1142 2972  Q 928 2841 928 2578  Q 928 2378 1081 2264  Q 1234 2150 1697 2047  L 1894 2003  Q 2506 1872 2764 1633  Q 3022 1394 3022 966  Q 3022 478 2636 193  Q 2250 -91 1575 -91  Q 1294 -91 989 -36  Q 684 19 347 128  L 347 722  Q 666 556 975 473  Q 1284 391 1588 391  Q 1994 391 2212 530  Q 2431 669 2431 922  Q 2431 1156 2273 1281  Q 2116 1406 1581 1522  L 1381 1569  Q 847 1681 609 1914  Q 372 2147 372 2553  Q 372 3047 722 3315  Q 1072 3584 1716 3584  Q 2034 3584 2315 3537  Q 2597 3491 2834 3397  z \" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-23\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"83.789062\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"115.576172\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-66\" x=\"176.757812\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"211.962891\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-63\" x=\"243.75\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"298.730469\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6c\" x=\"360.253906\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6c\" x=\"388.037109\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-73\" x=\"415.820312\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PolyCollection_1\">\n",
       "    <path d=\"M 80.733177 101.961246  L 80.733177 101.228155  L 100.582979 101.228155  L 100.582979 101.961246  z \" clip-path=\"url(#p46c9ec0a9e)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 102.788513 101.961246  L 102.788513 100.006337  L 122.638315 100.006337  L 122.638315 101.961246  z \" clip-path=\"url(#p46c9ec0a9e)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 124.843849 101.961246  L 124.843849 99.028882  L 144.693651 99.028882  L 144.693651 101.961246  z \" clip-path=\"url(#p46c9ec0a9e)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 146.899185 101.961246  L 146.899185 95.363428  L 166.748987 95.363428  L 166.748987 101.961246  z \" clip-path=\"url(#p46c9ec0a9e)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 168.954521 101.961246  L 168.954521 86.810701  L 188.804323 86.810701  L 188.804323 101.961246  z \" clip-path=\"url(#p46c9ec0a9e)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 191.009857 101.961246  L 191.009857 36.960519  L 210.859659 36.960519  L 210.859659 101.961246  z \" clip-path=\"url(#p46c9ec0a9e)\" style=\"fill: #595959\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 218.46875 105.626701  L 232.76875 105.626701  L 232.76875 24.986701  L 218.46875 24.986701  L 218.46875 105.626701  z \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 51.06875 105.626701  L 51.06875 24.986701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 51.06875 105.626701  L 218.46875 105.626701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_9\">\n",
       "    <!-- trial1 -->\n",
       "    <g style=\"fill: #1a1a1a\" transform=\"translate(223.1905 53.832326)rotate(-270)scale(0.088 -0.088)\">\n",
       "     <defs>\n",
       "      <path id=\"DejaVuSans-74\" d=\"M 1172 4494  L 1172 3500  L 2356 3500  L 2356 3053  L 1172 3053  L 1172 1153  Q 1172 725 1289 603  Q 1406 481 1766 481  L 2356 481  L 2356 0  L 1766 0  Q 1100 0 847 248  Q 594 497 594 1153  L 594 3053  L 172 3053  L 172 3500  L 594 3500  L 594 4494  L 1172 4494  z \" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"DejaVuSans-72\" d=\"M 2631 2963  Q 2534 3019 2420 3045  Q 2306 3072 2169 3072  Q 1681 3072 1420 2755  Q 1159 2438 1159 1844  L 1159 0  L 581 0  L 581 3500  L 1159 3500  L 1159 2956  Q 1341 3275 1631 3429  Q 1922 3584 2338 3584  Q 2397 3584 2469 3576  Q 2541 3569 2628 3553  L 2631 2963  z \" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"DejaVuSans-69\" d=\"M 603 3500  L 1178 3500  L 1178 0  L 603 0  L 603 3500  z M 603 4863  L 1178 4863  L 1178 4134  L 603 4134  L 603 4863  z \" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"DejaVuSans-61\" d=\"M 2194 1759  Q 1497 1759 1228 1600  Q 959 1441 959 1056  Q 959 750 1161 570  Q 1363 391 1709 391  Q 2188 391 2477 730  Q 2766 1069 2766 1631  L 2766 1759  L 2194 1759  z M 3341 1997  L 3341 0  L 2766 0  L 2766 531  Q 2569 213 2275 61  Q 1981 -91 1556 -91  Q 1019 -91 701 211  Q 384 513 384 1019  Q 384 1609 779 1909  Q 1175 2209 1959 2209  L 2766 2209  L 2766 2266  Q 2766 2663 2505 2880  Q 2244 3097 1772 3097  Q 1472 3097 1187 3025  Q 903 2953 641 2809  L 641 3341  Q 956 3463 1253 3523  Q 1550 3584 1831 3584  Q 2591 3584 2966 3190  Q 3341 2797 3341 1997  z \" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#DejaVuSans-74\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-72\" x=\"39.208984\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-69\" x=\"80.322266\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-61\" x=\"108.105469\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-6c\" x=\"169.384766\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-31\" x=\"197.167969\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       "  <g id=\"axes_2\">\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 51.06875 191.306701  L 218.46875 191.306701  L 218.46875 110.666701  L 51.06875 110.666701  z \" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_3\">\n",
       "    <g id=\"xtick_8\">\n",
       "     <g id=\"line2d_8\">\n",
       "      <path d=\"M 68.602742 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <defs>\n",
       "       <path id=\"m9cd550433d\" d=\"M 0 0  L 0 2.75  \" style=\"stroke: #333333\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m9cd550433d\" x=\"68.602742\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_10\">\n",
       "      <!-- 2 -->\n",
       "      <g transform=\"translate(64.785242 205.374826)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_9\">\n",
       "     <g id=\"line2d_9\">\n",
       "      <path d=\"M 112.713414 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m9cd550433d\" x=\"112.713414\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_11\">\n",
       "      <!-- 4 -->\n",
       "      <g transform=\"translate(108.895914 205.374826)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-34\" d=\"M 2419 4116  L 825 1625  L 2419 1625  L 2419 4116  z M 2253 4666  L 3047 4666  L 3047 1625  L 3713 1625  L 3713 1100  L 3047 1100  L 3047 0  L 2419 0  L 2419 1100  L 313 1100  L 313 1709  L 2253 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-34\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_10\">\n",
       "     <g id=\"line2d_10\">\n",
       "      <path d=\"M 156.824086 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m9cd550433d\" x=\"156.824086\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_12\">\n",
       "      <!-- 6 -->\n",
       "      <g transform=\"translate(153.006586 205.374826)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-36\" d=\"M 2113 2584  Q 1688 2584 1439 2293  Q 1191 2003 1191 1497  Q 1191 994 1439 701  Q 1688 409 2113 409  Q 2538 409 2786 701  Q 3034 994 3034 1497  Q 3034 2003 2786 2293  Q 2538 2584 2113 2584  z M 3366 4563  L 3366 3988  Q 3128 4100 2886 4159  Q 2644 4219 2406 4219  Q 1781 4219 1451 3797  Q 1122 3375 1075 2522  Q 1259 2794 1537 2939  Q 1816 3084 2150 3084  Q 2853 3084 3261 2657  Q 3669 2231 3669 1497  Q 3669 778 3244 343  Q 2819 -91 2113 -91  Q 1303 -91 875 529  Q 447 1150 447 2328  Q 447 3434 972 4092  Q 1497 4750 2381 4750  Q 2619 4750 2861 4703  Q 3103 4656 3366 4563  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-36\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_11\">\n",
       "     <g id=\"line2d_11\">\n",
       "      <path d=\"M 200.934758 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m9cd550433d\" x=\"200.934758\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_13\">\n",
       "      <!-- 8 -->\n",
       "      <g transform=\"translate(197.117258 205.374826)scale(0.12 -0.12)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-38\" d=\"M 2034 2216  Q 1584 2216 1326 1975  Q 1069 1734 1069 1313  Q 1069 891 1326 650  Q 1584 409 2034 409  Q 2484 409 2743 651  Q 3003 894 3003 1313  Q 3003 1734 2745 1975  Q 2488 2216 2034 2216  z M 1403 2484  Q 997 2584 770 2862  Q 544 3141 544 3541  Q 544 4100 942 4425  Q 1341 4750 2034 4750  Q 2731 4750 3128 4425  Q 3525 4100 3525 3541  Q 3525 3141 3298 2862  Q 3072 2584 2669 2484  Q 3125 2378 3379 2068  Q 3634 1759 3634 1313  Q 3634 634 3220 271  Q 2806 -91 2034 -91  Q 1263 -91 848 271  Q 434 634 434 1313  Q 434 1759 690 2068  Q 947 2378 1403 2484  z M 1172 3481  Q 1172 3119 1398 2916  Q 1625 2713 2034 2713  Q 2441 2713 2670 2916  Q 2900 3119 2900 3481  Q 2900 3844 2670 4047  Q 2441 4250 2034 4250  Q 1625 4250 1398 4047  Q 1172 3844 1172 3481  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-38\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_12\"/>\n",
       "    <g id=\"xtick_13\"/>\n",
       "    <g id=\"xtick_14\"/>\n",
       "    <g id=\"text_14\">\n",
       "     <!-- # of viral genes detected -->\n",
       "     <g transform=\"translate(55.783437 221.388576)scale(0.12 -0.12)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-76\" d=\"M 191 3500  L 800 3500  L 1894 563  L 2988 3500  L 3597 3500  L 2284 0  L 1503 0  L 191 3500  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-67\" d=\"M 2906 1791  Q 2906 2416 2648 2759  Q 2391 3103 1925 3103  Q 1463 3103 1205 2759  Q 947 2416 947 1791  Q 947 1169 1205 825  Q 1463 481 1925 481  Q 2391 481 2648 825  Q 2906 1169 2906 1791  z M 3481 434  Q 3481 -459 3084 -895  Q 2688 -1331 1869 -1331  Q 1566 -1331 1297 -1286  Q 1028 -1241 775 -1147  L 775 -588  Q 1028 -725 1275 -790  Q 1522 -856 1778 -856  Q 2344 -856 2625 -561  Q 2906 -266 2906 331  L 2906 616  Q 2728 306 2450 153  Q 2172 0 1784 0  Q 1141 0 747 490  Q 353 981 353 1791  Q 353 2603 747 3093  Q 1141 3584 1784 3584  Q 2172 3584 2450 3431  Q 2728 3278 2906 2969  L 2906 3500  L 3481 3500  L 3481 434  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6e\" d=\"M 3513 2113  L 3513 0  L 2938 0  L 2938 2094  Q 2938 2591 2744 2837  Q 2550 3084 2163 3084  Q 1697 3084 1428 2787  Q 1159 2491 1159 1978  L 1159 0  L 581 0  L 581 3500  L 1159 3500  L 1159 2956  Q 1366 3272 1645 3428  Q 1925 3584 2291 3584  Q 2894 3584 3203 3211  Q 3513 2838 3513 2113  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-64\" d=\"M 2906 2969  L 2906 4863  L 3481 4863  L 3481 0  L 2906 0  L 2906 525  Q 2725 213 2448 61  Q 2172 -91 1784 -91  Q 1150 -91 751 415  Q 353 922 353 1747  Q 353 2572 751 3078  Q 1150 3584 1784 3584  Q 2172 3584 2448 3432  Q 2725 3281 2906 2969  z M 947 1747  Q 947 1113 1208 752  Q 1469 391 1925 391  Q 2381 391 2643 752  Q 2906 1113 2906 1747  Q 2906 2381 2643 2742  Q 2381 3103 1925 3103  Q 1469 3103 1208 2742  Q 947 2381 947 1747  z \" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-23\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"83.789062\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"115.576172\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-66\" x=\"176.757812\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"211.962891\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-76\" x=\"243.75\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"302.929688\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-72\" x=\"330.712891\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-61\" x=\"371.826172\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6c\" x=\"433.105469\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"460.888672\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-67\" x=\"492.675781\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"556.152344\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" x=\"617.675781\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"681.054688\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-73\" x=\"742.578125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"794.677734\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-64\" x=\"826.464844\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"889.941406\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"951.464844\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"990.673828\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-63\" x=\"1052.197266\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"1107.177734\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"1146.386719\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-64\" x=\"1207.910156\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_4\">\n",
       "    <g id=\"ytick_14\">\n",
       "     <g id=\"line2d_12\">\n",
       "      <path d=\"M 51.06875 187.641246  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"187.641246\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_15\">\n",
       "      <!-- 0 -->\n",
       "      <g transform=\"translate(38.48375 190.952496)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_15\">\n",
       "     <g id=\"line2d_13\">\n",
       "      <path d=\"M 51.06875 175.423064  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"175.423064\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_16\">\n",
       "      <!-- 50 -->\n",
       "      <g transform=\"translate(30.84875 178.734314)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-35\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_16\">\n",
       "     <g id=\"line2d_14\">\n",
       "      <path d=\"M 51.06875 163.204882  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"163.204882\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_17\">\n",
       "      <!-- 100 -->\n",
       "      <g transform=\"translate(23.21375 166.516132)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_17\">\n",
       "     <g id=\"line2d_15\">\n",
       "      <path d=\"M 51.06875 150.986701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"150.986701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_18\">\n",
       "      <!-- 150 -->\n",
       "      <g transform=\"translate(23.21375 154.297951)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_18\">\n",
       "     <g id=\"line2d_16\">\n",
       "      <path d=\"M 51.06875 138.768519  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"138.768519\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_19\">\n",
       "      <!-- 200 -->\n",
       "      <g transform=\"translate(23.21375 142.079769)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_19\">\n",
       "     <g id=\"line2d_17\">\n",
       "      <path d=\"M 51.06875 126.550337  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"126.550337\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_20\">\n",
       "      <!-- 250 -->\n",
       "      <g transform=\"translate(23.21375 129.861587)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-32\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_20\">\n",
       "     <g id=\"line2d_18\">\n",
       "      <path d=\"M 51.06875 114.332155  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2ff2f96fda\" x=\"51.06875\" y=\"114.332155\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_21\">\n",
       "      <!-- 300 -->\n",
       "      <g transform=\"translate(23.21375 117.643405)scale(0.12 -0.12)\">\n",
       "       <use xlink:href=\"#DejaVuSans-33\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"127.246094\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_21\"/>\n",
       "    <g id=\"ytick_22\"/>\n",
       "    <g id=\"ytick_23\"/>\n",
       "    <g id=\"ytick_24\"/>\n",
       "    <g id=\"ytick_25\"/>\n",
       "    <g id=\"ytick_26\"/>\n",
       "   </g>\n",
       "   <g id=\"PolyCollection_2\">\n",
       "    <path d=\"M 58.677841 187.641246  L 58.677841 187.152519  L 78.527643 187.152519  L 78.527643 187.641246  z \" clip-path=\"url(#pa8ec8a3d40)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 80.733177 187.641246  L 80.733177 187.152519  L 100.582979 187.152519  L 100.582979 187.641246  z \" clip-path=\"url(#pa8ec8a3d40)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 102.788513 187.641246  L 102.788513 186.908155  L 122.638315 186.908155  L 122.638315 187.641246  z \" clip-path=\"url(#pa8ec8a3d40)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 124.843849 187.641246  L 124.843849 185.686337  L 144.693651 185.686337  L 144.693651 187.641246  z \" clip-path=\"url(#pa8ec8a3d40)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 146.899185 187.641246  L 146.899185 184.464519  L 166.748987 184.464519  L 166.748987 187.641246  z \" clip-path=\"url(#pa8ec8a3d40)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 168.954521 187.641246  L 168.954521 170.535792  L 188.804323 170.535792  L 188.804323 187.641246  z \" clip-path=\"url(#pa8ec8a3d40)\" style=\"fill: #595959\"/>\n",
       "    <path d=\"M 191.009857 187.641246  L 191.009857 149.520519  L 210.859659 149.520519  L 210.859659 187.641246  z \" clip-path=\"url(#pa8ec8a3d40)\" style=\"fill: #595959\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_7\">\n",
       "    <path d=\"M 218.46875 191.306701  L 232.76875 191.306701  L 232.76875 110.666701  L 218.46875 110.666701  L 218.46875 191.306701  z \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_8\">\n",
       "    <path d=\"M 51.06875 191.306701  L 51.06875 110.666701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_9\">\n",
       "    <path d=\"M 51.06875 191.306701  L 218.46875 191.306701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_22\">\n",
       "    <!-- trial3 -->\n",
       "    <g style=\"fill: #1a1a1a\" transform=\"translate(223.1905 139.512326)rotate(-270)scale(0.088 -0.088)\">\n",
       "     <use xlink:href=\"#DejaVuSans-74\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-72\" x=\"39.208984\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-69\" x=\"80.322266\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-61\" x=\"108.105469\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-6c\" x=\"169.384766\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-33\" x=\"197.167969\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       "  <g id=\"text_23\">\n",
       "   <!-- Number of viral genes per cell -->\n",
       "   <g transform=\"translate(41.20625 16.318125)scale(0.12 -0.12)\">\n",
       "    <defs>\n",
       "     <path id=\"DejaVuSans-4e\" d=\"M 628 4666  L 1478 4666  L 3547 763  L 3547 4666  L 4159 4666  L 4159 0  L 3309 0  L 1241 3903  L 1241 0  L 628 0  L 628 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "     <path id=\"DejaVuSans-75\" d=\"M 544 1381  L 544 3500  L 1119 3500  L 1119 1403  Q 1119 906 1312 657  Q 1506 409 1894 409  Q 2359 409 2629 706  Q 2900 1003 2900 1516  L 2900 3500  L 3475 3500  L 3475 0  L 2900 0  L 2900 538  Q 2691 219 2414 64  Q 2138 -91 1772 -91  Q 1169 -91 856 284  Q 544 659 544 1381  z M 1991 3584  L 1991 3584  z \" transform=\"scale(0.015625)\"/>\n",
       "     <path id=\"DejaVuSans-6d\" d=\"M 3328 2828  Q 3544 3216 3844 3400  Q 4144 3584 4550 3584  Q 5097 3584 5394 3201  Q 5691 2819 5691 2113  L 5691 0  L 5113 0  L 5113 2094  Q 5113 2597 4934 2840  Q 4756 3084 4391 3084  Q 3944 3084 3684 2787  Q 3425 2491 3425 1978  L 3425 0  L 2847 0  L 2847 2094  Q 2847 2600 2669 2842  Q 2491 3084 2119 3084  Q 1678 3084 1418 2786  Q 1159 2488 1159 1978  L 1159 0  L 581 0  L 581 3500  L 1159 3500  L 1159 2956  Q 1356 3278 1631 3431  Q 1906 3584 2284 3584  Q 2666 3584 2933 3390  Q 3200 3197 3328 2828  z \" transform=\"scale(0.015625)\"/>\n",
       "     <path id=\"DejaVuSans-62\" d=\"M 3116 1747  Q 3116 2381 2855 2742  Q 2594 3103 2138 3103  Q 1681 3103 1420 2742  Q 1159 2381 1159 1747  Q 1159 1113 1420 752  Q 1681 391 2138 391  Q 2594 391 2855 752  Q 3116 1113 3116 1747  z M 1159 2969  Q 1341 3281 1617 3432  Q 1894 3584 2278 3584  Q 2916 3584 3314 3078  Q 3713 2572 3713 1747  Q 3713 922 3314 415  Q 2916 -91 2278 -91  Q 1894 -91 1617 61  Q 1341 213 1159 525  L 1159 0  L 581 0  L 581 4863  L 1159 4863  L 1159 2969  z \" transform=\"scale(0.015625)\"/>\n",
       "     <path id=\"DejaVuSans-70\" d=\"M 1159 525  L 1159 -1331  L 581 -1331  L 581 3500  L 1159 3500  L 1159 2969  Q 1341 3281 1617 3432  Q 1894 3584 2278 3584  Q 2916 3584 3314 3078  Q 3713 2572 3713 1747  Q 3713 922 3314 415  Q 2916 -91 2278 -91  Q 1894 -91 1617 61  Q 1341 213 1159 525  z M 3116 1747  Q 3116 2381 2855 2742  Q 2594 3103 2138 3103  Q 1681 3103 1420 2742  Q 1159 2381 1159 1747  Q 1159 1113 1420 752  Q 1681 391 2138 391  Q 2594 391 2855 752  Q 3116 1113 3116 1747  z \" transform=\"scale(0.015625)\"/>\n",
       "    </defs>\n",
       "    <use xlink:href=\"#DejaVuSans-4e\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-75\" x=\"74.804688\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6d\" x=\"138.183594\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-62\" x=\"235.595703\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"299.072266\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-72\" x=\"360.595703\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"401.708984\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6f\" x=\"433.496094\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-66\" x=\"494.677734\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"529.882812\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-76\" x=\"561.669922\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-69\" x=\"620.849609\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-72\" x=\"648.632812\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-61\" x=\"689.746094\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6c\" x=\"751.025391\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"778.808594\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-67\" x=\"810.595703\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"874.072266\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6e\" x=\"935.595703\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"998.974609\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-73\" x=\"1060.498047\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"1112.597656\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-70\" x=\"1144.384766\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"1207.861328\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-72\" x=\"1269.384766\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"1310.498047\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-63\" x=\"1342.285156\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"1397.265625\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6c\" x=\"1458.789062\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6c\" x=\"1486.572266\"/>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"p46c9ec0a9e\">\n",
       "   <rect x=\"51.06875\" y=\"24.986701\" width=\"167.4\" height=\"80.64\"/>\n",
       "  </clipPath>\n",
       "  <clipPath id=\"pa8ec8a3d40\">\n",
       "   <rect x=\"51.06875\" y=\"110.666701\" width=\"167.4\" height=\"80.64\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</g>\n",
       "      <text x=\"5\" y=\"15\" font-size=\"14\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">B</text>\n",
       "    </g>\n",
       "    <g transform=\"translate(0, 470) scale(1) translate(0, 0) scale(2) \">\n",
       "      <g><metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2022-11-04T10:45:27.024626</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.5.2, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 253.146701  L 293.19375 253.146701  L 293.19375 -0  L 0 -0  z \" style=\"fill: #ffffff; stroke: #ffffff; stroke-linejoin: miter\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 48.49375 105.626701  L 271.69375 105.626701  L 271.69375 24.986701  L 48.49375 24.986701  z \" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\"/>\n",
       "    <g id=\"xtick_2\"/>\n",
       "    <g id=\"xtick_3\"/>\n",
       "    <g id=\"xtick_4\"/>\n",
       "    <g id=\"xtick_5\"/>\n",
       "    <g id=\"xtick_6\"/>\n",
       "    <g id=\"xtick_7\"/>\n",
       "    <g id=\"xtick_8\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\">\n",
       "     <g id=\"line2d_1\">\n",
       "      <path d=\"M 48.49375 101.961246  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <defs>\n",
       "       <path id=\"m3cc0bc3ded\" d=\"M 0 0  L -2.75 0  \" style=\"stroke: #333333\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m3cc0bc3ded\" x=\"48.49375\" y=\"101.961246\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_1\">\n",
       "      <!-- 0 -->\n",
       "      <g transform=\"translate(37.18125 104.720621)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-30\" d=\"M 2034 4250  Q 1547 4250 1301 3770  Q 1056 3291 1056 2328  Q 1056 1369 1301 889  Q 1547 409 2034 409  Q 2525 409 2770 889  Q 3016 1369 3016 2328  Q 3016 3291 2770 3770  Q 2525 4250 2034 4250  z M 2034 4750  Q 2819 4750 3233 4129  Q 3647 3509 3647 2328  Q 3647 1150 3233 529  Q 2819 -91 2034 -91  Q 1250 -91 836 529  Q 422 1150 422 2328  Q 422 3509 836 4129  Q 1250 4750 2034 4750  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_2\">\n",
       "     <g id=\"line2d_2\">\n",
       "      <path d=\"M 48.49375 86.024487  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m3cc0bc3ded\" x=\"48.49375\" y=\"86.024487\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_2\">\n",
       "      <!-- 0.25 -->\n",
       "      <g transform=\"translate(21.278125 88.783862)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-2e\" d=\"M 684 794  L 1344 794  L 1344 0  L 684 0  L 684 794  z \" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-32\" d=\"M 1228 531  L 3431 531  L 3431 0  L 469 0  L 469 531  Q 828 903 1448 1529  Q 2069 2156 2228 2338  Q 2531 2678 2651 2914  Q 2772 3150 2772 3378  Q 2772 3750 2511 3984  Q 2250 4219 1831 4219  Q 1534 4219 1204 4116  Q 875 4013 500 3803  L 500 4441  Q 881 4594 1212 4672  Q 1544 4750 1819 4750  Q 2544 4750 2975 4387  Q 3406 4025 3406 3419  Q 3406 3131 3298 2873  Q 3191 2616 2906 2266  Q 2828 2175 2409 1742  Q 1991 1309 1228 531  z \" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-35\" d=\"M 691 4666  L 3169 4666  L 3169 4134  L 1269 4134  L 1269 2991  Q 1406 3038 1543 3061  Q 1681 3084 1819 3084  Q 2600 3084 3056 2656  Q 3513 2228 3513 1497  Q 3513 744 3044 326  Q 2575 -91 1722 -91  Q 1428 -91 1123 -41  Q 819 9 494 109  L 494 744  Q 775 591 1075 516  Q 1375 441 1709 441  Q 2250 441 2565 725  Q 2881 1009 2881 1497  Q 2881 1984 2565 2268  Q 2250 2553 1709 2553  Q 1456 2553 1204 2497  Q 953 2441 691 2322  L 691 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-32\" x=\"95.410156\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"159.033203\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_3\">\n",
       "     <g id=\"line2d_3\">\n",
       "      <path d=\"M 48.49375 70.087728  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m3cc0bc3ded\" x=\"48.49375\" y=\"70.087728\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_3\">\n",
       "      <!-- 0.50 -->\n",
       "      <g transform=\"translate(21.278125 72.847103)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"95.410156\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"159.033203\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_4\">\n",
       "     <g id=\"line2d_4\">\n",
       "      <path d=\"M 48.49375 54.150969  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m3cc0bc3ded\" x=\"48.49375\" y=\"54.150969\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_4\">\n",
       "      <!-- 0.75 -->\n",
       "      <g transform=\"translate(21.278125 56.910344)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-37\" d=\"M 525 4666  L 3525 4666  L 3525 4397  L 1831 0  L 1172 0  L 2766 4134  L 525 4134  L 525 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-37\" x=\"95.410156\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"159.033203\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_5\">\n",
       "     <g id=\"line2d_5\">\n",
       "      <path d=\"M 48.49375 38.214211  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m3cc0bc3ded\" x=\"48.49375\" y=\"38.214211\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_5\">\n",
       "      <!-- 1 -->\n",
       "      <g transform=\"translate(37.18125 40.973586)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-31\" d=\"M 794 531  L 1825 531  L 1825 4091  L 703 3866  L 703 4441  L 1819 4666  L 2450 4666  L 2450 531  L 3481 531  L 3481 0  L 794 0  L 794 531  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_6\"/>\n",
       "    <g id=\"ytick_7\"/>\n",
       "    <g id=\"ytick_8\"/>\n",
       "    <g id=\"ytick_9\"/>\n",
       "    <g id=\"ytick_10\"/>\n",
       "    <g id=\"text_6\">\n",
       "     <!-- fraction of cells -->\n",
       "     <g transform=\"translate(14.798438 145.479201)rotate(-90)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-66\" d=\"M 2375 4863  L 2375 4384  L 1825 4384  Q 1516 4384 1395 4259  Q 1275 4134 1275 3809  L 1275 3500  L 2222 3500  L 2222 3053  L 1275 3053  L 1275 0  L 697 0  L 697 3053  L 147 3053  L 147 3500  L 697 3500  L 697 3744  Q 697 4328 969 4595  Q 1241 4863 1831 4863  L 2375 4863  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-72\" d=\"M 2631 2963  Q 2534 3019 2420 3045  Q 2306 3072 2169 3072  Q 1681 3072 1420 2755  Q 1159 2438 1159 1844  L 1159 0  L 581 0  L 581 3500  L 1159 3500  L 1159 2956  Q 1341 3275 1631 3429  Q 1922 3584 2338 3584  Q 2397 3584 2469 3576  Q 2541 3569 2628 3553  L 2631 2963  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-61\" d=\"M 2194 1759  Q 1497 1759 1228 1600  Q 959 1441 959 1056  Q 959 750 1161 570  Q 1363 391 1709 391  Q 2188 391 2477 730  Q 2766 1069 2766 1631  L 2766 1759  L 2194 1759  z M 3341 1997  L 3341 0  L 2766 0  L 2766 531  Q 2569 213 2275 61  Q 1981 -91 1556 -91  Q 1019 -91 701 211  Q 384 513 384 1019  Q 384 1609 779 1909  Q 1175 2209 1959 2209  L 2766 2209  L 2766 2266  Q 2766 2663 2505 2880  Q 2244 3097 1772 3097  Q 1472 3097 1187 3025  Q 903 2953 641 2809  L 641 3341  Q 956 3463 1253 3523  Q 1550 3584 1831 3584  Q 2591 3584 2966 3190  Q 3341 2797 3341 1997  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-63\" d=\"M 3122 3366  L 3122 2828  Q 2878 2963 2633 3030  Q 2388 3097 2138 3097  Q 1578 3097 1268 2742  Q 959 2388 959 1747  Q 959 1106 1268 751  Q 1578 397 2138 397  Q 2388 397 2633 464  Q 2878 531 3122 666  L 3122 134  Q 2881 22 2623 -34  Q 2366 -91 2075 -91  Q 1284 -91 818 406  Q 353 903 353 1747  Q 353 2603 823 3093  Q 1294 3584 2113 3584  Q 2378 3584 2631 3529  Q 2884 3475 3122 3366  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-74\" d=\"M 1172 4494  L 1172 3500  L 2356 3500  L 2356 3053  L 1172 3053  L 1172 1153  Q 1172 725 1289 603  Q 1406 481 1766 481  L 2356 481  L 2356 0  L 1766 0  Q 1100 0 847 248  Q 594 497 594 1153  L 594 3053  L 172 3053  L 172 3500  L 594 3500  L 594 4494  L 1172 4494  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-69\" d=\"M 603 3500  L 1178 3500  L 1178 0  L 603 0  L 603 3500  z M 603 4863  L 1178 4863  L 1178 4134  L 603 4134  L 603 4863  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6f\" d=\"M 1959 3097  Q 1497 3097 1228 2736  Q 959 2375 959 1747  Q 959 1119 1226 758  Q 1494 397 1959 397  Q 2419 397 2687 759  Q 2956 1122 2956 1747  Q 2956 2369 2687 2733  Q 2419 3097 1959 3097  z M 1959 3584  Q 2709 3584 3137 3096  Q 3566 2609 3566 1747  Q 3566 888 3137 398  Q 2709 -91 1959 -91  Q 1206 -91 779 398  Q 353 888 353 1747  Q 353 2609 779 3096  Q 1206 3584 1959 3584  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6e\" d=\"M 3513 2113  L 3513 0  L 2938 0  L 2938 2094  Q 2938 2591 2744 2837  Q 2550 3084 2163 3084  Q 1697 3084 1428 2787  Q 1159 2491 1159 1978  L 1159 0  L 581 0  L 581 3500  L 1159 3500  L 1159 2956  Q 1366 3272 1645 3428  Q 1925 3584 2291 3584  Q 2894 3584 3203 3211  Q 3513 2838 3513 2113  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-20\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-65\" d=\"M 3597 1894  L 3597 1613  L 953 1613  Q 991 1019 1311 708  Q 1631 397 2203 397  Q 2534 397 2845 478  Q 3156 559 3463 722  L 3463 178  Q 3153 47 2828 -22  Q 2503 -91 2169 -91  Q 1331 -91 842 396  Q 353 884 353 1716  Q 353 2575 817 3079  Q 1281 3584 2069 3584  Q 2775 3584 3186 3129  Q 3597 2675 3597 1894  z M 3022 2063  Q 3016 2534 2758 2815  Q 2500 3097 2075 3097  Q 1594 3097 1305 2825  Q 1016 2553 972 2059  L 3022 2063  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6c\" d=\"M 603 4863  L 1178 4863  L 1178 0  L 603 0  L 603 4863  z \" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-73\" d=\"M 2834 3397  L 2834 2853  Q 2591 2978 2328 3040  Q 2066 3103 1784 3103  Q 1356 3103 1142 2972  Q 928 2841 928 2578  Q 928 2378 1081 2264  Q 1234 2150 1697 2047  L 1894 2003  Q 2506 1872 2764 1633  Q 3022 1394 3022 966  Q 3022 478 2636 193  Q 2250 -91 1575 -91  Q 1294 -91 989 -36  Q 684 19 347 128  L 347 722  Q 666 556 975 473  Q 1284 391 1588 391  Q 1994 391 2212 530  Q 2431 669 2431 922  Q 2431 1156 2273 1281  Q 2116 1406 1581 1522  L 1381 1569  Q 847 1681 609 1914  Q 372 2147 372 2553  Q 372 3047 722 3315  Q 1072 3584 1716 3584  Q 2034 3584 2315 3537  Q 2597 3491 2834 3397  z \" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-66\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-72\" x=\"35.205078\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-61\" x=\"76.318359\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-63\" x=\"137.597656\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"192.578125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"231.787109\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"259.570312\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" x=\"320.751953\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"384.130859\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6f\" x=\"415.917969\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-66\" x=\"477.099609\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"512.304688\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-63\" x=\"544.091797\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"599.072266\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6c\" x=\"660.595703\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6c\" x=\"688.378906\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-73\" x=\"716.162109\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PolyCollection_1\">\n",
       "    <path d=\"M 52.576677 101.961246  L 52.576677 41.418427  L 77.074238 41.418427  L 77.074238 101.961246  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 52.576677 41.418427  L 52.576677 38.214211  L 77.074238 38.214211  L 77.074238 41.418427  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 79.796189 101.961246  L 79.796189 41.249784  L 104.29375 41.249784  L 104.29375 101.961246  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 79.796189 41.249784  L 79.796189 38.214211  L 104.29375 38.214211  L 104.29375 41.249784  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 107.015701 101.961246  L 107.015701 43.273499  L 131.513262 43.273499  L 131.513262 101.961246  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 107.015701 43.273499  L 107.015701 38.214211  L 131.513262 38.214211  L 131.513262 43.273499  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 134.235213 101.961246  L 134.235213 44.454  L 158.732774 44.454  L 158.732774 101.961246  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 134.235213 44.454  L 134.235213 38.214211  L 158.732774 38.214211  L 158.732774 44.454  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 161.454726 101.961246  L 161.454726 41.081141  L 185.952287 41.081141  L 185.952287 101.961246  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 161.454726 41.081141  L 161.454726 38.214211  L 185.952287 38.214211  L 185.952287 41.081141  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 188.674238 101.961246  L 188.674238 40.575212  L 213.171799 40.575212  L 213.171799 101.961246  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 188.674238 40.575212  L 188.674238 38.214211  L 213.171799 38.214211  L 213.171799 40.575212  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 215.89375 101.961246  L 215.89375 43.442142  L 240.391311 43.442142  L 240.391311 101.961246  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 215.89375 43.442142  L 215.89375 38.214211  L 240.391311 38.214211  L 240.391311 43.442142  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 243.113262 101.961246  L 243.113262 43.779428  L 267.610823 43.779428  L 267.610823 101.961246  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 243.113262 43.779428  L 243.113262 38.214211  L 267.610823 38.214211  L 267.610823 43.779428  z \" clip-path=\"url(#p2be22814a7)\" style=\"fill: #d55e00\"/>\n",
       "   </g>\n",
       "   <g id=\"text_7\">\n",
       "    <g clip-path=\"url(#p2be22814a7)\">\n",
       "     <!-- 0.05 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(53.692645 34.598882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-35\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_8\">\n",
       "    <g clip-path=\"url(#p2be22814a7)\">\n",
       "     <!-- 0.05 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(80.912157 34.598882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-35\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_9\">\n",
       "    <g clip-path=\"url(#p2be22814a7)\">\n",
       "     <!-- 0.08 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(108.131669 34.598882)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-38\" d=\"M 2034 2216  Q 1584 2216 1326 1975  Q 1069 1734 1069 1313  Q 1069 891 1326 650  Q 1584 409 2034 409  Q 2484 409 2743 651  Q 3003 894 3003 1313  Q 3003 1734 2745 1975  Q 2488 2216 2034 2216  z M 1403 2484  Q 997 2584 770 2862  Q 544 3141 544 3541  Q 544 4100 942 4425  Q 1341 4750 2034 4750  Q 2731 4750 3128 4425  Q 3525 4100 3525 3541  Q 3525 3141 3298 2862  Q 3072 2584 2669 2484  Q 3125 2378 3379 2068  Q 3634 1759 3634 1313  Q 3634 634 3220 271  Q 2806 -91 2034 -91  Q 1263 -91 848 271  Q 434 634 434 1313  Q 434 1759 690 2068  Q 947 2378 1403 2484  z M 1172 3481  Q 1172 3119 1398 2916  Q 1625 2713 2034 2713  Q 2441 2713 2670 2916  Q 2900 3119 2900 3481  Q 2900 3844 2670 4047  Q 2441 4250 2034 4250  Q 1625 4250 1398 4047  Q 1172 3844 1172 3481  z \" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-38\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_10\">\n",
       "    <g clip-path=\"url(#p2be22814a7)\">\n",
       "     <!-- 0.10 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(135.351181 34.598882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-31\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_11\">\n",
       "    <g clip-path=\"url(#p2be22814a7)\">\n",
       "     <!-- 0.04 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(162.570694 34.598882)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-34\" d=\"M 2419 4116  L 825 1625  L 2419 1625  L 2419 4116  z M 2253 4666  L 3047 4666  L 3047 1625  L 3713 1625  L 3713 1100  L 3047 1100  L 3047 0  L 2419 0  L 2419 1100  L 313 1100  L 313 1709  L 2253 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-34\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_12\">\n",
       "    <g clip-path=\"url(#p2be22814a7)\">\n",
       "     <!-- 0.04 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(189.790206 34.598882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-34\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_13\">\n",
       "    <g clip-path=\"url(#p2be22814a7)\">\n",
       "     <!-- 0.08 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(217.009718 34.598882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-38\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_14\">\n",
       "    <g clip-path=\"url(#p2be22814a7)\">\n",
       "     <!-- 0.09 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(244.22923 34.598882)scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-39\" d=\"M 703 97  L 703 672  Q 941 559 1184 500  Q 1428 441 1663 441  Q 2288 441 2617 861  Q 2947 1281 2994 2138  Q 2813 1869 2534 1725  Q 2256 1581 1919 1581  Q 1219 1581 811 2004  Q 403 2428 403 3163  Q 403 3881 828 4315  Q 1253 4750 1959 4750  Q 2769 4750 3195 4129  Q 3622 3509 3622 2328  Q 3622 1225 3098 567  Q 2575 -91 1691 -91  Q 1453 -91 1209 -44  Q 966 3 703 97  z M 1959 2075  Q 2384 2075 2632 2365  Q 2881 2656 2881 3163  Q 2881 3666 2632 3958  Q 2384 4250 1959 4250  Q 1534 4250 1286 3958  Q 1038 3666 1038 3163  Q 1038 2656 1286 2365  Q 1534 2075 1959 2075  z \" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-39\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 271.69375 105.626701  L 285.99375 105.626701  L 285.99375 24.986701  L 271.69375 24.986701  L 271.69375 105.626701  z \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 48.49375 105.626701  L 48.49375 24.986701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 48.49375 105.626701  L 271.69375 105.626701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_15\">\n",
       "    <!-- trial1 -->\n",
       "    <g style=\"fill: #1a1a1a\" transform=\"translate(276.4155 53.832326)rotate(-270)scale(0.088 -0.088)\">\n",
       "     <use xlink:href=\"#DejaVuSans-74\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-72\" x=\"39.208984\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-69\" x=\"80.322266\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-61\" x=\"108.105469\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-6c\" x=\"169.384766\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-31\" x=\"197.167969\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 57.331094 245.946701  L 255.656406 245.946701  L 255.656406 205.706701  L 57.331094 205.706701  z \" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_7\">\n",
       "    <path d=\"M 69.531094 233.746701  L 85.371094 233.746701  L 85.371094 217.906701  L 69.531094 217.906701  L 69.531094 233.746701  z \" clip-path=\"url(#p21c6b33a9d)\" style=\"fill: none\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_8\">\n",
       "    <path d=\"M 69.531094 233.746701  L 85.371094 233.746701  L 85.371094 217.906701  L 69.531094 217.906701  z \" clip-path=\"url(#p21c6b33a9d)\" style=\"fill: #d55e00\"/>\n",
       "   </g>\n",
       "   <g id=\"text_16\">\n",
       "    <!-- Gene absent -->\n",
       "    <g transform=\"translate(88.371094 228.586076)scale(0.1 -0.1)\">\n",
       "     <defs>\n",
       "      <path id=\"DejaVuSans-47\" d=\"M 3809 666  L 3809 1919  L 2778 1919  L 2778 2438  L 4434 2438  L 4434 434  Q 4069 175 3628 42  Q 3188 -91 2688 -91  Q 1594 -91 976 548  Q 359 1188 359 2328  Q 359 3472 976 4111  Q 1594 4750 2688 4750  Q 3144 4750 3555 4637  Q 3966 4525 4313 4306  L 4313 3634  Q 3963 3931 3569 4081  Q 3175 4231 2741 4231  Q 1884 4231 1454 3753  Q 1025 3275 1025 2328  Q 1025 1384 1454 906  Q 1884 428 2741 428  Q 3075 428 3337 486  Q 3600 544 3809 666  z \" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"DejaVuSans-62\" d=\"M 3116 1747  Q 3116 2381 2855 2742  Q 2594 3103 2138 3103  Q 1681 3103 1420 2742  Q 1159 2381 1159 1747  Q 1159 1113 1420 752  Q 1681 391 2138 391  Q 2594 391 2855 752  Q 3116 1113 3116 1747  z M 1159 2969  Q 1341 3281 1617 3432  Q 1894 3584 2278 3584  Q 2916 3584 3314 3078  Q 3713 2572 3713 1747  Q 3713 922 3314 415  Q 2916 -91 2278 -91  Q 1894 -91 1617 61  Q 1341 213 1159 525  L 1159 0  L 581 0  L 581 4863  L 1159 4863  L 1159 2969  z \" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#DejaVuSans-47\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-65\" x=\"77.490234\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-6e\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-65\" x=\"202.392578\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-20\" x=\"263.916016\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-61\" x=\"295.703125\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-62\" x=\"356.982422\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-73\" x=\"420.458984\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-65\" x=\"472.558594\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-6e\" x=\"534.082031\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-74\" x=\"597.460938\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"patch_9\">\n",
       "    <path d=\"M 157.038281 233.746701  L 172.878281 233.746701  L 172.878281 217.906701  L 157.038281 217.906701  L 157.038281 233.746701  z \" clip-path=\"url(#paf33f55a0e)\" style=\"fill: none\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_10\">\n",
       "    <path d=\"M 157.038281 233.746701  L 172.878281 233.746701  L 172.878281 217.906701  L 157.038281 217.906701  z \" clip-path=\"url(#paf33f55a0e)\" style=\"fill: #999999\"/>\n",
       "   </g>\n",
       "   <g id=\"text_17\">\n",
       "    <!-- Gene present -->\n",
       "    <g transform=\"translate(175.878281 228.586076)scale(0.1 -0.1)\">\n",
       "     <defs>\n",
       "      <path id=\"DejaVuSans-70\" d=\"M 1159 525  L 1159 -1331  L 581 -1331  L 581 3500  L 1159 3500  L 1159 2969  Q 1341 3281 1617 3432  Q 1894 3584 2278 3584  Q 2916 3584 3314 3078  Q 3713 2572 3713 1747  Q 3713 922 3314 415  Q 2916 -91 2278 -91  Q 1894 -91 1617 61  Q 1341 213 1159 525  z M 3116 1747  Q 3116 2381 2855 2742  Q 2594 3103 2138 3103  Q 1681 3103 1420 2742  Q 1159 2381 1159 1747  Q 1159 1113 1420 752  Q 1681 391 2138 391  Q 2594 391 2855 752  Q 3116 1113 3116 1747  z \" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#DejaVuSans-47\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-65\" x=\"77.490234\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-6e\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-65\" x=\"202.392578\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-20\" x=\"263.916016\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-70\" x=\"295.703125\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-72\" x=\"359.179688\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-65\" x=\"398.042969\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-73\" x=\"459.566406\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-65\" x=\"511.666016\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-6e\" x=\"573.189453\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-74\" x=\"636.568359\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       "  <g id=\"axes_2\">\n",
       "   <g id=\"patch_11\">\n",
       "    <path d=\"M 48.49375 191.306701  L 271.69375 191.306701  L 271.69375 110.666701  L 48.49375 110.666701  z \" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_3\">\n",
       "    <g id=\"xtick_9\">\n",
       "     <g id=\"line2d_6\">\n",
       "      <path d=\"M 64.825457 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <defs>\n",
       "       <path id=\"m4bee8a817f\" d=\"M 0 0  L 0 2.75  \" style=\"stroke: #333333\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m4bee8a817f\" x=\"64.825457\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_18\">\n",
       "      <!-- PB2 -->\n",
       "      <g transform=\"translate(55.198895 203.855138)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-50\" d=\"M 1259 4147  L 1259 2394  L 2053 2394  Q 2494 2394 2734 2622  Q 2975 2850 2975 3272  Q 2975 3691 2734 3919  Q 2494 4147 2053 4147  L 1259 4147  z M 628 4666  L 2053 4666  Q 2838 4666 3239 4311  Q 3641 3956 3641 3272  Q 3641 2581 3239 2228  Q 2838 1875 2053 1875  L 1259 1875  L 1259 0  L 628 0  L 628 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-42\" d=\"M 1259 2228  L 1259 519  L 2272 519  Q 2781 519 3026 730  Q 3272 941 3272 1375  Q 3272 1813 3026 2020  Q 2781 2228 2272 2228  L 1259 2228  z M 1259 4147  L 1259 2741  L 2194 2741  Q 2656 2741 2882 2914  Q 3109 3088 3109 3444  Q 3109 3797 2882 3972  Q 2656 4147 2194 4147  L 1259 4147  z M 628 4666  L 2241 4666  Q 2963 4666 3353 4366  Q 3744 4066 3744 3513  Q 3744 3084 3544 2831  Q 3344 2578 2956 2516  Q 3422 2416 3680 2098  Q 3938 1781 3938 1306  Q 3938 681 3513 340  Q 3088 0 2303 0  L 628 0  L 628 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-50\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-42\" x=\"60.302734\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-32\" x=\"128.90625\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_10\">\n",
       "     <g id=\"line2d_7\">\n",
       "      <path d=\"M 92.04497 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m4bee8a817f\" x=\"92.04497\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_19\">\n",
       "      <!-- PB1 -->\n",
       "      <g transform=\"translate(82.418407 203.855138)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-50\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-42\" x=\"60.302734\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-31\" x=\"128.90625\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_11\">\n",
       "     <g id=\"line2d_8\">\n",
       "      <path d=\"M 119.264482 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m4bee8a817f\" x=\"119.264482\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_20\">\n",
       "      <!-- PA -->\n",
       "      <g transform=\"translate(113.148075 203.855138)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-41\" d=\"M 2188 4044  L 1331 1722  L 3047 1722  L 2188 4044  z M 1831 4666  L 2547 4666  L 4325 0  L 3669 0  L 3244 1197  L 1141 1197  L 716 0  L 50 0  L 1831 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-50\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-41\" x=\"53.927734\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_12\">\n",
       "     <g id=\"line2d_9\">\n",
       "      <path d=\"M 146.483994 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m4bee8a817f\" x=\"146.483994\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_21\">\n",
       "      <!-- HA -->\n",
       "      <g transform=\"translate(139.303525 203.855138)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-48\" d=\"M 628 4666  L 1259 4666  L 1259 2753  L 3553 2753  L 3553 4666  L 4184 4666  L 4184 0  L 3553 0  L 3553 2222  L 1259 2222  L 1259 0  L 628 0  L 628 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-48\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-41\" x=\"75.195312\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_13\">\n",
       "     <g id=\"line2d_10\">\n",
       "      <path d=\"M 173.703506 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m4bee8a817f\" x=\"173.703506\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_22\">\n",
       "      <!-- NP -->\n",
       "      <g transform=\"translate(166.948037 203.855138)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-4e\" d=\"M 628 4666  L 1478 4666  L 3547 763  L 3547 4666  L 4159 4666  L 4159 0  L 3309 0  L 1241 3903  L 1241 0  L 628 0  L 628 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-4e\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-50\" x=\"74.804688\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_14\">\n",
       "     <g id=\"line2d_11\">\n",
       "      <path d=\"M 200.923018 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m4bee8a817f\" x=\"200.923018\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_23\">\n",
       "      <!-- NA -->\n",
       "      <g transform=\"translate(193.762081 203.855138)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-4e\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-41\" x=\"74.804688\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_15\">\n",
       "     <g id=\"line2d_12\">\n",
       "      <path d=\"M 228.14253 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m4bee8a817f\" x=\"228.14253\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_24\">\n",
       "      <!-- M -->\n",
       "      <g transform=\"translate(223.828468 203.855138)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-4d\" d=\"M 628 4666  L 1569 4666  L 2759 1491  L 3956 4666  L 4897 4666  L 4897 0  L 4281 0  L 4281 4097  L 3078 897  L 2444 897  L 1241 4097  L 1241 0  L 628 0  L 628 4666  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-4d\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_16\">\n",
       "     <g id=\"line2d_13\">\n",
       "      <path d=\"M 255.362043 191.306701  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m4bee8a817f\" x=\"255.362043\" y=\"191.306701\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_25\">\n",
       "      <!-- NS -->\n",
       "      <g transform=\"translate(248.447199 203.855138)scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-53\" d=\"M 3425 4513  L 3425 3897  Q 3066 4069 2747 4153  Q 2428 4238 2131 4238  Q 1616 4238 1336 4038  Q 1056 3838 1056 3469  Q 1056 3159 1242 3001  Q 1428 2844 1947 2747  L 2328 2669  Q 3034 2534 3370 2195  Q 3706 1856 3706 1288  Q 3706 609 3251 259  Q 2797 -91 1919 -91  Q 1588 -91 1214 -16  Q 841 59 441 206  L 441 856  Q 825 641 1194 531  Q 1563 422 1919 422  Q 2459 422 2753 634  Q 3047 847 3047 1241  Q 3047 1584 2836 1778  Q 2625 1972 2144 2069  L 1759 2144  Q 1053 2284 737 2584  Q 422 2884 422 3419  Q 422 4038 858 4394  Q 1294 4750 2059 4750  Q 2388 4750 2728 4690  Q 3069 4631 3425 4513  z \" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-4e\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-53\" x=\"74.804688\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_4\">\n",
       "    <g id=\"ytick_11\">\n",
       "     <g id=\"line2d_14\">\n",
       "      <path d=\"M 48.49375 187.641246  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m3cc0bc3ded\" x=\"48.49375\" y=\"187.641246\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_26\">\n",
       "      <!-- 0 -->\n",
       "      <g transform=\"translate(37.18125 190.400621)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_12\">\n",
       "     <g id=\"line2d_15\">\n",
       "      <path d=\"M 48.49375 171.704487  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m3cc0bc3ded\" x=\"48.49375\" y=\"171.704487\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_27\">\n",
       "      <!-- 0.25 -->\n",
       "      <g transform=\"translate(21.278125 174.463862)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-32\" x=\"95.410156\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"159.033203\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_13\">\n",
       "     <g id=\"line2d_16\">\n",
       "      <path d=\"M 48.49375 155.767728  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m3cc0bc3ded\" x=\"48.49375\" y=\"155.767728\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_28\">\n",
       "      <!-- 0.50 -->\n",
       "      <g transform=\"translate(21.278125 158.527103)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"95.410156\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"159.033203\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_14\">\n",
       "     <g id=\"line2d_17\">\n",
       "      <path d=\"M 48.49375 139.830969  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m3cc0bc3ded\" x=\"48.49375\" y=\"139.830969\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_29\">\n",
       "      <!-- 0.75 -->\n",
       "      <g transform=\"translate(21.278125 142.590344)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-37\" x=\"95.410156\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"159.033203\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_15\">\n",
       "     <g id=\"line2d_18\">\n",
       "      <path d=\"M 48.49375 123.894211  \" style=\"fill: none; stroke: #000000; stroke-width: 1.5; stroke-linecap: square\"/>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m3cc0bc3ded\" x=\"48.49375\" y=\"123.894211\" style=\"stroke: #333333\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_30\">\n",
       "      <!-- 1 -->\n",
       "      <g transform=\"translate(37.18125 126.653586)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_16\"/>\n",
       "    <g id=\"ytick_17\"/>\n",
       "    <g id=\"ytick_18\"/>\n",
       "    <g id=\"ytick_19\"/>\n",
       "    <g id=\"ytick_20\"/>\n",
       "   </g>\n",
       "   <g id=\"PolyCollection_2\">\n",
       "    <path d=\"M 52.576677 187.641246  L 52.576677 128.160744  L 77.074238 128.160744  L 77.074238 187.641246  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 52.576677 128.160744  L 52.576677 123.894211  L 77.074238 123.894211  L 77.074238 128.160744  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 79.796189 187.641246  L 79.796189 126.403936  L 104.29375 126.403936  L 104.29375 187.641246  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 79.796189 126.403936  L 79.796189 123.894211  L 104.29375 123.894211  L 104.29375 126.403936  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 107.015701 187.641246  L 107.015701 129.415607  L 131.513262 129.415607  L 131.513262 187.641246  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 107.015701 129.415607  L 107.015701 123.894211  L 131.513262 123.894211  L 131.513262 129.415607  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 134.235213 187.641246  L 134.235213 133.180196  L 158.732774 133.180196  L 158.732774 187.641246  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 134.235213 133.180196  L 134.235213 123.894211  L 158.732774 123.894211  L 158.732774 133.180196  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 161.454726 187.641246  L 161.454726 126.403936  L 185.952287 126.403936  L 185.952287 187.641246  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 161.454726 126.403936  L 161.454726 123.894211  L 185.952287 123.894211  L 185.952287 126.403936  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 188.674238 187.641246  L 188.674238 126.654909  L 213.171799 126.654909  L 213.171799 187.641246  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 188.674238 126.654909  L 188.674238 123.894211  L 213.171799 123.894211  L 213.171799 126.654909  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 215.89375 187.641246  L 215.89375 128.160744  L 240.391311 128.160744  L 240.391311 187.641246  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 215.89375 128.160744  L 215.89375 123.894211  L 240.391311 123.894211  L 240.391311 128.160744  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #d55e00\"/>\n",
       "    <path d=\"M 243.113262 187.641246  L 243.113262 131.423388  L 267.610823 131.423388  L 267.610823 187.641246  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #999999\"/>\n",
       "    <path d=\"M 243.113262 131.423388  L 243.113262 123.894211  L 267.610823 123.894211  L 267.610823 131.423388  z \" clip-path=\"url(#pa38d82f77d)\" style=\"fill: #d55e00\"/>\n",
       "   </g>\n",
       "   <g id=\"text_31\">\n",
       "    <g clip-path=\"url(#pa38d82f77d)\">\n",
       "     <!-- 0.07 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(53.692645 120.278882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-37\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_32\">\n",
       "    <g clip-path=\"url(#pa38d82f77d)\">\n",
       "     <!-- 0.04 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(80.912157 120.278882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-34\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_33\">\n",
       "    <g clip-path=\"url(#pa38d82f77d)\">\n",
       "     <!-- 0.09 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(108.131669 120.278882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-39\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_34\">\n",
       "    <g clip-path=\"url(#pa38d82f77d)\">\n",
       "     <!-- 0.15 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(135.351181 120.278882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-31\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-35\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_35\">\n",
       "    <g clip-path=\"url(#pa38d82f77d)\">\n",
       "     <!-- 0.04 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(162.570694 120.278882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-34\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_36\">\n",
       "    <g clip-path=\"url(#pa38d82f77d)\">\n",
       "     <!-- 0.04 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(189.790206 120.278882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-34\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_37\">\n",
       "    <g clip-path=\"url(#pa38d82f77d)\">\n",
       "     <!-- 0.07 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(217.009718 120.278882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-37\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_38\">\n",
       "    <g clip-path=\"url(#pa38d82f77d)\">\n",
       "     <!-- 0.12 -->\n",
       "     <g style=\"fill: #d55e00\" transform=\"translate(244.22923 120.278882)scale(0.1 -0.1)\">\n",
       "      <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-31\" x=\"95.410156\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-32\" x=\"159.033203\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"patch_12\">\n",
       "    <path d=\"M 271.69375 191.306701  L 285.99375 191.306701  L 285.99375 110.666701  L 271.69375 110.666701  L 271.69375 191.306701  z \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_13\">\n",
       "    <path d=\"M 48.49375 191.306701  L 48.49375 110.666701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_14\">\n",
       "    <path d=\"M 48.49375 191.306701  L 271.69375 191.306701  \" style=\"fill: none; stroke: #000000; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_39\">\n",
       "    <!-- trial3 -->\n",
       "    <g style=\"fill: #1a1a1a\" transform=\"translate(276.4155 139.512326)rotate(-270)scale(0.088 -0.088)\">\n",
       "     <defs>\n",
       "      <path id=\"DejaVuSans-33\" d=\"M 2597 2516  Q 3050 2419 3304 2112  Q 3559 1806 3559 1356  Q 3559 666 3084 287  Q 2609 -91 1734 -91  Q 1441 -91 1130 -33  Q 819 25 488 141  L 488 750  Q 750 597 1062 519  Q 1375 441 1716 441  Q 2309 441 2620 675  Q 2931 909 2931 1356  Q 2931 1769 2642 2001  Q 2353 2234 1838 2234  L 1294 2234  L 1294 2753  L 1863 2753  Q 2328 2753 2575 2939  Q 2822 3125 2822 3475  Q 2822 3834 2567 4026  Q 2313 4219 1838 4219  Q 1578 4219 1281 4162  Q 984 4106 628 3988  L 628 4550  Q 988 4650 1302 4700  Q 1616 4750 1894 4750  Q 2613 4750 3031 4423  Q 3450 4097 3450 3541  Q 3450 3153 3228 2886  Q 3006 2619 2597 2516  z \" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#DejaVuSans-74\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-72\" x=\"39.208984\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-69\" x=\"80.322266\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-61\" x=\"108.105469\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-6c\" x=\"169.384766\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-33\" x=\"197.167969\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       "  <g id=\"text_40\">\n",
       "   <!-- Viral gene presence in infected cells -->\n",
       "   <g transform=\"translate(47.72875 16.318125)scale(0.12 -0.12)\">\n",
       "    <defs>\n",
       "     <path id=\"DejaVuSans-56\" d=\"M 1831 0  L 50 4666  L 709 4666  L 2188 738  L 3669 4666  L 4325 4666  L 2547 0  L 1831 0  z \" transform=\"scale(0.015625)\"/>\n",
       "     <path id=\"DejaVuSans-67\" d=\"M 2906 1791  Q 2906 2416 2648 2759  Q 2391 3103 1925 3103  Q 1463 3103 1205 2759  Q 947 2416 947 1791  Q 947 1169 1205 825  Q 1463 481 1925 481  Q 2391 481 2648 825  Q 2906 1169 2906 1791  z M 3481 434  Q 3481 -459 3084 -895  Q 2688 -1331 1869 -1331  Q 1566 -1331 1297 -1286  Q 1028 -1241 775 -1147  L 775 -588  Q 1028 -725 1275 -790  Q 1522 -856 1778 -856  Q 2344 -856 2625 -561  Q 2906 -266 2906 331  L 2906 616  Q 2728 306 2450 153  Q 2172 0 1784 0  Q 1141 0 747 490  Q 353 981 353 1791  Q 353 2603 747 3093  Q 1141 3584 1784 3584  Q 2172 3584 2450 3431  Q 2728 3278 2906 2969  L 2906 3500  L 3481 3500  L 3481 434  z \" transform=\"scale(0.015625)\"/>\n",
       "     <path id=\"DejaVuSans-64\" d=\"M 2906 2969  L 2906 4863  L 3481 4863  L 3481 0  L 2906 0  L 2906 525  Q 2725 213 2448 61  Q 2172 -91 1784 -91  Q 1150 -91 751 415  Q 353 922 353 1747  Q 353 2572 751 3078  Q 1150 3584 1784 3584  Q 2172 3584 2448 3432  Q 2725 3281 2906 2969  z M 947 1747  Q 947 1113 1208 752  Q 1469 391 1925 391  Q 2381 391 2643 752  Q 2906 1113 2906 1747  Q 2906 2381 2643 2742  Q 2381 3103 1925 3103  Q 1469 3103 1208 2742  Q 947 2381 947 1747  z \" transform=\"scale(0.015625)\"/>\n",
       "    </defs>\n",
       "    <use xlink:href=\"#DejaVuSans-56\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-69\" x=\"66.158203\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-72\" x=\"93.941406\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-61\" x=\"135.054688\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6c\" x=\"196.333984\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"224.117188\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-67\" x=\"255.904297\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"319.380859\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6e\" x=\"380.904297\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"444.283203\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"505.806641\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-70\" x=\"537.59375\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-72\" x=\"601.070312\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"639.933594\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-73\" x=\"701.457031\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"753.556641\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6e\" x=\"815.080078\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-63\" x=\"878.458984\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"933.439453\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"994.962891\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-69\" x=\"1026.75\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6e\" x=\"1054.533203\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"1117.912109\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-69\" x=\"1149.699219\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6e\" x=\"1177.482422\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-66\" x=\"1240.861328\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"1276.066406\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-63\" x=\"1337.589844\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-74\" x=\"1392.570312\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"1431.779297\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-64\" x=\"1493.302734\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-20\" x=\"1556.779297\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-63\" x=\"1588.566406\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-65\" x=\"1643.546875\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6c\" x=\"1705.070312\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-6c\" x=\"1732.853516\"/>\n",
       "    <use xlink:href=\"#DejaVuSans-73\" x=\"1760.636719\"/>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"p2be22814a7\">\n",
       "   <rect x=\"48.49375\" y=\"24.986701\" width=\"223.2\" height=\"80.64\"/>\n",
       "  </clipPath>\n",
       "  <clipPath id=\"p21c6b33a9d\">\n",
       "   <path d=\"M 69.531094 233.746701  L 69.531094 217.906701  L 85.371094 217.906701  L 85.371094 233.746701  \"/>\n",
       "  </clipPath>\n",
       "  <clipPath id=\"paf33f55a0e\">\n",
       "   <path d=\"M 157.038281 233.746701  L 157.038281 217.906701  L 172.878281 217.906701  L 172.878281 233.746701  \"/>\n",
       "  </clipPath>\n",
       "  <clipPath id=\"pa38d82f77d\">\n",
       "   <rect x=\"48.49375\" y=\"110.666701\" width=\"223.2\" height=\"80.64\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</g>\n",
       "      <text x=\"5\" y=\"15\" font-size=\"14\" font-family=\"Verdana\" font-weight=\"normal\" letter-spacing=\"0\" text-anchor=\"start\" fill=\"black\">C</text>\n",
       "    </g>\n",
       "  </g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<svgutils.compose.Figure at 0x7fc5d9b78fd0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig2_file = figure_dir+'/'+'fig2.svg'\n",
    "\n",
    "fig2 = (Figure(\"32cm\", \"26cm\",\n",
    "       Panel(\n",
    "            Grid(20, 20),\n",
    "           SVG(viral_transcription_file),\n",
    "           Text(\"A\", 5, 15, size=14)).scale(2),\n",
    "       Panel(\n",
    "           SVG(genes_per_cell_file),\n",
    "           Text(\"B\", 5, 15, size=14)).scale(2).move(720,0),\n",
    "       Panel(\n",
    "           SVG(gene_absence_file),\n",
    "           Text(\"C\", 5, 15, size=14)).scale(2).move(0,470),\n",
    "       ))\n",
    "display(fig2)\n",
    "fig2.save(fig2_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot viral gene presence or absence raw data\n",
    "For each single infected cell, plot the expression of each viral gene relative to total viral expression. Color by whether gene was called as present or absent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression_file = figure_dir+'/'+'gene_expression_plot.svg'\n",
    "\n",
    "gene_expression_plot = (\n",
    "    p9.ggplot(viral_genes_by_cell,\n",
    "               p9.aes(x='frac_viral_UMIs',\n",
    "                      y='frac_gene_UMIs',\n",
    "                      color='gene_present')) +\n",
    "    p9.geom_point(alpha=0.3, size=3) +\n",
    "    p9.facet_wrap('~gene', ncol=4) +\n",
    "    p9.labs(x='fraction of mRNA from virus',\n",
    "            y='fraction of mRNA\\nfrom specific viral gene') +\n",
    "    p9.scale_x_log10(breaks=[0.01,0.03,0.1,0.3]) +\n",
    "    p9.scale_y_log10(breaks=[1e-4,1e-2,1]) +\n",
    "    p9.theme(figure_size=(8, 4),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text=p9.element_text(size=12, color=\"black\"),\n",
    "             axis_text_x=p9.element_text(rotation=90),\n",
    "             legend_position='top',\n",
    "             legend_box_spacing=0,\n",
    "             legend_title=p9.element_blank(),\n",
    "             legend_text=p9.element_text(size=12),\n",
    "             strip_text=p9.element_text(size=12)) +\n",
    "    p9.scale_color_manual([CBPALETTE[6],CBPALETTE[0]]))\n",
    "\n",
    "gene_expression_plot.save(filename = gene_expression_file)\n",
    "display(gene_expression_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Viral genome sequencing\n",
    "### Plot viral transcription in fully-sequenced cells\n",
    "For later analyses, I will require the cells to have PacBio sequencing information for all expressed viral genes, so that we can call consensus genomes. I want to see how viral transcription activity is associated with our ability to fully sequence a viral genome.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fully_sequenced_file = figure_dir+'/'+'fully_sequenced_plot.svg'\n",
    "\n",
    "fully_sequenced_plot = (\n",
    "    p9.ggplot(wt_infected_cells,\n",
    "               p9.aes(x='virus_fully_sequenced',\n",
    "                      y='frac_viral_UMIs')) +\n",
    "    p9.geom_jitter(alpha=0.3, size=3, width=0.2) +\n",
    "    p9.stat_summary(fun_y=np.mean, \n",
    "                    fun_ymin=np.mean,\n",
    "                    fun_ymax=np.mean,\n",
    "                    geom='crossbar',\n",
    "                    width=0.9,\n",
    "                    color=CBPALETTE[2]) +\n",
    "    p9.geom_text(data=wt_infected_cells.groupby('virus_fully_sequenced').nunique().reset_index(),\n",
    "                 mapping=p9.aes(y=0.74,\n",
    "                                label='cell_barcode'),\n",
    "                 size=12) +\n",
    "    p9.coord_flip() +\n",
    "    p9.labs(x='',\n",
    "            y='fraction mRNA from virus') +\n",
    "    p9.scale_y_continuous(limits=[0,0.75], breaks=[0,0.25,0.5,0.75]) +\n",
    "    p9.scale_x_discrete(labels=['Incomplete sequencing', 'Complete sequencing']) +\n",
    "    p9.theme(figure_size=(3.5, 1.75),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text=p9.element_text(size=12, color=\"black\")))\n",
    "\n",
    "fully_sequenced_plot.save(filename = fully_sequenced_file)\n",
    "display(fully_sequenced_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, generate a dataframe that is a subset of `wt_infected_cells`, containing just cells with complete PacBio sequencing of all expressed viral genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fully_sequenced_cells = (\n",
    "    wt_infected_cells\n",
    "    .query('virus_fully_sequenced == True'))\n",
    "display(fully_sequenced_cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aggregate mutation information\n",
    "Viral mutation information from the PacBio sequencing is currently spread across many columns (one for each gene). I want to aggregate all viral mutaiton information into a single column so that I can display this on plots in the future. I will also annotate cells by whether they have non-synonymous mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def viral_mutations(row):\n",
    "    row_dict = row.to_dict()\n",
    "    mutations = []\n",
    "    for key, val in row_dict.items():\n",
    "        if re.fullmatch(\"mutations_flu\\w+\", key) and val:\n",
    "            gene = key.replace(\"mutations_flu\", \"\")\n",
    "            if val == \"Not Detected\":\n",
    "                assert not row_dict[f\"present_flu{gene}\"]\n",
    "            elif val != \"WT\":\n",
    "                mutations.append(f\"{gene}: {val}\")\n",
    "    return \", \".join(mutations)\n",
    "\n",
    "def nonsynonymous_mutations(row):\n",
    "    row_dict = row.to_dict()\n",
    "    nonsynonymous_mutations = []\n",
    "    for key, val in row_dict.items():\n",
    "        if re.fullmatch(\"mutations_flu\\w+\", key) and val:\n",
    "            gene = key.replace(\"mutations_flu\", \"\")\n",
    "            if val == \"Not Detected\":\n",
    "                assert not row_dict[f\"present_flu{gene}\"]\n",
    "            elif val != \"WT\":\n",
    "                for x in val.split(' '):\n",
    "                    if \"synonymous\" in x:\n",
    "                        pass\n",
    "                    elif \"noncoding\" in x:\n",
    "                        pass\n",
    "                    else:\n",
    "                        nonsynonymous_mutations.append(f\"{gene}: {x}\")\n",
    "    return \", \".join(nonsynonymous_mutations)\n",
    "\n",
    "def nonsynonymous_mutation_count(row):\n",
    "    row_dict = row.to_dict()\n",
    "    nonsynonymous_mutation_count = 0\n",
    "    for key, val in row_dict.items():\n",
    "        if re.fullmatch(\"mutations_flu\\w+\", key) and val:\n",
    "            gene = key.replace(\"mutations_flu\", \"\")\n",
    "            if val == \"Not Detected\":\n",
    "                assert not row_dict[f\"present_flu{gene}\"]\n",
    "            elif val != \"WT\":\n",
    "                for x in val.split(' '):\n",
    "                    if \"synonymous\" in x:\n",
    "                        pass\n",
    "                    elif \"noncoding\" in x:\n",
    "                        pass\n",
    "                    else:\n",
    "                        nonsynonymous_mutation_count += 1\n",
    "    return nonsynonymous_mutation_count\n",
    "\n",
    "def mutation_support(row):\n",
    "    row_dict = row.to_dict()\n",
    "    support_vals = []\n",
    "    for key, val in row_dict.items():\n",
    "        if re.fullmatch(\"mutation_support_flu\\w+\", key) and val:\n",
    "            gene = key.replace(\"mutation_support_flu\", \"\")\n",
    "            if val != \"Not Applicable\":\n",
    "                support_vals.append(f\"{gene}: {val}\")\n",
    "    return \", \".join(support_vals)\n",
    "\n",
    "def mutated_genes(row):\n",
    "    row_dict = row.to_dict()\n",
    "    mutated_genes = []\n",
    "    for key, val in row_dict.items():\n",
    "        if re.fullmatch(\"mutations_flu\\w+\", key) and val:\n",
    "            gene = key.replace(\"mutations_flu\", \"\")\n",
    "            if val == \"Not Detected\":\n",
    "                assert not row_dict[f\"present_flu{gene}\"]\n",
    "            elif val != \"WT\":\n",
    "                mutated_genes.append(gene)\n",
    "    return \", \".join(mutated_genes)\n",
    "\n",
    "def label_cells(row):\n",
    "    cell_label = None\n",
    "    if row['all_viral_genes'] == True:\n",
    "        if row['has_nonsynonymous_mutations'] == False:\n",
    "            cell_label = \"Complete unmutated genome\"\n",
    "        else:\n",
    "            cell_label = \"Complete mutated genome\"\n",
    "    elif row['all_viral_genes'] == False:\n",
    "        if row['has_nonsynonymous_mutations'] == False:\n",
    "            cell_label = \"Missing viral gene(s)\"\n",
    "        else:\n",
    "            cell_label = \"Missing viral gene(s) and\\nnon-synonymous mutations\"\n",
    "    return cell_label\n",
    "\n",
    "fully_sequenced_cells = (\n",
    "    fully_sequenced_cells\n",
    "    .assign(\n",
    "        viral_mutations=lambda x: x.apply(viral_mutations, axis=1),\n",
    "        nonsynonymous_mutations=lambda x: x.apply(nonsynonymous_mutations, axis=1),\n",
    "        mutation_support=lambda x: x.apply(mutation_support, axis=1),\n",
    "        mutated_genes=lambda x: x.apply(mutated_genes, axis=1),\n",
    "        has_mutations=lambda x: x[\"mutated_genes\"].astype(bool),\n",
    "        has_nonsynonymous_mutations=lambda x: x[\"nonsynonymous_mutations\"].astype(bool),\n",
    "        nonsynonymous_mutation_count=lambda x: x.apply(nonsynonymous_mutation_count, axis=1),\n",
    "        cell_label=lambda x: x.apply(label_cells, axis=1).astype('category')\n",
    "    )\n",
    ")\n",
    "display(fully_sequenced_cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Viral genotype statistics\n",
    "Calculate how many infected cells we characterized, how many conensus viral sequences have mutations, and what kinds of mutations they carry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutated_cell_counts = (\n",
    "    pd.DataFrame(\n",
    "        fully_sequenced_cells.groupby(['cell_label'])\n",
    "        ['cell_barcode'].nunique()).reset_index()\n",
    "    .rename(columns={'cell_barcode': 'n_cells'}))\n",
    "\n",
    "mutated_cell_counts['cell_label'] = (\n",
    "    mutated_cell_counts['cell_label']\n",
    "    .cat.reorder_categories(\n",
    "        ['Missing viral gene(s) and\\nnon-synonymous mutations', 'Missing viral gene(s)', \n",
    "        'Complete mutated genome', 'Complete unmutated genome']))\n",
    "\n",
    "display(mutated_cell_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot viral genotype statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutated_cell_counts_file = figure_dir+'/'+'mutated_cell_counts_plot.svg'\n",
    "\n",
    "mutated_cell_counts_plot = (\n",
    "    p9.ggplot(mutated_cell_counts,\n",
    "              p9.aes(x='cell_label',\n",
    "                     y='n_cells',\n",
    "                     width=0.85)) +\n",
    "    p9.geom_bar(stat='identity') +\n",
    "    p9.coord_flip() +\n",
    "    p9.ggtitle('Consensus viral genome sequences                                 ') +\n",
    "    p9.labs(x = '', y='# of infected cells') +\n",
    "    p9.ylim([0,45]) +\n",
    "    p9.theme(figure_size=(3, 2),\n",
    "       plot_title=p9.element_text(size=12),\n",
    "       axis_title=p9.element_text(size=12),\n",
    "       axis_text=p9.element_text(size=11, color=\"black\")))\n",
    "\n",
    "mutated_cell_counts_plot.save(filename = mutated_cell_counts_file)\n",
    "display(mutated_cell_counts_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot number of non-synonymous mutations per cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_per_cell_file = figure_dir+'/'+'mutations_per_cell.svg'\n",
    "\n",
    "mutations_per_cell_plot = (\n",
    "    p9.ggplot(\n",
    "        fully_sequenced_cells,\n",
    "        p9.aes(x='nonsynonymous_mutation_count')) +\n",
    "    p9.geom_bar() +\n",
    "    p9.ggtitle(f'Number of mutations per viral genome') +\n",
    "    p9.labs(x='# of non-synonymous\\nmutations',\n",
    "            y='# of cells') +\n",
    "    p9.scale_y_continuous(limits=[0,75], breaks=[0,25,50,75]) +\n",
    "    p9.theme(figure_size=(3, 2),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text=p9.element_text(size=11, color=\"black\"),\n",
    "             legend_box_spacing=0.2,\n",
    "             legend_title=p9.element_blank()))\n",
    "mutations_per_cell_plot.save(filename = mutations_per_cell_file)\n",
    "display(mutations_per_cell_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make composite Fig 3\n",
    "Make a composite figure showing viral genotypes and number of mutations per genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fig3_file = figure_dir+'/'+'fig3.svg'\n",
    "\n",
    "fig3 = (Figure(\"25cm\", \"7cm\",\n",
    "       Panel(\n",
    "#           Grid(20, 20),\n",
    "           SVG(mutated_cell_counts_file),\n",
    "           Text(\"A\", 5, 15, size=14)).scale(1.5),\n",
    "       Panel(\n",
    "           SVG(mutations_per_cell_file).move(15,0),\n",
    "           Text(\"B\", 5, 15, size=14)).scale(1.5).move(500,0),\n",
    "       ))\n",
    "display(fig3)\n",
    "fig3.save(fig3_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Progeny production measurements\n",
    "### Plot viral transcription in cells that express both barcoded genes\n",
    "How does total viral transcription correlate with expressing both viral genes needed for progeny contribution meausrements?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcoded_cells = wt_infected_cells.query(\"all_barcoded_genes\")\n",
    "display(barcoded_cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcoded_cells_file = figure_dir+'/'+'barcoded_cells_plot.svg'\n",
    "\n",
    "barcoded_cells_plot = (\n",
    "    p9.ggplot(wt_infected_cells,\n",
    "               p9.aes(x='all_barcoded_genes',\n",
    "                      y='frac_viral_UMIs')) +\n",
    "    p9.geom_jitter(alpha=0.3, size=3, width=0.2) +\n",
    "    p9.stat_summary(fun_y=np.mean, \n",
    "                    fun_ymin=np.mean,\n",
    "                    fun_ymax=np.mean,\n",
    "                    geom='crossbar',\n",
    "                    width=0.9,\n",
    "                    color=CBPALETTE[2]) +\n",
    "    p9.geom_text(data=wt_infected_cells.groupby('all_barcoded_genes').nunique().reset_index(),\n",
    "                 mapping=p9.aes(y=0.74,\n",
    "                                label='cell_barcode'),\n",
    "                 size=12) +\n",
    "    p9.coord_flip() +\n",
    "    p9.labs(x='',\n",
    "            y='fraction mRNA from virus') +\n",
    "    p9.scale_y_continuous(limits=[0,0.75], breaks=[0,0.25,0.5,0.75]) +\n",
    "    p9.scale_x_discrete(labels=['Missing barcoded gene(s)', 'Both barcoded genes']) +\n",
    "    p9.theme(figure_size=(3.5, 1.75),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text=p9.element_text(size=12, color=\"black\")))\n",
    "\n",
    "barcoded_cells_plot.save(filename = barcoded_cells_file)\n",
    "display(barcoded_cells_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cells with all measurements\n",
    "### Select subset of cells with complete sequencing and both barcoded viral genes\n",
    "In the next section, I will calculate progeny contributions for cells that have complete long-read sequencing and express both barcoded viral genes. I want to count how many cells meet these criteria, and how many fail one or both of these criteria."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequenced_barcoded_cell_counts = (wt_infected_cells\n",
    "        .groupby(\"sequenced_and_barcoded\")\n",
    "        .aggregate(n_cells=pd.NamedAgg(\"cell_barcode\", \"count\"))\n",
    "        .reset_index()\n",
    ")\n",
    "sequenced_barcoded_cell_counts['sequenced_and_barcoded'] = (\n",
    "    sequenced_barcoded_cell_counts['sequenced_and_barcoded']\n",
    "    .astype('category')\n",
    "    .cat\n",
    "    .reorder_categories(['not sequenced or barcoded',\n",
    "              'barcoded',\n",
    "              'sequenced',\n",
    "              'sequenced and barcoded'])\n",
    ")\n",
    "sequenced_barcoded_cell_counts['selected_cells'] = (\n",
    "    sequenced_barcoded_cell_counts['sequenced_and_barcoded'] == \\\n",
    "    \"sequenced and barcoded\"\n",
    ")\n",
    "sequenced_barcoded_cells = fully_sequenced_cells.query('sequenced_and_barcoded == \"sequenced and barcoded\"')\n",
    "display(sequenced_barcoded_cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot number of cells with or without complete sequencing and both barcoded genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_cells_file = figure_dir+'/'+'selected_cells_plot.svg'\n",
    "\n",
    "selected_cells_plot = (\n",
    "    p9.ggplot(sequenced_barcoded_cell_counts,\n",
    "               p9.aes(x='sequenced_and_barcoded',\n",
    "                      y='n_cells',\n",
    "                      width=0.85)) +\n",
    "    p9.geom_bar(stat='identity', position='dodge') +\n",
    "    p9.geom_text(p9.aes(y=98, label='n_cells'), size=11) +\n",
    "    p9.labs(x='',\n",
    "            y='# of cells') +\n",
    "    p9.coord_flip() +\n",
    "    p9.scale_y_continuous(limits=[0,98], breaks=[0,20,40,60,80]) +\n",
    "    p9.scale_x_discrete(labels=[\"Incomplete sequencing and\\nmissing barcoded gene(s)\",\n",
    "                                \"Both barcoded genes\",\n",
    "                                \"Complete sequencing\",\n",
    "                                \"Complete sequencing and\\nboth barcoded genes\"]) +\n",
    "    p9.theme(figure_size=(3.6, 1.75),\n",
    "       plot_title=p9.element_text(size=12),\n",
    "       axis_title=p9.element_text(size=12),\n",
    "       axis_text=p9.element_text(size=11, color=\"black\")))\n",
    "\n",
    "selected_cells_plot.save(filename = selected_cells_file)\n",
    "display(selected_cells_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot viral transcription in cells with complete sequencing and both barcoded viral genes\n",
    "Do the cells that meet these criteria have a different distribution of viral transcription, or are we subsampling the data in a fairly non-biased way?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequenced_and_barcoded_file = figure_dir+'/'+'sequenced_and_barcoded_plot.svg'\n",
    "\n",
    "sequenced_and_barcoded_plot = (\n",
    "    p9.ggplot(wt_infected_cells,\n",
    "               p9.aes(x='factor(sequenced_and_barcoded == \"sequenced and barcoded\")',\n",
    "                      y='frac_viral_UMIs')) +\n",
    "    p9.geom_jitter(alpha=0.3, size=3, width=0.2) +\n",
    "    p9.stat_summary(fun_y=np.mean, \n",
    "                    fun_ymin=np.mean,\n",
    "                    fun_ymax=np.mean,\n",
    "                    geom='crossbar',\n",
    "                    width=0.9,\n",
    "                    color=CBPALETTE[2]) +\n",
    "    p9.geom_text(data=pd.DataFrame(sequenced_barcoded_cell_counts.groupby('selected_cells')['n_cells'].sum().reset_index()),\n",
    "                 mapping=p9.aes(\n",
    "                     x='selected_cells',\n",
    "                     y=0.74,\n",
    "                     label='n_cells'),\n",
    "                 size=12) +\n",
    "    p9.coord_flip() +\n",
    "    p9.labs(x='',\n",
    "            y='fraction mRNA from virus') +\n",
    "    p9.scale_y_continuous(limits=[0,0.75], breaks=[0,0.25,0.5,0.75]) +\n",
    "    p9.scale_x_discrete(labels=['Incomplete sequencing or\\nmissing barcoded gene(s)', 'Complete sequencing and\\nboth barcoded genes']) +\n",
    "    p9.theme(figure_size=(3.5, 1.75),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text=p9.element_text(size=12, color=\"black\")))\n",
    "\n",
    "sequenced_and_barcoded_plot.save(filename = sequenced_and_barcoded_file)\n",
    "display(sequenced_and_barcoded_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make composite Fig S4\n",
    "Assemble a composite figure showing how many cells are sequenced completely, have both barcoded viral genes, and the relationship of these criteria to total viral transcription."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figS4_file = figure_dir+'/'+'figS4.svg'\n",
    "\n",
    "figS4 = (Figure(\"27cm\", \"11cm\",\n",
    "       Panel(\n",
    "#           Grid(20, 20),\n",
    "           SVG(fully_sequenced_file).move(15,10),\n",
    "           Text(\"A\", 0, 20, size=18)),\n",
    "       Panel(\n",
    "           SVG(barcoded_cells_file).move(0,10),\n",
    "           Text(\"B\", 0, 20, size=18)).move(380,0),\n",
    "       Panel(\n",
    "           SVG(selected_cells_file).move(0,20),\n",
    "           Text(\"C\", 0, 20, size=18)).move(0,150),\n",
    "       Panel(\n",
    "           SVG(sequenced_and_barcoded_file).move(10,20),\n",
    "           Text(\"D\", 0, 20, size=18)).move(380,150),\n",
    "       )).scale(1.3)\n",
    "display(figS4)\n",
    "figS4.save(figS4_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many cells are contributing progeny?\n",
    "With these data, we can ask a simple question about the progeny population. Among the sequenced, barcoded cells, how many contribute any progeny?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_sequenced_barcoded_cells = (sequenced_barcoded_cells['cell_barcode'].nunique())\n",
    "print(f'There are {n_sequenced_barcoded_cells} in the final dataset with complete progeny measurements and complete sequencing.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(integrated_data[integrated_data['cell_barcode']\n",
    "                 .isin(sequenced_barcoded_cells[\"cell_barcode\"].to_list())]\n",
    " .groupby(['contributes_progeny_supernatant', 'contributes_progeny_second_infection'])\n",
    " ['cell_barcode']\n",
    " .nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate progeny contribution of cells with complete sequencing and both barcoded viral genes\n",
    "Next, I will calculate the contribution of each infected cell to the progeny population. I will normalize the the progeny viral barcode frequences among source and gene (e.g. 'supernatant fluHA'). Then, I will take the geometric mean of both barcoded genes to estimate the progeny contribution for each infected cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for condition in [\"supernatant\", \"second_infection\"]:\n",
    "    for gene in barcoded_genes:\n",
    "        # normalize among selected cells and within barcoded gene\n",
    "        col = f\"freq_{condition}_{gene}\"\n",
    "        sequenced_barcoded_cells[col] = (\n",
    "            sequenced_barcoded_cells[col] /\n",
    "            sequenced_barcoded_cells[col].sum()\n",
    "        ).astype(float)\n",
    "    # calculate geometric mean of normalized values\n",
    "    mean_col = f\"freq_{condition}\"\n",
    "    sequenced_barcoded_cells[mean_col] = (\n",
    "        sequenced_barcoded_cells[f\"{mean_col}_fluHA\"]\n",
    "        * sequenced_barcoded_cells[f\"{mean_col}_fluNA\"]\n",
    "    )**(1 / 2)\n",
    "    sequenced_barcoded_cells[mean_col] = (\n",
    "        sequenced_barcoded_cells[mean_col] /\n",
    "        sequenced_barcoded_cells[mean_col].sum()\n",
    "    ).astype(float)\n",
    "\n",
    "display(sequenced_barcoded_cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export CSV of cells with complete sequencing data\n",
    "To generate this plot, we needed to provide progeny contribution measurements where possible. This is why this task is performed **after** estimating progeny contributions. Cells with complete PacBio data without progeny contribution measurements will have null value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bring in progeny production measurements when available\n",
    "fully_sequenced_cells_CSV = 'results/pacbio/fully_sequenced_cells.csv'\n",
    "\n",
    "fully_sequenced_cells = pd.merge(\n",
    "    left=fully_sequenced_cells,\n",
    "    right=sequenced_barcoded_cells[['cell_barcode', 'freq_supernatant', 'freq_second_infection']],\n",
    "    on='cell_barcode',\n",
    "    how='left',\n",
    "    validate='one_to_one'\n",
    ")\n",
    "display(fully_sequenced_cells)\n",
    "print(f\"Writing fully sequenced cells data to CSV: {fully_sequenced_cells_CSV}\")\n",
    "fully_sequenced_cells.to_csv(fully_sequenced_cells_CSV)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate viral genome plot using R script\n",
    "There is an R script that generates the viral genomes plot using `gggenes`. We call that here using `subprocess.call`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# call R script to generate viral genomes plot\n",
    "viral_genomes_file = figure_dir+'/'+'viral_genomes_plot.pdf'\n",
    "rscript_file = \"scripts/viral_genomes_plot.R\"\n",
    "subprocess.call(\" \".join([\"Rscript\", rscript_file, fully_sequenced_cells_CSV, viral_genomes_file]), shell = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Heterogeneity in viral products\n",
    "### Calculate Gini coefficients\n",
    "Calculate Gini coefficient for viral transcription, physical progeny production, and infectious progeny production."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_array = wt_infected_cells['frac_viral_UMIs'].to_numpy()\n",
    "total_gini = gini(total_array)\n",
    "print(f\"The Gini coefficient for viral transcription in all {n_wt_infected_cells} infected cells is {total_gini:0f}\\n\")\n",
    "\n",
    "\n",
    "distributions = {'viral_transcription': 'frac_viral_UMIs',\n",
    "                 'physical_progeny_virions': 'freq_supernatant',\n",
    "                 'infectious_progeny_virions': 'freq_second_infection'}\n",
    "gini_coeffs = {'viral_transcription': None,\n",
    "               'physical_progeny_virions': None,\n",
    "               'infectious_progeny_virions': None,}\n",
    "\n",
    "print(f\"For the {n_sequenced_barcoded_cells} cells with complete progeny measurements and complete sequencing:\")\n",
    "for distribution, variable in distributions.items():\n",
    "    dist_array = sequenced_barcoded_cells[variable].to_numpy()\n",
    "    gini_coeff = gini(dist_array)\n",
    "    gini_coeffs[distribution] = gini_coeff\n",
    "    print(f\"{distribution} has a Gini coefficient of {gini_coeff:0f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot histograms of viral transcription, physical progeny (supernatant), and infectious progeny (second infection) frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "physical_progeny_file = figure_dir+'/'+'physical_progeny_plot.svg'\n",
    "\n",
    "physical_progeny_plot = (\n",
    "    p9.ggplot(\n",
    "        sequenced_barcoded_cells,\n",
    "        p9.aes(x='freq_supernatant')) +\n",
    "    p9.geom_histogram(breaks=[0,0.02,0.04,0.06,0.08,0.1,0.12,0.14,\n",
    "                              0.16,0.18,0.2,0.22,0.24,0.26,0.28]) +\n",
    "    p9.ylim(0,85) +\n",
    "    p9.ggtitle(f'Physical progeny production') +\n",
    "    p9.labs(x='fraction physical virions',\n",
    "            y='# of cells') +\n",
    "    p9.annotate('text',\n",
    "                label=f'Gini={round(gini_coeffs[\"physical_progeny_virions\"], 2)}',\n",
    "                x=0.24, y=75,\n",
    "                size=12) +\n",
    "    p9.theme(figure_size=(3, 2.5),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text=p9.element_text(size=10, color=\"black\"),\n",
    "             legend_title=p9.element_text(size=12),\n",
    "             legend_title_align='center'))\n",
    "physical_progeny_plot.save(filename = physical_progeny_file)\n",
    "display(physical_progeny_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "infectious_progeny_file = figure_dir+'/'+'infectious_progeny_plot.svg'\n",
    "\n",
    "infectious_progeny_plot = (\n",
    "    p9.ggplot(\n",
    "        sequenced_barcoded_cells,\n",
    "        p9.aes(x='freq_second_infection')) +\n",
    "    p9.geom_histogram(breaks=[0,0.02,0.04,0.06,0.08,0.1,0.12,0.14,\n",
    "                              0.16,0.18,0.2,0.22,0.24,0.26,0.28]) +\n",
    "    p9.ylim(0,85) +\n",
    "    p9.ggtitle(f'Infectious progeny production') +\n",
    "    p9.labs(x='fraction infectious virions',\n",
    "            y='') +\n",
    "    p9.annotate('text',\n",
    "                label=f'Gini={round(gini_coeffs[\"infectious_progeny_virions\"], 2)}',\n",
    "                x=0.24, y=75,\n",
    "                size=12) +\n",
    "    p9.theme(figure_size=(3, 2.5),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text=p9.element_text(size=10, color=\"black\"),\n",
    "             legend_title=p9.element_text(size=12),\n",
    "             legend_title_align='center'))\n",
    "infectious_progeny_plot.save(filename = infectious_progeny_file)\n",
    "display(infectious_progeny_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcription_selected_cells_file = figure_dir+'/'+'transcription_selected_cells_plot.svg'\n",
    "\n",
    "transcription_selected_cells_plot = (\n",
    "    p9.ggplot(\n",
    "        sequenced_barcoded_cells,\n",
    "        p9.aes(x='frac_viral_UMIs')) +\n",
    "    p9.geom_histogram(breaks=[0,0.05,0.1,0.15,0.2,0.25,0.3,0.35,\n",
    "                              0.4,0.45,0.5,0.55,0.6,0.65,0.7]) +\n",
    "    p9.ylim(0,85) +\n",
    "    p9.ggtitle(f'Viral transcription') +\n",
    "    p9.labs(x='fraction mRNA from virus',\n",
    "            y='') +\n",
    "    p9.annotate('text',\n",
    "                label=f'Gini={round(gini_coeffs[\"viral_transcription\"], 2)}',\n",
    "                x=0.6, y=75,\n",
    "                size=12) +\n",
    "    p9.theme(figure_size=(3, 2.5),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text=p9.element_text(size=10, color=\"black\"),\n",
    "             legend_title=p9.element_text(size=12),\n",
    "             legend_title_align='center'))\n",
    "transcription_selected_cells_plot.save(filename = transcription_selected_cells_file)\n",
    "display(transcription_selected_cells_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make composite Fig 4\n",
    "Make composite figure of viral transcription, physical progeny production, and infectious progeny production."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig4_file = figure_dir+'/'+'fig4.svg'\n",
    "\n",
    "fig4 = (Figure(\"33cm\", \"12cm\",\n",
    "       Panel(\n",
    "#           Grid(20, 20),\n",
    "           SVG(physical_progeny_file).move(0,15),\n",
    "           Text(\"A\", 5, 15, size=14)).scale(2),\n",
    "       Panel(\n",
    "           SVG(infectious_progeny_file).move(0,15),\n",
    "           Text(\"B\", 5, 15, size=14)).scale(2).move(420,0),\n",
    "       Panel(\n",
    "           SVG(transcription_selected_cells_file).move(0,15),\n",
    "           Text(\"C\", 5, 15, size=14)).scale(2).move(820,0),\n",
    "       ))\n",
    "display(fig4)\n",
    "fig4.save(fig4_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate cumulative sum of viral products\n",
    "How many cells does it take to contribute half of the transcripts or progeny? Note that in this dataframe I will calculate the cumulative sum on the absolute number of viral UMIs across all cells, rather than the fraction of viral UMIs within a cell (which is the metric I use for other plots)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cumulative_df = sequenced_barcoded_cells[['cell_barcode', 'viral_UMIs', 'freq_supernatant', 'freq_second_infection']]\n",
    "cumulative_df['viral_UMIs'] = (\n",
    "    cumulative_df['viral_UMIs'] /\n",
    "    cumulative_df['viral_UMIs'].sum())\n",
    "cumsum_list = list()\n",
    "for var in cumulative_df.columns:\n",
    "    if var == 'cell_barcode':\n",
    "        pass\n",
    "    else:\n",
    "        cumsum_list.append(\n",
    "            pd.DataFrame(cumulative_df\n",
    "                         .sort_values(by=var, ascending=False)\n",
    "                         .reset_index()\n",
    "                         [var]\n",
    "                         .cumsum()))\n",
    "cumsum_df = (\n",
    "    pd.melt(pd.concat(cumsum_list, axis=1).reset_index(),\n",
    "            id_vars='index',\n",
    "            var_name='source',\n",
    "            value_name='cumulative_sum'))\n",
    "cumsum_df['index'] = cumsum_df['index'] + 1 # 1-base indexing for cell number\n",
    "cumsum_df['source'] = (\n",
    "    cumsum_df['source']\n",
    "    .astype('category')\n",
    "    .cat\n",
    "    .reorder_categories(['viral_UMIs',\n",
    "                         'freq_supernatant',\n",
    "                         'freq_second_infection'])\n",
    ")\n",
    "display(cumsum_df.query('cumulative_sum > 0.45 and cumulative_sum < 0.6'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot cumulative sum curves"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cumulative_sum_file = figure_dir+'/'+'cumulative_sum_plot.svg'\n",
    "\n",
    "cumulative_sum_plot = (\n",
    "    p9.ggplot() +\n",
    "    p9.geom_point(cumsum_df,\n",
    "                  p9.aes(x='index',\n",
    "                         y='cumulative_sum',\n",
    "                         color='source'),\n",
    "                  alpha=0.5,\n",
    "                  size=3) +\n",
    "    p9.geom_hline(yintercept=0.5, linetype='dashed') +\n",
    "    p9.labs(x='cell rank',\n",
    "            y='cumulative fraction',\n",
    "            color='Viral product') +\n",
    "    p9.scale_x_continuous(breaks=[0,20,40,60,80]) +\n",
    "    p9.scale_color_manual(values=CBPALETTE,\n",
    "                          labels=['Viral mRNA',\n",
    "                                  'Physical progeny virions',\n",
    "                                  'Infectious progeny virions']) +\n",
    "    p9.theme(figure_size=(6, 4),\n",
    "             legend_position='right',\n",
    "             legend_title=p9.element_text(size=14),\n",
    "             legend_text=p9.element_text(size=14),\n",
    "             axis_title=p9.element_text(size=14),\n",
    "             axis_text=p9.element_text(size=14, color=\"black\")))\n",
    "cumulative_sum_plot.save(filename = cumulative_sum_file)\n",
    "display(cumulative_sum_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Transcription-progeny production relationships\n",
    "### Modify df with sequenced and barcoded cells data for better plotting\n",
    "I will make modifications to a copy of the dataframe with sequenced and barcoded cells for optimal plotting. Specifically, I will manually adjust the viral transcription values of a few cells by a small amount (jitter) so that they can be plot without overlaying each other."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcription_progeny_df = sequenced_barcoded_cells.copy()\n",
    "transcription_progeny_df['frac_viral_UMIs_adjusted'] = (\n",
    "    transcription_progeny_df['frac_viral_UMIs']\n",
    ")\n",
    "# Jitter two cells right around 0.3\n",
    "transcription_progeny_df.loc[261, 'frac_viral_UMIs_adjusted'] = (\n",
    "    transcription_progeny_df.loc[261]['frac_viral_UMIs'] * 1.02\n",
    ")\n",
    "transcription_progeny_df.loc[315, 'frac_viral_UMIs_adjusted'] = (\n",
    "    transcription_progeny_df.loc[261]['frac_viral_UMIs'] * 0.98\n",
    ")\n",
    "\n",
    "display(transcription_progeny_df[['cell_barcode',\n",
    "                                  'frac_viral_UMIs',\n",
    "                                  'frac_viral_UMIs_adjusted',\n",
    "                                  'present_fluNS']]\n",
    "        .sort_values(by='frac_viral_UMIs', ascending=False)\n",
    "        .head(12))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will also add annotations that I will use for future plots that label the viral genotypes and whether or not the cells express the influenza NS gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def label_genome_plot(row):\n",
    "    if row['cell_label'] == \"Complete unmutated genome\":\n",
    "        return row['cell_label']\n",
    "    elif row['cell_label'] == \"Complete mutated genome\":\n",
    "        return row['cell_label']\n",
    "    else:\n",
    "        return \"Missing viral gene(s)\"\n",
    "\n",
    "transcription_progeny_df = (\n",
    "    transcription_progeny_df\n",
    "    .assign(genome_label=lambda x: x.apply(label_genome_plot, axis=1))\n",
    ")\n",
    "transcription_progeny_df['genome_label'] = (\n",
    "    transcription_progeny_df['genome_label']\n",
    "    .astype('category')\n",
    "    .cat\n",
    "    .reorder_categories(['Complete unmutated genome',\n",
    "                         'Complete mutated genome',\n",
    "                         'Missing viral gene(s)'])\n",
    ")\n",
    "display(transcription_progeny_df['genome_label'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will copy the `present_fluNS` column with descriptive strings instead of True/False:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def present_fluNS_string(row):\n",
    "    if row['present_fluNS'] == True:\n",
    "        return \"Expresses NS\"\n",
    "    else:\n",
    "        return \"Does not express NS\"\n",
    "\n",
    "transcription_progeny_df = (\n",
    "    transcription_progeny_df\n",
    "    .assign(present_fluNS_string=lambda x: x.apply(present_fluNS_string, axis = 1))\n",
    ")\n",
    "\n",
    "transcription_progeny_df.groupby(['present_fluNS', 'present_fluNS_string'])['cell_barcode'].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot transcription-progeny production relationship\n",
    "Plot the relationship between viral transcription and progeny produciton. I will produce both an SVG version for the manuscript and an HTML version for uploading to the web.\n",
    "\n",
    "**I need to add [altair_saver](https://github.com/altair-viz/altair_saver/) to the environment to save an SVG version of this plot. For now, I will manually download one and add to the results folder.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files and variables\n",
    "transcription_progeny_file = figure_dir+'/'+'transcription_progeny_plot.svg'\n",
    "transcription_progeny_interactive_file = 'docs/_includes/transcription_progeny_interactive.html'\n",
    "\n",
    "freq_cols = [f\"freq_{condition}\" for condition in [\"supernatant\", \"second_infection\"]]\n",
    "y_label = {\"freq_supernatant\": \"fraction physical progeny\",\n",
    "           \"freq_second_infection\": \"fraction infectious progeny\"}\n",
    "\n",
    "cell_selector = alt.selection_single(on='mouseover', empty='none')\n",
    "\n",
    "# Colors and shapes\n",
    "genome_label_values = [\"Complete unmutated genome\",\n",
    "                 \"Complete mutated genome\",\n",
    "                 \"Missing viral gene(s)\"]\n",
    "genome_label_colors = [\"#0072B2\", \"#999999\", \"#FF5349\"] # Verified cb-readable\n",
    "\n",
    "present_fluNS_values = [\"Does not express NS\", \"Expresses NS\"]\n",
    "present_fluNS_shapes = [\"triangle\", \"circle\"]\n",
    "\n",
    "# Selectors\n",
    "genotype_selector = alt.selection_single(\n",
    "    fields=[\"genome_label\"],\n",
    "    bind=alt.binding_select(\n",
    "        name=\"Viral genotype\\t\",\n",
    "        options=[None,\n",
    "                 \"Complete unmutated genome\",\n",
    "                 \"Complete mutated genome\",\n",
    "                 \"Missing viral gene(s)\"],\n",
    "        labels=[\"All genomes\",\n",
    "                \"Complete unmutated genome\",\n",
    "                \"Complete mutated genome\",\n",
    "                \"Missing viral gene(s)\"]\n",
    "    ),\n",
    ")\n",
    "\n",
    "NS_selector = alt.selection_single(\n",
    "    fields=[\"present_fluNS_string\"],\n",
    "    bind=alt.binding_select(\n",
    "        name=\"Expresses NS\\t\",\n",
    "        options=[None,\n",
    "                 \"Does not express NS\",\n",
    "                 \"Expresses NS\"],\n",
    "        labels=[\"All\",\n",
    "                \"Does not express NS\",\n",
    "                \"Expresses NS\"]\n",
    "    ),\n",
    ")\n",
    "\n",
    "# Build charts\n",
    "transcription_progeny_charts = []\n",
    "for freq_col in freq_cols:\n",
    "    transcription_progeny_charts.append(\n",
    "        alt.Chart(transcription_progeny_df)\n",
    "        .encode(\n",
    "            x=alt.X(\"frac_viral_UMIs_adjusted\",\n",
    "                    scale=alt.Scale(domain=(0, 0.65)),\n",
    "                    axis=alt.Axis(title='fraction mRNA from virus')),\n",
    "            y=alt.Y(freq_col,\n",
    "                    sort=[False, True],\n",
    "                    scale=alt.Scale(domain=(0, 0.22), bins=[0,0.05,0.1,0.15,0.2]),\n",
    "                    axis=alt.Axis(title=f'{y_label[freq_col]}')),\n",
    "            tooltip=[\"cell_barcode\", alt.Tooltip(\"frac_viral_UMIs\", format='.2g'),\n",
    "                     \"total_UMIs\", \"missing_viral_genes\", \"mutated_genes\",\n",
    "                     *[alt.Tooltip(c, format='.2g') for c in freq_cols],\n",
    "                     \"viral_mutations\", \"mutation_support\",\n",
    "                    ],\n",
    "            color=alt.Color(\"genome_label\",\n",
    "                            legend=alt.Legend(title=\"Viral genotype\",\n",
    "                                              orient='none',\n",
    "                                              legendX=-80,\n",
    "                                              legendY=-45,\n",
    "                                              direction='horizontal',\n",
    "                                              titleFontSize=18,\n",
    "                                              titleFontWeight=\"normal\",\n",
    "                                              titleAnchor='start',\n",
    "                                              labelFontSize=14,\n",
    "                                              symbolStrokeWidth=0,\n",
    "                                              symbolSize=120,\n",
    "                                              labelLimit=200),\n",
    "                            scale=alt.Scale(domain=genome_label_values,\n",
    "                                            range=genome_label_colors)),\n",
    "            shape=alt.Shape(\"present_fluNS_string\",\n",
    "                            legend=alt.Legend(title=\"Expresses NS\",\n",
    "                                              orient='none',\n",
    "                                              legendX=520,\n",
    "                                              legendY=-45,\n",
    "                                              direction='horizontal',\n",
    "                                              titleFontSize=18,\n",
    "                                              titleFontWeight=\"normal\",\n",
    "                                              titleAnchor='start',\n",
    "                                              labelFontSize=14,\n",
    "                                              symbolStrokeWidth=0,\n",
    "                                              symbolSize=120),\n",
    "                            scale=alt.Scale(domain=present_fluNS_values,\n",
    "                                            range=present_fluNS_shapes)),\n",
    "            strokeWidth=alt.condition(cell_selector, alt.value(2), alt.value(0)),\n",
    "            opacity=alt.condition(\n",
    "                cell_selector,\n",
    "                alt.value(1),\n",
    "                alt.value(0.75),\n",
    "            ),\n",
    "            size=alt.value(100),\n",
    "        )\n",
    "        .mark_point(filled=True, stroke=\"black\")\n",
    "        .properties(width=300, height=300)\n",
    "    )\n",
    "\n",
    "transcription_progeny_plot = (\n",
    "    alt.hconcat(*transcription_progeny_charts, spacing=60)\n",
    "    .configure_view(strokeWidth=0)\n",
    "    .add_selection(\n",
    "        cell_selector,\n",
    "        genotype_selector,\n",
    "        NS_selector\n",
    "    )\n",
    "    .transform_filter(genotype_selector)\n",
    "    .transform_filter(NS_selector)\n",
    "    .configure_axis(grid=False,\n",
    "                    titleFontSize=16,\n",
    "                    titleFontWeight=\"normal\",\n",
    "                    labelFontSize=14,\n",
    "                    titlePadding=10)\n",
    ")\n",
    "\n",
    "display(transcription_progeny_plot)\n",
    "#transcription_progeny_plot.save(transcription_progeny_file)\n",
    "transcription_progeny_plot.save(transcription_progeny_interactive_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Examine effect of missing viral genes on transcription and progeny production\n",
    "### Label cells based on missing viral genes\n",
    "Simplify missing viral gene information. Categorize cells as expressing all viral genes, missing NS, or missing another gene(s)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def label_missing_NS(row):\n",
    "    if row['all_viral_genes'] == True:\n",
    "        return \"All viral genes\"\n",
    "    elif row['present_fluNS'] == False:\n",
    "        return \"Missing NS gene\"\n",
    "    else:\n",
    "        return \"Missing other viral gene(s)\"\n",
    "\n",
    "transcription_progeny_df = (\n",
    "    transcription_progeny_df\n",
    "    .assign(\n",
    "        missing_NS=lambda x: x.apply(label_missing_NS, axis=1)))\n",
    "\n",
    "transcription_progeny_df['missing_NS'] = (\n",
    "    transcription_progeny_df['missing_NS']\n",
    "    .astype('category')\n",
    "    .cat\n",
    "    .reorder_categories(['Missing other viral gene(s)',\n",
    "                         'Missing NS gene',\n",
    "                         'All viral genes'])\n",
    ")\n",
    "\n",
    "display(transcription_progeny_df.groupby('missing_NS')['cell_barcode'].nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot the viral transcription based on missing viral genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_NS_transcription_file = figure_dir+'/'+'missing_NS_transcription_plot.svg'\n",
    "\n",
    "missing_NS_transcription_plot = (\n",
    "    p9.ggplot(transcription_progeny_df,\n",
    "               p9.aes(x='missing_NS',\n",
    "                      y='frac_viral_UMIs',)) +\n",
    "    p9.geom_jitter(alpha=0.4, size=3, width=0.15) +\n",
    "    p9.stat_summary(fun_y=np.mean, \n",
    "                    fun_ymin=np.mean,\n",
    "                    fun_ymax=np.mean,\n",
    "                    geom='crossbar',\n",
    "                    width=0.9,\n",
    "                    color=CBPALETTE[2]) +\n",
    "    p9.ggtitle('Viral transcription') +\n",
    "    p9.coord_flip() +\n",
    "    p9.labs(x='',\n",
    "            y='fraction mRNA from virus') +\n",
    "    p9.theme(figure_size=(3.5, 1.75),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text=p9.element_text(size=12, color=\"black\")))\n",
    "\n",
    "missing_NS_transcription_plot.save(filename = missing_NS_transcription_file)\n",
    "display(missing_NS_transcription_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_NS_transcription_swarm_file = figure_dir+'/'+'missing_NS_transcription_swarm_plot.svg'\n",
    "\n",
    "# Figure theme\n",
    "fig, ax = plt.subplots()\n",
    "fig.set_size_inches(5, 2.2)\n",
    "\n",
    "missing_NS_transcription_swarm_plot = (\n",
    "    sns.swarmplot(data=transcription_progeny_df,\n",
    "                  x='frac_viral_UMIs',\n",
    "                  y='missing_NS',\n",
    "                  order=['All viral genes', 'Missing NS gene', 'Missing other viral gene(s)'],\n",
    "                  palette=['#4F4F4F', '#4F4F4F', '#4F4F4F'],\n",
    "                  size=4))\n",
    "# Add mean lines\n",
    "sns.boxplot(\n",
    "    ax=missing_NS_transcription_swarm_plot,\n",
    "    data=transcription_progeny_df,\n",
    "    x=\"frac_viral_UMIs\",\n",
    "    y=\"missing_NS\",\n",
    "    order=['All viral genes', 'Missing NS gene', 'Missing other viral gene(s)'],\n",
    "    showmeans=True,\n",
    "    meanline=True,\n",
    "    meanprops={'color': CBPALETTE[2], 'ls': '-', 'lw': 1},\n",
    "    medianprops={'visible': False},\n",
    "    whiskerprops={'visible': False},\n",
    "    zorder=10,\n",
    "    showfliers=False,\n",
    "    showbox=False,\n",
    "    showcaps=False,)\n",
    "\n",
    "# Figure theme\n",
    "(missing_NS_transcription_swarm_plot\n",
    " .set(ylabel=None,\n",
    "      title='Viral transcription'))\n",
    "sns.despine()\n",
    "\n",
    "plt.tight_layout()\n",
    "fig.savefig(fname = missing_NS_transcription_swarm_file)\n",
    "display(missing_NS_transcription_swarm_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate statistical significance of missing viral genes on viral transcription\n",
    "Check whether cells missing NS have statistically significant difference in viral transcription.  \n",
    "Since these data are not normally distributed, I will use a non-parametric test. First, I will perform a [Kruskal-Wallis test](http://www.biostathandbook.com/kruskalwallis.html) to see if any of the three groups come from a different distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np_arrays = list()\n",
    "for group in transcription_progeny_df['missing_NS'].unique():\n",
    "    np_arrays.append(\n",
    "        transcription_progeny_df\n",
    "        .query(f'missing_NS == \"{group}\"')\n",
    "        ['frac_viral_UMIs']\n",
    "        .to_numpy())\n",
    "stats.kruskal(np_arrays[0], np_arrays[1], np_arrays[2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The result is significant. Now, I will perform pairwise comparisons of each group using a post-hoc [Dunn's test](https://rpkgs.datanovia.com/rstatix/reference/dunn_test.html) to see which group(s) have statistically signifcant difference in medians. I believe Dunn's test is preferred to the unpaired t-test because Dunn's test accounts for pooled variance calculated across all samples, while the unpaired t-test does not.\n",
    "\n",
    "Use Bonferroni correction as a conservative way of accounting for multiple hypothesis testing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sp.posthoc_dunn(transcription_progeny_df[['missing_NS', 'frac_viral_UMIs']],\n",
    "                group_col='missing_NS',\n",
    "                val_col='frac_viral_UMIs',\n",
    "                p_adjust='bonferroni')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot the progeny production based on missing viral genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_NS_progeny_file = figure_dir+'/'+'missing_NS_progeny_plot.svg'\n",
    "\n",
    "missing_NS_progeny_plot = (\n",
    "    p9.ggplot(transcription_progeny_df,\n",
    "               p9.aes(x='missing_NS',\n",
    "                      y='freq_supernatant',)) +\n",
    "    p9.geom_jitter(alpha=0.4, size=3, width=0.15) +\n",
    "    p9.stat_summary(fun_y=np.mean, \n",
    "                    fun_ymin=np.mean,\n",
    "                    fun_ymax=np.mean,\n",
    "                    geom='crossbar',\n",
    "                    width=0.9,\n",
    "                    color=CBPALETTE[2]) +\n",
    "    p9.ggtitle('Viral progeny production') +\n",
    "    p9.coord_flip() +\n",
    "    p9.labs(x='',\n",
    "            y='fraction physical progeny') +\n",
    "    p9.theme(figure_size=(3.5, 1.75),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text_x=p9.element_text(size=12, color=\"black\"),\n",
    "             axis_text_y=p9.element_blank()))\n",
    "\n",
    "missing_NS_progeny_plot.save(filename = missing_NS_progeny_file)\n",
    "display(missing_NS_progeny_plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_NS_progeny_swarm_file = figure_dir+'/'+'missing_NS_progeny_swarm_plot.svg'\n",
    "\n",
    "# Figure theme\n",
    "fig, ax = plt.subplots()\n",
    "fig.set_size_inches(3.1, 2.2)\n",
    "\n",
    "missing_NS_progeny_swarm_plot = (\n",
    "    sns.swarmplot(data=transcription_progeny_df,\n",
    "                  x='freq_supernatant',\n",
    "                  y='missing_NS',\n",
    "                  order=['All viral genes', 'Missing NS gene', 'Missing other viral gene(s)'],\n",
    "                  palette=['#4F4F4F', '#4F4F4F', '#4F4F4F'],\n",
    "                  size=4))\n",
    "# Add mean lines\n",
    "sns.boxplot(\n",
    "    ax=missing_NS_progeny_swarm_plot,\n",
    "    data=transcription_progeny_df,\n",
    "    x=\"freq_supernatant\",\n",
    "    y=\"missing_NS\",\n",
    "    order=['All viral genes', 'Missing NS gene', 'Missing other viral gene(s)'],\n",
    "    showmeans=True,\n",
    "    meanline=True,\n",
    "    meanprops={'color': CBPALETTE[2], 'ls': '-', 'lw': 1},\n",
    "    medianprops={'visible': False},\n",
    "    whiskerprops={'visible': False},\n",
    "    zorder=10,\n",
    "    showfliers=False,\n",
    "    showbox=False,\n",
    "    showcaps=False,)\n",
    "\n",
    "# Figure theme\n",
    "(missing_NS_progeny_swarm_plot\n",
    " .set(ylabel=None,\n",
    "      yticklabels=[],\n",
    "      title='Viral progeny production'))\n",
    "sns.despine()\n",
    "\n",
    "plt.tight_layout()\n",
    "fig.savefig(fname = missing_NS_progeny_swarm_file)\n",
    "display(missing_NS_progeny_swarm_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare the progeny production distributions of all infected cells to cells with complete, unmutated viral genomes\n",
    "### Calculate Gini coefficients for cells with complete, unmuated viral genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmutated_distributions = {'viral_transcription': 'frac_viral_UMIs',\n",
    "                 'physical_progeny_virions': 'freq_supernatant',\n",
    "                 'infectious_progeny_virions': 'freq_second_infection'}\n",
    "unmutated_gini_coeffs = {'viral_transcription': None,\n",
    "               'physical_progeny_virions': None,\n",
    "               'infectious_progeny_virions': None,}\n",
    "\n",
    "print(\"For the \"+str(transcription_progeny_df.query('genome_label == \"Complete unmutated genome\"')['cell_barcode'].nunique())+\" unmutated cells with complete progeny measurements and complete sequencing:\")\n",
    "for distribution, variable in distributions.items():\n",
    "    dist_array = (transcription_progeny_df\n",
    "                  .query('genome_label == \"Complete unmutated genome\"')\n",
    "                  [variable]\n",
    "                  .to_numpy())\n",
    "    gini_coeff = gini(dist_array)\n",
    "    unmutated_gini_coeffs[distribution] = gini_coeff\n",
    "    print(f\"{distribution} has a Gini coefficient of {gini_coeff:0f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot physical progeny production in all infected cells\n",
    "Make another plot of progeny production in infected cells, formatted for use in a composite figure. This is the same information that is displayed in figure 4."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_physical_progeny_file = figure_dir+'/'+'all_physical_progeny_plot.svg'\n",
    "\n",
    "all_physical_progeny_plot = (\n",
    "    p9.ggplot(transcription_progeny_df,\n",
    "              p9.aes(x='freq_supernatant')) +\n",
    "    p9.geom_histogram(breaks=[0,0.02,0.04,0.06,0.08,0.1,0.12,0.14,\n",
    "                              0.16,0.18,0.2,0.22,0.24,0.26,0.28]) +\n",
    "    p9.ylim(0,85) +\n",
    "    p9.ggtitle(f'All viral genomes') +\n",
    "    p9.labs(x='fraction physical virions',\n",
    "            y='# of cells') +\n",
    "    p9.annotate('text',\n",
    "                label=f'Gini={round(gini_coeffs[\"physical_progeny_virions\"], 2)}',\n",
    "                x=0.22, y=75,\n",
    "                size=16) +\n",
    "    p9.theme(figure_size=(3, 2.5),\n",
    "             plot_title=p9.element_text(size=16),\n",
    "             axis_title=p9.element_text(size=16),\n",
    "             axis_text=p9.element_text(size=16, color=\"black\"),\n",
    "             legend_title=p9.element_text(size=16),\n",
    "             legend_title_align='center'))\n",
    "all_physical_progeny_plot.save(filename = all_physical_progeny_file)\n",
    "display(all_physical_progeny_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot physical progeny produciton in cells with complete, unmutated viral genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmutated_physical_progeny_file = figure_dir+'/'+'unmutated_physical_progeny_plot.svg'\n",
    "\n",
    "unmutated_physical_progeny_plot = (\n",
    "    p9.ggplot(transcription_progeny_df.query('genome_label == \"Complete unmutated genome\"'),\n",
    "              p9.aes(x='freq_supernatant')) +\n",
    "    p9.geom_histogram(breaks=[0,0.02,0.04,0.06,0.08,0.1,0.12,0.14,\n",
    "                              0.16,0.18,0.2,0.22,0.24,0.26,0.28],\n",
    "                      fill=\"#0072B2\") +\n",
    "    p9.ylim(0,35) +\n",
    "    p9.ggtitle(f'Complete unmutated genomes') +\n",
    "    p9.labs(x='fraction physical virions',\n",
    "            y='') +\n",
    "    p9.annotate('text',\n",
    "                label=f'Gini={round(unmutated_gini_coeffs[\"physical_progeny_virions\"], 2)}',\n",
    "                x=0.22, y=29,\n",
    "                size=16) +\n",
    "    p9.theme(figure_size=(3, 2.5),\n",
    "             plot_title=p9.element_text(size=16),\n",
    "             axis_title=p9.element_text(size=16),\n",
    "             axis_text=p9.element_text(size=16, color=\"black\"),\n",
    "             legend_title=p9.element_text(size=16),\n",
    "             legend_title_align='center'))\n",
    "unmutated_physical_progeny_plot.save(filename = unmutated_physical_progeny_file)\n",
    "display(unmutated_physical_progeny_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot infectious progeny production in all infected cells\n",
    "Make another plot of progeny production in infected cells, formatted for use in a composite figure. This is the same information that is displayed in figure 4."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_infectious_progeny_file = figure_dir+'/'+'all_infectious_progeny_plot.svg'\n",
    "\n",
    "all_infectious_progeny_plot = (\n",
    "    p9.ggplot(transcription_progeny_df,\n",
    "              p9.aes(x='freq_second_infection')) +\n",
    "    p9.geom_histogram(breaks=[0,0.02,0.04,0.06,0.08,0.1,0.12,0.14,\n",
    "                              0.16,0.18,0.2,0.22,0.24,0.26,0.28]) +\n",
    "    p9.ylim(0,85) +\n",
    "    p9.ggtitle(f'All viral genomes') +\n",
    "    p9.labs(x='fraction infectious virions',\n",
    "            y='# of cells') +\n",
    "    p9.annotate('text',\n",
    "                label=f'Gini={round(gini_coeffs[\"infectious_progeny_virions\"], 2)}',\n",
    "                x=0.22, y=75,\n",
    "                size=16) +\n",
    "    p9.theme(figure_size=(3, 2.5),\n",
    "             plot_title=p9.element_text(size=16),\n",
    "             axis_title=p9.element_text(size=16),\n",
    "             axis_text=p9.element_text(size=16, color=\"black\"),\n",
    "             legend_title=p9.element_text(size=16),\n",
    "             legend_title_align='center'))\n",
    "all_infectious_progeny_plot.save(filename = all_infectious_progeny_file)\n",
    "display(all_infectious_progeny_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot infectious progeny produciton in cells with complete, unmutated viral genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmutated_infectious_progeny_file = figure_dir+'/'+'unmutated_infectious_progeny_plot.svg'\n",
    "\n",
    "unmutated_infectious_progeny_plot = (\n",
    "    p9.ggplot(transcription_progeny_df.query('genome_label == \"Complete unmutated genome\"'),\n",
    "              p9.aes(x='freq_second_infection')) +\n",
    "    p9.geom_histogram(breaks=[0,0.02,0.04,0.06,0.08,0.1,0.12,0.14,\n",
    "                              0.16,0.18,0.2,0.22,0.24,0.26,0.28],\n",
    "                      fill=\"#0072B2\") +\n",
    "    p9.ylim(0,35) +\n",
    "    p9.ggtitle(f'Complete unmutated genomes') +\n",
    "    p9.labs(x='fraction infectious virions',\n",
    "            y='') +\n",
    "    p9.annotate('text',\n",
    "                label=f'Gini={round(unmutated_gini_coeffs[\"infectious_progeny_virions\"], 2)}',\n",
    "                x=0.22, y=29,\n",
    "                size=16) +\n",
    "    p9.theme(figure_size=(3, 2.5),\n",
    "             plot_title=p9.element_text(size=16),\n",
    "             axis_title=p9.element_text(size=16),\n",
    "             axis_text=p9.element_text(size=16, color=\"black\"),\n",
    "             legend_title=p9.element_text(size=16),\n",
    "             legend_title_align='center'))\n",
    "unmutated_infectious_progeny_plot.save(filename = unmutated_infectious_progeny_file)\n",
    "display(unmutated_infectious_progeny_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make composite Fig 5\n",
    "Make composite figure showing viral transcription and progeny production, the effect of missing viral genes, and the heterogeneity of progeny production in cells with complete, unmutated viral genomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig5_file = figure_dir+'/'+'fig5.svg'\n",
    "\n",
    "fig5 = (Figure(\"26cm\", \"25cm\",\n",
    "       Panel(\n",
    "#           Grid(20, 20),\n",
    "           SVG(transcription_progeny_file).scale(1.1).move(0,30),\n",
    "           Text(\"A\", 5, 15, size=18).scale(1.3)),\n",
    "       Panel(\n",
    "           SVG(missing_NS_transcription_swarm_file),\n",
    "           Text(\"B\", 5, 15, size=18)\n",
    "       ).scale(1.3).move(0,495),\n",
    "       Panel(\n",
    "           SVG(missing_NS_progeny_swarm_file),\n",
    "       ).scale(1.3).move(500,495),\n",
    "       Panel(\n",
    "           SVG(all_physical_progeny_file).scale(0.95).move(0,20),\n",
    "           Text(\"C\", 5, 15, size=18).scale(1.3),\n",
    "           Text(\"Physical progeny production\", 120, 15, size=16)\n",
    "       ).move(0,720),\n",
    "       Panel(\n",
    "           SVG(unmutated_physical_progeny_file).scale(0.95).move(0,20),\n",
    "       ).move(225,720),\n",
    "       Panel(\n",
    "           SVG(all_infectious_progeny_file).scale(0.95).move(0,20),\n",
    "           Text(\"D\", 5, 15, size=18).scale(1.3),\n",
    "           Text(\"Infectious progeny production\", 120, 15, size=16)\n",
    "       ).move(480,720),\n",
    "       Panel(\n",
    "           SVG(unmutated_infectious_progeny_file).scale(0.95).move(0,20),\n",
    "       ).move(710,720),\n",
    "       ))\n",
    "display(fig5)\n",
    "fig5.save(fig5_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot effect of viral mutations of physical and infectious progeny production\n",
    "Plot correlation between physical progeny virion (supernatant) frequency and infectious progeny virion (second infection) frequency. Annotate by genome labels used in transcription-progeny plot above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "progeny_freq_corr_file = figure_dir+'/'+'progeny_freq_corr_plot.svg'\n",
    "\n",
    "progeny_freq_corr_plot = (\n",
    "    p9.ggplot(transcription_progeny_df,\n",
    "               p9.aes(x='freq_supernatant',\n",
    "                      y='freq_second_infection',\n",
    "                      color='genome_label')) +\n",
    "    p9.geom_point(alpha=0.4, size=4) +\n",
    "    p9.ggtitle('Progeny associated with\\nsingle infected cells') +\n",
    "    p9.labs(x='fraction physical progeny',\n",
    "            y='fraction \\ninfectious progeny') +\n",
    "    p9.scale_x_continuous(limits=[0,0.22], breaks=[0,0.1,0.2]) +\n",
    "    p9.scale_y_continuous(limits=[0,0.22], breaks=[0,0.1,0.2]) +\n",
    "    p9.theme(figure_size=(3, 3),\n",
    "             plot_title=p9.element_text(size=12),\n",
    "             axis_title=p9.element_text(size=12),\n",
    "             axis_text=p9.element_text(size=12, color=\"black\"),) +\n",
    "    p9.scale_color_manual(genome_label_colors, name=\"Viral mutations\"))\n",
    "\n",
    "progeny_freq_corr_plot.save(filename = progeny_freq_corr_file)\n",
    "display(progeny_freq_corr_plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assess correlations\n",
    "### What is the correlation of transcription and progeny production?\n",
    "I will calculate a Pearson's R value of transcription and physical progeny production, an R value of transcription and physical progeny production, and an R value of physical progeny production and infectious progeny production."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "correlation_df = (\n",
    "    transcription_progeny_df[['frac_viral_UMIs', 'freq_supernatant', 'freq_second_infection']]\n",
    "    .corr())\n",
    "display(correlation_df)\n",
    "\n",
    "print(f'The R value between viral transcription and physical progeny production is {correlation_df.iloc[0,1]:2g}')\n",
    "print(f'The R value between viral transcription and infectious progeny production is {correlation_df.iloc[0,2]:2g}')\n",
    "print(f'The R value between physical progeny prodcution and infectious progeny production is {correlation_df.iloc[1,2]:2g}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export processed data to CSV\n",
    "Export dataframe with information for 92 infected cells with long-read sequencing for all expressed viral genes and both barcoded viral genes expressed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_measurement_cells_data_CSV = (\n",
    "    'results/viral_fastq10x/scProgenyProduction_trial3_complete_measurements_cells_data.csv')\n",
    "sequenced_barcoded_cells.to_csv(complete_measurement_cells_data_CSV)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
