{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Integrate data\n",
    "This notebook takes processed data and integrates it into a single dataframe.\n",
    "\n",
    "The general structure of the dataframe is that each cell barcode is listed on a row, and features of that cell are listed in columns. Cells that have more than one valid viral barcode identified may have multiple rows--one for each valid viral barcode."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import python modules:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Variables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input sources\n",
    "cell_annotations_csv = 'results/viral_tags_bcs_in_cells/scProgenyProduction_trial3_cell_barcodes_with_viral_tags.csv.gz' #snakemake.input.cell_annotations\n",
    "viral_genes_by_cell_csv = 'results/viral_fastq10x/scProgenyProduction_trial3_viral_genes_by_cell.csv.gz' #snakemake.input.viral_genes_by_cell_csv\n",
    "viral_barcodes_valid_csv = 'results/viral_fastq10x/scProgenyProduction_trial3_viral_bc_by_cell_valid.csv.gz' #snakemake.input.viral_barcodes_valid_csv\n",
    "filtered_progeny_viral_bc_csv = 'results/viral_progeny/scProgenyProduction_trial3_filtered_progeny_viral_bc.csv.gz' #snakemake.input.filtered_progeny_viral_bc_csv\n",
    "contributes_progeny_by_cell_csv = 'results/viral_fastq10x/scProgenyProduction_trial3_contributes_progeny_by_cell.csv.gz' #snakemake.output.contributes_progeny_by_cell_csv\n",
    "\n",
    "# # Params and wildcards\n",
    "barcoded_viral_genes = ['fluHA', 'fluNA'] #snakemake.params.barcoded_viral_genes\n",
    "# expt = expt = #snakemake.wildcards.expt\n",
    "\n",
    "# # Output\n",
    "# integrated_data_csv = #snakemake.output.integrated_data_csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data and transform into wide format\n",
    "Load data from the following sources:  \n",
    "* `cell_annotations_csv` contains a list of all cell barcodes, their infection status/tag, total_UMIs, viral_UMIs, and frac_viral_UMIs  \n",
    "\n",
    "* `viral_genes_by_cell_csv` contains every cell barcode, each viral gene, whether the viral gene is present, the fraction of UMIs from that gene, and the total number of viral genes detected in that cell  \n",
    "\n",
    "* `viral_barcodes_valid_csv` contains each infected cell, the valid viral barcodes detected in that cell, viral_bc_UMIs, and frac_viral_bc_UMIs  \n",
    "\n",
    "* `filtered_progeny_viral_bc_csv` contains the progeny source, the tag of the progeny source, the valid viral barcodes detected in that progeny source, and the frequency of the viral barcodes in that progeny source  \n",
    "\n",
    "Some of these dataframes are in a long format, with many rows per cell barcode. I will transform them into a wide format, so that each cell barcode has only one row. The only exception is the viral barcodes, which will have one row per cell barcode-viral barcode pair."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load cell barcodes, tags, and basic metrics for every cell in transcriptome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAGTAACAAGT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>47873</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCATCATTGCTT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>90114</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAAGATGTTGA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>111630</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>infected</td>\n",
       "      <td>both</td>\n",
       "      <td>56828</td>\n",
       "      <td>24082</td>\n",
       "      <td>0.423770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>124341</td>\n",
       "      <td>4654</td>\n",
       "      <td>0.037429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>63150</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368</th>\n",
       "      <td>TTTGATCTCGCATTGA</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>170914</td>\n",
       "      <td>10415</td>\n",
       "      <td>0.060937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>65941</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>135009</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000059</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3372 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode    infected infecting_viral_tag  total_UMIs  \\\n",
       "0     AAACCCAGTAACAAGT  uninfected                none       47873   \n",
       "1     AAACCCATCATTGCTT  uninfected                none       90114   \n",
       "2     AAACGAAAGATGTTGA  uninfected                none      111630   \n",
       "3     AAACGAAGTACTTCCC    infected                both       56828   \n",
       "4     AAACGAAGTAGACGTG    infected                  wt      124341   \n",
       "...                ...         ...                 ...         ...   \n",
       "3367  TTTGATCTCCCGTTCA  uninfected                none       63150   \n",
       "3368  TTTGATCTCGCATTGA    infected                  wt      170914   \n",
       "3369  TTTGGAGAGTTGCCTA  uninfected                none       65941   \n",
       "3370  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "3371  TTTGTTGAGTGCAGCA  uninfected                none      135009   \n",
       "\n",
       "      viral_UMIs  frac_viral_UMIs  \n",
       "0              6         0.000125  \n",
       "1             10         0.000111  \n",
       "2             18         0.000161  \n",
       "3          24082         0.423770  \n",
       "4           4654         0.037429  \n",
       "...          ...              ...  \n",
       "3367           3         0.000048  \n",
       "3368       10415         0.060937  \n",
       "3369          12         0.000182  \n",
       "3370        3526         0.023486  \n",
       "3371           8         0.000059  \n",
       "\n",
       "[3372 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cell_annotations = pd.read_csv(cell_annotations_csv)\n",
    "cell_annotations = (\n",
    "    cell_annotations\n",
    "    [['cell_barcode',\n",
    "      'infected',\n",
    "      'infecting_viral_tag',\n",
    "      'total_UMIs',\n",
    "      'viral_UMIs',\n",
    "      'frac_viral_UMIs']])\n",
    "\n",
    "display(cell_annotations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load viral genes detected in each cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>n_viral_genes</th>\n",
       "      <th>frac_UMIs_fluHA</th>\n",
       "      <th>frac_UMIs_fluM</th>\n",
       "      <th>frac_UMIs_fluNA</th>\n",
       "      <th>frac_UMIs_fluNP</th>\n",
       "      <th>frac_UMIs_fluNS</th>\n",
       "      <th>frac_UMIs_fluPA</th>\n",
       "      <th>frac_UMIs_fluPB1</th>\n",
       "      <th>frac_UMIs_fluPB2</th>\n",
       "      <th>present_fluHA</th>\n",
       "      <th>present_fluM</th>\n",
       "      <th>present_fluNA</th>\n",
       "      <th>present_fluNP</th>\n",
       "      <th>present_fluNS</th>\n",
       "      <th>present_fluPA</th>\n",
       "      <th>present_fluPB1</th>\n",
       "      <th>present_fluPB2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAGTAACAAGT</td>\n",
       "      <td>0</td>\n",
       "      <td>4.17772e-05</td>\n",
       "      <td>6.26658e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.08886e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCATCATTGCTT</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5.54853e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>3.32912e-05</td>\n",
       "      <td>1.10971e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.10971e-05</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAAGATGTTGA</td>\n",
       "      <td>0</td>\n",
       "      <td>2.68745e-05</td>\n",
       "      <td>8.95817e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>8.95817e-06</td>\n",
       "      <td>3.58327e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>7</td>\n",
       "      <td>3.51939e-05</td>\n",
       "      <td>0.229816</td>\n",
       "      <td>0.00739072</td>\n",
       "      <td>0.0728338</td>\n",
       "      <td>0.10157</td>\n",
       "      <td>0.00040473</td>\n",
       "      <td>0.00890406</td>\n",
       "      <td>0.00281551</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00256553</td>\n",
       "      <td>0.0127231</td>\n",
       "      <td>0.000249314</td>\n",
       "      <td>0.00488978</td>\n",
       "      <td>0.0165271</td>\n",
       "      <td>8.84664e-05</td>\n",
       "      <td>0.000361908</td>\n",
       "      <td>2.41272e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.58353e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>1.58353e-05</td>\n",
       "      <td>1.58353e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368</th>\n",
       "      <td>TTTGATCTCGCATTGA</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00360415</td>\n",
       "      <td>0.0283242</td>\n",
       "      <td>0.0078636</td>\n",
       "      <td>0.00723756</td>\n",
       "      <td>0.013147</td>\n",
       "      <td>8.77634e-05</td>\n",
       "      <td>0.000497326</td>\n",
       "      <td>0.000175527</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9.09904e-05</td>\n",
       "      <td>1.51651e-05</td>\n",
       "      <td>1.51651e-05</td>\n",
       "      <td>6.06603e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000619463</td>\n",
       "      <td>0.0111104</td>\n",
       "      <td>0.00029974</td>\n",
       "      <td>0.00278425</td>\n",
       "      <td>0.0081929</td>\n",
       "      <td>4.66263e-05</td>\n",
       "      <td>0.000379671</td>\n",
       "      <td>5.32872e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>1</td>\n",
       "      <td>1.48138e-05</td>\n",
       "      <td>1.48138e-05</td>\n",
       "      <td>7.40691e-06</td>\n",
       "      <td>0</td>\n",
       "      <td>2.22207e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3372 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode  n_viral_genes frac_UMIs_fluHA frac_UMIs_fluM  \\\n",
       "0     AAACCCAGTAACAAGT              0     4.17772e-05    6.26658e-05   \n",
       "1     AAACCCATCATTGCTT              1               0    5.54853e-05   \n",
       "2     AAACGAAAGATGTTGA              0     2.68745e-05    8.95817e-05   \n",
       "3     AAACGAAGTACTTCCC              7     3.51939e-05       0.229816   \n",
       "4     AAACGAAGTAGACGTG              8      0.00256553      0.0127231   \n",
       "...                ...            ...             ...            ...   \n",
       "3367  TTTGATCTCCCGTTCA              0               0    1.58353e-05   \n",
       "3368  TTTGATCTCGCATTGA              8      0.00360415      0.0283242   \n",
       "3369  TTTGGAGAGTTGCCTA              1               0    9.09904e-05   \n",
       "3370  TTTGGAGGTATCGTTG              8     0.000619463      0.0111104   \n",
       "3371  TTTGTTGAGTGCAGCA              1     1.48138e-05    1.48138e-05   \n",
       "\n",
       "     frac_UMIs_fluNA frac_UMIs_fluNP frac_UMIs_fluNS frac_UMIs_fluPA  \\\n",
       "0                  0               0     2.08886e-05               0   \n",
       "1                  0     3.32912e-05     1.10971e-05               0   \n",
       "2                  0     8.95817e-06     3.58327e-05               0   \n",
       "3         0.00739072       0.0728338         0.10157      0.00040473   \n",
       "4        0.000249314      0.00488978       0.0165271     8.84664e-05   \n",
       "...              ...             ...             ...             ...   \n",
       "3367               0     1.58353e-05     1.58353e-05               0   \n",
       "3368       0.0078636      0.00723756        0.013147     8.77634e-05   \n",
       "3369     1.51651e-05     1.51651e-05     6.06603e-05               0   \n",
       "3370      0.00029974      0.00278425       0.0081929     4.66263e-05   \n",
       "3371     7.40691e-06               0     2.22207e-05               0   \n",
       "\n",
       "     frac_UMIs_fluPB1 frac_UMIs_fluPB2 present_fluHA present_fluM  \\\n",
       "0                   0                0         False        False   \n",
       "1                   0      1.10971e-05         False        False   \n",
       "2                   0                0         False        False   \n",
       "3          0.00890406       0.00281551         False         True   \n",
       "4         0.000361908      2.41272e-05          True         True   \n",
       "...               ...              ...           ...          ...   \n",
       "3367                0                0         False        False   \n",
       "3368      0.000497326      0.000175527          True         True   \n",
       "3369                0                0         False        False   \n",
       "3370      0.000379671      5.32872e-05          True         True   \n",
       "3371                0                0         False        False   \n",
       "\n",
       "     present_fluNA present_fluNP present_fluNS present_fluPA present_fluPB1  \\\n",
       "0            False         False         False         False          False   \n",
       "1            False         False         False         False          False   \n",
       "2            False         False         False         False          False   \n",
       "3             True          True          True          True           True   \n",
       "4             True          True          True          True           True   \n",
       "...            ...           ...           ...           ...            ...   \n",
       "3367         False         False         False         False          False   \n",
       "3368          True          True          True          True           True   \n",
       "3369          True         False         False         False          False   \n",
       "3370          True          True          True          True           True   \n",
       "3371          True         False         False         False          False   \n",
       "\n",
       "     present_fluPB2  \n",
       "0             False  \n",
       "1              True  \n",
       "2             False  \n",
       "3              True  \n",
       "4              True  \n",
       "...             ...  \n",
       "3367          False  \n",
       "3368           True  \n",
       "3369          False  \n",
       "3370           True  \n",
       "3371          False  \n",
       "\n",
       "[3372 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_genes = pd.read_csv(viral_genes_by_cell_csv)\n",
    "viral_genes = viral_genes.rename(columns={'frac_gene_UMIs': 'frac_UMIs',\n",
    "                                          'gene_present': 'present'})\n",
    "viral_genes = (viral_genes\n",
    "               .pivot(\n",
    "    index=['cell_barcode', 'n_viral_genes'],\n",
    "    columns=['gene'],\n",
    "    values=['frac_UMIs', 'present']))\n",
    "viral_genes.columns = ['_'.join(col).strip() for col in viral_genes.columns.values]\n",
    "viral_genes = viral_genes.reset_index()\n",
    "display(viral_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load valid viral barcodes in each infected cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>barcoded_gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>viral_bc_UMIs</th>\n",
       "      <th>frac_viral_bc_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAGTAAGCGACATGAG</td>\n",
       "      <td>251</td>\n",
       "      <td>0.002019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAGGATTCTGATGGT</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>GTGGAGTCGCCAGTTC</td>\n",
       "      <td>114</td>\n",
       "      <td>0.001424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAGGGCCAGGCTACC</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAGTGATCCCCATAC</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAGGGCCAGGCTACC</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>CATTTAACGCTGTGAG</td>\n",
       "      <td>15</td>\n",
       "      <td>0.000741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAGGGCCAGGCTACC</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>CGTAGGATGTTGCGTC</td>\n",
       "      <td>31</td>\n",
       "      <td>0.001532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>TTTACCAGTCGCTTAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTGGAGGAGACCCGTG</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>TTTAGTCCATCATCCC</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>AGAAACCTCGACATAT</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000</th>\n",
       "      <td>TTTAGTCCATCATCCC</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTGGACGCATTGCAAA</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1001</th>\n",
       "      <td>TTTCACAAGCCAAGCA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GGTATCAGTTATTGTT</td>\n",
       "      <td>186</td>\n",
       "      <td>0.002718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1002</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>ACATCTTATTTACACG</td>\n",
       "      <td>39</td>\n",
       "      <td>0.000260</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1003 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode barcoded_gene     viral_barcode  viral_bc_UMIs  \\\n",
       "0     AAACGAAGTAGACGTG         fluHA  AAGTAAGCGACATGAG            251   \n",
       "1     AAAGGATTCTGATGGT         fluHA  GTGGAGTCGCCAGTTC            114   \n",
       "2     AAAGGGCCAGGCTACC         fluHA  AAAGTGATCCCCATAC              8   \n",
       "3     AAAGGGCCAGGCTACC         fluHA  CATTTAACGCTGTGAG             15   \n",
       "4     AAAGGGCCAGGCTACC         fluHA  CGTAGGATGTTGCGTC             31   \n",
       "...                ...           ...               ...            ...   \n",
       "998   TTTACCAGTCGCTTAA         fluNA  TTGGAGGAGACCCGTG              7   \n",
       "999   TTTAGTCCATCATCCC         fluNA  AGAAACCTCGACATAT             11   \n",
       "1000  TTTAGTCCATCATCCC         fluNA  TTGGACGCATTGCAAA             18   \n",
       "1001  TTTCACAAGCCAAGCA         fluNA  GGTATCAGTTATTGTT            186   \n",
       "1002  TTTGGAGGTATCGTTG         fluNA  ACATCTTATTTACACG             39   \n",
       "\n",
       "      frac_viral_bc_UMIs  \n",
       "0               0.002019  \n",
       "1               0.001424  \n",
       "2               0.000395  \n",
       "3               0.000741  \n",
       "4               0.001532  \n",
       "...                  ...  \n",
       "998             0.000061  \n",
       "999             0.000463  \n",
       "1000            0.000757  \n",
       "1001            0.002718  \n",
       "1002            0.000260  \n",
       "\n",
       "[1003 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "transcriptome_viral_barcodes = pd.read_csv(viral_barcodes_valid_csv)\n",
    "transcriptome_viral_barcodes = transcriptome_viral_barcodes.drop(columns=['valid_viral_bc'])\n",
    "transcriptome_viral_barcodes = transcriptome_viral_barcodes.rename(columns={'gene': 'barcoded_gene'})\n",
    "assert set(transcriptome_viral_barcodes['barcoded_gene']) == set(barcoded_viral_genes), \\\n",
    "       \"Barcoded genes in barcode counts do not match expectation.\"\n",
    "display(transcriptome_viral_barcodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load progeny viral barcode frequencies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>barcoded_gene</th>\n",
       "      <th colspan=\"2\" halign=\"left\">progeny_freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>source</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>second_infection</th>\n",
       "      <th>supernatant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAACACGTTCTATA</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAACATGATGACGCC</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAACTAGTTAGAGCA</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAGCCATTCGGAGA</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAGTTCTTGGATGT</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>TTTGCCAGAAAATCTT</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>TTTGTCGGCAGTCACT</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864</th>\n",
       "      <td>TTTTAACGTTATACTA</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865</th>\n",
       "      <td>TTTTACCTACGTAGTT</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866</th>\n",
       "      <td>TTTTGCTTCGAGATAG</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.093595</td>\n",
       "      <td>0.012644</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>867 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           viral_barcode infecting_viral_tag barcoded_gene     progeny_freq  \\\n",
       "source                                                     second_infection   \n",
       "0       AAAAACACGTTCTATA                  wt         fluNA         0.000010   \n",
       "1       AAAACATGATGACGCC                  wt         fluNA         0.000010   \n",
       "2       AAAACTAGTTAGAGCA                  wt         fluHA         0.000010   \n",
       "3       AAAAGCCATTCGGAGA                  wt         fluNA         0.000010   \n",
       "4       AAAAGTTCTTGGATGT                  wt         fluNA         0.000010   \n",
       "..                   ...                 ...           ...              ...   \n",
       "862     TTTGCCAGAAAATCTT                  wt         fluHA         0.000010   \n",
       "863     TTTGTCGGCAGTCACT                  wt         fluNA         0.000010   \n",
       "864     TTTTAACGTTATACTA                  wt         fluNA         0.000010   \n",
       "865     TTTTACCTACGTAGTT                  wt         fluNA         0.000010   \n",
       "866     TTTTGCTTCGAGATAG                  wt         fluNA         0.093595   \n",
       "\n",
       "                    \n",
       "source supernatant  \n",
       "0         0.000010  \n",
       "1         0.000010  \n",
       "2         0.000010  \n",
       "3         0.000010  \n",
       "4         0.000010  \n",
       "..             ...  \n",
       "862       0.000010  \n",
       "863       0.000010  \n",
       "864       0.000010  \n",
       "865       0.000010  \n",
       "866       0.012644  \n",
       "\n",
       "[867 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "progeny_viral_barcodes = pd.read_csv(filtered_progeny_viral_bc_csv)\n",
    "progeny_viral_barcodes = progeny_viral_barcodes.drop(columns=['Unnamed: 0'])\n",
    "progeny_viral_barcodes = (progeny_viral_barcodes\n",
    "                          .rename(columns={'tag': 'infecting_viral_tag',\n",
    "                                           'average_freq': 'progeny_freq',\n",
    "                                           'gene': 'barcoded_gene'}))\n",
    "assert set(progeny_viral_barcodes['barcoded_gene']) == set(barcoded_viral_genes), \\\n",
    "       \"Barcoded genes in barcode counts do not match expectation.\"\n",
    "progeny_viral_barcodes = (\n",
    "    progeny_viral_barcodes\n",
    "    .pivot(index=['viral_barcode', 'infecting_viral_tag', 'barcoded_gene'],\n",
    "           columns=['source'],\n",
    "           values=['progeny_freq'])\n",
    "    .reset_index())\n",
    "progeny_viral_barcodes.columns = ['_'.join(col).strip() for col in viral_genes.columns.values]\n",
    "progeny_viral_barcodes = progeny_viral_barcodes.reset_index()\n",
    "display(progeny_viral_barcodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load whether each infected cell contributes any progeny:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>source</th>\n",
       "      <th>max_progeny_freq</th>\n",
       "      <th>contributes_progeny</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>second_infection</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAGGATTCTGATGGT</td>\n",
       "      <td>second_infection</td>\n",
       "      <td>0.077933</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAGGGCCAGGCTACC</td>\n",
       "      <td>second_infection</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAGGGCTCCGCACTT</td>\n",
       "      <td>second_infection</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAGTCCAGTAGAGTT</td>\n",
       "      <td>second_infection</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>815</th>\n",
       "      <td>TTTAGTCCATCATCCC</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>TTTAGTCGTGCTCCGA</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>TTTCACAAGCCAAGCA</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>TTTGATCTCGCATTGA</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>819</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>0.001790</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>820 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         cell_barcode            source  max_progeny_freq contributes_progeny\n",
       "0    AAACGAAGTAGACGTG  second_infection          0.000010               False\n",
       "1    AAAGGATTCTGATGGT  second_infection          0.077933                True\n",
       "2    AAAGGGCCAGGCTACC  second_infection          0.000010               False\n",
       "3    AAAGGGCTCCGCACTT  second_infection          0.000124                True\n",
       "4    AAAGTCCAGTAGAGTT  second_infection          0.000010               False\n",
       "..                ...               ...               ...                 ...\n",
       "815  TTTAGTCCATCATCCC       supernatant          0.000010               False\n",
       "816  TTTAGTCGTGCTCCGA       supernatant          0.000010               False\n",
       "817  TTTCACAAGCCAAGCA       supernatant          0.000010               False\n",
       "818  TTTGATCTCGCATTGA       supernatant          0.000010               False\n",
       "819  TTTGGAGGTATCGTTG       supernatant          0.001790                True\n",
       "\n",
       "[820 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "contributes_progeny = pd.read_csv(contributes_progeny_by_cell_csv)\n",
    "contributes_progeny = contributes_progeny.rename(columns={'gene': 'barcoded_gene'})\n",
    "assert set(transcriptome_viral_barcodes['barcoded_gene']) == set(barcoded_viral_genes), \\\n",
    "       \"Barcoded genes in barcode counts do not match expectation.\"\n",
    "display(contributes_progeny)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Integrate data into single dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Revise integrated dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output integrated dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
