{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Integrate data\n",
    "This notebook takes processed data and integrates it into a single dataframe.\n",
    "\n",
    "The general structure of the dataframe is that each cell barcode is listed on a row, and features of that cell are listed in columns. Cells that have more than one valid viral barcode identified may have multiple rows--one for each valid viral barcode."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import python modules:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Variables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input sources\n",
    "cell_annotations_csv = 'results/viral_tags_bcs_in_cells/scProgenyProduction_trial3_cell_barcodes_with_viral_tags.csv.gz' #snakemake.input.cell_annotations\n",
    "viral_genes_by_cell_csv = 'results/viral_fastq10x/scProgenyProduction_trial3_viral_genes_by_cell.csv.gz' #snakemake.input.viral_genes_by_cell_csv\n",
    "viral_barcodes_valid_csv = 'results/viral_fastq10x/scProgenyProduction_trial3_viral_bc_by_cell_valid.csv.gz' #snakemake.input.viral_barcodes_valid_csv\n",
    "filtered_progeny_viral_bc_csv = 'results/viral_progeny/scProgenyProduction_trial3_filtered_progeny_viral_bc.csv.gz' #snakemake.input.filtered_progeny_viral_bc_csv\n",
    "\n",
    "# # Params and wildcards\n",
    "barcoded_viral_genes = ['fluHA', 'fluNA'] #snakemake.params.barcoded_viral_genes\n",
    "# expt = expt = #snakemake.wildcards.expt\n",
    "\n",
    "# # Output\n",
    "# integrated_data_csv = #snakemake.output.integrated_data_csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data\n",
    "Load data from the following sources:  \n",
    "* `cell_annotations_csv` contains a list of all cell barcodes, their infection status/tag, total_UMIs, viral_UMIs, and frac_viral_UMIs  \n",
    "\n",
    "* `viral_genes_by_cell_csv` contains every cell barcode, each viral gene, whether the viral gene is present, the fraction of UMIs from that gene, and the total number of viral genes detected in that cell  \n",
    "\n",
    "* `viral_barcodes_valid_csv` contains each infected cell, the valid viral barcodes detected in that cell, viral_bc_UMIs, and frac_viral_bc_UMIs  \n",
    "\n",
    "* `filtered_progeny_viral_bc_csv` contains the progeny source, the tag of the progeny source, the valid viral barcodes detected in that progeny source, and the frequency of the viral barcodes in that progeny source"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load cell barcodes, tags, and basic metrics for every cell in transcriptome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>cellular_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>viral_tag_doublet</th>\n",
       "      <th>viral_tag_major</th>\n",
       "      <th>viral_tag_minor</th>\n",
       "      <th>viral_tag_major_infected</th>\n",
       "      <th>viral_tag_minor_infected</th>\n",
       "      <th>viral_tag_major_counts</th>\n",
       "      <th>viral_tag_minor_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAGTAACAAGT</td>\n",
       "      <td>47873</td>\n",
       "      <td>6</td>\n",
       "      <td>47867</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>syn</td>\n",
       "      <td>wt</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCATCATTGCTT</td>\n",
       "      <td>90114</td>\n",
       "      <td>10</td>\n",
       "      <td>90104</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>wt</td>\n",
       "      <td>syn</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAAGATGTTGA</td>\n",
       "      <td>111630</td>\n",
       "      <td>18</td>\n",
       "      <td>111612</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>wt</td>\n",
       "      <td>syn</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>56828</td>\n",
       "      <td>24082</td>\n",
       "      <td>32746</td>\n",
       "      <td>0.423770</td>\n",
       "      <td>infected</td>\n",
       "      <td>both</td>\n",
       "      <td>True</td>\n",
       "      <td>wt</td>\n",
       "      <td>syn</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>18769</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>124341</td>\n",
       "      <td>4654</td>\n",
       "      <td>119687</td>\n",
       "      <td>0.037429</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>False</td>\n",
       "      <td>wt</td>\n",
       "      <td>syn</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>3694</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>63150</td>\n",
       "      <td>3</td>\n",
       "      <td>63147</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>wt</td>\n",
       "      <td>syn</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368</th>\n",
       "      <td>TTTGATCTCGCATTGA</td>\n",
       "      <td>170914</td>\n",
       "      <td>10415</td>\n",
       "      <td>160499</td>\n",
       "      <td>0.060937</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>False</td>\n",
       "      <td>wt</td>\n",
       "      <td>syn</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>6883</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>65941</td>\n",
       "      <td>12</td>\n",
       "      <td>65929</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>wt</td>\n",
       "      <td>syn</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>146604</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>False</td>\n",
       "      <td>wt</td>\n",
       "      <td>syn</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>2784</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3371</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>135009</td>\n",
       "      <td>8</td>\n",
       "      <td>135001</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>False</td>\n",
       "      <td>wt</td>\n",
       "      <td>syn</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3372 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode  total_UMIs  viral_UMIs  cellular_UMIs  \\\n",
       "0     AAACCCAGTAACAAGT       47873           6          47867   \n",
       "1     AAACCCATCATTGCTT       90114          10          90104   \n",
       "2     AAACGAAAGATGTTGA      111630          18         111612   \n",
       "3     AAACGAAGTACTTCCC       56828       24082          32746   \n",
       "4     AAACGAAGTAGACGTG      124341        4654         119687   \n",
       "...                ...         ...         ...            ...   \n",
       "3367  TTTGATCTCCCGTTCA       63150           3          63147   \n",
       "3368  TTTGATCTCGCATTGA      170914       10415         160499   \n",
       "3369  TTTGGAGAGTTGCCTA       65941          12          65929   \n",
       "3370  TTTGGAGGTATCGTTG      150130        3526         146604   \n",
       "3371  TTTGTTGAGTGCAGCA      135009           8         135001   \n",
       "\n",
       "      frac_viral_UMIs    infected infecting_viral_tag  viral_tag_doublet  \\\n",
       "0            0.000125  uninfected                none              False   \n",
       "1            0.000111  uninfected                none              False   \n",
       "2            0.000161  uninfected                none              False   \n",
       "3            0.423770    infected                both               True   \n",
       "4            0.037429    infected                  wt              False   \n",
       "...               ...         ...                 ...                ...   \n",
       "3367         0.000048  uninfected                none              False   \n",
       "3368         0.060937    infected                  wt              False   \n",
       "3369         0.000182  uninfected                none              False   \n",
       "3370         0.023486    infected                  wt              False   \n",
       "3371         0.000059  uninfected                none              False   \n",
       "\n",
       "     viral_tag_major viral_tag_minor  viral_tag_major_infected  \\\n",
       "0                syn              wt                     False   \n",
       "1                 wt             syn                     False   \n",
       "2                 wt             syn                     False   \n",
       "3                 wt             syn                      True   \n",
       "4                 wt             syn                      True   \n",
       "...              ...             ...                       ...   \n",
       "3367              wt             syn                     False   \n",
       "3368              wt             syn                      True   \n",
       "3369              wt             syn                     False   \n",
       "3370              wt             syn                      True   \n",
       "3371              wt             syn                     False   \n",
       "\n",
       "      viral_tag_minor_infected  viral_tag_major_counts  viral_tag_minor_counts  \n",
       "0                        False                       1                       1  \n",
       "1                        False                       5                       2  \n",
       "2                        False                       9                       2  \n",
       "3                         True                   18769                      28  \n",
       "4                        False                    3694                       9  \n",
       "...                        ...                     ...                     ...  \n",
       "3367                     False                       3                       0  \n",
       "3368                     False                    6883                      10  \n",
       "3369                     False                       4                       3  \n",
       "3370                     False                    2784                       2  \n",
       "3371                     False                       4                       2  \n",
       "\n",
       "[3372 rows x 14 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cell_annotations = pd.read_csv(cell_annotations_csv)\n",
    "display(cell_annotations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load viral genes detected in each cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>n_viral_genes</th>\n",
       "      <th>gene</th>\n",
       "      <th>frac_gene_UMIs</th>\n",
       "      <th>gene_present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAGTAACAAGT</td>\n",
       "      <td>0</td>\n",
       "      <td>fluPB2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCATCATTGCTT</td>\n",
       "      <td>1</td>\n",
       "      <td>fluPB2</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAAGATGTTGA</td>\n",
       "      <td>0</td>\n",
       "      <td>fluPB2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>7</td>\n",
       "      <td>fluPB2</td>\n",
       "      <td>0.002816</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>8</td>\n",
       "      <td>fluPB2</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26971</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>0</td>\n",
       "      <td>fluNS</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26972</th>\n",
       "      <td>TTTGATCTCGCATTGA</td>\n",
       "      <td>8</td>\n",
       "      <td>fluNS</td>\n",
       "      <td>0.013147</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26973</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>1</td>\n",
       "      <td>fluNS</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26974</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>8</td>\n",
       "      <td>fluNS</td>\n",
       "      <td>0.008193</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26975</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>1</td>\n",
       "      <td>fluNS</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26976 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           cell_barcode  n_viral_genes    gene  frac_gene_UMIs  gene_present\n",
       "0      AAACCCAGTAACAAGT              0  fluPB2        0.000000         False\n",
       "1      AAACCCATCATTGCTT              1  fluPB2        0.000011          True\n",
       "2      AAACGAAAGATGTTGA              0  fluPB2        0.000000         False\n",
       "3      AAACGAAGTACTTCCC              7  fluPB2        0.002816          True\n",
       "4      AAACGAAGTAGACGTG              8  fluPB2        0.000024          True\n",
       "...                 ...            ...     ...             ...           ...\n",
       "26971  TTTGATCTCCCGTTCA              0   fluNS        0.000016         False\n",
       "26972  TTTGATCTCGCATTGA              8   fluNS        0.013147          True\n",
       "26973  TTTGGAGAGTTGCCTA              1   fluNS        0.000061         False\n",
       "26974  TTTGGAGGTATCGTTG              8   fluNS        0.008193          True\n",
       "26975  TTTGTTGAGTGCAGCA              1   fluNS        0.000022         False\n",
       "\n",
       "[26976 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_genes = pd.read_csv(viral_genes_by_cell_csv)\n",
    "display(viral_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load valid viral barcodes in each infected cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>viral_bc_UMIs</th>\n",
       "      <th>frac_viral_bc_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAGTAAGCGACATGAG</td>\n",
       "      <td>251</td>\n",
       "      <td>0.002019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAGGATTCTGATGGT</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>GTGGAGTCGCCAGTTC</td>\n",
       "      <td>114</td>\n",
       "      <td>0.001424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAGGGCCAGGCTACC</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAGTGATCCCCATAC</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAGGGCCAGGCTACC</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>CATTTAACGCTGTGAG</td>\n",
       "      <td>15</td>\n",
       "      <td>0.000741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAGGGCCAGGCTACC</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>CGTAGGATGTTGCGTC</td>\n",
       "      <td>31</td>\n",
       "      <td>0.001532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>TTTACCAGTCGCTTAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTGGAGGAGACCCGTG</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>TTTAGTCCATCATCCC</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>AGAAACCTCGACATAT</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000</th>\n",
       "      <td>TTTAGTCCATCATCCC</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTGGACGCATTGCAAA</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1001</th>\n",
       "      <td>TTTCACAAGCCAAGCA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GGTATCAGTTATTGTT</td>\n",
       "      <td>186</td>\n",
       "      <td>0.002718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1002</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>ACATCTTATTTACACG</td>\n",
       "      <td>39</td>\n",
       "      <td>0.000260</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1003 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode   gene     viral_barcode  viral_bc_UMIs  \\\n",
       "0     AAACGAAGTAGACGTG  fluHA  AAGTAAGCGACATGAG            251   \n",
       "1     AAAGGATTCTGATGGT  fluHA  GTGGAGTCGCCAGTTC            114   \n",
       "2     AAAGGGCCAGGCTACC  fluHA  AAAGTGATCCCCATAC              8   \n",
       "3     AAAGGGCCAGGCTACC  fluHA  CATTTAACGCTGTGAG             15   \n",
       "4     AAAGGGCCAGGCTACC  fluHA  CGTAGGATGTTGCGTC             31   \n",
       "...                ...    ...               ...            ...   \n",
       "998   TTTACCAGTCGCTTAA  fluNA  TTGGAGGAGACCCGTG              7   \n",
       "999   TTTAGTCCATCATCCC  fluNA  AGAAACCTCGACATAT             11   \n",
       "1000  TTTAGTCCATCATCCC  fluNA  TTGGACGCATTGCAAA             18   \n",
       "1001  TTTCACAAGCCAAGCA  fluNA  GGTATCAGTTATTGTT            186   \n",
       "1002  TTTGGAGGTATCGTTG  fluNA  ACATCTTATTTACACG             39   \n",
       "\n",
       "      frac_viral_bc_UMIs  \n",
       "0               0.002019  \n",
       "1               0.001424  \n",
       "2               0.000395  \n",
       "3               0.000741  \n",
       "4               0.001532  \n",
       "...                  ...  \n",
       "998             0.000061  \n",
       "999             0.000463  \n",
       "1000            0.000757  \n",
       "1001            0.002718  \n",
       "1002            0.000260  \n",
       "\n",
       "[1003 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "transcriptome_viral_barcodes = pd.read_csv(viral_barcodes_valid_csv)\n",
    "transcriptome_viral_barcodes = transcriptome_viral_barcodes.drop(columns=['valid_viral_bc'])\n",
    "assert set(transcriptome_viral_barcodes['gene']) == set(barcoded_viral_genes), \\\n",
    "       \"Barcoded genes in barcode counts do not match expectation.\"\n",
    "display(transcriptome_viral_barcodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load progeny viral barcode frequencies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>progeny_freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAATTGTCCAATATA</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAACCGAGGAAATCCC</td>\n",
       "      <td>0.001310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAATATCGCACCAAGA</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AACACACGATACAGAC</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AACAGTACGAAGCCGA</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1729</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTCTTTACTCAGAAT</td>\n",
       "      <td>0.002060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1730</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTGTCGGCAGTCACT</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1731</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTAACGTTATACTA</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1732</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTACCTACGTAGTT</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTGCTTCGAGATAG</td>\n",
       "      <td>0.012644</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1734 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                source infecting_viral_tag   gene     viral_barcode  \\\n",
       "0     second_infection                 syn  fluHA  AAAATTGTCCAATATA   \n",
       "1     second_infection                 syn  fluHA  AAACCGAGGAAATCCC   \n",
       "2     second_infection                 syn  fluHA  AAATATCGCACCAAGA   \n",
       "3     second_infection                 syn  fluHA  AACACACGATACAGAC   \n",
       "4     second_infection                 syn  fluHA  AACAGTACGAAGCCGA   \n",
       "...                ...                 ...    ...               ...   \n",
       "1729       supernatant                  wt  fluNA  TTTCTTTACTCAGAAT   \n",
       "1730       supernatant                  wt  fluNA  TTTGTCGGCAGTCACT   \n",
       "1731       supernatant                  wt  fluNA  TTTTAACGTTATACTA   \n",
       "1732       supernatant                  wt  fluNA  TTTTACCTACGTAGTT   \n",
       "1733       supernatant                  wt  fluNA  TTTTGCTTCGAGATAG   \n",
       "\n",
       "      progeny_freq  \n",
       "0         0.000010  \n",
       "1         0.001310  \n",
       "2         0.000010  \n",
       "3         0.000010  \n",
       "4         0.000010  \n",
       "...            ...  \n",
       "1729      0.002060  \n",
       "1730      0.000010  \n",
       "1731      0.000010  \n",
       "1732      0.000010  \n",
       "1733      0.012644  \n",
       "\n",
       "[1734 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "progeny_viral_barcodes = pd.read_csv(filtered_progeny_viral_bc_csv)\n",
    "progeny_viral_barcodes = progeny_viral_barcodes.drop(columns=['Unnamed: 0'])\n",
    "progeny_viral_barcodes = (progeny_viral_barcodes\n",
    "                          .rename(columns={'tag': 'infecting_viral_tag',\n",
    "                                           'average_freq': 'progeny_freq'}))\n",
    "assert set(progeny_viral_barcodes['gene']) == set(barcoded_viral_genes), \\\n",
    "       \"Barcoded genes in barcode counts do not match expectation.\"\n",
    "display(progeny_viral_barcodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Integrate data into single dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Revise integrated dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output integrated dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
