# flu sequences

This subdirectory has the flu sequences used in the experiments.
That virus is a variant of the A/California/4/2009 (H1N1) virus, abbreviated "CA09".
Note that the experiments used both wildtype and synonymously tagged variants (these variants differ by defined nucleotide changes near the termini of each gene.)
The directory contains:

 - FASTA file with the full vRNAs for the wildtype tag variant: [flu-CA09.fasta](flu-CA09.fasta)

 - GTF file annotating the full vRNA FASTA file for the wildtype tag variant: [flu-CA09.gtf](flu-CA09.gtf)

 - FASTA file with the viral mRNAs for the wildtype tag variant: [flu-CA09-mRNA.fasta](flu-CA09-mRNA.fasta)

 - Genbank file with the full vRNAs with annotated features including the tag locations as ambiguous nucleotides indicating the possible tag identities: [flu-CA09.gb](flu-CA09.gb)


These files are automatically generated from the Bloom lab plasmid maps in [./plasmid_maps](plasmid_maps), which give the sequences of the plasmids used to create teh viruses by reverse genetics.
The plasmid maps that create the files are generated by the Python script [make_seqs_and_gtf.py](make_seqs_and_gtf.py), which you can run with simply:

    python make_seqs_and_gtf.py

## Notes

### Upstream start codon in NP segment
The vRNA for the NP from CA09 contains an upstream translation-initiation site that adds a 6 amino-acid N-terminal extension to the protein, increasing its length from 498 to 504 residues.
Based on the results described in [Wise et al](https://www.biorxiv.org/content/biorxiv/early/2019/09/04/738427) (which are admittedly for a slightly different viral strain), it appears that the upstream start is less used than the canonical one, so we have annotated the NP CDS using the canonical start codon to give a 498-residue rather than a 504-residue protein.

### Barcoded Segments
The plasmid maps for the barcoded segments (with `'bc'` in their Genbank file names) have some differences from the rest of the plasmids. First, these segments are carried on a pHH, rather pHW backbone. Second, these segments have a duplicated packaging signal at the 5' and 3' end of the coding sequence. Third, these segments have a 16N viral barcode inserted between the end of the coding sequence and the duplicated packaging signal.

Briefly, these plasmids were produced by amplifying the coding sequence from plasmids `2232 (HA_G155E)`, `2303 (HA_G155E-dblSyn)`, `368 (NA)`, or `2304 (NA-dblSyn)` with a primer that appends a 16N barcode to the 3' end of the CDS. They were then cloned into a backbone with an appropriate duplicated packaging signal: `2227 (HA)` or `2228 (NA)`. For complete details on generating the barcoded plasmid libraries for these segments, see David Bacsik's notebook entries: [Cloning Barcoded CA09 HA and NA](https://benchling.com/s/etr-rkHAIdbirA7Qo3WrTvN9).
