{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Correct viral barcodes to reduce replication, PCR, and sequencing errors\n",
    "This Python Jupyter notebook uses `UMI_tools` directional adjacency method to correct viral barcodes that are likely derived from errors in the replication or sequencing library preparation process.\n",
    "\n",
    "## Notes about UMI_tools\n",
    "* Using directional adjacency method. This has [been demonstrated on simulated data](https://cgatoxford.wordpress.com/2015/08/14/unique-molecular-identifiers-the-problem-the-solution-and-the-proof/) to produce a more accurate estimate of true UMI number than other heuristics.\n",
    "* Sequence must be input as byte. See definition here: https://stackoverflow.com/questions/6269765/what-does-the-b-character-do-in-front-of-a-string-literal\n",
    "* The corrected barcode is returned as the first barcode in the group list. See umi_tools API documentation: https://umi-tools.readthedocs.io/en/latest/API.html\n",
    "\n",
    "Import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "from IPython.display import display\n",
    "\n",
    "from dms_variants.constants import CBPALETTE\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import plotnine as p9\n",
    "\n",
    "from umi_tools import UMIClusterer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get `snakemake` variables [as described here](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#jupyter-notebook-integration):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_bc_in_progeny_freq_csv = snakemake.input.viral_bc_in_progeny_freq_csv\n",
    "viral_bc_in_progeny_corrected = (snakemake\n",
    "                                 .output\n",
    "                                 .viral_bc_in_progeny_corrected_csv)\n",
    "plot = snakemake.output.plot\n",
    "expt = snakemake.wildcards.expt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import barcode frequency data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>tag</th>\n",
       "      <th>gene</th>\n",
       "      <th>barcode</th>\n",
       "      <th>mean_freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAAAGTAAATCTT</td>\n",
       "      <td>5.255458e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAAATCTTAATAA</td>\n",
       "      <td>2.031765e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAAATCTTAATGA</td>\n",
       "      <td>8.673284e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAACCCAATTATT</td>\n",
       "      <td>1.061100e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAACCCGATTATT</td>\n",
       "      <td>3.408782e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130646</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTGTTAGCGTCCTG</td>\n",
       "      <td>3.067874e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130647</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTTTAGAAAACGTA</td>\n",
       "      <td>8.987181e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130648</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTTTAGAAAACGTC</td>\n",
       "      <td>1.680603e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130649</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTTTCACTGCCATT</td>\n",
       "      <td>8.987181e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130650</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTTTCGGCGAGTGC</td>\n",
       "      <td>1.204282e-04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130651 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  source  tag   gene           barcode     mean_freq\n",
       "0       second_infection  syn  fluHA  AAAAAAAGTAAATCTT  5.255458e-05\n",
       "1       second_infection  syn  fluHA  AAAAAAATCTTAATAA  2.031765e-06\n",
       "2       second_infection  syn  fluHA  AAAAAAATCTTAATGA  8.673284e-05\n",
       "3       second_infection  syn  fluHA  AAAAAACCCAATTATT  1.061100e-06\n",
       "4       second_infection  syn  fluHA  AAAAAACCCGATTATT  3.408782e-05\n",
       "...                  ...  ...    ...               ...           ...\n",
       "130646       supernatant   wt  fluNA  TTTTGTTAGCGTCCTG  3.067874e-04\n",
       "130647       supernatant   wt  fluNA  TTTTTTAGAAAACGTA  8.987181e-07\n",
       "130648       supernatant   wt  fluNA  TTTTTTAGAAAACGTC  1.680603e-04\n",
       "130649       supernatant   wt  fluNA  TTTTTTCACTGCCATT  8.987181e-07\n",
       "130650       supernatant   wt  fluNA  TTTTTTCGGCGAGTGC  1.204282e-04\n",
       "\n",
       "[130651 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_bc_df = pd.read_csv(gzip.open(viral_bc_in_progeny_freq_csv))\n",
    "display(viral_bc_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cluster barcodes within each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 21467 unique barcodes in the raw data for second_infection syn fluHA\n",
      "5276 barcodes were corrected.\n",
      "There are 16191 corrected barcodes for second_infection syn fluHA\n",
      "\n",
      "There are 22686 unique barcodes in the raw data for second_infection syn fluNA\n",
      "3440 barcodes were corrected.\n",
      "There are 19246 corrected barcodes for second_infection syn fluNA\n",
      "\n",
      "There are 25980 unique barcodes in the raw data for second_infection wt fluHA\n",
      "6063 barcodes were corrected.\n",
      "There are 19917 corrected barcodes for second_infection wt fluHA\n",
      "\n",
      "There are 20583 unique barcodes in the raw data for second_infection wt fluNA\n",
      "3885 barcodes were corrected.\n",
      "There are 16698 corrected barcodes for second_infection wt fluNA\n",
      "\n",
      "There are 11147 unique barcodes in the raw data for supernatant syn fluHA\n",
      "3609 barcodes were corrected.\n",
      "There are 7538 corrected barcodes for supernatant syn fluHA\n",
      "\n",
      "There are 8884 unique barcodes in the raw data for supernatant syn fluNA\n",
      "3404 barcodes were corrected.\n",
      "There are 5480 corrected barcodes for supernatant syn fluNA\n",
      "\n",
      "There are 12856 unique barcodes in the raw data for supernatant wt fluHA\n",
      "3855 barcodes were corrected.\n",
      "There are 9001 corrected barcodes for supernatant wt fluHA\n",
      "\n",
      "There are 7048 unique barcodes in the raw data for supernatant wt fluNA\n",
      "3155 barcodes were corrected.\n",
      "There are 3893 corrected barcodes for supernatant wt fluNA\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>corrected_bc</th>\n",
       "      <th>original_bc</th>\n",
       "      <th>source</th>\n",
       "      <th>tag</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>130651</td>\n",
       "      <td>130651</td>\n",
       "      <td>130651</td>\n",
       "      <td>130651</td>\n",
       "      <td>130651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>71020</td>\n",
       "      <td>102870</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>TGTTATATTTGTATTG</td>\n",
       "      <td>ACTGTTCTAACAAACA</td>\n",
       "      <td>second_infection</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluHA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>108</td>\n",
       "      <td>6</td>\n",
       "      <td>90716</td>\n",
       "      <td>66467</td>\n",
       "      <td>71450</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            corrected_bc       original_bc            source     tag    gene\n",
       "count             130651            130651            130651  130651  130651\n",
       "unique             71020            102870                 2       2       2\n",
       "top     TGTTATATTTGTATTG  ACTGTTCTAACAAACA  second_infection      wt   fluHA\n",
       "freq                 108                 6             90716   66467   71450"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clusterer = UMIClusterer(cluster_method=\"directional\")\n",
    "\n",
    "lookup_list = []\n",
    "\n",
    "for (source, tag, gene), df in (viral_bc_df\n",
    "                                .groupby(['source',\n",
    "                                          'tag',\n",
    "                                          'gene'])):\n",
    "    n_unique_bc = df['barcode'].nunique()\n",
    "    print(f\"There are {n_unique_bc} unique barcodes \"\n",
    "          f\"in the raw data for {source} {tag} {gene}\")\n",
    "\n",
    "    # Convert dataframe to dictionary. Dict requried input for umi_tools.\n",
    "    viral_bc_dict = (df[['barcode', 'mean_freq']]\n",
    "                     .set_index('barcode')\n",
    "                     .to_dict(orient='dict'))\n",
    "    viral_bc_dict = viral_bc_dict['mean_freq']\n",
    "\n",
    "    # Convert barcode strings to byte. Byte required dtype for umi_tools.\n",
    "    byte_dict = {}\n",
    "    for key, value in viral_bc_dict.items():\n",
    "        byte_dict[key.encode(\"utf-8\")] = float(value)\n",
    "\n",
    "    # Cluster barcodes\n",
    "    bc_groups = clusterer(byte_dict, threshold=1)\n",
    "    groups_df = pd.DataFrame(bc_groups)\n",
    "    groups_df = (groups_df\n",
    "                 .stack()\n",
    "                 .str\n",
    "                 .decode('utf-8')\n",
    "                 .unstack())  # Convert bytes back to string\n",
    "    groups_df = groups_df.rename(columns={0: 'corrected_bc'})\n",
    "    groups_df = groups_df.set_index('corrected_bc', drop=False)\n",
    "    n_corrected_bc = groups_df['corrected_bc'].nunique()\n",
    "    print(f\"{n_unique_bc - n_corrected_bc} barcodes were corrected.\")\n",
    "    print(f\"There are {n_corrected_bc} corrected barcodes \"\n",
    "          f\"for {source} {tag} {gene}\\n\")\n",
    "\n",
    "    # Generate lookup table for this sample\n",
    "    temp_lookup_df = (groups_df.melt(ignore_index=False,\n",
    "                                     value_name='original_bc')\n",
    "                      ['original_bc']\n",
    "                      .dropna()\n",
    "                      .reset_index())\n",
    "    temp_lookup_df['source'] = source\n",
    "    temp_lookup_df['tag'] = tag\n",
    "    temp_lookup_df['gene'] = gene\n",
    "    lookup_list.append(temp_lookup_df)\n",
    "\n",
    "lookup_df = pd.concat(lookup_list)\n",
    "display(lookup_df.describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge corrected barcode data with barcode frequency data. Aggregate frequency on corrected barcodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>tag</th>\n",
       "      <th>gene</th>\n",
       "      <th>mean_freq_original_bc</th>\n",
       "      <th>corrected_bc</th>\n",
       "      <th>original_bc</th>\n",
       "      <th>mean_freq_corrected_bc</th>\n",
       "      <th>diff</th>\n",
       "      <th>diff_ratio</th>\n",
       "      <th>corrected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>AAAAAAAGTAAATCTT</td>\n",
       "      <td>AAAAAAAGTAAATCTT</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>AAAAAAATCTTAATGA</td>\n",
       "      <td>AAAAAAATCTTAATAA</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>42.688424</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>AAAAAAATCTTAATGA</td>\n",
       "      <td>AAAAAAATCTTAATGA</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.023426</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>AAAAAACCCGATTATT</td>\n",
       "      <td>AAAAAACCCAATTATT</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>32.124979</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>AAAAAACCCGATTATT</td>\n",
       "      <td>AAAAAACCCGATTATT</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.031128</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130646</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>TTTTGTCATGATAGCC</td>\n",
       "      <td>TTTTGTCATGATAGCC</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130647</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>TTTTGTCATGATAGCC</td>\n",
       "      <td>TTTTGTCATGATAGCT</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>156.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130648</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>TTTTGTTAGCGTCCTG</td>\n",
       "      <td>TTTTGTTAGCGTCATG</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.000307</td>\n",
       "      <td>261.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130649</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000307</td>\n",
       "      <td>TTTTGTTAGCGTCCTG</td>\n",
       "      <td>TTTTGTTAGCGTCCTG</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.003831</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130650</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>TTTTTTCGGCGAGTGC</td>\n",
       "      <td>TTTTTTCGGCGAGTGC</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130651 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  source  tag   gene  mean_freq_original_bc      corrected_bc  \\\n",
       "0       second_infection  syn  fluHA               0.000053  AAAAAAAGTAAATCTT   \n",
       "1       second_infection  syn  fluHA               0.000002  AAAAAAATCTTAATGA   \n",
       "2       second_infection  syn  fluHA               0.000087  AAAAAAATCTTAATGA   \n",
       "3       second_infection  syn  fluHA               0.000001  AAAAAACCCGATTATT   \n",
       "4       second_infection  syn  fluHA               0.000034  AAAAAACCCGATTATT   \n",
       "...                  ...  ...    ...                    ...               ...   \n",
       "130646       supernatant   wt  fluNA               0.000183  TTTTGTCATGATAGCC   \n",
       "130647       supernatant   wt  fluNA               0.000001  TTTTGTCATGATAGCC   \n",
       "130648       supernatant   wt  fluNA               0.000001  TTTTGTTAGCGTCCTG   \n",
       "130649       supernatant   wt  fluNA               0.000307  TTTTGTTAGCGTCCTG   \n",
       "130650       supernatant   wt  fluNA               0.000120  TTTTTTCGGCGAGTGC   \n",
       "\n",
       "             original_bc  mean_freq_corrected_bc      diff  diff_ratio  \\\n",
       "0       AAAAAAAGTAAATCTT                0.000053  0.000000    0.000000   \n",
       "1       AAAAAAATCTTAATAA                0.000089  0.000087   42.688424   \n",
       "2       AAAAAAATCTTAATGA                0.000089  0.000002    0.023426   \n",
       "3       AAAAAACCCAATTATT                0.000035  0.000034   32.124979   \n",
       "4       AAAAAACCCGATTATT                0.000035  0.000001    0.031128   \n",
       "...                  ...                     ...       ...         ...   \n",
       "130646  TTTTGTCATGATAGCC                0.000185  0.000001    0.006410   \n",
       "130647  TTTTGTCATGATAGCT                0.000185  0.000183  156.000000   \n",
       "130648  TTTTGTTAGCGTCATG                0.000308  0.000307  261.000000   \n",
       "130649  TTTTGTTAGCGTCCTG                0.000308  0.000001    0.003831   \n",
       "130650  TTTTTTCGGCGAGTGC                0.000120  0.000000    0.000000   \n",
       "\n",
       "        corrected  \n",
       "0           False  \n",
       "1            True  \n",
       "2           False  \n",
       "3            True  \n",
       "4           False  \n",
       "...           ...  \n",
       "130646      False  \n",
       "130647       True  \n",
       "130648       True  \n",
       "130649      False  \n",
       "130650      False  \n",
       "\n",
       "[130651 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_bc_df = pd.merge(viral_bc_df,\n",
    "                       lookup_df,\n",
    "                       left_on=['source', 'tag', 'gene', 'barcode'],\n",
    "                       right_on=['source', 'tag', 'gene', 'original_bc'])\n",
    "viral_bc_df = viral_bc_df.drop('barcode', axis=1)\n",
    "\n",
    "viral_bc_df = pd.merge(viral_bc_df,\n",
    "                       (viral_bc_df\n",
    "                        .groupby(['source',\n",
    "                                  'tag',\n",
    "                                  'gene',\n",
    "                                  'corrected_bc'])\n",
    "                        .sum()\n",
    "                        .reset_index()),\n",
    "                       on=['source', 'tag', 'gene', 'corrected_bc'],\n",
    "                       suffixes=['_original_bc', '_corrected_bc'])\n",
    "\n",
    "viral_bc_df['diff'] = (viral_bc_df['mean_freq_corrected_bc']\n",
    "                       - viral_bc_df['mean_freq_original_bc'])\n",
    "viral_bc_df['diff_ratio'] = (viral_bc_df['diff']\n",
    "                             / viral_bc_df['mean_freq_original_bc'])\n",
    "viral_bc_df['corrected'] = (viral_bc_df['corrected_bc']\n",
    "                            != viral_bc_df['original_bc'])\n",
    "display(viral_bc_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot fraction of reads that underwent correction:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'expt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-29-12a793e197ad>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m                                      fill='corrected')) +\n\u001b[1;32m      4\u001b[0m        \u001b[0mp9\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgeom_bar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstat\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'identity'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m        p9.ggtitle('Barcode correction rates '\n\u001b[0m\u001b[1;32m      6\u001b[0m                   f'for experiment {expt}') +\n\u001b[1;32m      7\u001b[0m        \u001b[0mp9\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mylab\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'frac reads corrected'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'expt' is not defined"
     ]
    }
   ],
   "source": [
    "fig = (p9.ggplot(viral_bc_df, p9.aes(x='corrected',\n",
    "                                     y='mean_freq_original_bc',\n",
    "                                     fill='corrected')) +\n",
    "       p9.geom_bar(stat='identity') +\n",
    "       p9.ggtitle('Barcode correction rates '\n",
    "                  f'for experiment {expt}') +\n",
    "       p9.ylab('frac reads corrected') +\n",
    "       p9.facet_grid('source~tag+gene') +\n",
    "       p9.scale_x_discrete(limits=(viral_bc_df['corrected']\n",
    "                                   .unique()\n",
    "                                   .tolist())) +  # Order categorical values\n",
    "       p9.theme(figure_size=(1.5*((viral_bc_df['gene'].nunique() *\n",
    "                                   viral_bc_df['tag'].nunique())),\n",
    "                             1.6*viral_bc_df['source'].nunique()),\n",
    "                axis_text_x=p9.element_text(angle=90),\n",
    "                plot_title=p9.element_text(size=10),\n",
    "                axis_title=p9.element_text(size=10),\n",
    "                legend_title=p9.element_text(size=9),\n",
    "                legend_title_align='center') +\n",
    "       p9.scale_fill_manual(CBPALETTE[0:]))\n",
    "\n",
    "# save plot\n",
    "print(f\"Saving plot to {plot}\")\n",
    "p9.ggsave(plot=fig, filename=plot, verbose=False)\n",
    "\n",
    "# show plot\n",
    "fig.draw()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For (second_infection, syn, fluHA) sample, 0.70% of reads were corrected.\n",
      "For (second_infection, syn, fluNA) sample, 0.51% of reads were corrected.\n",
      "For (second_infection, wt, fluHA) sample, 0.62% of reads were corrected.\n",
      "For (second_infection, wt, fluNA) sample, 0.72% of reads were corrected.\n",
      "For (supernatant, syn, fluHA) sample, 0.88% of reads were corrected.\n",
      "For (supernatant, syn, fluNA) sample, 0.77% of reads were corrected.\n",
      "For (supernatant, wt, fluHA) sample, 1.19% of reads were corrected.\n",
      "For (supernatant, wt, fluNA) sample, 0.98% of reads were corrected.\n"
     ]
    }
   ],
   "source": [
    "for (source, tag, gene), df in (viral_bc_df\n",
    "                                .groupby(['source',\n",
    "                                          'tag',\n",
    "                                          'gene'])):\n",
    "    frac_corrected_df = (df.groupby(['corrected'])\n",
    "                         .sum()\n",
    "                         .reset_index())\n",
    "    frac_corrected = (frac_corrected_df\n",
    "                      .query(\"corrected == True\")\n",
    "                      [\"mean_freq_original_bc\"])\n",
    "\n",
    "    print(f'For ({source}, {tag}, {gene}) sample, '\n",
    "          f'{float(frac_corrected) * 100:.2f}'\n",
    "          f'% of reads were corrected.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot effect of correction for each barcode:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'expt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-28-a349ef2932b0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m                                      color='corrected')) +\n\u001b[1;32m      4\u001b[0m        \u001b[0mp9\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgeom_point\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malpha\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.2\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m        p9.ggtitle(f'Barcode frequency changes during correction\\n'\n\u001b[0m\u001b[1;32m      6\u001b[0m                   f'for experiment {expt}') +\n\u001b[1;32m      7\u001b[0m        \u001b[0mp9\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfacet_grid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'gene+tag~source'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'expt' is not defined"
     ]
    }
   ],
   "source": [
    "fig = (p9.ggplot(viral_bc_df, p9.aes(x='mean_freq_original_bc',\n",
    "                                     y='diff_ratio',\n",
    "                                     color='corrected')) +\n",
    "       p9.geom_point(alpha=0.2) +\n",
    "       p9.ggtitle(f'Barcode frequency changes during correction\\n'\n",
    "                  f'for experiment {expt}') +\n",
    "       p9.facet_grid('gene+tag~source') +\n",
    "       p9.scale_x_log10() +\n",
    "       p9.scale_y_log10() +\n",
    "       p9.theme(figure_size=(4 * viral_bc_df['source'].nunique(),\n",
    "                             1.5 * ((viral_bc_df['gene'].nunique()\n",
    "                                     * viral_bc_df['tag'].nunique()\n",
    "                                     ))),\n",
    "                axis_text_x=p9.element_text(angle=90),\n",
    "                plot_title=p9.element_text(size=10),\n",
    "                axis_title=p9.element_text(size=10),\n",
    "                legend_title=p9.element_text(size=9),\n",
    "                legend_title_align='center') +\n",
    "       p9.scale_fill_manual(CBPALETTE[0:]))\n",
    "\n",
    "fig.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export CSV of corrected barcode frequencies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>tag</th>\n",
       "      <th>gene</th>\n",
       "      <th>corrected_bc</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAAAGTAAATCTT</td>\n",
       "      <td>0.000053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAAATCTTAATGA</td>\n",
       "      <td>0.000089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAACCCGATTATT</td>\n",
       "      <td>0.000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAACGAATGGATT</td>\n",
       "      <td>0.000253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>second_infection</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAAAAAGGGGTGTATG</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130640</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTGATGGTAAGTAC</td>\n",
       "      <td>0.000223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130642</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTGCTTCGAGATAG</td>\n",
       "      <td>0.001143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130646</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTGTCATGATAGCC</td>\n",
       "      <td>0.000185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130648</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTGTTAGCGTCCTG</td>\n",
       "      <td>0.000308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130650</th>\n",
       "      <td>supernatant</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTTTTTCGGCGAGTGC</td>\n",
       "      <td>0.000120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97964 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  source  tag   gene      corrected_bc      freq\n",
       "0       second_infection  syn  fluHA  AAAAAAAGTAAATCTT  0.000053\n",
       "1       second_infection  syn  fluHA  AAAAAAATCTTAATGA  0.000089\n",
       "3       second_infection  syn  fluHA  AAAAAACCCGATTATT  0.000035\n",
       "5       second_infection  syn  fluHA  AAAAAACGAATGGATT  0.000253\n",
       "8       second_infection  syn  fluHA  AAAAAAGGGGTGTATG  0.000005\n",
       "...                  ...  ...    ...               ...       ...\n",
       "130640       supernatant   wt  fluNA  TTTTGATGGTAAGTAC  0.000223\n",
       "130642       supernatant   wt  fluNA  TTTTGCTTCGAGATAG  0.001143\n",
       "130646       supernatant   wt  fluNA  TTTTGTCATGATAGCC  0.000185\n",
       "130648       supernatant   wt  fluNA  TTTTGTTAGCGTCCTG  0.000308\n",
       "130650       supernatant   wt  fluNA  TTTTTTCGGCGAGTGC  0.000120\n",
       "\n",
       "[97964 rows x 5 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_bc_df = (viral_bc_df[['source',\n",
    "                            'tag',\n",
    "                            'gene',\n",
    "                            'corrected_bc',\n",
    "                            'mean_freq_corrected_bc']]\n",
    "               .drop_duplicates()\n",
    "               .rename(columns={'mean_freq_corrected_bc':\n",
    "                                'freq'}))\n",
    "\n",
    "\n",
    "viral_bc_df.to_csv(viral_bc_in_progeny_corrected,\n",
    "                   columns=['source',\n",
    "                            'tag',\n",
    "                            'gene',\n",
    "                            'corrected_bc',\n",
    "                            'freq'],\n",
    "                   index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
