{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:51.557425Z",
     "iopub.status.busy": "2022-05-25T18:48:51.556831Z",
     "iopub.status.idle": "2022-05-25T18:48:51.682705Z",
     "shell.execute_reply": "2022-05-25T18:48:51.683744Z"
    },
    "tags": [
     "snakemake-job-properties"
    ]
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/fh/fast/bloom_j/software/miniconda3/envs/barcoded_flu_pdmH1N1/lib/python3.8/site-packages', '/fh/fast/bloom_j/computational_notebooks/dbacsik/2019/flu_pdmH1N1_barcode_hashing/notebooks']); import pickle; snakemake = pickle.loads(b'\\x80\\x04\\x95\\x86 \\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c_results/viral_tags_bcs_in_cells/scProgenyProduction_trial3_cell_barcodes_with_viral_tags.csv.gz\\x94\\x8cLresults/viral_fastq10x/scProgenyProduction_trial3_viral_genes_by_cell.csv.gz\\x94\\x8c?results/pacbio/scProgenyProduction_trial3_consensus_gene.csv.gz\\x94\\x8cOresults/viral_fastq10x/scProgenyProduction_trial3_viral_bc_by_cell_valid.csv.gz\\x94\\x8cQresults/viral_progeny/scProgenyProduction_trial3_filtered_progeny_viral_bc.csv.gz\\x94\\x8cTresults/viral_fastq10x/scProgenyProduction_trial3_contributes_progeny_by_cell.csv.gz\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\x10cell_annotations\\x94K\\x00N\\x86\\x94\\x8c\\x17viral_genes_by_cell_csv\\x94K\\x01N\\x86\\x94\\x8c\\x19pacbio_consensus_gene_csv\\x94K\\x02N\\x86\\x94\\x8c\\x18viral_barcodes_valid_csv\\x94K\\x03N\\x86\\x94\\x8c\\x1dfiltered_progeny_viral_bc_csv\\x94K\\x04N\\x86\\x94\\x8c\\x1fcontributes_progeny_by_cell_csv\\x94K\\x05N\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94eh!\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(h\\')}\\x94\\x8c\\x05_name\\x94h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bh\\x13h\\nh\\x15h\\x0bh\\x17h\\x0ch\\x19h\\rh\\x1bh\\x0eh\\x1dh\\x0fub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94\\x8cDresults/viral_fastq10x/scProgenyProduction_trial3_integrate_data.csv\\x94a}\\x94(h\\x11}\\x94\\x8c\\x13integrated_data_csv\\x94K\\x00N\\x86\\x94sh\\x1f]\\x94(h!h\"eh!h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bh8h5ub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94]\\x94(\\x8c\\x05fluHA\\x94\\x8c\\x05fluNA\\x94ea}\\x94(h\\x11}\\x94\\x8c\\x14barcoded_viral_genes\\x94K\\x00N\\x86\\x94sh\\x1f]\\x94(h!h\"eh!h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bhLhGub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94\\x8c\\x1ascProgenyProduction_trial3\\x94a}\\x94(h\\x11}\\x94\\x8c\\x04expt\\x94K\\x00N\\x86\\x94sh\\x1f]\\x94(h!h\"eh!h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94b\\x8c\\x04expt\\x94h[ub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01e}\\x94(h\\x11}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94uh\\x1f]\\x94(h!h\"eh!h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bhqK\\x01hsK\\x01ub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8c<results/logs/integrate_data_scProgenyProduction_trial3.ipynb\\x94a}\\x94(h\\x11}\\x94\\x8c\\x08notebook\\x94K\\x00N\\x86\\x94sh\\x1f]\\x94(h!h\"eh!h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h!sNt\\x94bh\"h%h\\'\\x85\\x94R\\x94(h\\')}\\x94h+h\"sNt\\x94bh\\x85h\\x82ub\\x8c\\x06config\\x94}\\x94(\\x8c\\x08max_cpus\\x94K\\x10\\x8c\\x07log_dir\\x94\\x8c\\x0cresults/logs\\x94\\x8c\\x0cfastq10x_dir\\x94\\x8c\\x10results/fastq10x\\x94\\x8c\\npacbio_dir\\x94\\x8c\\x0eresults/pacbio\\x94\\x8c\\x0emkfastq10x_dir\\x94\\x8c\\x1fresults/fastq10x/mkfastq_output\\x94\\x8c\\ngenome_dir\\x94\\x8c\\x0fresults/genomes\\x94\\x8c\\trefgenome\\x94\\x8c\\x19results/genomes/refgenome\\x94\\x8c\\x14aligned_fastq10x_dir\\x94\\x8c\\x18results/aligned_fastq10x\\x94\\x8c\\x12viral_fastq10x_dir\\x94\\x8c\\x16results/viral_fastq10x\\x94\\x8c\\x1bviral_tags_bcs_in_cells_dir\\x94\\x8c\\x1fresults/viral_tags_bcs_in_cells\\x94\\x8c\\x11viral_progeny_dir\\x94\\x8c\\x15results/viral_progeny\\x94\\x8c\\x13align_and_parse_dir\\x94\\x8c#results/pacbio/CA09_align_and_parse\\x94\\x8c\\x0fcell_genome_ftp\\x94\\x8cmftp://ftp.ensembl.org/pub/release-98/fasta/canis_familiaris/dna/Canis_familiaris.CanFam3.1.dna.toplevel.fa.gz\\x94\\x8c\\x0ccell_gtf_ftp\\x94\\x8c^ftp://ftp.ensembl.org/pub/release-98/gtf/canis_familiaris/Canis_familiaris.CanFam3.1.98.gtf.gz\\x94\\x8c\\x0cviral_genome\\x94\\x8c!data/flu_sequences/flu-CA09.fasta\\x94\\x8c\\tviral_gtf\\x94\\x8c\\x1fdata/flu_sequences/flu-CA09.gtf\\x94\\x8c\\rviral_genbank\\x94\\x8c\\x1edata/flu_sequences/flu-CA09.gb\\x94\\x8c\\x14viral_tag_identities\\x94\\x8c+data/flu_sequences/flu-CA09_viral_tags.yaml\\x94\\x8c\\x0fviral_amplicons\\x94\\x8c.data/flu_sequences/pacbio_amplicons/fluCA09.gb\\x94\\x8c\\x11amplicon_features\\x94\\x8c9data/flu_sequences/pacbio_amplicons/fluCA09_features.yaml\\x94\\x8c\\x1eamplicon_linearization_primers\\x94\\x8c=data/flu_sequences/pacbio_amplicons/linearization_primers.tsv\\x94\\x8c\\x15amplicon_to_reference\\x94\\x8c=data/flu_sequences/pacbio_amplicons/amplicon_to_reference.csv\\x94\\x8c\\x0escoreGapNoncan\\x94J\\xfc\\xff\\xff\\xff\\x8c\\x0cscoreGapGCAG\\x94J\\xfc\\xff\\xff\\xff\\x8c\\x0cscoreGapATAC\\x94J\\xfc\\xff\\xff\\xff\\x8c\\x14cb_whitelist_10x_url\\x94\\x8cvhttps://raw.githubusercontent.com/10XGenomics/cellranger/master/lib/python/cellranger/barcodes/3M-february-2018.txt.gz\\x94\\x8c\\x10cb_whitelist_10x\\x94\\x8c-results/aligned_fastq10x/cb_whitelist_10x.txt\\x94\\x8c\\ncb_len_10x\\x94K\\x10\\x8c\\x0bumi_len_10x\\x94K\\x0c\\x8c\\x14total_UMI_deviations\\x94G@\\x04\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x13infection_threshold\\x94G?\\x84z\\xe1G\\xae\\x14{\\x8c\\rviral_tag_fdr\\x94G?\\x1a6\\xe2\\xeb\\x1cC-\\x8c\\x1dviral_barcode_upstream_length\\x94K\\x1c\\x8c\\x16viral_barcode_mismatch\\x94K\\x00\\x8c\\x12viral_barcode_minq\\x94K\\x1e\\x8c\\x17progeny_detection_limit\\x94\\x8c\\x041e-5\\x94\\x8c\\x0cviral_bc_fdr\\x94G?\\x1a6\\xe2\\xeb\\x1cC-\\x8c\\x0eccs_min_length\\x94K2\\x8c\\x0eccs_max_length\\x94M\\x88\\x13\\x8c\\nccs_min_rq\\x94G?\\xef\\xf7\\xce\\xd9\\x16\\x87+\\x8c\\x0bexperiments\\x94}\\x94\\x8c\\x1ascProgenyProduction_trial3\\x94}\\x94(\\x8c\\x0bdescription\\x94X\\x1b\\x01\\x00\\x00Single-cell transcriptomics performed on even lower MOI-infected sample. Innoculum volumes based on `scProgenyProduction_trial1` and `scProgenyProduction_trial2` data. Loaded all collected cells from two 24-well wells, each infected with either wildtype or dblySn viral tag variants.\\x94\\x8c\\tlab_notes\\x94\\x8c0https://benchling.com/s/etr-N4laiIAy3AUePJdig7wR\\x94\\x8c\\rexpect_ncells\\x94M\\x10\\'\\x8c\\x0ftranscriptomics\\x94}\\x94(\\x8c\\x08datetime\\x94\\x8c\\x04date\\x94\\x93\\x94C\\x04\\x07\\xe5\\x03\\x0c\\x94\\x85\\x94R\\x94}\\x94(\\x8c\\x05index\\x94\\x8c\\x08SI-TT-A3\\x94\\x8c\\nbcl_folder\\x94\\x8c_/shared/ngs/illumina/bloom_lab/210312_D00300_1192_BHKHNYBCX3/raw/210312_D00300_1192_BHKHNYBCX3/\\x94\\x8c\\x04lane\\x94K\\x01\\x8c\\x10index_sequencing\\x94\\x8c\\x06single\\x94uh\\xdfC\\x04\\x07\\xe5\\x04\\t\\x94\\x85\\x94R\\x94}\\x94(\\x8c\\x05index\\x94\\x8c\\x08SI-TT-A3\\x94\\x8c\\nbcl_folder\\x94\\x8c_/shared/ngs/illumina/bloom_lab/210409_D00300_1210_BHKC7KBCX3/raw/210409_D00300_1210_BHKC7KBCX3/\\x94\\x8c\\x04lane\\x94\\x8c\\x01*\\x94\\x8c\\x10index_sequencing\\x94\\x8c\\x06single\\x94uh\\xdfC\\x04\\x07\\xe5\\x08\\x19\\x94\\x85\\x94R\\x94}\\x94(\\x8c\\x05index\\x94\\x8c\\x08SI-TT-A3\\x94\\x8c\\nbcl_folder\\x94\\x8c_/shared/ngs/illumina/bloom_lab/210825_D00300_1310_BHMH5TBCX3/raw/210825_D00300_1310_BHMH5TBCX3/\\x94\\x8c\\x04lane\\x94h\\xf4\\x8c\\x10index_sequencing\\x94\\x8c\\x04none\\x94uh\\xdfC\\x04\\x07\\xe5\\t\\x0f\\x94\\x85\\x94R\\x94}\\x94(\\x8c\\x05index\\x94\\x8c\\x08SI-TT-A3\\x94\\x8c\\nbcl_folder\\x94\\x8cY/shared/ngs/illumina/bloom_lab/210915_VH00699_2_AAAMCL7M5/raw/210915_VH00699_2_AAAMCL7M5/\\x94\\x8c\\x04lane\\x94h\\xf4\\x8c\\x10index_sequencing\\x94\\x8c\\x04none\\x94uu\\x8c\\x17pacbio_viral_sequencing\\x94}\\x94(\\x8c\\x1c2021-08-05_all_segments_run1\\x94\\x8cp/fh/fast/bloom_j/SR/ngs/pacbio/210730_SequelIIe/r64272e_20210730_182729/1_A01/m64272e_210730_193026.subreads.bam\\x94\\x8c\\x1c2021-09-17_all_segments_run2\\x94\\x8cp/fh/fast/bloom_j/SR/ngs/pacbio/210917_SequelIIe/r64272e_20210917_174421/1_A01/m64272e_210917_175514.subreads.bam\\x94u\\x8c\\x0eviral_barcodes\\x94}\\x94(\\x8c\\x0bsupernatant\\x94}\\x94(\\x8c\\x02wt\\x94}\\x94(\\x8c\\x05fluHA\\x94}\\x94(\\x8c\\x0breplicate_1\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x04\\x08\\x94\\x85\\x94R\\x94\\x8c\\x90/shared/ngs/illumina/bloom_lab/210408_M04866_0464_000000000-JL5M4_new-demux/Data/Intensities/BaseCalls/WT-sup-fluHA-rep1_S9_L001_R1_001.fastq.gz\\x94s\\x8c\\x0breplicate_2\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x04\\x08\\x94\\x85\\x94R\\x94\\x8c\\x91/shared/ngs/illumina/bloom_lab/210408_M04866_0464_000000000-JL5M4_new-demux/Data/Intensities/BaseCalls/WT-sup-fluHA-rep2_S13_L001_R1_001.fastq.gz\\x94su\\x8c\\x05fluNA\\x94}\\x94(\\x8c\\x0breplicate_1\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x04\\x08\\x94\\x85\\x94R\\x94\\x8c\\x90/shared/ngs/illumina/bloom_lab/210408_M04866_0464_000000000-JL5M4_new-demux/Data/Intensities/BaseCalls/WT-sup-fluNA-rep1_S1_L001_R1_001.fastq.gz\\x94s\\x8c\\x0breplicate_2\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x05\\x11\\x94\\x85\\x94R\\x94\\x8c\\x86/shared/ngs/illumina/bloom_lab/210517_M03100_0681_000000000-JL4TM/Data/Intensities/BaseCalls/WT-sup-fluNA-rep2_S1_L001_R1_001.fastq.gz\\x94suu\\x8c\\x03syn\\x94}\\x94(\\x8c\\x05fluHA\\x94}\\x94(\\x8c\\x0breplicate_1\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x04\\x08\\x94\\x85\\x94R\\x94\\x8c\\x95/shared/ngs/illumina/bloom_lab/210408_M04866_0464_000000000-JL5M4_new-demux/Data/Intensities/BaseCalls/dblSyn-sup-fluHA-rep1_S10_L001_R1_001.fastq.gz\\x94s\\x8c\\x0breplicate_2\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x04\\x08\\x94\\x85\\x94R\\x94\\x8c\\x95/shared/ngs/illumina/bloom_lab/210408_M04866_0464_000000000-JL5M4_new-demux/Data/Intensities/BaseCalls/dblSyn-sup-fluHA-rep2_S14_L001_R1_001.fastq.gz\\x94su\\x8c\\x05fluNA\\x94}\\x94(\\x8c\\x0breplicate_1\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x04\\x08\\x94\\x85\\x94R\\x94\\x8c\\x94/shared/ngs/illumina/bloom_lab/210408_M04866_0464_000000000-JL5M4_new-demux/Data/Intensities/BaseCalls/dblSyn-sup-fluNA-rep1_S2_L001_R1_001.fastq.gz\\x94s\\x8c\\x0breplicate_2\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x04\\x08\\x94\\x85\\x94R\\x94\\x8c\\x94/shared/ngs/illumina/bloom_lab/210408_M04866_0464_000000000-JL5M4_new-demux/Data/Intensities/BaseCalls/dblSyn-sup-fluNA-rep2_S6_L001_R1_001.fastq.gz\\x94suuu\\x8c\\x10second_infection\\x94}\\x94(\\x8c\\x02wt\\x94}\\x94(\\x8c\\x05fluHA\\x94}\\x94(\\x8c\\x0breplicate_1\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x05\\x11\\x94\\x85\\x94R\\x94\\x8c\\x85/shared/ngs/illumina/bloom_lab/210517_M03100_0681_000000000-JL4TM/Data/Intensities/BaseCalls/WT-SI-fluHA-rep1_S2_L001_R1_001.fastq.gz\\x94s\\x8c\\x0breplicate_2\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x05\\x11\\x94\\x85\\x94R\\x94\\x8c\\x85/shared/ngs/illumina/bloom_lab/210517_M03100_0681_000000000-JL4TM/Data/Intensities/BaseCalls/WT-SI-fluHA-rep2_S3_L001_R1_001.fastq.gz\\x94su\\x8c\\x05fluNA\\x94}\\x94(\\x8c\\x0breplicate_1\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x05\\x11\\x94\\x85\\x94R\\x94\\x8c\\x86/shared/ngs/illumina/bloom_lab/210517_M03100_0681_000000000-JL4TM/Data/Intensities/BaseCalls/WT-SI-fluNA-rep1_S11_L001_R1_001.fastq.gz\\x94s\\x8c\\x0breplicate_2\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x05\\x11\\x94\\x85\\x94R\\x94\\x8c\\x86/shared/ngs/illumina/bloom_lab/210517_M03100_0681_000000000-JL4TM/Data/Intensities/BaseCalls/WT-SI-fluNA-rep2_S12_L001_R1_001.fastq.gz\\x94suu\\x8c\\x03syn\\x94}\\x94(\\x8c\\x05fluHA\\x94}\\x94(\\x8c\\x0breplicate_1\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x05\\x11\\x94\\x85\\x94R\\x94\\x8c\\x89/shared/ngs/illumina/bloom_lab/210517_M03100_0681_000000000-JL4TM/Data/Intensities/BaseCalls/dblSyn-SI-fluHA-rep1_S4_L001_R1_001.fastq.gz\\x94s\\x8c\\x0breplicate_2\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x05\\x11\\x94\\x85\\x94R\\x94\\x8c\\x89/shared/ngs/illumina/bloom_lab/210517_M03100_0681_000000000-JL4TM/Data/Intensities/BaseCalls/dblSyn-SI-fluHA-rep2_S5_L001_R1_001.fastq.gz\\x94su\\x8c\\x05fluNA\\x94}\\x94(\\x8c\\x0breplicate_1\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x05\\x11\\x94\\x85\\x94R\\x94\\x8c\\x8a/shared/ngs/illumina/bloom_lab/210517_M03100_0681_000000000-JL4TM/Data/Intensities/BaseCalls/dblSyn-SI-fluNA-rep1_S13_L001_R1_001.fastq.gz\\x94s\\x8c\\x0breplicate_2\\x94}\\x94h\\xdfC\\x04\\x07\\xe5\\x05\\x11\\x94\\x85\\x94R\\x94\\x8c\\x8a/shared/ngs/illumina/bloom_lab/210517_M03100_0681_000000000-JL4TM/Data/Intensities/BaseCalls/dblSyn-SI-fluNA-rep2_S14_L001_R1_001.fastq.gz\\x94suuuuusu\\x8c\\x04rule\\x94\\x8c\\x0eintegrate_data\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8c[/fh/fast/bloom_j/computational_notebooks/dbacsik/2019/flu_pdmH1N1_barcode_hashing/notebooks\\x94ub.'); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir('/fh/fast/bloom_j/computational_notebooks/dbacsik/2019/flu_pdmH1N1_barcode_hashing');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Integrate data\n",
    "This notebook takes processed data and integrates it into a single dataframe.\n",
    "\n",
    "The general structure of the dataframe is that each cell barcode is listed on a row, and features of that cell are listed in columns. Cells that have more than one valid viral barcode identified may have multiple rows--one for each valid viral barcode."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import python modules:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:51.687637Z",
     "iopub.status.busy": "2022-05-25T18:48:51.686639Z",
     "iopub.status.idle": "2022-05-25T18:48:52.202464Z",
     "shell.execute_reply": "2022-05-25T18:48:52.203284Z"
    }
   },
   "outputs": [],
   "source": [
    "from IPython.display import display\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Variables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.207168Z",
     "iopub.status.busy": "2022-05-25T18:48:52.206024Z",
     "iopub.status.idle": "2022-05-25T18:48:52.211343Z",
     "shell.execute_reply": "2022-05-25T18:48:52.211857Z"
    }
   },
   "outputs": [],
   "source": [
    "# Input sources\n",
    "cell_annotations_csv = snakemake.input.cell_annotations\n",
    "viral_genes_by_cell_csv = snakemake.input.viral_genes_by_cell_csv\n",
    "pacbio_consensus_gene_csv = snakemake.input.pacbio_consensus_gene_csv\n",
    "viral_barcodes_valid_csv = snakemake.input.viral_barcodes_valid_csv\n",
    "filtered_progeny_viral_bc_csv = snakemake.input.filtered_progeny_viral_bc_csv\n",
    "contributes_progeny_by_cell_csv = snakemake.input.contributes_progeny_by_cell_csv\n",
    "\n",
    "# # Params and wildcards\n",
    "barcoded_viral_genes = snakemake.params.barcoded_viral_genes\n",
    "expt = snakemake.wildcards.expt\n",
    "\n",
    "# Output\n",
    "integrated_data_csv = snakemake.output.integrated_data_csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data and transform into wide format\n",
    "Load data from the following sources:  \n",
    "* `cell_annotations_csv` contains a list of all cell barcodes, their infection status/tag, total_UMIs, viral_UMIs, and frac_viral_UMIs  \n",
    "\n",
    "* `viral_genes_by_cell_csv` contains every cell barcode, each viral gene, whether the viral gene is present, the fraction of UMIs from that gene, and the total number of viral genes detected in that cell  \n",
    "\n",
    "* `viral_barcodes_valid_csv` contains each infected cell, the valid viral barcodes detected in that cell, viral_bc_UMIs, and frac_viral_bc_UMIs  \n",
    "\n",
    "* `filtered_progeny_viral_bc_csv` contains the progeny source, the tag of the progeny source, the valid viral barcodes detected in that progeny source, and the frequency of the viral barcodes in that progeny source  \n",
    "\n",
    "* `pacbio_consensus_gene_csv` contains the cell barcode, each viral gene, and the consensus genotype of the viral gene in that cell\n",
    "\n",
    "Some of these dataframes are in a long format, with many rows per cell barcode. I will transform them into a wide format, so that each cell barcode has only one row. The only exception is the viral barcodes, which will have one row per cell barcode-viral barcode pair."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load cell barcodes, tags, and basic metrics for every cell in transcriptome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.214468Z",
     "iopub.status.busy": "2022-05-25T18:48:52.213665Z",
     "iopub.status.idle": "2022-05-25T18:48:52.252206Z",
     "shell.execute_reply": "2022-05-25T18:48:52.252778Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAGTAACAAGT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>47873</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCATCATTGCTT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>90114</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAAGATGTTGA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>111630</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>infected</td>\n",
       "      <td>both</td>\n",
       "      <td>56828</td>\n",
       "      <td>24082</td>\n",
       "      <td>0.423770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>124341</td>\n",
       "      <td>4654</td>\n",
       "      <td>0.037429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3124</th>\n",
       "      <td>TTTGATCTCAGTCATG</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>44890</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3125</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>63150</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3126</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>65941</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3127</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3128</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>135009</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000059</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3129 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode    infected infecting_viral_tag  total_UMIs  \\\n",
       "0     AAACCCAGTAACAAGT  uninfected                none       47873   \n",
       "1     AAACCCATCATTGCTT  uninfected                none       90114   \n",
       "2     AAACGAAAGATGTTGA  uninfected                none      111630   \n",
       "3     AAACGAAGTACTTCCC    infected                both       56828   \n",
       "4     AAACGAAGTAGACGTG    infected                  wt      124341   \n",
       "...                ...         ...                 ...         ...   \n",
       "3124  TTTGATCTCAGTCATG  uninfected                none       44890   \n",
       "3125  TTTGATCTCCCGTTCA  uninfected                none       63150   \n",
       "3126  TTTGGAGAGTTGCCTA  uninfected                none       65941   \n",
       "3127  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "3128  TTTGTTGAGTGCAGCA  uninfected                none      135009   \n",
       "\n",
       "      viral_UMIs  frac_viral_UMIs  \n",
       "0              6         0.000125  \n",
       "1             10         0.000111  \n",
       "2             18         0.000161  \n",
       "3          24082         0.423770  \n",
       "4           4654         0.037429  \n",
       "...          ...              ...  \n",
       "3124           4         0.000089  \n",
       "3125           3         0.000048  \n",
       "3126          12         0.000182  \n",
       "3127        3526         0.023486  \n",
       "3128           8         0.000059  \n",
       "\n",
       "[3129 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cell_annotations = pd.read_csv(cell_annotations_csv)\n",
    "cell_annotations = (\n",
    "    cell_annotations\n",
    "    [['cell_barcode',\n",
    "      'infected',\n",
    "      'infecting_viral_tag',\n",
    "      'total_UMIs',\n",
    "      'viral_UMIs',\n",
    "      'frac_viral_UMIs']])\n",
    "\n",
    "display(cell_annotations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load viral genes detected in each cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.255685Z",
     "iopub.status.busy": "2022-05-25T18:48:52.254805Z",
     "iopub.status.idle": "2022-05-25T18:48:52.375961Z",
     "shell.execute_reply": "2022-05-25T18:48:52.376479Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>n_viral_genes</th>\n",
       "      <th>transcriptome_UMIs_fluHA</th>\n",
       "      <th>transcriptome_UMIs_fluM</th>\n",
       "      <th>transcriptome_UMIs_fluNA</th>\n",
       "      <th>transcriptome_UMIs_fluNP</th>\n",
       "      <th>transcriptome_UMIs_fluNS</th>\n",
       "      <th>transcriptome_UMIs_fluPA</th>\n",
       "      <th>transcriptome_UMIs_fluPB1</th>\n",
       "      <th>transcriptome_UMIs_fluPB2</th>\n",
       "      <th>...</th>\n",
       "      <th>frac_UMIs_fluPB1</th>\n",
       "      <th>frac_UMIs_fluPB2</th>\n",
       "      <th>present_fluHA</th>\n",
       "      <th>present_fluM</th>\n",
       "      <th>present_fluNA</th>\n",
       "      <th>present_fluNP</th>\n",
       "      <th>present_fluNS</th>\n",
       "      <th>present_fluPA</th>\n",
       "      <th>present_fluPB1</th>\n",
       "      <th>present_fluPB2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAGTAACAAGT</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCATCATTGCTT</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1.10971e-05</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAAGATGTTGA</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>13060</td>\n",
       "      <td>420</td>\n",
       "      <td>4139</td>\n",
       "      <td>5772</td>\n",
       "      <td>23</td>\n",
       "      <td>506</td>\n",
       "      <td>160</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00890406</td>\n",
       "      <td>0.00281551</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>8</td>\n",
       "      <td>319</td>\n",
       "      <td>1582</td>\n",
       "      <td>31</td>\n",
       "      <td>608</td>\n",
       "      <td>2055</td>\n",
       "      <td>11</td>\n",
       "      <td>45</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000361908</td>\n",
       "      <td>2.41272e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3124</th>\n",
       "      <td>TTTGATCTCAGTCATG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3125</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3126</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3127</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>45</td>\n",
       "      <td>418</td>\n",
       "      <td>1230</td>\n",
       "      <td>7</td>\n",
       "      <td>57</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000379671</td>\n",
       "      <td>5.32872e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3128</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3129 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode  n_viral_genes transcriptome_UMIs_fluHA  \\\n",
       "0     AAACCCAGTAACAAGT              0                        2   \n",
       "1     AAACCCATCATTGCTT              1                        0   \n",
       "2     AAACGAAAGATGTTGA              0                        3   \n",
       "3     AAACGAAGTACTTCCC              7                        2   \n",
       "4     AAACGAAGTAGACGTG              8                      319   \n",
       "...                ...            ...                      ...   \n",
       "3124  TTTGATCTCAGTCATG              0                        0   \n",
       "3125  TTTGATCTCCCGTTCA              0                        0   \n",
       "3126  TTTGGAGAGTTGCCTA              1                        0   \n",
       "3127  TTTGGAGGTATCGTTG              8                       93   \n",
       "3128  TTTGTTGAGTGCAGCA              1                        2   \n",
       "\n",
       "     transcriptome_UMIs_fluM transcriptome_UMIs_fluNA  \\\n",
       "0                          3                        0   \n",
       "1                          5                        0   \n",
       "2                         10                        0   \n",
       "3                      13060                      420   \n",
       "4                       1582                       31   \n",
       "...                      ...                      ...   \n",
       "3124                       4                        0   \n",
       "3125                       1                        0   \n",
       "3126                       6                        1   \n",
       "3127                    1668                       45   \n",
       "3128                       2                        1   \n",
       "\n",
       "     transcriptome_UMIs_fluNP transcriptome_UMIs_fluNS  \\\n",
       "0                           0                        1   \n",
       "1                           3                        1   \n",
       "2                           1                        4   \n",
       "3                        4139                     5772   \n",
       "4                         608                     2055   \n",
       "...                       ...                      ...   \n",
       "3124                        0                        0   \n",
       "3125                        1                        1   \n",
       "3126                        1                        4   \n",
       "3127                      418                     1230   \n",
       "3128                        0                        3   \n",
       "\n",
       "     transcriptome_UMIs_fluPA transcriptome_UMIs_fluPB1  \\\n",
       "0                           0                         0   \n",
       "1                           0                         0   \n",
       "2                           0                         0   \n",
       "3                          23                       506   \n",
       "4                          11                        45   \n",
       "...                       ...                       ...   \n",
       "3124                        0                         0   \n",
       "3125                        0                         0   \n",
       "3126                        0                         0   \n",
       "3127                        7                        57   \n",
       "3128                        0                         0   \n",
       "\n",
       "     transcriptome_UMIs_fluPB2  ... frac_UMIs_fluPB1 frac_UMIs_fluPB2  \\\n",
       "0                            0  ...                0                0   \n",
       "1                            1  ...                0      1.10971e-05   \n",
       "2                            0  ...                0                0   \n",
       "3                          160  ...       0.00890406       0.00281551   \n",
       "4                            3  ...      0.000361908      2.41272e-05   \n",
       "...                        ...  ...              ...              ...   \n",
       "3124                         0  ...                0                0   \n",
       "3125                         0  ...                0                0   \n",
       "3126                         0  ...                0                0   \n",
       "3127                         8  ...      0.000379671      5.32872e-05   \n",
       "3128                         0  ...                0                0   \n",
       "\n",
       "     present_fluHA present_fluM present_fluNA present_fluNP present_fluNS  \\\n",
       "0            False        False         False         False         False   \n",
       "1            False        False         False         False         False   \n",
       "2            False        False         False         False         False   \n",
       "3            False         True          True          True          True   \n",
       "4             True         True          True          True          True   \n",
       "...            ...          ...           ...           ...           ...   \n",
       "3124         False        False         False         False         False   \n",
       "3125         False        False         False         False         False   \n",
       "3126         False        False          True         False         False   \n",
       "3127          True         True          True          True          True   \n",
       "3128         False        False          True         False         False   \n",
       "\n",
       "     present_fluPA present_fluPB1 present_fluPB2  \n",
       "0            False          False          False  \n",
       "1            False          False           True  \n",
       "2            False          False          False  \n",
       "3             True           True           True  \n",
       "4             True           True           True  \n",
       "...            ...            ...            ...  \n",
       "3124         False          False          False  \n",
       "3125         False          False          False  \n",
       "3126         False          False          False  \n",
       "3127          True           True           True  \n",
       "3128         False          False          False  \n",
       "\n",
       "[3129 rows x 26 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_genes = pd.read_csv(viral_genes_by_cell_csv)\n",
    "viral_genes = viral_genes.rename(columns={'gene_UMIs': 'transcriptome_UMIs',\n",
    "                                          'frac_gene_UMIs': 'frac_UMIs',\n",
    "                                          'gene_present': 'present'})\n",
    "viral_genes = (viral_genes\n",
    "               .pivot(\n",
    "    index=['cell_barcode', 'n_viral_genes'],\n",
    "    columns=['gene'],\n",
    "    values=['transcriptome_UMIs', 'frac_UMIs', 'present']))\n",
    "viral_genes.columns = ['_'.join(col).strip() for col in viral_genes.columns.values]\n",
    "viral_genes = viral_genes.reset_index()\n",
    "display(viral_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load viral gene genotypes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.379192Z",
     "iopub.status.busy": "2022-05-25T18:48:52.378384Z",
     "iopub.status.idle": "2022-05-25T18:48:52.409100Z",
     "shell.execute_reply": "2022-05-25T18:48:52.409605Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>mutations_fluHA</th>\n",
       "      <th>mutations_fluM</th>\n",
       "      <th>mutations_fluNA</th>\n",
       "      <th>mutations_fluNP</th>\n",
       "      <th>mutations_fluNS</th>\n",
       "      <th>mutations_fluPA</th>\n",
       "      <th>mutations_fluPB1</th>\n",
       "      <th>mutations_fluPB2</th>\n",
       "      <th>pacbio_UMIs_fluHA</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio_UMIs_fluPB1</th>\n",
       "      <th>pacbio_UMIs_fluPB2</th>\n",
       "      <th>mutation_support_fluHA</th>\n",
       "      <th>mutation_support_fluM</th>\n",
       "      <th>mutation_support_fluNA</th>\n",
       "      <th>mutation_support_fluNP</th>\n",
       "      <th>mutation_support_fluNS</th>\n",
       "      <th>mutation_support_fluPA</th>\n",
       "      <th>mutation_support_fluPB1</th>\n",
       "      <th>mutation_support_fluPB2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>46</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAGGATTCTGATGGT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAGGGCTCCGCACTT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>del168to2007 del2097to2106</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>1611</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1597/1611 1601/1611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAGTCCAGTAGAGTT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>88</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>TTGGGATGTAACGATA</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>36</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>TTGGGATTCCGATTAG</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>30</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>TTTACCAGTCGCTTAA</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>TTTCACAAGCCAAGCA</td>\n",
       "      <td>T465A_H155Q</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WT</td>\n",
       "      <td>90</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>104</td>\n",
       "      <td>85/90</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>del47to47</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31/31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         cell_barcode mutations_fluHA mutations_fluM mutations_fluNA  \\\n",
       "0    AAACGAAGTACTTCCC             NaN             WT              WT   \n",
       "1    AAACGAAGTAGACGTG              WT             WT              WT   \n",
       "2    AAAGGATTCTGATGGT              WT             WT              WT   \n",
       "3    AAAGGGCTCCGCACTT              WT             WT              WT   \n",
       "4    AAAGTCCAGTAGAGTT              WT             WT              WT   \n",
       "..                ...             ...            ...             ...   \n",
       "405  TTGGGATGTAACGATA              WT             WT              WT   \n",
       "406  TTGGGATTCCGATTAG              WT             WT              WT   \n",
       "407  TTTACCAGTCGCTTAA              WT             WT              WT   \n",
       "408  TTTCACAAGCCAAGCA     T465A_H155Q             WT              WT   \n",
       "409  TTTGGAGGTATCGTTG              WT             WT       del47to47   \n",
       "\n",
       "    mutations_fluNP mutations_fluNS mutations_fluPA mutations_fluPB1  \\\n",
       "0                WT              WT              WT               WT   \n",
       "1                WT              WT              WT               WT   \n",
       "2                WT              WT              WT               WT   \n",
       "3                WT              WT              WT               WT   \n",
       "4                WT              WT              WT               WT   \n",
       "..              ...             ...             ...              ...   \n",
       "405              WT              WT              WT               WT   \n",
       "406              WT              WT              WT               WT   \n",
       "407              WT              WT             NaN               WT   \n",
       "408              WT             NaN              WT              NaN   \n",
       "409              WT              WT              WT               WT   \n",
       "\n",
       "               mutations_fluPB2 pacbio_UMIs_fluHA  ... pacbio_UMIs_fluPB1  \\\n",
       "0                            WT               NaN  ...                 46   \n",
       "1                           NaN                36  ...                 15   \n",
       "2                           NaN                 4  ...                  2   \n",
       "3    del168to2007 del2097to2106                11  ...                  5   \n",
       "4                            WT                19  ...                 88   \n",
       "..                          ...               ...  ...                ...   \n",
       "405                          WT                36  ...                 28   \n",
       "406                          WT                30  ...                 18   \n",
       "407                         NaN                 8  ...                  4   \n",
       "408                          WT                90  ...                NaN   \n",
       "409                          WT                12  ...                 23   \n",
       "\n",
       "    pacbio_UMIs_fluPB2 mutation_support_fluHA mutation_support_fluM  \\\n",
       "0                   21                    NaN                   NaN   \n",
       "1                  NaN                    NaN                   NaN   \n",
       "2                  NaN                    NaN                   NaN   \n",
       "3                 1611                    NaN                   NaN   \n",
       "4                   55                    NaN                   NaN   \n",
       "..                 ...                    ...                   ...   \n",
       "405                  5                    NaN                   NaN   \n",
       "406                  4                    NaN                   NaN   \n",
       "407                NaN                    NaN                   NaN   \n",
       "408                104                  85/90                   NaN   \n",
       "409                  7                    NaN                   NaN   \n",
       "\n",
       "    mutation_support_fluNA mutation_support_fluNP mutation_support_fluNS  \\\n",
       "0                      NaN                    NaN                    NaN   \n",
       "1                      NaN                    NaN                    NaN   \n",
       "2                      NaN                    NaN                    NaN   \n",
       "3                      NaN                    NaN                    NaN   \n",
       "4                      NaN                    NaN                    NaN   \n",
       "..                     ...                    ...                    ...   \n",
       "405                    NaN                    NaN                    NaN   \n",
       "406                    NaN                    NaN                    NaN   \n",
       "407                    NaN                    NaN                    NaN   \n",
       "408                    NaN                    NaN                    NaN   \n",
       "409                  31/31                    NaN                    NaN   \n",
       "\n",
       "    mutation_support_fluPA mutation_support_fluPB1 mutation_support_fluPB2  \n",
       "0                      NaN                     NaN                     NaN  \n",
       "1                      NaN                     NaN                     NaN  \n",
       "2                      NaN                     NaN                     NaN  \n",
       "3                      NaN                     NaN     1597/1611 1601/1611  \n",
       "4                      NaN                     NaN                     NaN  \n",
       "..                     ...                     ...                     ...  \n",
       "405                    NaN                     NaN                     NaN  \n",
       "406                    NaN                     NaN                     NaN  \n",
       "407                    NaN                     NaN                     NaN  \n",
       "408                    NaN                     NaN                     NaN  \n",
       "409                    NaN                     NaN                     NaN  \n",
       "\n",
       "[410 rows x 25 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "viral_genotypes = pd.read_csv(pacbio_consensus_gene_csv)\n",
    "viral_genotypes = viral_genotypes.rename(columns={'total_UMI': 'pacbio_UMIs',\n",
    "                                                  'consensus_mutations_annotated': 'mutations',\n",
    "                                                  'support': 'mutation_support'})\n",
    "viral_genotypes = (\n",
    "    viral_genotypes\n",
    "    .pivot(index=['cell_barcode'],\n",
    "           columns=['gene'],\n",
    "           values=['mutations', 'pacbio_UMIs', 'mutation_support']))\n",
    "viral_genotypes.columns = ['_'.join(col).strip() for col in viral_genotypes.columns.values]\n",
    "viral_genotypes = viral_genotypes.reset_index()\n",
    "display(viral_genotypes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load valid viral barcodes in each infected cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.412206Z",
     "iopub.status.busy": "2022-05-25T18:48:52.411419Z",
     "iopub.status.idle": "2022-05-25T18:48:52.427691Z",
     "shell.execute_reply": "2022-05-25T18:48:52.428207Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>barcoded_gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>viral_bc_UMIs</th>\n",
       "      <th>frac_viral_bc_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAGTAAGCGACATGAG</td>\n",
       "      <td>251</td>\n",
       "      <td>0.002019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAGGATTCTGATGGT</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>GTGGAGTCGCCAGTTC</td>\n",
       "      <td>114</td>\n",
       "      <td>0.001424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAGGGCTCCGCACTT</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AATCTACGAGGGAAAC</td>\n",
       "      <td>33</td>\n",
       "      <td>0.000257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAGGGCTCCGCACTT</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>ATGGATCAGATTTCCT</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAGGGCTCCGCACTT</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>GCAAAAATACAAAGTA</td>\n",
       "      <td>28</td>\n",
       "      <td>0.000218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>895</th>\n",
       "      <td>TTGGGATTCCGATTAG</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GTTTCGTGAGCATGTA</td>\n",
       "      <td>25</td>\n",
       "      <td>0.000182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>896</th>\n",
       "      <td>TTTACCAGTCGCTTAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>ATTTGGATGTAGGGGC</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>897</th>\n",
       "      <td>TTTACCAGTCGCTTAA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>TTGGAGGAGACCCGTG</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898</th>\n",
       "      <td>TTTCACAAGCCAAGCA</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>GGTATCAGTTATTGTT</td>\n",
       "      <td>186</td>\n",
       "      <td>0.002718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>ACATCTTATTTACACG</td>\n",
       "      <td>39</td>\n",
       "      <td>0.000260</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>900 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         cell_barcode barcoded_gene     viral_barcode  viral_bc_UMIs  \\\n",
       "0    AAACGAAGTAGACGTG         fluHA  AAGTAAGCGACATGAG            251   \n",
       "1    AAAGGATTCTGATGGT         fluHA  GTGGAGTCGCCAGTTC            114   \n",
       "2    AAAGGGCTCCGCACTT         fluHA  AATCTACGAGGGAAAC             33   \n",
       "3    AAAGGGCTCCGCACTT         fluHA  ATGGATCAGATTTCCT             10   \n",
       "4    AAAGGGCTCCGCACTT         fluHA  GCAAAAATACAAAGTA             28   \n",
       "..                ...           ...               ...            ...   \n",
       "895  TTGGGATTCCGATTAG         fluNA  GTTTCGTGAGCATGTA             25   \n",
       "896  TTTACCAGTCGCTTAA         fluNA  ATTTGGATGTAGGGGC              4   \n",
       "897  TTTACCAGTCGCTTAA         fluNA  TTGGAGGAGACCCGTG              7   \n",
       "898  TTTCACAAGCCAAGCA         fluNA  GGTATCAGTTATTGTT            186   \n",
       "899  TTTGGAGGTATCGTTG         fluNA  ACATCTTATTTACACG             39   \n",
       "\n",
       "     frac_viral_bc_UMIs  \n",
       "0              0.002019  \n",
       "1              0.001424  \n",
       "2              0.000257  \n",
       "3              0.000078  \n",
       "4              0.000218  \n",
       "..                  ...  \n",
       "895            0.000182  \n",
       "896            0.000035  \n",
       "897            0.000061  \n",
       "898            0.002718  \n",
       "899            0.000260  \n",
       "\n",
       "[900 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "transcriptome_viral_barcodes = pd.read_csv(viral_barcodes_valid_csv)\n",
    "transcriptome_viral_barcodes = transcriptome_viral_barcodes.drop(columns=['valid_viral_bc'])\n",
    "transcriptome_viral_barcodes = transcriptome_viral_barcodes.rename(columns={'gene': 'barcoded_gene'})\n",
    "assert set(transcriptome_viral_barcodes['barcoded_gene']) == set(barcoded_viral_genes), \\\n",
    "       \"Barcoded genes in barcode counts do not match expectation.\"\n",
    "display(transcriptome_viral_barcodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load progeny viral barcode frequencies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.430837Z",
     "iopub.status.busy": "2022-05-25T18:48:52.430011Z",
     "iopub.status.idle": "2022-05-25T18:48:52.457157Z",
     "shell.execute_reply": "2022-05-25T18:48:52.457664Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>barcoded_gene</th>\n",
       "      <th>freq_second_infection</th>\n",
       "      <th>freq_supernatant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAACACGTTCTATA</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAACATGATGACGCC</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAACTAGTTAGAGCA</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAGTTCTTGGATGT</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAATTGTCCAATATA</td>\n",
       "      <td>syn</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>790</th>\n",
       "      <td>TTTCGTGAATCACTGC</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>791</th>\n",
       "      <td>TTTCTTTACTCAGAAT</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.003833</td>\n",
       "      <td>0.002060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>792</th>\n",
       "      <td>TTTGCACACGTCTAAT</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>TTTTAACGTTATACTA</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>TTTTGCTTCGAGATAG</td>\n",
       "      <td>wt</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>0.093595</td>\n",
       "      <td>0.012644</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>795 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        viral_barcode infecting_viral_tag barcoded_gene  \\\n",
       "0    AAAAACACGTTCTATA                  wt         fluNA   \n",
       "1    AAAACATGATGACGCC                  wt         fluNA   \n",
       "2    AAAACTAGTTAGAGCA                  wt         fluHA   \n",
       "3    AAAAGTTCTTGGATGT                  wt         fluNA   \n",
       "4    AAAATTGTCCAATATA                 syn         fluHA   \n",
       "..                ...                 ...           ...   \n",
       "790  TTTCGTGAATCACTGC                  wt         fluHA   \n",
       "791  TTTCTTTACTCAGAAT                  wt         fluNA   \n",
       "792  TTTGCACACGTCTAAT                  wt         fluHA   \n",
       "793  TTTTAACGTTATACTA                  wt         fluNA   \n",
       "794  TTTTGCTTCGAGATAG                  wt         fluNA   \n",
       "\n",
       "     freq_second_infection  freq_supernatant  \n",
       "0                 0.000010          0.000010  \n",
       "1                 0.000010          0.000010  \n",
       "2                 0.000010          0.000010  \n",
       "3                 0.000010          0.000010  \n",
       "4                 0.000010          0.000010  \n",
       "..                     ...               ...  \n",
       "790               0.000010          0.000073  \n",
       "791               0.003833          0.002060  \n",
       "792               0.000010          0.000010  \n",
       "793               0.000010          0.000010  \n",
       "794               0.093595          0.012644  \n",
       "\n",
       "[795 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "progeny_viral_barcodes = pd.read_csv(filtered_progeny_viral_bc_csv)\n",
    "progeny_viral_barcodes = progeny_viral_barcodes.drop(columns=['Unnamed: 0'])\n",
    "progeny_viral_barcodes = (progeny_viral_barcodes\n",
    "                          .rename(columns={'tag': 'infecting_viral_tag',\n",
    "                                           'average_freq': 'freq',\n",
    "                                           'gene': 'barcoded_gene'}))\n",
    "assert set(progeny_viral_barcodes['barcoded_gene']) == set(barcoded_viral_genes), \\\n",
    "       \"Barcoded genes in barcode counts do not match expectation.\"\n",
    "progeny_viral_barcodes = (\n",
    "    progeny_viral_barcodes\n",
    "    .pivot(index=['viral_barcode', 'infecting_viral_tag', 'barcoded_gene'],\n",
    "           columns=['source'],\n",
    "           values=['freq']))\n",
    "progeny_viral_barcodes.columns = ['_'.join(col).strip() for col in progeny_viral_barcodes.columns.values]\n",
    "progeny_viral_barcodes = progeny_viral_barcodes.reset_index()\n",
    "display(progeny_viral_barcodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load whether each infected cell contributes any progeny:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.460387Z",
     "iopub.status.busy": "2022-05-25T18:48:52.459591Z",
     "iopub.status.idle": "2022-05-25T18:48:52.478640Z",
     "shell.execute_reply": "2022-05-25T18:48:52.479141Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>max_freq_second_infection</th>\n",
       "      <th>max_freq_supernatant</th>\n",
       "      <th>contributes_progeny_second_infection</th>\n",
       "      <th>contributes_progeny_supernatant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>1e-05</td>\n",
       "      <td>7.62399e-05</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAGGATTCTGATGGT</td>\n",
       "      <td>0.077933</td>\n",
       "      <td>0.0131834</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAGGGCTCCGCACTT</td>\n",
       "      <td>0.000124415</td>\n",
       "      <td>0.00195048</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAGTCCAGTAGAGTT</td>\n",
       "      <td>1e-05</td>\n",
       "      <td>1e-05</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AACCAACAGGGTCTTT</td>\n",
       "      <td>2.17873e-05</td>\n",
       "      <td>0.000971284</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>TTGGGATGTAACGATA</td>\n",
       "      <td>0.000153675</td>\n",
       "      <td>0.00128394</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>TTGGGATTCCGATTAG</td>\n",
       "      <td>1e-05</td>\n",
       "      <td>1e-05</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>TTTACCAGTCGCTTAA</td>\n",
       "      <td>1.8581e-05</td>\n",
       "      <td>0.00121169</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>TTTCACAAGCCAAGCA</td>\n",
       "      <td>1e-05</td>\n",
       "      <td>1e-05</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>2.81273e-05</td>\n",
       "      <td>0.00179022</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>359 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         cell_barcode max_freq_second_infection max_freq_supernatant  \\\n",
       "0    AAACGAAGTAGACGTG                     1e-05          7.62399e-05   \n",
       "1    AAAGGATTCTGATGGT                  0.077933            0.0131834   \n",
       "2    AAAGGGCTCCGCACTT               0.000124415           0.00195048   \n",
       "3    AAAGTCCAGTAGAGTT                     1e-05                1e-05   \n",
       "4    AACCAACAGGGTCTTT               2.17873e-05          0.000971284   \n",
       "..                ...                       ...                  ...   \n",
       "354  TTGGGATGTAACGATA               0.000153675           0.00128394   \n",
       "355  TTGGGATTCCGATTAG                     1e-05                1e-05   \n",
       "356  TTTACCAGTCGCTTAA                1.8581e-05           0.00121169   \n",
       "357  TTTCACAAGCCAAGCA                     1e-05                1e-05   \n",
       "358  TTTGGAGGTATCGTTG               2.81273e-05           0.00179022   \n",
       "\n",
       "    contributes_progeny_second_infection contributes_progeny_supernatant  \n",
       "0                                  False                            True  \n",
       "1                                   True                            True  \n",
       "2                                   True                            True  \n",
       "3                                  False                           False  \n",
       "4                                   True                            True  \n",
       "..                                   ...                             ...  \n",
       "354                                 True                            True  \n",
       "355                                False                           False  \n",
       "356                                 True                            True  \n",
       "357                                False                           False  \n",
       "358                                 True                            True  \n",
       "\n",
       "[359 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "contributes_progeny = pd.read_csv(contributes_progeny_by_cell_csv)\n",
    "contributes_progeny = contributes_progeny.rename(columns={'gene': 'barcoded_gene',\n",
    "                                                          'max_progeny_freq': 'max_freq'})\n",
    "assert set(transcriptome_viral_barcodes['barcoded_gene']) == set(barcoded_viral_genes), \\\n",
    "       \"Barcoded genes in barcode counts do not match expectation.\"\n",
    "contributes_progeny = (\n",
    "    contributes_progeny\n",
    "    .pivot(index=['cell_barcode'],\n",
    "           columns=['source'],\n",
    "           values=['max_freq', 'contributes_progeny']))\n",
    "contributes_progeny.columns = ['_'.join(col).strip() for col in contributes_progeny.columns.values]\n",
    "contributes_progeny = contributes_progeny.reset_index()\n",
    "display(contributes_progeny)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Integrate data into single dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Start with `cell_annotations` dataframe. Bring in viral gene expression information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.481634Z",
     "iopub.status.busy": "2022-05-25T18:48:52.480855Z",
     "iopub.status.idle": "2022-05-25T18:48:52.519794Z",
     "shell.execute_reply": "2022-05-25T18:48:52.520321Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "      <th>n_viral_genes</th>\n",
       "      <th>transcriptome_UMIs_fluHA</th>\n",
       "      <th>transcriptome_UMIs_fluM</th>\n",
       "      <th>transcriptome_UMIs_fluNA</th>\n",
       "      <th>...</th>\n",
       "      <th>frac_UMIs_fluPB1</th>\n",
       "      <th>frac_UMIs_fluPB2</th>\n",
       "      <th>present_fluHA</th>\n",
       "      <th>present_fluM</th>\n",
       "      <th>present_fluNA</th>\n",
       "      <th>present_fluNP</th>\n",
       "      <th>present_fluNS</th>\n",
       "      <th>present_fluPA</th>\n",
       "      <th>present_fluPB1</th>\n",
       "      <th>present_fluPB2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAGTAACAAGT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>47873</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCATCATTGCTT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>90114</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1.10971e-05</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAAGATGTTGA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>111630</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>infected</td>\n",
       "      <td>both</td>\n",
       "      <td>56828</td>\n",
       "      <td>24082</td>\n",
       "      <td>0.423770</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>13060</td>\n",
       "      <td>420</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00890406</td>\n",
       "      <td>0.00281551</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>124341</td>\n",
       "      <td>4654</td>\n",
       "      <td>0.037429</td>\n",
       "      <td>8</td>\n",
       "      <td>319</td>\n",
       "      <td>1582</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000361908</td>\n",
       "      <td>2.41272e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3124</th>\n",
       "      <td>TTTGATCTCAGTCATG</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>44890</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3125</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>63150</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3126</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>65941</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3127</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000379671</td>\n",
       "      <td>5.32872e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3128</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>135009</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3129 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode    infected infecting_viral_tag  total_UMIs  \\\n",
       "0     AAACCCAGTAACAAGT  uninfected                none       47873   \n",
       "1     AAACCCATCATTGCTT  uninfected                none       90114   \n",
       "2     AAACGAAAGATGTTGA  uninfected                none      111630   \n",
       "3     AAACGAAGTACTTCCC    infected                both       56828   \n",
       "4     AAACGAAGTAGACGTG    infected                  wt      124341   \n",
       "...                ...         ...                 ...         ...   \n",
       "3124  TTTGATCTCAGTCATG  uninfected                none       44890   \n",
       "3125  TTTGATCTCCCGTTCA  uninfected                none       63150   \n",
       "3126  TTTGGAGAGTTGCCTA  uninfected                none       65941   \n",
       "3127  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "3128  TTTGTTGAGTGCAGCA  uninfected                none      135009   \n",
       "\n",
       "      viral_UMIs  frac_viral_UMIs  n_viral_genes transcriptome_UMIs_fluHA  \\\n",
       "0              6         0.000125              0                        2   \n",
       "1             10         0.000111              1                        0   \n",
       "2             18         0.000161              0                        3   \n",
       "3          24082         0.423770              7                        2   \n",
       "4           4654         0.037429              8                      319   \n",
       "...          ...              ...            ...                      ...   \n",
       "3124           4         0.000089              0                        0   \n",
       "3125           3         0.000048              0                        0   \n",
       "3126          12         0.000182              1                        0   \n",
       "3127        3526         0.023486              8                       93   \n",
       "3128           8         0.000059              1                        2   \n",
       "\n",
       "     transcriptome_UMIs_fluM transcriptome_UMIs_fluNA  ... frac_UMIs_fluPB1  \\\n",
       "0                          3                        0  ...                0   \n",
       "1                          5                        0  ...                0   \n",
       "2                         10                        0  ...                0   \n",
       "3                      13060                      420  ...       0.00890406   \n",
       "4                       1582                       31  ...      0.000361908   \n",
       "...                      ...                      ...  ...              ...   \n",
       "3124                       4                        0  ...                0   \n",
       "3125                       1                        0  ...                0   \n",
       "3126                       6                        1  ...                0   \n",
       "3127                    1668                       45  ...      0.000379671   \n",
       "3128                       2                        1  ...                0   \n",
       "\n",
       "     frac_UMIs_fluPB2 present_fluHA present_fluM present_fluNA present_fluNP  \\\n",
       "0                   0         False        False         False         False   \n",
       "1         1.10971e-05         False        False         False         False   \n",
       "2                   0         False        False         False         False   \n",
       "3          0.00281551         False         True          True          True   \n",
       "4         2.41272e-05          True         True          True          True   \n",
       "...               ...           ...          ...           ...           ...   \n",
       "3124                0         False        False         False         False   \n",
       "3125                0         False        False         False         False   \n",
       "3126                0         False        False          True         False   \n",
       "3127      5.32872e-05          True         True          True          True   \n",
       "3128                0         False        False          True         False   \n",
       "\n",
       "     present_fluNS present_fluPA present_fluPB1 present_fluPB2  \n",
       "0            False         False          False          False  \n",
       "1            False         False          False           True  \n",
       "2            False         False          False          False  \n",
       "3             True          True           True           True  \n",
       "4             True          True           True           True  \n",
       "...            ...           ...            ...            ...  \n",
       "3124         False         False          False          False  \n",
       "3125         False         False          False          False  \n",
       "3126         False         False          False          False  \n",
       "3127          True          True           True           True  \n",
       "3128         False         False          False          False  \n",
       "\n",
       "[3129 rows x 31 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "integrated_df = pd.merge(\n",
    "    left=cell_annotations,\n",
    "    right=viral_genes,\n",
    "    on='cell_barcode',\n",
    "    how='outer',\n",
    "    validate='one_to_one'\n",
    ")\n",
    "display(integrated_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Bring in viral gene genotypes. Many cells do not have pacbio data. For all missing data, fill all counts with 0 and fill all mutations with \"Not Detected\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.523098Z",
     "iopub.status.busy": "2022-05-25T18:48:52.522296Z",
     "iopub.status.idle": "2022-05-25T18:48:52.573468Z",
     "shell.execute_reply": "2022-05-25T18:48:52.573997Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "      <th>n_viral_genes</th>\n",
       "      <th>transcriptome_UMIs_fluHA</th>\n",
       "      <th>transcriptome_UMIs_fluM</th>\n",
       "      <th>transcriptome_UMIs_fluNA</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio_UMIs_fluPB1_y</th>\n",
       "      <th>pacbio_UMIs_fluPB2_y</th>\n",
       "      <th>mutation_support_fluHA_y</th>\n",
       "      <th>mutation_support_fluM_y</th>\n",
       "      <th>mutation_support_fluNA_y</th>\n",
       "      <th>mutation_support_fluNP_y</th>\n",
       "      <th>mutation_support_fluNS_y</th>\n",
       "      <th>mutation_support_fluPA_y</th>\n",
       "      <th>mutation_support_fluPB1_y</th>\n",
       "      <th>mutation_support_fluPB2_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAGTAACAAGT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>47873</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCATCATTGCTT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>90114</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAAGATGTTGA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>111630</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>infected</td>\n",
       "      <td>both</td>\n",
       "      <td>56828</td>\n",
       "      <td>24082</td>\n",
       "      <td>0.423770</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>13060</td>\n",
       "      <td>420</td>\n",
       "      <td>...</td>\n",
       "      <td>46</td>\n",
       "      <td>21</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>124341</td>\n",
       "      <td>4654</td>\n",
       "      <td>0.037429</td>\n",
       "      <td>8</td>\n",
       "      <td>319</td>\n",
       "      <td>1582</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3124</th>\n",
       "      <td>TTTGATCTCAGTCATG</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>44890</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3125</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>63150</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3126</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>65941</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3127</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>7</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>31/31</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3128</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>135009</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3129 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode    infected infecting_viral_tag  total_UMIs  \\\n",
       "0     AAACCCAGTAACAAGT  uninfected                none       47873   \n",
       "1     AAACCCATCATTGCTT  uninfected                none       90114   \n",
       "2     AAACGAAAGATGTTGA  uninfected                none      111630   \n",
       "3     AAACGAAGTACTTCCC    infected                both       56828   \n",
       "4     AAACGAAGTAGACGTG    infected                  wt      124341   \n",
       "...                ...         ...                 ...         ...   \n",
       "3124  TTTGATCTCAGTCATG  uninfected                none       44890   \n",
       "3125  TTTGATCTCCCGTTCA  uninfected                none       63150   \n",
       "3126  TTTGGAGAGTTGCCTA  uninfected                none       65941   \n",
       "3127  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "3128  TTTGTTGAGTGCAGCA  uninfected                none      135009   \n",
       "\n",
       "      viral_UMIs  frac_viral_UMIs  n_viral_genes transcriptome_UMIs_fluHA  \\\n",
       "0              6         0.000125              0                        2   \n",
       "1             10         0.000111              1                        0   \n",
       "2             18         0.000161              0                        3   \n",
       "3          24082         0.423770              7                        2   \n",
       "4           4654         0.037429              8                      319   \n",
       "...          ...              ...            ...                      ...   \n",
       "3124           4         0.000089              0                        0   \n",
       "3125           3         0.000048              0                        0   \n",
       "3126          12         0.000182              1                        0   \n",
       "3127        3526         0.023486              8                       93   \n",
       "3128           8         0.000059              1                        2   \n",
       "\n",
       "     transcriptome_UMIs_fluM transcriptome_UMIs_fluNA  ...  \\\n",
       "0                          3                        0  ...   \n",
       "1                          5                        0  ...   \n",
       "2                         10                        0  ...   \n",
       "3                      13060                      420  ...   \n",
       "4                       1582                       31  ...   \n",
       "...                      ...                      ...  ...   \n",
       "3124                       4                        0  ...   \n",
       "3125                       1                        0  ...   \n",
       "3126                       6                        1  ...   \n",
       "3127                    1668                       45  ...   \n",
       "3128                       2                        1  ...   \n",
       "\n",
       "     pacbio_UMIs_fluPB1_y pacbio_UMIs_fluPB2_y mutation_support_fluHA_y  \\\n",
       "0                       0                    0           Not Applicable   \n",
       "1                       0                    0           Not Applicable   \n",
       "2                       0                    0           Not Applicable   \n",
       "3                      46                   21           Not Applicable   \n",
       "4                      15                    0           Not Applicable   \n",
       "...                   ...                  ...                      ...   \n",
       "3124                    0                    0           Not Applicable   \n",
       "3125                    0                    0           Not Applicable   \n",
       "3126                    0                    0           Not Applicable   \n",
       "3127                   23                    7           Not Applicable   \n",
       "3128                    0                    0           Not Applicable   \n",
       "\n",
       "     mutation_support_fluM_y mutation_support_fluNA_y  \\\n",
       "0             Not Applicable           Not Applicable   \n",
       "1             Not Applicable           Not Applicable   \n",
       "2             Not Applicable           Not Applicable   \n",
       "3             Not Applicable           Not Applicable   \n",
       "4             Not Applicable           Not Applicable   \n",
       "...                      ...                      ...   \n",
       "3124          Not Applicable           Not Applicable   \n",
       "3125          Not Applicable           Not Applicable   \n",
       "3126          Not Applicable           Not Applicable   \n",
       "3127          Not Applicable                    31/31   \n",
       "3128          Not Applicable           Not Applicable   \n",
       "\n",
       "     mutation_support_fluNP_y mutation_support_fluNS_y  \\\n",
       "0              Not Applicable           Not Applicable   \n",
       "1              Not Applicable           Not Applicable   \n",
       "2              Not Applicable           Not Applicable   \n",
       "3              Not Applicable           Not Applicable   \n",
       "4              Not Applicable           Not Applicable   \n",
       "...                       ...                      ...   \n",
       "3124           Not Applicable           Not Applicable   \n",
       "3125           Not Applicable           Not Applicable   \n",
       "3126           Not Applicable           Not Applicable   \n",
       "3127           Not Applicable           Not Applicable   \n",
       "3128           Not Applicable           Not Applicable   \n",
       "\n",
       "     mutation_support_fluPA_y mutation_support_fluPB1_y  \\\n",
       "0              Not Applicable            Not Applicable   \n",
       "1              Not Applicable            Not Applicable   \n",
       "2              Not Applicable            Not Applicable   \n",
       "3              Not Applicable            Not Applicable   \n",
       "4              Not Applicable            Not Applicable   \n",
       "...                       ...                       ...   \n",
       "3124           Not Applicable            Not Applicable   \n",
       "3125           Not Applicable            Not Applicable   \n",
       "3126           Not Applicable            Not Applicable   \n",
       "3127           Not Applicable            Not Applicable   \n",
       "3128           Not Applicable            Not Applicable   \n",
       "\n",
       "     mutation_support_fluPB2_y  \n",
       "0               Not Applicable  \n",
       "1               Not Applicable  \n",
       "2               Not Applicable  \n",
       "3               Not Applicable  \n",
       "4               Not Applicable  \n",
       "...                        ...  \n",
       "3124            Not Applicable  \n",
       "3125            Not Applicable  \n",
       "3126            Not Applicable  \n",
       "3127            Not Applicable  \n",
       "3128            Not Applicable  \n",
       "\n",
       "[3129 rows x 79 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "integrated_df = pd.merge(\n",
    "    left=integrated_df,\n",
    "    right=viral_genotypes,\n",
    "    on='cell_barcode',\n",
    "    how='left',\n",
    "    validate='one_to_one'\n",
    ")\n",
    "\n",
    "# Fill empty mutations columns with \"Not Detected\"\n",
    "integrated_df[[col for col in integrated_df if col.startswith('mutations')]] = (\n",
    "     integrated_df[[col for col in integrated_df if col.startswith('mutations')]]\n",
    "     .fillna('Not Detected'))\n",
    "# Fill empty pacbio_UMIs columns with 0\n",
    "integrated_df[[col for col in integrated_df if col.startswith('pacbio_UMIs')]] = (\n",
    "     integrated_df[[col for col in integrated_df if col.startswith('pacbio_UMIs')]]\n",
    "     .fillna(0))\n",
    "# Fill NaN support columns with \"Not Applicable\"\n",
    "integrated_df[[col for col in integrated_df if col.startswith('mutation_support')]] = (\n",
    "     integrated_df[[col for col in integrated_df if col.startswith('mutation_support')]]\n",
    "     .fillna('Not Applicable'))\n",
    "\n",
    "display(integrated_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Bring in valid viral barcodes for cells that have them:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.576527Z",
     "iopub.status.busy": "2022-05-25T18:48:52.575741Z",
     "iopub.status.idle": "2022-05-25T18:48:52.611863Z",
     "shell.execute_reply": "2022-05-25T18:48:52.612363Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "      <th>n_viral_genes</th>\n",
       "      <th>transcriptome_UMIs_fluHA</th>\n",
       "      <th>transcriptome_UMIs_fluM</th>\n",
       "      <th>transcriptome_UMIs_fluNA</th>\n",
       "      <th>...</th>\n",
       "      <th>mutation_support_fluNA_y</th>\n",
       "      <th>mutation_support_fluNP_y</th>\n",
       "      <th>mutation_support_fluNS_y</th>\n",
       "      <th>mutation_support_fluPA_y</th>\n",
       "      <th>mutation_support_fluPB1_y</th>\n",
       "      <th>mutation_support_fluPB2_y</th>\n",
       "      <th>barcoded_gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>viral_bc_UMIs</th>\n",
       "      <th>frac_viral_bc_UMIs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAGTAACAAGT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>47873</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCATCATTGCTT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>90114</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAAGATGTTGA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>111630</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>infected</td>\n",
       "      <td>both</td>\n",
       "      <td>56828</td>\n",
       "      <td>24082</td>\n",
       "      <td>0.423770</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>13060</td>\n",
       "      <td>420</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>124341</td>\n",
       "      <td>4654</td>\n",
       "      <td>0.037429</td>\n",
       "      <td>8</td>\n",
       "      <td>319</td>\n",
       "      <td>1582</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAGTAAGCGACATGAG</td>\n",
       "      <td>251.0</td>\n",
       "      <td>0.002019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3677</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>63150</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3678</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>65941</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3679</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>31/31</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGTAAACTTCCTCGCG</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.000433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3680</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>31/31</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>ACATCTTATTTACACG</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.000260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3681</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>135009</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3682 rows × 83 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode    infected infecting_viral_tag  total_UMIs  \\\n",
       "0     AAACCCAGTAACAAGT  uninfected                none       47873   \n",
       "1     AAACCCATCATTGCTT  uninfected                none       90114   \n",
       "2     AAACGAAAGATGTTGA  uninfected                none      111630   \n",
       "3     AAACGAAGTACTTCCC    infected                both       56828   \n",
       "4     AAACGAAGTAGACGTG    infected                  wt      124341   \n",
       "...                ...         ...                 ...         ...   \n",
       "3677  TTTGATCTCCCGTTCA  uninfected                none       63150   \n",
       "3678  TTTGGAGAGTTGCCTA  uninfected                none       65941   \n",
       "3679  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "3680  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "3681  TTTGTTGAGTGCAGCA  uninfected                none      135009   \n",
       "\n",
       "      viral_UMIs  frac_viral_UMIs  n_viral_genes transcriptome_UMIs_fluHA  \\\n",
       "0              6         0.000125              0                        2   \n",
       "1             10         0.000111              1                        0   \n",
       "2             18         0.000161              0                        3   \n",
       "3          24082         0.423770              7                        2   \n",
       "4           4654         0.037429              8                      319   \n",
       "...          ...              ...            ...                      ...   \n",
       "3677           3         0.000048              0                        0   \n",
       "3678          12         0.000182              1                        0   \n",
       "3679        3526         0.023486              8                       93   \n",
       "3680        3526         0.023486              8                       93   \n",
       "3681           8         0.000059              1                        2   \n",
       "\n",
       "     transcriptome_UMIs_fluM transcriptome_UMIs_fluNA  ...  \\\n",
       "0                          3                        0  ...   \n",
       "1                          5                        0  ...   \n",
       "2                         10                        0  ...   \n",
       "3                      13060                      420  ...   \n",
       "4                       1582                       31  ...   \n",
       "...                      ...                      ...  ...   \n",
       "3677                       1                        0  ...   \n",
       "3678                       6                        1  ...   \n",
       "3679                    1668                       45  ...   \n",
       "3680                    1668                       45  ...   \n",
       "3681                       2                        1  ...   \n",
       "\n",
       "     mutation_support_fluNA_y mutation_support_fluNP_y  \\\n",
       "0              Not Applicable           Not Applicable   \n",
       "1              Not Applicable           Not Applicable   \n",
       "2              Not Applicable           Not Applicable   \n",
       "3              Not Applicable           Not Applicable   \n",
       "4              Not Applicable           Not Applicable   \n",
       "...                       ...                      ...   \n",
       "3677           Not Applicable           Not Applicable   \n",
       "3678           Not Applicable           Not Applicable   \n",
       "3679                    31/31           Not Applicable   \n",
       "3680                    31/31           Not Applicable   \n",
       "3681           Not Applicable           Not Applicable   \n",
       "\n",
       "     mutation_support_fluNS_y mutation_support_fluPA_y  \\\n",
       "0              Not Applicable           Not Applicable   \n",
       "1              Not Applicable           Not Applicable   \n",
       "2              Not Applicable           Not Applicable   \n",
       "3              Not Applicable           Not Applicable   \n",
       "4              Not Applicable           Not Applicable   \n",
       "...                       ...                      ...   \n",
       "3677           Not Applicable           Not Applicable   \n",
       "3678           Not Applicable           Not Applicable   \n",
       "3679           Not Applicable           Not Applicable   \n",
       "3680           Not Applicable           Not Applicable   \n",
       "3681           Not Applicable           Not Applicable   \n",
       "\n",
       "     mutation_support_fluPB1_y mutation_support_fluPB2_y barcoded_gene  \\\n",
       "0               Not Applicable            Not Applicable           NaN   \n",
       "1               Not Applicable            Not Applicable           NaN   \n",
       "2               Not Applicable            Not Applicable           NaN   \n",
       "3               Not Applicable            Not Applicable           NaN   \n",
       "4               Not Applicable            Not Applicable         fluHA   \n",
       "...                        ...                       ...           ...   \n",
       "3677            Not Applicable            Not Applicable           NaN   \n",
       "3678            Not Applicable            Not Applicable           NaN   \n",
       "3679            Not Applicable            Not Applicable         fluHA   \n",
       "3680            Not Applicable            Not Applicable         fluNA   \n",
       "3681            Not Applicable            Not Applicable           NaN   \n",
       "\n",
       "         viral_barcode viral_bc_UMIs frac_viral_bc_UMIs  \n",
       "0                  NaN           NaN                NaN  \n",
       "1                  NaN           NaN                NaN  \n",
       "2                  NaN           NaN                NaN  \n",
       "3                  NaN           NaN                NaN  \n",
       "4     AAGTAAGCGACATGAG         251.0           0.002019  \n",
       "...                ...           ...                ...  \n",
       "3677               NaN           NaN                NaN  \n",
       "3678               NaN           NaN                NaN  \n",
       "3679  AGTAAACTTCCTCGCG          65.0           0.000433  \n",
       "3680  ACATCTTATTTACACG          39.0           0.000260  \n",
       "3681               NaN           NaN                NaN  \n",
       "\n",
       "[3682 rows x 83 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "integrated_df = pd.merge(\n",
    "    left=integrated_df,\n",
    "    right=transcriptome_viral_barcodes,\n",
    "    on='cell_barcode',\n",
    "    how='left',\n",
    "    validate='one_to_many'\n",
    ")\n",
    "display(integrated_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Bring in progeny frequencies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.614898Z",
     "iopub.status.busy": "2022-05-25T18:48:52.614123Z",
     "iopub.status.idle": "2022-05-25T18:48:52.650691Z",
     "shell.execute_reply": "2022-05-25T18:48:52.651204Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "      <th>n_viral_genes</th>\n",
       "      <th>transcriptome_UMIs_fluHA</th>\n",
       "      <th>transcriptome_UMIs_fluM</th>\n",
       "      <th>transcriptome_UMIs_fluNA</th>\n",
       "      <th>...</th>\n",
       "      <th>mutation_support_fluNS_y</th>\n",
       "      <th>mutation_support_fluPA_y</th>\n",
       "      <th>mutation_support_fluPB1_y</th>\n",
       "      <th>mutation_support_fluPB2_y</th>\n",
       "      <th>barcoded_gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>viral_bc_UMIs</th>\n",
       "      <th>frac_viral_bc_UMIs</th>\n",
       "      <th>freq_second_infection</th>\n",
       "      <th>freq_supernatant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAGTAACAAGT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>47873</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCATCATTGCTT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>90114</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAAGATGTTGA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>111630</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>infected</td>\n",
       "      <td>both</td>\n",
       "      <td>56828</td>\n",
       "      <td>24082</td>\n",
       "      <td>0.423770</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>13060</td>\n",
       "      <td>420</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>124341</td>\n",
       "      <td>4654</td>\n",
       "      <td>0.037429</td>\n",
       "      <td>8</td>\n",
       "      <td>319</td>\n",
       "      <td>1582</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAGTAAGCGACATGAG</td>\n",
       "      <td>251.0</td>\n",
       "      <td>0.002019</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3677</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>63150</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3678</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>65941</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3679</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGTAAACTTCCTCGCG</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.000433</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3680</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>ACATCTTATTTACACG</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>0.001790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3681</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>135009</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3682 rows × 85 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode    infected infecting_viral_tag  total_UMIs  \\\n",
       "0     AAACCCAGTAACAAGT  uninfected                none       47873   \n",
       "1     AAACCCATCATTGCTT  uninfected                none       90114   \n",
       "2     AAACGAAAGATGTTGA  uninfected                none      111630   \n",
       "3     AAACGAAGTACTTCCC    infected                both       56828   \n",
       "4     AAACGAAGTAGACGTG    infected                  wt      124341   \n",
       "...                ...         ...                 ...         ...   \n",
       "3677  TTTGATCTCCCGTTCA  uninfected                none       63150   \n",
       "3678  TTTGGAGAGTTGCCTA  uninfected                none       65941   \n",
       "3679  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "3680  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "3681  TTTGTTGAGTGCAGCA  uninfected                none      135009   \n",
       "\n",
       "      viral_UMIs  frac_viral_UMIs  n_viral_genes transcriptome_UMIs_fluHA  \\\n",
       "0              6         0.000125              0                        2   \n",
       "1             10         0.000111              1                        0   \n",
       "2             18         0.000161              0                        3   \n",
       "3          24082         0.423770              7                        2   \n",
       "4           4654         0.037429              8                      319   \n",
       "...          ...              ...            ...                      ...   \n",
       "3677           3         0.000048              0                        0   \n",
       "3678          12         0.000182              1                        0   \n",
       "3679        3526         0.023486              8                       93   \n",
       "3680        3526         0.023486              8                       93   \n",
       "3681           8         0.000059              1                        2   \n",
       "\n",
       "     transcriptome_UMIs_fluM transcriptome_UMIs_fluNA  ...  \\\n",
       "0                          3                        0  ...   \n",
       "1                          5                        0  ...   \n",
       "2                         10                        0  ...   \n",
       "3                      13060                      420  ...   \n",
       "4                       1582                       31  ...   \n",
       "...                      ...                      ...  ...   \n",
       "3677                       1                        0  ...   \n",
       "3678                       6                        1  ...   \n",
       "3679                    1668                       45  ...   \n",
       "3680                    1668                       45  ...   \n",
       "3681                       2                        1  ...   \n",
       "\n",
       "     mutation_support_fluNS_y mutation_support_fluPA_y  \\\n",
       "0              Not Applicable           Not Applicable   \n",
       "1              Not Applicable           Not Applicable   \n",
       "2              Not Applicable           Not Applicable   \n",
       "3              Not Applicable           Not Applicable   \n",
       "4              Not Applicable           Not Applicable   \n",
       "...                       ...                      ...   \n",
       "3677           Not Applicable           Not Applicable   \n",
       "3678           Not Applicable           Not Applicable   \n",
       "3679           Not Applicable           Not Applicable   \n",
       "3680           Not Applicable           Not Applicable   \n",
       "3681           Not Applicable           Not Applicable   \n",
       "\n",
       "     mutation_support_fluPB1_y mutation_support_fluPB2_y barcoded_gene  \\\n",
       "0               Not Applicable            Not Applicable           NaN   \n",
       "1               Not Applicable            Not Applicable           NaN   \n",
       "2               Not Applicable            Not Applicable           NaN   \n",
       "3               Not Applicable            Not Applicable           NaN   \n",
       "4               Not Applicable            Not Applicable         fluHA   \n",
       "...                        ...                       ...           ...   \n",
       "3677            Not Applicable            Not Applicable           NaN   \n",
       "3678            Not Applicable            Not Applicable           NaN   \n",
       "3679            Not Applicable            Not Applicable         fluHA   \n",
       "3680            Not Applicable            Not Applicable         fluNA   \n",
       "3681            Not Applicable            Not Applicable           NaN   \n",
       "\n",
       "         viral_barcode viral_bc_UMIs frac_viral_bc_UMIs freq_second_infection  \\\n",
       "0                  NaN           NaN                NaN                   NaN   \n",
       "1                  NaN           NaN                NaN                   NaN   \n",
       "2                  NaN           NaN                NaN                   NaN   \n",
       "3                  NaN           NaN                NaN                   NaN   \n",
       "4     AAGTAAGCGACATGAG         251.0           0.002019              0.000010   \n",
       "...                ...           ...                ...                   ...   \n",
       "3677               NaN           NaN                NaN                   NaN   \n",
       "3678               NaN           NaN                NaN                   NaN   \n",
       "3679  AGTAAACTTCCTCGCG          65.0           0.000433              0.000010   \n",
       "3680  ACATCTTATTTACACG          39.0           0.000260              0.000028   \n",
       "3681               NaN           NaN                NaN                   NaN   \n",
       "\n",
       "     freq_supernatant  \n",
       "0                 NaN  \n",
       "1                 NaN  \n",
       "2                 NaN  \n",
       "3                 NaN  \n",
       "4            0.000076  \n",
       "...               ...  \n",
       "3677              NaN  \n",
       "3678              NaN  \n",
       "3679         0.000695  \n",
       "3680         0.001790  \n",
       "3681              NaN  \n",
       "\n",
       "[3682 rows x 85 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "integrated_df = pd.merge(\n",
    "    left=integrated_df,\n",
    "    right=progeny_viral_barcodes,\n",
    "    on=['infecting_viral_tag', 'barcoded_gene', 'viral_barcode'],\n",
    "    how='left',\n",
    "    validate='many_to_one'\n",
    ")\n",
    "display(integrated_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Bring in annotation of whether each infected cell contributes any progeny:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.653721Z",
     "iopub.status.busy": "2022-05-25T18:48:52.652928Z",
     "iopub.status.idle": "2022-05-25T18:48:52.690819Z",
     "shell.execute_reply": "2022-05-25T18:48:52.691324Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>infected</th>\n",
       "      <th>infecting_viral_tag</th>\n",
       "      <th>total_UMIs</th>\n",
       "      <th>viral_UMIs</th>\n",
       "      <th>frac_viral_UMIs</th>\n",
       "      <th>n_viral_genes</th>\n",
       "      <th>transcriptome_UMIs_fluHA</th>\n",
       "      <th>transcriptome_UMIs_fluM</th>\n",
       "      <th>transcriptome_UMIs_fluNA</th>\n",
       "      <th>...</th>\n",
       "      <th>barcoded_gene</th>\n",
       "      <th>viral_barcode</th>\n",
       "      <th>viral_bc_UMIs</th>\n",
       "      <th>frac_viral_bc_UMIs</th>\n",
       "      <th>freq_second_infection</th>\n",
       "      <th>freq_supernatant</th>\n",
       "      <th>max_freq_second_infection</th>\n",
       "      <th>max_freq_supernatant</th>\n",
       "      <th>contributes_progeny_second_infection</th>\n",
       "      <th>contributes_progeny_supernatant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAGTAACAAGT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>47873</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCATCATTGCTT</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>90114</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAAAGATGTTGA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>111630</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAAGTACTTCCC</td>\n",
       "      <td>infected</td>\n",
       "      <td>both</td>\n",
       "      <td>56828</td>\n",
       "      <td>24082</td>\n",
       "      <td>0.423770</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>13060</td>\n",
       "      <td>420</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGAAGTAGACGTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>124341</td>\n",
       "      <td>4654</td>\n",
       "      <td>0.037429</td>\n",
       "      <td>8</td>\n",
       "      <td>319</td>\n",
       "      <td>1582</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AAGTAAGCGACATGAG</td>\n",
       "      <td>251.0</td>\n",
       "      <td>0.002019</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>1e-05</td>\n",
       "      <td>7.62399e-05</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3677</th>\n",
       "      <td>TTTGATCTCCCGTTCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>63150</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3678</th>\n",
       "      <td>TTTGGAGAGTTGCCTA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>65941</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3679</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>fluHA</td>\n",
       "      <td>AGTAAACTTCCTCGCG</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.000433</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000695</td>\n",
       "      <td>2.81273e-05</td>\n",
       "      <td>0.00179022</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3680</th>\n",
       "      <td>TTTGGAGGTATCGTTG</td>\n",
       "      <td>infected</td>\n",
       "      <td>wt</td>\n",
       "      <td>150130</td>\n",
       "      <td>3526</td>\n",
       "      <td>0.023486</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>1668</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>fluNA</td>\n",
       "      <td>ACATCTTATTTACACG</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>0.001790</td>\n",
       "      <td>2.81273e-05</td>\n",
       "      <td>0.00179022</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3681</th>\n",
       "      <td>TTTGTTGAGTGCAGCA</td>\n",
       "      <td>uninfected</td>\n",
       "      <td>none</td>\n",
       "      <td>135009</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3682 rows × 89 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cell_barcode    infected infecting_viral_tag  total_UMIs  \\\n",
       "0     AAACCCAGTAACAAGT  uninfected                none       47873   \n",
       "1     AAACCCATCATTGCTT  uninfected                none       90114   \n",
       "2     AAACGAAAGATGTTGA  uninfected                none      111630   \n",
       "3     AAACGAAGTACTTCCC    infected                both       56828   \n",
       "4     AAACGAAGTAGACGTG    infected                  wt      124341   \n",
       "...                ...         ...                 ...         ...   \n",
       "3677  TTTGATCTCCCGTTCA  uninfected                none       63150   \n",
       "3678  TTTGGAGAGTTGCCTA  uninfected                none       65941   \n",
       "3679  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "3680  TTTGGAGGTATCGTTG    infected                  wt      150130   \n",
       "3681  TTTGTTGAGTGCAGCA  uninfected                none      135009   \n",
       "\n",
       "      viral_UMIs  frac_viral_UMIs  n_viral_genes transcriptome_UMIs_fluHA  \\\n",
       "0              6         0.000125              0                        2   \n",
       "1             10         0.000111              1                        0   \n",
       "2             18         0.000161              0                        3   \n",
       "3          24082         0.423770              7                        2   \n",
       "4           4654         0.037429              8                      319   \n",
       "...          ...              ...            ...                      ...   \n",
       "3677           3         0.000048              0                        0   \n",
       "3678          12         0.000182              1                        0   \n",
       "3679        3526         0.023486              8                       93   \n",
       "3680        3526         0.023486              8                       93   \n",
       "3681           8         0.000059              1                        2   \n",
       "\n",
       "     transcriptome_UMIs_fluM transcriptome_UMIs_fluNA  ... barcoded_gene  \\\n",
       "0                          3                        0  ...           NaN   \n",
       "1                          5                        0  ...           NaN   \n",
       "2                         10                        0  ...           NaN   \n",
       "3                      13060                      420  ...           NaN   \n",
       "4                       1582                       31  ...         fluHA   \n",
       "...                      ...                      ...  ...           ...   \n",
       "3677                       1                        0  ...           NaN   \n",
       "3678                       6                        1  ...           NaN   \n",
       "3679                    1668                       45  ...         fluHA   \n",
       "3680                    1668                       45  ...         fluNA   \n",
       "3681                       2                        1  ...           NaN   \n",
       "\n",
       "         viral_barcode viral_bc_UMIs frac_viral_bc_UMIs freq_second_infection  \\\n",
       "0                  NaN           NaN                NaN                   NaN   \n",
       "1                  NaN           NaN                NaN                   NaN   \n",
       "2                  NaN           NaN                NaN                   NaN   \n",
       "3                  NaN           NaN                NaN                   NaN   \n",
       "4     AAGTAAGCGACATGAG         251.0           0.002019              0.000010   \n",
       "...                ...           ...                ...                   ...   \n",
       "3677               NaN           NaN                NaN                   NaN   \n",
       "3678               NaN           NaN                NaN                   NaN   \n",
       "3679  AGTAAACTTCCTCGCG          65.0           0.000433              0.000010   \n",
       "3680  ACATCTTATTTACACG          39.0           0.000260              0.000028   \n",
       "3681               NaN           NaN                NaN                   NaN   \n",
       "\n",
       "     freq_supernatant max_freq_second_infection max_freq_supernatant  \\\n",
       "0                 NaN                       NaN                  NaN   \n",
       "1                 NaN                       NaN                  NaN   \n",
       "2                 NaN                       NaN                  NaN   \n",
       "3                 NaN                       NaN                  NaN   \n",
       "4            0.000076                     1e-05          7.62399e-05   \n",
       "...               ...                       ...                  ...   \n",
       "3677              NaN                       NaN                  NaN   \n",
       "3678              NaN                       NaN                  NaN   \n",
       "3679         0.000695               2.81273e-05           0.00179022   \n",
       "3680         0.001790               2.81273e-05           0.00179022   \n",
       "3681              NaN                       NaN                  NaN   \n",
       "\n",
       "     contributes_progeny_second_infection contributes_progeny_supernatant  \n",
       "0                                     NaN                             NaN  \n",
       "1                                     NaN                             NaN  \n",
       "2                                     NaN                             NaN  \n",
       "3                                     NaN                             NaN  \n",
       "4                                   False                            True  \n",
       "...                                   ...                             ...  \n",
       "3677                                  NaN                             NaN  \n",
       "3678                                  NaN                             NaN  \n",
       "3679                                 True                            True  \n",
       "3680                                 True                            True  \n",
       "3681                                  NaN                             NaN  \n",
       "\n",
       "[3682 rows x 89 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "integrated_df = pd.merge(\n",
    "    left=integrated_df,\n",
    "    right=contributes_progeny,\n",
    "    on=['cell_barcode'],\n",
    "    how='left',\n",
    "    validate='many_to_one'\n",
    ")\n",
    "display(integrated_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check integrated dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check that the number of total cells has not changed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.694001Z",
     "iopub.status.busy": "2022-05-25T18:48:52.693179Z",
     "iopub.status.idle": "2022-05-25T18:48:52.697216Z",
     "shell.execute_reply": "2022-05-25T18:48:52.697729Z"
    }
   },
   "outputs": [],
   "source": [
    "assert integrated_df['cell_barcode'].nunique() == \\\n",
    "    len(cell_annotations['cell_barcode']), \\\n",
    "    \"Total number of cells changed\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check that every viral barcode has a `barcoded_gene`, `frac_viral_bc_UMIS`, `freq_supernatant`, and `freq_second_infection`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.700232Z",
     "iopub.status.busy": "2022-05-25T18:48:52.699430Z",
     "iopub.status.idle": "2022-05-25T18:48:52.792821Z",
     "shell.execute_reply": "2022-05-25T18:48:52.793434Z"
    }
   },
   "outputs": [],
   "source": [
    "assert len(integrated_df.query('viral_barcode.notnull()', engine='python')) == \\\n",
    "    len(integrated_df.query('barcoded_gene.notnull()', engine='python')) == \\\n",
    "    len(integrated_df.query('frac_viral_bc_UMIs.notnull()', engine='python')) == \\\n",
    "    len(integrated_df.query('freq_supernatant.notnull()', engine='python')) == \\\n",
    "    len(integrated_df.query('freq_second_infection.notnull()', engine='python')), \\\n",
    "    \"Mismatch in viral barcode data.\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output integrated dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-05-25T18:48:52.796243Z",
     "iopub.status.busy": "2022-05-25T18:48:52.795440Z",
     "iopub.status.idle": "2022-05-25T18:48:52.887491Z",
     "shell.execute_reply": "2022-05-25T18:48:52.888176Z"
    }
   },
   "outputs": [],
   "source": [
    "# save CSV\n",
    "integrated_df.to_csv(integrated_data_csv, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
