{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# parse_barcodes\n",
    "This notebook attempts to parse cell barcodes, UMIs, and viral barcodes from *only VIRAL reads*. It uses Read1 data to do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import\n",
    "import glob\n",
    "import os\n",
    "import gzip\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "#from ggplot import *\n",
    "\n",
    "from IPython.display import display, HTML\n",
    "\n",
    "from Bio import SeqIO\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constants\n",
    "sample_file = 'samples.csv'\n",
    "\n",
    "out_folder = 'outs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>data_type</th>\n",
       "      <th>index</th>\n",
       "      <th>data_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>transcripts</td>\n",
       "      <td>A1</td>\n",
       "      <td>out/virus_hashing/outs/fastq_path/CL3WR/HUNDREDuM_primer/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TENuM_primer</td>\n",
       "      <td>transcripts</td>\n",
       "      <td>A2</td>\n",
       "      <td>out/virus_hashing/outs/fastq_path/CL3WR/TENuM_primer/</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load samples\n",
    "samples = pd.read_csv(sample_file, comment='#')\n",
    "\n",
    "display(HTML(samples.to_html(index=False)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Handle Read 1 Data\n",
    "\n",
    "### Split features based on position in Read 1\n",
    "Now, I must parse out the features I expect to be present in each of these reads. For this, I will simply split out the sequence based on position, since each feature should be some exact length. The lengths I anticipate are designated in a cell below.\n",
    "\n",
    "The order of the features looks like this, depending on whether the adapter was appended in a way that retained the Cell Barcode and UMI or not:  \n",
    "With Cell Barcode/UMI = `TruSeq Read 1 - Cell Barcode - UMI - PolyA`  \n",
    "Without Cell Barcode/UMI = `TruSeq Read 1 - Viral Barcode - CDS`; an important note: if the molecule looks like this, the CDS will start with the 10 bp sequence `GCGGCCGCCT`.\n",
    "\n",
    "I will parse the `TruSeq Read 1` feature as the first `22 bp`.  \n",
    "I will parse the `Cell Barcode` OR `Viral Barcode` as the next `16 bp`.  \n",
    "Finally, I will parse the `UMI` OR `CDS` as the final `12 bp`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Feature Lengths\n",
    "truseq_len = 22\n",
    "bc_len = 16\n",
    "umi_cds_len = 12\n",
    "\n",
    "#HA and NA CDS sequences\n",
    "virus_constant = 'GCGGCCGCCT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing reads for sample \"HUNDREDuM_primer\"\n",
      "Parsing file out/virus_hashing/outs/fastq_path/CL3WR/HUNDREDuM_primer/HUNDREDuM_primer_S1_L001_R1_001.fastq.gz\n",
      "Done.\n",
      "\n",
      "Processing reads for sample \"TENuM_primer\"\n",
      "Parsing file out/virus_hashing/outs/fastq_path/CL3WR/TENuM_primer/TENuM_primer_S2_L001_R1_001.fastq.gz\n",
      "Done.\n",
      "\n",
      "Done loading FASTQ files.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "r1_dict = {'record_id': [],\n",
    "               'primer': [],\n",
    "               'truseq': [],\n",
    "               'bc': [],\n",
    "               'umi_or_cds': [],\n",
    "               'has_cell_bc': []\n",
    "              }\n",
    "\n",
    "# Load Read 1 Files\n",
    "for tup in samples.itertuples(index=False):\n",
    "    print(f'Processing reads for sample \"{tup.sample}\"')\n",
    "    r1files = glob.glob(os.path.join(tup.data_file, '*R1*.fastq.gz'))\n",
    "    for file in r1files:\n",
    "        print(f'Parsing file {file}')\n",
    "        with gzip.open(file, \"rt\") as gunzip_file:\n",
    "            for record in SeqIO.parse(gunzip_file, \"fastq\"):\n",
    "                # Start parsing features\n",
    "                r1_dict['record_id'].append(record.id)\n",
    "                r1_dict['primer'].append(tup.sample)\n",
    "                r1_dict['truseq'].append(str(record.seq[0:22]))\n",
    "                r1_dict['bc'].append(str(record.seq[22:38]))\n",
    "                r1_dict['umi_or_cds'].append(str(record.seq[38:50]))\n",
    "                if (virus_constant in str(record.seq[38:50])):\n",
    "                    r1_dict['has_cell_bc'].append(False)\n",
    "                else:\n",
    "                    r1_dict['has_cell_bc'].append(True)\n",
    "        print('Done.\\n')\n",
    "\n",
    "print('Done loading FASTQ files.\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converting to dataframe.\n",
      "There are 537641 reads in this data set.\n",
      "\n",
      "Dataframe head:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>primer</th>\n",
       "      <th>truseq</th>\n",
       "      <th>bc</th>\n",
       "      <th>umi_or_cds</th>\n",
       "      <th>has_cell_bc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:16081:1664</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>ACGACCGCGTCCTCTT</td>\n",
       "      <td>CCTCTTCCCCGG</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:21218:1667</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>CTCTCCTTCCTCCCTC</td>\n",
       "      <td>CCCTCTCCCCCT</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:17297:1672</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>CTCTGCCTCCCGTGTA</td>\n",
       "      <td>CAAACATTTCTC</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:21080:1736</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>TCAACGATCGCCATTA</td>\n",
       "      <td>TTCCAACCTGTT</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:15840:1750</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>AACCGCCTCCCGTTTC</td>\n",
       "      <td>CCCCTCCCTCTT</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      record_id            primer  \\\n",
       "0  M03100:474:000000000-CL3WR:1:2116:16081:1664  HUNDREDuM_primer   \n",
       "1  M03100:474:000000000-CL3WR:1:2116:21218:1667  HUNDREDuM_primer   \n",
       "2  M03100:474:000000000-CL3WR:1:2116:17297:1672  HUNDREDuM_primer   \n",
       "3  M03100:474:000000000-CL3WR:1:2116:21080:1736  HUNDREDuM_primer   \n",
       "4  M03100:474:000000000-CL3WR:1:2116:15840:1750  HUNDREDuM_primer   \n",
       "\n",
       "                   truseq                bc    umi_or_cds  has_cell_bc  \n",
       "0  CTACACGACGCTCTTCNGATCT  ACGACCGCGTCCTCTT  CCTCTTCCCCGG         True  \n",
       "1  CTACACGACGCTCTTCNGATCT  CTCTCCTTCCTCCCTC  CCCTCTCCCCCT         True  \n",
       "2  CTACACGACGCTCTTCNGATCT  CTCTGCCTCCCGTGTA  CAAACATTTCTC         True  \n",
       "3  CTACACGACGCTCTTCNGATCT  TCAACGATCGCCATTA  TTCCAACCTGTT         True  \n",
       "4  CTACACGACGCTCTTCNGATCT  AACCGCCTCCCGTTTC  CCCCTCCCTCTT         True  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Converting to dataframe.')\n",
    "r1_reads = pd.DataFrame.from_dict(r1_dict)\n",
    "\n",
    "print(f\"There are {len(r1_reads['record_id'])} reads in this data set.\\n\")\n",
    "      \n",
    "print('Dataframe head:')\n",
    "r1_reads.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a check, I just want to see that the reads being called as missing a cell barcode have the appropriate constant sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>primer</th>\n",
       "      <th>truseq</th>\n",
       "      <th>bc</th>\n",
       "      <th>umi_or_cds</th>\n",
       "      <th>has_cell_bc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1886</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:21363:5878</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>GGATTGCGGGCCTTCC</td>\n",
       "      <td>CGCGGCCGCCTA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4676</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:8072:7817</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>TAAGGCCCTGATATTC</td>\n",
       "      <td>GCGGCCGCCTGT</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8246</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:17775:9854</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>GATGGTGGTACCTGCC</td>\n",
       "      <td>GCGGCCGCCTAT</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8947</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:20841:10190</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>GATGGTGGTACCTGCC</td>\n",
       "      <td>GCGGCCGCCTAT</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9054</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:15390:10235</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>GAAATTTGCGGTTCCG</td>\n",
       "      <td>GCGGCCGCCTTG</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          record_id            primer  \\\n",
       "1886   M03100:474:000000000-CL3WR:1:2116:21363:5878  HUNDREDuM_primer   \n",
       "4676    M03100:474:000000000-CL3WR:1:2116:8072:7817  HUNDREDuM_primer   \n",
       "8246   M03100:474:000000000-CL3WR:1:2116:17775:9854  HUNDREDuM_primer   \n",
       "8947  M03100:474:000000000-CL3WR:1:2116:20841:10190  HUNDREDuM_primer   \n",
       "9054  M03100:474:000000000-CL3WR:1:2116:15390:10235  HUNDREDuM_primer   \n",
       "\n",
       "                      truseq                bc    umi_or_cds  has_cell_bc  \n",
       "1886  CTACACGACGCTCTTCNGATCT  GGATTGCGGGCCTTCC  CGCGGCCGCCTA        False  \n",
       "4676  CTACACGACGCTCTTCNGATCT  TAAGGCCCTGATATTC  GCGGCCGCCTGT        False  \n",
       "8246  CTACACGACGCTCTTCNGATCT  GATGGTGGTACCTGCC  GCGGCCGCCTAT        False  \n",
       "8947  CTACACGACGCTCTTCNGATCT  GATGGTGGTACCTGCC  GCGGCCGCCTAT        False  \n",
       "9054  CTACACGACGCTCTTCNGATCT  GAAATTTGCGGTTCCG  GCGGCCGCCTTG        False  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r1_reads[r1_reads['has_cell_bc']==False].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many molecules are missing their cell barcode?\n",
    "\n",
    "Now, I want to ask how many reads from each sample are missing the cell barcode? This will be indicated by having the expected constant region from the virus CDS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>truseq</th>\n",
       "      <th>bc</th>\n",
       "      <th>umi_or_cds</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>primer</th>\n",
       "      <th>has_cell_bc</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">HUNDREDuM_primer</td>\n",
       "      <td>False</td>\n",
       "      <td>276</td>\n",
       "      <td>276</td>\n",
       "      <td>276</td>\n",
       "      <td>276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>300037</td>\n",
       "      <td>300037</td>\n",
       "      <td>300037</td>\n",
       "      <td>300037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">TENuM_primer</td>\n",
       "      <td>False</td>\n",
       "      <td>266</td>\n",
       "      <td>266</td>\n",
       "      <td>266</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>237062</td>\n",
       "      <td>237062</td>\n",
       "      <td>237062</td>\n",
       "      <td>237062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              record_id  truseq      bc  umi_or_cds\n",
       "primer           has_cell_bc                                       \n",
       "HUNDREDuM_primer False              276     276     276         276\n",
       "                 True            300037  300037  300037      300037\n",
       "TENuM_primer     False              266     266     266         266\n",
       "                 True            237062  237062  237062      237062"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r1_reads.groupby(['primer','has_cell_bc']).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "About ~270 reads of each sample appear to be missing a cell barcode. To know what proportion of the viral reads this is, we'll have to map each read to virus or host."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Handle Read 2 Data\n",
    "\n",
    "### Use Python to Call a read as viral/not viral\n",
    "The fastest way to figure out how many of our viral reads retain their cell barcode is just to search for a snippet of read 2 in the viral genome.\n",
    "\n",
    "First, I will define the viral genome.  I have directly copied a script which automatically generates FASTA files for this from another project (see [here](https://github.com/jbloomlab/pdmH1N1_flu_single_cell/tree/master/data/flu_sequences)). \n",
    "Then, I will trim Read 2 back to ~50 bp.\n",
    "Finally, I will check to see if each read is in the viral genome. These will be called viral molecules, and all others will be assumed to be host.\n",
    "\n",
    "As a technical note, I checked to see if strings which span two viral segments (e.g. str(End of PB2 + End of PB1)) are returned as being valid parts of the viral genome, and they are not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constants\n",
    "flu_genome_folder = 'flu_sequences'\n",
    "flu_files = ['flu-CA09.fasta', 'flu-CA09-dblSyn.fasta']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in flu genome from file: flu-CA09.fasta\n",
      "Reading in sequence for fluPB2.\n",
      "Reading in sequence for fluPB1.\n",
      "Reading in sequence for fluPA.\n",
      "Reading in sequence for fluHA.\n",
      "Reading in sequence for fluNP.\n",
      "Reading in sequence for fluNA.\n",
      "Reading in sequence for fluM.\n",
      "Reading in sequence for fluNS.\n",
      "Done with file.\n",
      "\n",
      "Reading in flu genome from file: flu-CA09-dblSyn.fasta\n",
      "Reading in sequence for fluPB2.\n",
      "Reading in sequence for fluPB1.\n",
      "Reading in sequence for fluPA.\n",
      "Reading in sequence for fluHA.\n",
      "Reading in sequence for fluNP.\n",
      "Reading in sequence for fluNA.\n",
      "Reading in sequence for fluM.\n",
      "Reading in sequence for fluNS.\n",
      "Done with file.\n",
      "\n",
      "Finished loading flu genomes.\n",
      "There were 16 sequences loaded from 2 files.\n"
     ]
    }
   ],
   "source": [
    "# Load in viral genomes and make into a single string that can be searched\n",
    "seqs = list()\n",
    "\n",
    "for file in flu_files:\n",
    "    print(f'Reading in flu genome from file: {file}')\n",
    "    file_path = flu_genome_folder + '/' + file\n",
    "    with open(file_path) as open_file:\n",
    "        for record in SeqIO.parse(open_file, \"fasta\"):\n",
    "            print(f\"Reading in sequence for {record.id}.\")\n",
    "            seqs.append(str(record.seq))\n",
    "    print('Done with file.\\n')\n",
    "\n",
    "print(\"Finished loading flu genomes.\")\n",
    "print(f\"There were {len(seqs)} sequences loaded from {len(flu_files)} files.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read 2 trim length\n",
    "trim_len = 100 #91 recommended as minimum by 10X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing reads for sample \"HUNDREDuM_primer\"\n",
      "Parsing file out/virus_hashing/outs/fastq_path/CL3WR/HUNDREDuM_primer/HUNDREDuM_primer_S1_L001_R2_001.fastq.gz\n",
      "50000 sequences processed.\n",
      "100000 sequences processed.\n",
      "150000 sequences processed.\n",
      "200000 sequences processed.\n",
      "250000 sequences processed.\n",
      "300000 sequences processed.\n",
      "Done.\n",
      "\n",
      "Processing reads for sample \"TENuM_primer\"\n",
      "Parsing file out/virus_hashing/outs/fastq_path/CL3WR/TENuM_primer/TENuM_primer_S2_L001_R2_001.fastq.gz\n",
      "50000 sequences processed.\n",
      "100000 sequences processed.\n",
      "150000 sequences processed.\n",
      "200000 sequences processed.\n",
      "Done.\n",
      "\n",
      "Done loading FASTQ files.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "r2_simple_dict = {'record_id': [],\n",
    "               'primer': [],\n",
    "               'cds': [],\n",
    "               'host_or_virus': []\n",
    "              }\n",
    "\n",
    "# Load Read 2 Files\n",
    "for tup in samples.itertuples(index=False):\n",
    "    print(f'Processing reads for sample \"{tup.sample}\"')\n",
    "    r1files = glob.glob(os.path.join(tup.data_file, '*R2*.fastq.gz'))\n",
    "    for file in r1files:\n",
    "        if ('qualfiltered' in file) or ('trimmed' in file):\n",
    "            pass\n",
    "        else:\n",
    "            print(f'Parsing file {file}')\n",
    "            with gzip.open(file, \"rt\") as gunzip_file:\n",
    "                count = 0\n",
    "                for record in SeqIO.parse(gunzip_file, \"fastq\"):\n",
    "                    count += 1\n",
    "                    if count % 50000 == 0:\n",
    "                        print(f\"{count} sequences processed.\")\n",
    "                    # Start parsing features\n",
    "                    r2_simple_dict['record_id'].append(record.id)\n",
    "                    r2_simple_dict['primer'].append(tup.sample)\n",
    "                    r2_simple_dict['cds'].append(str(record.seq[0:100]))\n",
    "                    if str(record.seq[0:100]) in str(seqs):\n",
    "                        r2_simple_dict['host_or_virus'].append('Virus')\n",
    "                    else:\n",
    "                        r2_simple_dict['host_or_virus'].append('Host')\n",
    "            print('Done.\\n')\n",
    "\n",
    "print('Done loading FASTQ files.\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converting to dataframe.\n",
      "There are 537641 reads in this data set.\n",
      "\n",
      "Dataframe head:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>primer</th>\n",
       "      <th>cds</th>\n",
       "      <th>host_or_virus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:16081:1664</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>ACGCAGTCGTATCAACCCCCCGTACATCCCCCCCCTCCTCCCCCTC...</td>\n",
       "      <td>Host</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:21218:1667</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTCCCGCCGTCTCTCCCCCCCCTCCCTCCCCCCCCCCCCCCCCCCC...</td>\n",
       "      <td>Host</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:17297:1672</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>ACCCAGTCCTATCAACCCCCCCTACCTCCCCCCCCCCCCCCCTCCC...</td>\n",
       "      <td>Host</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:21080:1736</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>ATTCCTTACACTCACCCACTCGTCTCTCCTTTCCCCCCCTCTTCCC...</td>\n",
       "      <td>Host</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:15840:1750</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTTGACATCTCCATCCCACACCACCCCAACCTCCCCCCTCCCACTT...</td>\n",
       "      <td>Host</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      record_id            primer  \\\n",
       "0  M03100:474:000000000-CL3WR:1:2116:16081:1664  HUNDREDuM_primer   \n",
       "1  M03100:474:000000000-CL3WR:1:2116:21218:1667  HUNDREDuM_primer   \n",
       "2  M03100:474:000000000-CL3WR:1:2116:17297:1672  HUNDREDuM_primer   \n",
       "3  M03100:474:000000000-CL3WR:1:2116:21080:1736  HUNDREDuM_primer   \n",
       "4  M03100:474:000000000-CL3WR:1:2116:15840:1750  HUNDREDuM_primer   \n",
       "\n",
       "                                                 cds host_or_virus  \n",
       "0  ACGCAGTCGTATCAACCCCCCGTACATCCCCCCCCTCCTCCCCCTC...          Host  \n",
       "1  CTCCCGCCGTCTCTCCCCCCCCTCCCTCCCCCCCCCCCCCCCCCCC...          Host  \n",
       "2  ACCCAGTCCTATCAACCCCCCCTACCTCCCCCCCCCCCCCCCTCCC...          Host  \n",
       "3  ATTCCTTACACTCACCCACTCGTCTCTCCTTTCCCCCCCTCTTCCC...          Host  \n",
       "4  CTTGACATCTCCATCCCACACCACCCCAACCTCCCCCCTCCCACTT...          Host  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Converting to dataframe.')\n",
    "r2_simple_reads = pd.DataFrame.from_dict(r2_simple_dict)\n",
    "\n",
    "print(f\"There are {len(r2_simple_reads['record_id'])} reads in this data set.\\n\")\n",
    "\n",
    "print('Dataframe head:')\n",
    "r2_simple_reads.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>cds</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>primer</th>\n",
       "      <th>host_or_virus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">HUNDREDuM_primer</td>\n",
       "      <td>Host</td>\n",
       "      <td>300164</td>\n",
       "      <td>300164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Virus</td>\n",
       "      <td>149</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">TENuM_primer</td>\n",
       "      <td>Host</td>\n",
       "      <td>237213</td>\n",
       "      <td>237213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Virus</td>\n",
       "      <td>115</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                record_id     cds\n",
       "primer           host_or_virus                   \n",
       "HUNDREDuM_primer Host              300164  300164\n",
       "                 Virus                149     149\n",
       "TENuM_primer     Host              237213  237213\n",
       "                 Virus                115     115"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r2_simple_reads.groupby(['primer','host_or_virus']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 537641 reads in this data set.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>primer</th>\n",
       "      <th>truseq</th>\n",
       "      <th>bc</th>\n",
       "      <th>umi_or_cds</th>\n",
       "      <th>has_cell_bc</th>\n",
       "      <th>cds</th>\n",
       "      <th>host_or_virus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:16081:1664</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>ACGACCGCGTCCTCTT</td>\n",
       "      <td>CCTCTTCCCCGG</td>\n",
       "      <td>True</td>\n",
       "      <td>ACGCAGTCGTATCAACCCCCCGTACATCCCCCCCCTCCTCCCCCTC...</td>\n",
       "      <td>Host</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:21218:1667</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>CTCTCCTTCCTCCCTC</td>\n",
       "      <td>CCCTCTCCCCCT</td>\n",
       "      <td>True</td>\n",
       "      <td>CTCCCGCCGTCTCTCCCCCCCCTCCCTCCCCCCCCCCCCCCCCCCC...</td>\n",
       "      <td>Host</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:17297:1672</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>CTCTGCCTCCCGTGTA</td>\n",
       "      <td>CAAACATTTCTC</td>\n",
       "      <td>True</td>\n",
       "      <td>ACCCAGTCCTATCAACCCCCCCTACCTCCCCCCCCCCCCCCCTCCC...</td>\n",
       "      <td>Host</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:21080:1736</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>TCAACGATCGCCATTA</td>\n",
       "      <td>TTCCAACCTGTT</td>\n",
       "      <td>True</td>\n",
       "      <td>ATTCCTTACACTCACCCACTCGTCTCTCCTTTCCCCCCCTCTTCCC...</td>\n",
       "      <td>Host</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:15840:1750</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>AACCGCCTCCCGTTTC</td>\n",
       "      <td>CCCCTCCCTCTT</td>\n",
       "      <td>True</td>\n",
       "      <td>CTTGACATCTCCATCCCACACCACCCCAACCTCCCCCCTCCCACTT...</td>\n",
       "      <td>Host</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      record_id            primer  \\\n",
       "0  M03100:474:000000000-CL3WR:1:2116:16081:1664  HUNDREDuM_primer   \n",
       "1  M03100:474:000000000-CL3WR:1:2116:21218:1667  HUNDREDuM_primer   \n",
       "2  M03100:474:000000000-CL3WR:1:2116:17297:1672  HUNDREDuM_primer   \n",
       "3  M03100:474:000000000-CL3WR:1:2116:21080:1736  HUNDREDuM_primer   \n",
       "4  M03100:474:000000000-CL3WR:1:2116:15840:1750  HUNDREDuM_primer   \n",
       "\n",
       "                   truseq                bc    umi_or_cds  has_cell_bc  \\\n",
       "0  CTACACGACGCTCTTCNGATCT  ACGACCGCGTCCTCTT  CCTCTTCCCCGG         True   \n",
       "1  CTACACGACGCTCTTCNGATCT  CTCTCCTTCCTCCCTC  CCCTCTCCCCCT         True   \n",
       "2  CTACACGACGCTCTTCNGATCT  CTCTGCCTCCCGTGTA  CAAACATTTCTC         True   \n",
       "3  CTACACGACGCTCTTCNGATCT  TCAACGATCGCCATTA  TTCCAACCTGTT         True   \n",
       "4  CTACACGACGCTCTTCNGATCT  AACCGCCTCCCGTTTC  CCCCTCCCTCTT         True   \n",
       "\n",
       "                                                 cds host_or_virus  \n",
       "0  ACGCAGTCGTATCAACCCCCCGTACATCCCCCCCCTCCTCCCCCTC...          Host  \n",
       "1  CTCCCGCCGTCTCTCCCCCCCCTCCCTCCCCCCCCCCCCCCCCCCC...          Host  \n",
       "2  ACCCAGTCCTATCAACCCCCCCTACCTCCCCCCCCCCCCCCCTCCC...          Host  \n",
       "3  ATTCCTTACACTCACCCACTCGTCTCTCCTTTCCCCCCCTCTTCCC...          Host  \n",
       "4  CTTGACATCTCCATCCCACACCACCCCAACCTCCCCCCTCCCACTT...          Host  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_simple_reads = pd.merge(r1_reads, r2_simple_reads, on=[\"record_id\",\"primer\"])\n",
    "\n",
    "print(f\"There are {len(merged_simple_reads['record_id'])} reads in this data set.\\n\")\n",
    "\n",
    "merged_simple_reads.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many viral reads are missing cell barcodes?\n",
    "\n",
    "Next, I want to ask how many viral reads, across both samples, are missing their cell_barcodes in Read 1. These reads may have to be thrown out, or they could possibly be parsed fully from read 2 if the molecule is short enough."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>primer</th>\n",
       "      <th>truseq</th>\n",
       "      <th>bc</th>\n",
       "      <th>umi_or_cds</th>\n",
       "      <th>cds</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>host_or_virus</th>\n",
       "      <th>has_cell_bc</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">Host</td>\n",
       "      <td>False</td>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>536844</td>\n",
       "      <td>536844</td>\n",
       "      <td>536844</td>\n",
       "      <td>536844</td>\n",
       "      <td>536844</td>\n",
       "      <td>536844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">Virus</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>255</td>\n",
       "      <td>255</td>\n",
       "      <td>255</td>\n",
       "      <td>255</td>\n",
       "      <td>255</td>\n",
       "      <td>255</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           record_id  primer  truseq      bc  umi_or_cds  \\\n",
       "host_or_virus has_cell_bc                                                  \n",
       "Host          False              533     533     533     533         533   \n",
       "              True            536844  536844  536844  536844      536844   \n",
       "Virus         False                9       9       9       9           9   \n",
       "              True               255     255     255     255         255   \n",
       "\n",
       "                              cds  \n",
       "host_or_virus has_cell_bc          \n",
       "Host          False           533  \n",
       "              True         536844  \n",
       "Virus         False             9  \n",
       "              True            255  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_simple_reads.groupby(['host_or_virus','has_cell_bc']).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On the plus side, most reads that are called as viral have a cell barcode.\n",
    "\n",
    "**Suprisingly, however, I am seeing a bunch of reads which I cannot map back to the viral genome which are missing cell barcodes. I would be surprised if these are really host reads. Perhaps they have mutations and were not exact matches?** I will print a few of these reads and blast them to see what they are."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTTGTATAGAAGAGAGGAGCACCAGGAAGTGGGCGAACGGCGCGTTGCAATGCCGGACTAGCATAAAATGATGCACCATAGGCGGGCGCGAGGAGGCGAA\n",
      "CTGTCCTCCACTCTTCCGCACTGTACAACATACCAAAAACTCCAACACTTATAGCCACGAAAGCTAAGTAGGGTTGGTAGAAGTGTGCACTGGGCGAGCC\n",
      "AAGCAGTGGTATCAACGCAGAGTACATGGGGACCCCCGAGCCGCAAAAGCAGGGGAAAACAAAAGCAACCAAATTGAAAGCAATACTAGTAGTTCTGCTA\n",
      "AAGCAGTGGTATCAACGCAGAGTACACGGGGGCCGGCCGCGCAGCAAAAGCAGGGGAAAACAAAAGCAACCAAATTGAAAGCAATACTAGTAGTTCTGCT\n",
      "AGCAGTGGTATCAACGCAGAGTACATGGGGGCAGTAAGATCGCGAAAGCAGGAGTTTAAACCGAATCCAAACCAGAAAATAATAACCATTGGGTCAATCT\n",
      "CAGCAGTCGCATCAACGCAGAGTACACGGGAGATTACTCGCAGCAAAAGCAGGGGAAAACAAAAGCAACCAAATTGAAAGCAAGACAAGTAGTTCTGCTA\n",
      "AAGCAGTGGTATCAACGCAGAGTACATGGGGGGTAGAGGCTGCAAAAGCAGGGGAAAACAAAAGCAACCAAATTGAAAGCAATACTAGTAGTTCTGCTAT\n",
      "AAATGAGTGGACCGGCTCTAGCGGGAGTTTTGTTCAGCATCCAGAACTAACAGGGCTGGATTGTATAAGACCTTGCTTCTGGGTTGAACTAATCCGAGGG\n",
      "AAGCAGTGGTATCAACGCAGAGTACATGGGAGGCTTGAGCAGCAAAAGCAGGGGAAAACAAAAGCAACCAAATTGAAAGCAATACTAGTAGTTCTGCTAT\n",
      "CAGTGGTATCAACGCAGAGTACATGGGAAAACAAAAGCAACCAAATTGAAAGCAATACTAGTAGTTCTGCTATATACATTTGCAACCGCAATTGCAGACA\n"
     ]
    }
   ],
   "source": [
    "for item in merged_simple_reads[(merged_simple_reads['host_or_virus'] == 'Host') & (merged_simple_reads['has_cell_bc'] == False)].iloc[0:10]['cds']:\n",
    "    print(item)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I then blasted the 10 sequences printed above. The results follow here:  \n",
    "1 did not map to any sequence in the BLASTn database.  \n",
    "2 mapped to a contemporary pdmH1N1 sequences, with mutations.  \n",
    "3 mapped to several contemporary pdmH1N1 sequences, with mutations.  \n",
    "4 mapped to a variety of historic H1N1 sequences (e.g. WSN) and several avian H3N8 sequences, all with mutatations.  \n",
    "5 mapped to several contemporary pdmH1N1 sequences, with mutations.  \n",
    "6 mapped to a contemporary pdmH1N1 sequences, with mutations.  \n",
    "7 mapped to several contemporary pdmH1N1 sequences, with mutations.  \n",
    "8 mapped to several contemporary pdmH1N1 sequences, with mutations.  \n",
    "9 mapped to several contemporary pdmH1N1 sequences, with mutations.  \n",
    "10 mapped to several contemporary pdmH1N1 sequences, with mutations.  \n",
    "\n",
    "Based on these results, I conclude that **most of these sequences are truly virus sequences, but not exact matches.** For that reason, I will have to align these using bowtie instead to allow some flexibility."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use Bowtie2 to call a read as viral/not viral\n",
    "\n",
    "Since the simple search strategy I used above was not very successful -- it broke on mutations -- I have used bowtie2 to _locally_ align trimmed R2 reads to the influenza reference genome.\n",
    "\n",
    "The following actions took place manually in the command line. This is not ideal, and I will eventually go back and automate these.\n",
    "\n",
    "First, I concatenated the flu-CA09 and flu-CA09-doublesyn fasta files into one flu_genome.fasta.  \n",
    "Then, I used `bowtie2-build` to make a flu reference genome out of this.\n",
    "\n",
    "I trimmed my R2 reads for quality (q score = 10+33), then trimmed them down to 100 bp using `cutadapt`.\n",
    "\n",
    "Finally, I used `bowtie2 -local` to align the reads from each sample to the reference genome. This allows for deletions.\n",
    "\n",
    "I converted the SAM files to BAM files and sorted them. They are stored in the `out` folder.\n",
    "\n",
    "Next, I will load these BAM files of all reads that mapped to a flu transcript. I will extract the record ID and merge this on the R1 dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing reads for sample \"HUNDREDuM_primer\"\n",
      "Reading in BAM file: out/HUNDREDuM_aligned.sam\n",
      "Done.\n",
      "\n",
      "Processing reads for sample \"TENuM_primer\"\n",
      "Reading in BAM file: out/TENuM_aligned.sam\n",
      "Done.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Load BAM Files\n",
    "r2_bowtie_dict = {'record_id': [],\n",
    "               'primer': [],\n",
    "               'cds': [],\n",
    "               'host_or_virus': [],\n",
    "               'gene_name': []\n",
    "              }\n",
    "\n",
    "for tup in samples.itertuples(index=False):\n",
    "    print(f'Processing reads for sample \"{tup.sample}\"')\n",
    "    sample_prefix = tup.sample.split('_')\n",
    "    bamfile = os.path.join('out', sample_prefix[0] + '_aligned.sam')\n",
    "    print(f'Reading in BAM file: {bamfile}')\n",
    "    bam_object = pysam.AlignmentFile(bamfile, \"r\")\n",
    "    for record in bam_object:\n",
    "            # Start parsing features\n",
    "            r2_bowtie_dict['record_id'].append(record.query_name)\n",
    "            r2_bowtie_dict['primer'].append(tup.sample)\n",
    "            r2_bowtie_dict['cds'].append(record.query_sequence)\n",
    "            if record.is_unmapped == True:\n",
    "                r2_bowtie_dict['host_or_virus'].append('Host')\n",
    "                r2_bowtie_dict['gene_name'].append('unmapped')\n",
    "            else:\n",
    "                r2_bowtie_dict['host_or_virus'].append('Virus')\n",
    "                r2_bowtie_dict['gene_name'].append(record.reference_name)\n",
    "    print('Done.\\n')\n",
    "        \n",
    "bam_object.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converting to dataframe.\n",
      "There are 537641 reads in this data set.\n",
      "\n",
      "Dataframe head:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>primer</th>\n",
       "      <th>cds</th>\n",
       "      <th>host_or_virus</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:16081:1664</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>ACGCAGTCGTATCAACCCCCCGTACATCCCCCCCCTCCTCCCCCTC...</td>\n",
       "      <td>Host</td>\n",
       "      <td>unmapped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:21218:1667</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTCCCGCCGTCTCTCCCCCCCCTCCCTCCCCCCCCCCCCCCCCCCC...</td>\n",
       "      <td>Host</td>\n",
       "      <td>unmapped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:17297:1672</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>ACCCAGTCCTATCAACCCCCCCTACCTCCCCCCCCCCCCCCCTCCC...</td>\n",
       "      <td>Host</td>\n",
       "      <td>unmapped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:21080:1736</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>ATTCCTTACACTCACCCACTCGTCTCTCCTTTCCCCCCCTCTTCCC...</td>\n",
       "      <td>Host</td>\n",
       "      <td>unmapped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:15840:1750</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTTGACATCTCCATCCCACACCACCCCAACCTCCCCCCTCCCACTT...</td>\n",
       "      <td>Host</td>\n",
       "      <td>unmapped</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      record_id            primer  \\\n",
       "0  M03100:474:000000000-CL3WR:1:2116:16081:1664  HUNDREDuM_primer   \n",
       "1  M03100:474:000000000-CL3WR:1:2116:21218:1667  HUNDREDuM_primer   \n",
       "2  M03100:474:000000000-CL3WR:1:2116:17297:1672  HUNDREDuM_primer   \n",
       "3  M03100:474:000000000-CL3WR:1:2116:21080:1736  HUNDREDuM_primer   \n",
       "4  M03100:474:000000000-CL3WR:1:2116:15840:1750  HUNDREDuM_primer   \n",
       "\n",
       "                                                 cds host_or_virus gene_name  \n",
       "0  ACGCAGTCGTATCAACCCCCCGTACATCCCCCCCCTCCTCCCCCTC...          Host  unmapped  \n",
       "1  CTCCCGCCGTCTCTCCCCCCCCTCCCTCCCCCCCCCCCCCCCCCCC...          Host  unmapped  \n",
       "2  ACCCAGTCCTATCAACCCCCCCTACCTCCCCCCCCCCCCCCCTCCC...          Host  unmapped  \n",
       "3  ATTCCTTACACTCACCCACTCGTCTCTCCTTTCCCCCCCTCTTCCC...          Host  unmapped  \n",
       "4  CTTGACATCTCCATCCCACACCACCCCAACCTCCCCCCTCCCACTT...          Host  unmapped  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Converting to dataframe.')\n",
    "r2_bowtie_reads = pd.DataFrame.from_dict(r2_bowtie_dict)\n",
    "\n",
    "print(f\"There are {len(r2_bowtie_reads['record_id'])} reads in this data set.\\n\")\n",
    "\n",
    "print('Dataframe head:')\n",
    "r2_bowtie_reads.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>cds</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>primer</th>\n",
       "      <th>host_or_virus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">HUNDREDuM_primer</td>\n",
       "      <td>Host</td>\n",
       "      <td>299598</td>\n",
       "      <td>299598</td>\n",
       "      <td>299598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Virus</td>\n",
       "      <td>715</td>\n",
       "      <td>715</td>\n",
       "      <td>715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">TENuM_primer</td>\n",
       "      <td>Host</td>\n",
       "      <td>236807</td>\n",
       "      <td>236807</td>\n",
       "      <td>236807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Virus</td>\n",
       "      <td>521</td>\n",
       "      <td>521</td>\n",
       "      <td>521</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                record_id     cds  gene_name\n",
       "primer           host_or_virus                              \n",
       "HUNDREDuM_primer Host              299598  299598     299598\n",
       "                 Virus                715     715        715\n",
       "TENuM_primer     Host              236807  236807     236807\n",
       "                 Virus                521     521        521"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r2_bowtie_reads.groupby(['primer','host_or_virus']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 537641 reads in this data set.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>primer</th>\n",
       "      <th>truseq</th>\n",
       "      <th>bc</th>\n",
       "      <th>umi_or_cds</th>\n",
       "      <th>has_cell_bc</th>\n",
       "      <th>cds</th>\n",
       "      <th>host_or_virus</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:16081:1664</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>ACGACCGCGTCCTCTT</td>\n",
       "      <td>CCTCTTCCCCGG</td>\n",
       "      <td>True</td>\n",
       "      <td>ACGCAGTCGTATCAACCCCCCGTACATCCCCCCCCTCCTCCCCCTC...</td>\n",
       "      <td>Host</td>\n",
       "      <td>unmapped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:21218:1667</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>CTCTCCTTCCTCCCTC</td>\n",
       "      <td>CCCTCTCCCCCT</td>\n",
       "      <td>True</td>\n",
       "      <td>CTCCCGCCGTCTCTCCCCCCCCTCCCTCCCCCCCCCCCCCCCCCCC...</td>\n",
       "      <td>Host</td>\n",
       "      <td>unmapped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:17297:1672</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>CTCTGCCTCCCGTGTA</td>\n",
       "      <td>CAAACATTTCTC</td>\n",
       "      <td>True</td>\n",
       "      <td>ACCCAGTCCTATCAACCCCCCCTACCTCCCCCCCCCCCCCCCTCCC...</td>\n",
       "      <td>Host</td>\n",
       "      <td>unmapped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:21080:1736</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>TCAACGATCGCCATTA</td>\n",
       "      <td>TTCCAACCTGTT</td>\n",
       "      <td>True</td>\n",
       "      <td>ATTCCTTACACTCACCCACTCGTCTCTCCTTTCCCCCCCTCTTCCC...</td>\n",
       "      <td>Host</td>\n",
       "      <td>unmapped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>M03100:474:000000000-CL3WR:1:2116:15840:1750</td>\n",
       "      <td>HUNDREDuM_primer</td>\n",
       "      <td>CTACACGACGCTCTTCNGATCT</td>\n",
       "      <td>AACCGCCTCCCGTTTC</td>\n",
       "      <td>CCCCTCCCTCTT</td>\n",
       "      <td>True</td>\n",
       "      <td>CTTGACATCTCCATCCCACACCACCCCAACCTCCCCCCTCCCACTT...</td>\n",
       "      <td>Host</td>\n",
       "      <td>unmapped</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      record_id            primer  \\\n",
       "0  M03100:474:000000000-CL3WR:1:2116:16081:1664  HUNDREDuM_primer   \n",
       "1  M03100:474:000000000-CL3WR:1:2116:21218:1667  HUNDREDuM_primer   \n",
       "2  M03100:474:000000000-CL3WR:1:2116:17297:1672  HUNDREDuM_primer   \n",
       "3  M03100:474:000000000-CL3WR:1:2116:21080:1736  HUNDREDuM_primer   \n",
       "4  M03100:474:000000000-CL3WR:1:2116:15840:1750  HUNDREDuM_primer   \n",
       "\n",
       "                   truseq                bc    umi_or_cds  has_cell_bc  \\\n",
       "0  CTACACGACGCTCTTCNGATCT  ACGACCGCGTCCTCTT  CCTCTTCCCCGG         True   \n",
       "1  CTACACGACGCTCTTCNGATCT  CTCTCCTTCCTCCCTC  CCCTCTCCCCCT         True   \n",
       "2  CTACACGACGCTCTTCNGATCT  CTCTGCCTCCCGTGTA  CAAACATTTCTC         True   \n",
       "3  CTACACGACGCTCTTCNGATCT  TCAACGATCGCCATTA  TTCCAACCTGTT         True   \n",
       "4  CTACACGACGCTCTTCNGATCT  AACCGCCTCCCGTTTC  CCCCTCCCTCTT         True   \n",
       "\n",
       "                                                 cds host_or_virus gene_name  \n",
       "0  ACGCAGTCGTATCAACCCCCCGTACATCCCCCCCCTCCTCCCCCTC...          Host  unmapped  \n",
       "1  CTCCCGCCGTCTCTCCCCCCCCTCCCTCCCCCCCCCCCCCCCCCCC...          Host  unmapped  \n",
       "2  ACCCAGTCCTATCAACCCCCCCTACCTCCCCCCCCCCCCCCCTCCC...          Host  unmapped  \n",
       "3  ATTCCTTACACTCACCCACTCGTCTCTCCTTTCCCCCCCTCTTCCC...          Host  unmapped  \n",
       "4  CTTGACATCTCCATCCCACACCACCCCAACCTCCCCCCTCCCACTT...          Host  unmapped  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_bowtie_reads = pd.merge(r1_reads, r2_bowtie_reads, on=[\"record_id\",\"primer\"])\n",
    "\n",
    "print(f\"There are {len(merged_bowtie_reads['record_id'])} reads in this data set.\\n\")\n",
    "\n",
    "merged_bowtie_reads.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many viral reads are missing cell barcodes?\n",
    "\n",
    "Next, I want to ask how many viral reads, across both samples, are missing their cell_barcodes in Read 1. These reads may have to be thrown out, or they could possibly be parsed fully from read 2 if the molecule is short enough."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>primer</th>\n",
       "      <th>truseq</th>\n",
       "      <th>bc</th>\n",
       "      <th>umi_or_cds</th>\n",
       "      <th>cds</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>host_or_virus</th>\n",
       "      <th>has_cell_bc</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">Host</td>\n",
       "      <td>False</td>\n",
       "      <td>222</td>\n",
       "      <td>222</td>\n",
       "      <td>222</td>\n",
       "      <td>222</td>\n",
       "      <td>222</td>\n",
       "      <td>222</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>536183</td>\n",
       "      <td>536183</td>\n",
       "      <td>536183</td>\n",
       "      <td>536183</td>\n",
       "      <td>536183</td>\n",
       "      <td>536183</td>\n",
       "      <td>536183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">Virus</td>\n",
       "      <td>False</td>\n",
       "      <td>320</td>\n",
       "      <td>320</td>\n",
       "      <td>320</td>\n",
       "      <td>320</td>\n",
       "      <td>320</td>\n",
       "      <td>320</td>\n",
       "      <td>320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>916</td>\n",
       "      <td>916</td>\n",
       "      <td>916</td>\n",
       "      <td>916</td>\n",
       "      <td>916</td>\n",
       "      <td>916</td>\n",
       "      <td>916</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           record_id  primer  truseq      bc  umi_or_cds  \\\n",
       "host_or_virus has_cell_bc                                                  \n",
       "Host          False              222     222     222     222         222   \n",
       "              True            536183  536183  536183  536183      536183   \n",
       "Virus         False              320     320     320     320         320   \n",
       "              True               916     916     916     916         916   \n",
       "\n",
       "                              cds  gene_name  \n",
       "host_or_virus has_cell_bc                     \n",
       "Host          False           222        222  \n",
       "              True         536183     536183  \n",
       "Virus         False           320        320  \n",
       "              True            916        916  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_bowtie_reads.groupby(['host_or_virus','has_cell_bc']).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Based on these results, it looks like there are approximately `1200 virus reads` in this dataset. About `3/4 (~900)` of these have a cell barcode! This is promising and suggests that we can use this data to study multiple infection!**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are also some reads that fail to map to the virus genome (which are called \"Host\" by default), but have a virus constant sequence at the expected position in read 1. These are a small minority of \"Host\" reads, so could totally be thrown out there. However, if they are actually \"Virus\" reads that failed to map, it would be nice to figure out how to map them succesfully, so we can include them in our analysis. As I did before, I will BLast them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTTGTATAGAAGAGAGGAGCACCAGGAAGTGGGCGAACGGCGCGTTGCAATGCCGGACTAGCATAAAATGATGCACCATAGGCGGGCGCGAGGAGGCGAA\n",
      "CTGTCCTCCACTCTTCCGCACTGTACAACATACCAAAAACTCCAACACTTATAGCCACGAAAGCTAAGTAGGGTTGGTAGAAGTGTGCACTGGGCGAGCC\n",
      "AGCAGTGGTATCAACGCAGAGTACATGGGGGCAGTAAGATCGCGAAAGCAGGAGTTTAAACCGAATCCAAACCAGAAAATAATAACCATTGGGTCAATCT\n",
      "CAGCAGTCGCATCAACGCAGAGTACACGGGAGATTACTCGCAGCAAAAGCAGGGGAAAACAAAAGCAACCAAATTGAAAGCAAGACAAGTAGTTCTGCTA\n",
      "AGTGGTATCAACGCAGAGTACATGGGTAGTGTGAAACGCGAAAGCAGGAGTTTAAACCGAATCCAAACCAGAAAATAATAACCATTGGGTCAATCTGTAT\n",
      "AAGCAGTGGTATCAACGCAGAGTACATGGGGACACTCACGGCGCGAAAGCAGGAGTTTAAACCGAATCCAAACCAGAAAATAATAACCATTGGGTCAATC\n",
      "AAGCAGTGGTATCAACGCAGAGTACATGGGGGCCTCCTGACGGGCGAACGCATGAGTATAGACCGAATCCAAACCAGAAAATAATAACCACTGGGTCTAT\n",
      "AAGCAGTGGTATCAACGCAGAGTACATGGGGACGCGAGGACGCGAAAGCAGGAGTTTAAACCGAATCCAAACCAGAAAATAATAACCATTGGGTCAATCT\n",
      "AAGCAGTGGTATCAACGCAGAGTACATGGGCTTCCATTCAGTAGCAAAAGCAGGGGAAACCAAAAGCAACCAAATTGAAAGCAATACTAGTAGTTGTGCT\n",
      "AAGCAGTGGTATCAACGCACAGTACATGGGGGCCTAGGTGCCCGGCAAAAGCAGGGGAAAACAAAAGCAACCAAATTGAAAGCAATACTAATAGTTCTGC\n"
     ]
    }
   ],
   "source": [
    "for item in merged_bowtie_reads[(merged_bowtie_reads['host_or_virus'] == 'Host') & (merged_bowtie_reads['has_cell_bc'] == False)].iloc[0:10]['cds']:\n",
    "    print(item)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "BLAST results:\n",
    "\n",
    "Read 1 mapped to a random smattering of bacterial sequences.  \n",
    "Read 2 maps to WSN and avian influenza sequences.  \n",
    "Read 3 maps to modern pdmH1N1 sequences.  \n",
    "Read 4 maps to WSN and other historic H1N1 (not pdm) sequences.  \n",
    "Read 5 maps to WSN and avian influenza sequences. \n",
    "Read 6 maps to modern pdmH1N1 seuqences.\n",
    "\n",
    "\n",
    "Based on these results, I think these are mostly viral reads that did not map by the bowtie alignment. I may have to modify those settings a bit to make them less conservative."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Do the barcoded viral segments -- HA and NA -- have cell barcodes?\n",
    "\n",
    "Many of the viral reads have a cell barcodes. However, at least 1/4 do not. I want to check that these are not somehow exclusively transcrtips from the barcoded segments: HA and NA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HA reads\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>primer</th>\n",
       "      <th>truseq</th>\n",
       "      <th>bc</th>\n",
       "      <th>umi_or_cds</th>\n",
       "      <th>cds</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>host_or_virus</th>\n",
       "      <th>has_cell_bc</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">Virus</td>\n",
       "      <td>False</td>\n",
       "      <td>305</td>\n",
       "      <td>305</td>\n",
       "      <td>305</td>\n",
       "      <td>305</td>\n",
       "      <td>305</td>\n",
       "      <td>305</td>\n",
       "      <td>305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>170</td>\n",
       "      <td>170</td>\n",
       "      <td>170</td>\n",
       "      <td>170</td>\n",
       "      <td>170</td>\n",
       "      <td>170</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           record_id  primer  truseq   bc  umi_or_cds  cds  \\\n",
       "host_or_virus has_cell_bc                                                    \n",
       "Virus         False              305     305     305  305         305  305   \n",
       "              True               170     170     170  170         170  170   \n",
       "\n",
       "                           gene_name  \n",
       "host_or_virus has_cell_bc             \n",
       "Virus         False              305  \n",
       "              True               170  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('HA reads')\n",
    "merged_bowtie_reads[merged_bowtie_reads['gene_name']=='fluHA'].groupby(['host_or_virus','has_cell_bc']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NA reads\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>record_id</th>\n",
       "      <th>primer</th>\n",
       "      <th>truseq</th>\n",
       "      <th>bc</th>\n",
       "      <th>umi_or_cds</th>\n",
       "      <th>cds</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>host_or_virus</th>\n",
       "      <th>has_cell_bc</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td rowspan=\"2\" valign=\"top\">Virus</td>\n",
       "      <td>False</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>True</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           record_id  primer  truseq  bc  umi_or_cds  cds  \\\n",
       "host_or_virus has_cell_bc                                                   \n",
       "Virus         False               15      15      15  15          15   15   \n",
       "              True                29      29      29  29          29   29   \n",
       "\n",
       "                           gene_name  \n",
       "host_or_virus has_cell_bc             \n",
       "Virus         False               15  \n",
       "              True                29  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('NA reads')\n",
    "merged_bowtie_reads[merged_bowtie_reads['gene_name']=='fluNA'].groupby(['host_or_virus','has_cell_bc']).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the HA reads, most molecules have lost their cell barcode. For the NA reads, most molecules retain their cell barcodes. There are fewer NA reads than HA reads, but ultimately, we can use information from both segments to estimate how many virions entered a cell. While these numbers are not ideal, with greater sequencing depth, they may provide reliable estimates of virion entry for many cells.\n",
    "\n",
    "**I still think these data give me enough information to order more sequencing! At least half the virus reads in my data (maybe more) will have cell barcodes. This proportion is slightly less for HA and NA, but still workable. It is likely that the R2 sequences will contain the virus barcodes for any molecule that has a cell barcode. Therefore, with this sequencing scheme, we should be able to study multiple infection in individual cells.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook parse_barcodes.ipynb to html\n",
      "[NbConvertApp] Writing 330398 bytes to parse_barcodes.html\n"
     ]
    }
   ],
   "source": [
    "#Save to HTML for easy viewing on GitHub\n",
    "!jupyter nbconvert parse_barcodes.ipynb --to html --output parse_barcodes.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
